{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process and barcode sequencing reads\n",
    "This script takes paired read (read 1: barcode + UMI (27bp), read 2: staggers + UMI + partial barcode (49bp)) fastq files and does the following:\n",
    "1. trims read 2 adapter sequences to recover barcode+UMI sequence\n",
    "2. merges read 1 and 2 with FLASH\n",
    "3. identifies and counts barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "import os\n",
    "import subprocess\n",
    "import regex\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import multiprocessing as mp\n",
    "import numpy as np\n",
    "\n",
    "# check number of available cores\n",
    "len(os.sched_getaffinity(0))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspect fastq files for quality control using FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/home/users/rang/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/fastqc/’: File exists\r\n"
     ]
    }
   ],
   "source": [
    "# !mkdir ~/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/fastqc/\n",
    "# !fastqc -q -o ~/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/fastqc/ ~/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/*fastq.gz \n",
    "!mkdir ~/scratch/yeast/crispey3/gxg_bya_aug2021/fastq/fastqc/\n",
    "!fastqc -q -o ~/scratch/yeast/crispey3/gxg_bya_aug2021/fastq/fastqc/ ~/scratch/yeast/crispey3/gxg_bya_aug2021/fastq/*fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summarize FastQC output with MultiQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;30m[WARNING]\u001b[0m         multiqc : \u001b[33mMultiQC Version v1.11 now available!\u001b[0m\n",
      "\u001b[1;30m[INFO   ]\u001b[0m         multiqc : This is MultiQC v1.9\n",
      "\u001b[1;30m[INFO   ]\u001b[0m         multiqc : Template    : default\n",
      "\u001b[1;30m[INFO   ]\u001b[0m         multiqc : Searching   : /home/users/rang/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/fastqc\n",
      "\u001b[?25lSearching 145 files..  [####################################]  100%          \u001b[?25h\n",
      "\u001b[1;30m[INFO   ]\u001b[0m          fastqc : Found 72 reports\n",
      "\u001b[1;30m[INFO   ]\u001b[0m         multiqc : Compressing plot data\n",
      "\u001b[1;30m[WARNING]\u001b[0m         multiqc : \u001b[33mPrevious MultiQC output found! Adjusting filenames..\u001b[0m\n",
      "\u001b[1;30m[WARNING]\u001b[0m         multiqc : \u001b[33mUse -f or --force to overwrite existing reports instead\u001b[0m\n",
      "\u001b[1;30m[INFO   ]\u001b[0m         multiqc : Report      : ../../../scratch/yeast/crispey3/pool1_scm_aug2021/fastq/fastqc/multiqc_report_1.html\n",
      "\u001b[1;30m[INFO   ]\u001b[0m         multiqc : Data        : ../../../scratch/yeast/crispey3/pool1_scm_aug2021/fastq/fastqc/multiqc_data_1\n",
      "\u001b[1;30m[INFO   ]\u001b[0m         multiqc : MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "# !multiqc -o ~/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/fastqc/ ~/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/fastqc/\n",
    "!multiqc -o ~/scratch/yeast/crispey3/gxg_bya_aug2021/fastq/fastqc/ ~/scratch/yeast/crispey3/gxg_bya_aug2021/fastq/fastqc/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map fastq file names to sample names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# key to map fastq names to output names\n",
    "seqID_to_sampleName = {}\n",
    "# sample_key_file = \"/home/users/rang/scratch/yeast/crispey3/pool1_scm_aug2021/SampleKey-18146-36.txt\"\n",
    "sample_key_file = \"/home/users/rang/scratch/yeast/crispey3/gxg_bya_aug2021/SampleKey-18146-35.txt\"\n",
    "with open(sample_key_file, 'r') as sample_key:\n",
    "    sample_key.readline() # skip header\n",
    "    for line in sample_key:\n",
    "        seqID, sampleName = line.rstrip().split(\"\\t\")\n",
    "        sampleName = sampleName.replace(\"_\",\"-\")\n",
    "        seqID_to_sampleName[seqID] = sampleName\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trim read 2 adapters with cutadapt\n",
    "Remove staggers, leaving the UMI+partial barcode sequence (19-26bp, may be shorter depending on quality trimming)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# working directory with fastq files\n",
    "# working_dir=\"/home/users/rang/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/\"\n",
    "working_dir=\"/home/users/rang/scratch/yeast/crispey3/gxg_bya_aug2021/fastq/\"\n",
    "os.chdir(working_dir)\n",
    "\n",
    "# get read 2 files for trimming\n",
    "fastq_list = sorted([os.path.abspath(x) for x in glob.glob(\"*R2_001.fastq.gz\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cutadapt parameters to trim read 2 to get barcode+UMI (27bp)\n",
    "adapter_5prime = 'GGCCAGTTTAAACTT'\n",
    "adapter_3prime = 'GCATGGC'\n",
    "\n",
    "num_of_cores = len(os.sched_getaffinity(0))\n",
    "err = 0.2 # fraction tolerated for adapter matching\n",
    "min_r2_length = 12 # R2 must contain at least UMI sequence and some of SphI linker\n",
    "output_dir_name = 'trimmed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-01_S1_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t1-1_S1_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t1-1_S1_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-02_S2_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t1-2_S2_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t1-2_S2_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-03_S3_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t1-3_S3_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t1-3_S3_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-04_S4_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t4-1_S4_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t4-1_S4_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-05_S5_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t4-2_S5_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t4-2_S5_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-06_S6_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t4-3_S6_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t4-3_S6_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-07_S7_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t7-1_S7_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t7-1_S7_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-08_S8_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t7-2_S8_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t7-2_S8_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-09_S9_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t7-3_S9_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t7-3_S9_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-10_S10_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t10-1_S10_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t10-1_S10_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-11_S11_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t10-2_S11_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t10-2_S11_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-12_S12_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t10-3_S12_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t10-3_S12_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-13_S13_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t13-1_S13_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t13-1_S13_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-14_S14_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t13-2_S14_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t13-2_S14_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Trimming: /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/18146FL-35-01V1-15_S15_L001_R2_001.fastq.gz\n",
      "Output files:\n",
      "BYA-t13-3_S15_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t13-3_S15_L001_R1_001_trimmed.fastq.gz\n",
      "\n",
      "Done trimming!\n"
     ]
    }
   ],
   "source": [
    "# store sample key in regex pattern\n",
    "pattern = regex.compile('|'.join(seqID_to_sampleName.keys()))\n",
    "\n",
    "# trim read 2, filter untrimmed read pairs\n",
    "for fastq_path in fastq_list:\n",
    "    fastq_dir = os.path.dirname(fastq_path)\n",
    "    output_dir = fastq_dir + \"/\"+output_dir_name+\"/\"\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    # rename output files by sample key stored in seqID_to_sampleName \n",
    "    fastq_file = os.path.basename(fastq_path)\n",
    "    output_file_r2 = pattern.sub(lambda x: seqID_to_sampleName[x.group()], fastq_file).replace(\"_001.fastq.gz\", \"_001_trimmed.fastq.gz\")\n",
    "    output_file_r1 = output_file_r2.replace(\"_R2_\", \"_R1_\")\n",
    "\n",
    "    print('Trimming: ' + fastq_path)\n",
    "    \n",
    "    cutadapt_cmd = [\"cutadapt\", \"-g\", adapter_5prime+\"...\"+adapter_3prime, #adapter_5prime, \n",
    "                    \"-j\", str(num_of_cores), \n",
    "                    \"-e\", str(err), \n",
    "                    #\"-q\", \"20\", # use -q for miseq/hiseq quality trimming\n",
    "                    \"--nextseq-trim\", \"20\", # use this option for nextseq quality trimming\n",
    "                    \"--discard-untrimmed\", \"-m\", str(min_r2_length), \n",
    "                    \"--pair-filter=first\", \n",
    "                    \"-o\", output_dir+output_file_r2, \"-p\", output_dir+output_file_r1,\n",
    "                    fastq_path, fastq_path.replace(\"L001_R2_\", \"L001_R1_\")]\n",
    "    \n",
    "    subprocess.run(cutadapt_cmd)\n",
    "\n",
    "    print(\"Output files:\")\n",
    "    print(output_file_r2)\n",
    "    print(output_file_r1)\n",
    "    print()\n",
    "    \n",
    "print('Done trimming!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge read 1 and read 2 with FLASH to produce final barcode+UMI sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir=\"/home/users/rang/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/trimmed/\"\n",
    "working_dir=\"/home/users/rang/scratch/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/\"\n",
    "os.chdir(working_dir)\n",
    "\n",
    "fastq_list = sorted([os.path.abspath(x) for x in glob.glob(\"*R1_001_trimmed.fastq.gz\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FLASH parameters\n",
    "min_overlap = 8 # cannot be longer than the shorter read.\n",
    "max_overlap = 12 # cannot be longer than the shorter read.\n",
    "max_mismatch = 0.25 #default is 0.25, set 0.4 to tolerate lower seq quality\n",
    "num_of_cores = len(os.sched_getaffinity(0))\n",
    "output_dir_name = 'merged'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t1-1_S1_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t1-1_S1_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t1-1_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t1-2_S2_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t1-2_S2_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t1-2_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t1-3_S3_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t1-3_S3_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t1-3_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t10-1_S10_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t10-1_S10_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t10-1_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t10-2_S11_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t10-2_S11_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t10-2_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t10-3_S12_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t10-3_S12_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t10-3_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t13-1_S13_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t13-1_S13_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t13-1_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t13-2_S14_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t13-2_S14_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t13-2_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t13-3_S15_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t13-3_S15_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t13-3_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t4-1_S4_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t4-1_S4_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t4-1_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t4-2_S5_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t4-2_S5_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t4-2_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t4-3_S6_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t4-3_S6_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t4-3_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t7-1_S7_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t7-1_S7_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t7-1_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t7-2_S8_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t7-2_S8_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t7-2_barcode merged\n",
      "Merging /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t7-3_S9_L001_R1_001_trimmed.fastq.gz and /scratch/users/rang/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/BYA-t7-3_S9_L001_R2_001_trimmed.fastq.gz\n",
      "BYA-t7-3_barcode merged\n",
      "Done merging!\n"
     ]
    }
   ],
   "source": [
    "# use FLASH to merge trimmed-filtered read 2 and read 1 data to produce final 27bp sequence containing barcode and UMI data\n",
    "for fastq_path in fastq_list:\n",
    "    fastq_dir = os.path.dirname(fastq_path)\n",
    "    output_dir = fastq_dir + \"/\"+output_dir_name+\"/\"\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    # check output file naming \n",
    "    output_prefix = os.path.basename(fastq_path).split(\"_\")[0]+\"_barcode\"\n",
    "    print('Merging', fastq_path, 'and', fastq_path.replace(\"_R1_\", \"_R2_\"))\n",
    "    \n",
    "    flash_cmd = [\"flash\", \"-m\", str(min_overlap), \"-M\", str(max_overlap),\n",
    "                 \"-x\", str(max_mismatch), #\"-O\", # use -O if innie-only merging does not work\n",
    "                 \"-t\", str(num_of_cores),\n",
    "                 \"-o\", output_prefix, \"-d\", output_dir, \n",
    "                 \"--compress\", \n",
    "                 fastq_path, fastq_path.replace(\"_R1_\", \"_R2_\")]\n",
    "    subprocess.run(flash_cmd)\n",
    "    print(output_prefix, \"merged\")\n",
    "\n",
    "print('Done merging!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (optional) Downsample reads for analysis\n",
    "Use seqtk in command line to downsample fastq files prior to assembling counts matrix.<br>\n",
    "e.g. seqtk sample -s100 read1.fq 10000 > sub1.fq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count barcodes\n",
    "Counting barcodes consists of several steps. First, parse each fastq file and count all sequences. After assembling into an initial sequences counts matrix, extract the barcode and UMI sequences and map them to a reference table of barcodes and UMIs. Counts for ID-able sequences are onsolidated into a final counts matrix for input to DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_seqs(fastq_file, min_seq_length, max_seq_length):\n",
    "    '''\n",
    "    Parses a fastq file and counts sequences. Returns dict of counts\n",
    "    '''\n",
    "    seq_counts_dict = {}\n",
    "    # parse fastq\n",
    "    with gzip.open(fastq_file, 'rt') as fastq:\n",
    "        for read in SeqIO.parse(fastq, \"fastq\"):\n",
    "            # filter for sequences within min/max length and contains no N's\n",
    "            if min_seq_length <= len(read.seq) <= max_seq_length and read.seq.count(\"N\")==0:\n",
    "                sequence = str(read.seq)\n",
    "                # count sequence\n",
    "                try:\n",
    "                    seq_counts_dict[sequence] += 1\n",
    "                except KeyError:\n",
    "                    seq_counts_dict[sequence] = 1\n",
    "    \n",
    "    return seq_counts_dict\n",
    "\n",
    "    \n",
    "def map_seq_to_barcode_umi(seq, barcode_table, umi_list, barcode_length, umi_length, linker_seq):\n",
    "    '''\n",
    "    splits a sequence into barcode and UMI, maps to barcode table and UMI list to assign ID\n",
    "    does NOT do UMI mapping if umi_list is empty.\n",
    "    '''\n",
    "    final_id = 'UNKNOWN'\n",
    "    \n",
    "    barcode, umi = split_barcode_umi_from_seq(seq, barcode_length, umi_length, linker_seq)\n",
    "    \n",
    "    if len(barcode)<barcode_length/2 or len(umi)<umi_length/2:\n",
    "        # Barcode/UMI too short\n",
    "        return final_id\n",
    "\n",
    "    # assign barcode ID\n",
    "    barcode_id = assign_barcode(barcode=barcode, barcode_table=barcode_table, error=1)\n",
    "    if barcode_id == 'UNKNOWN':\n",
    "        # Barcode cannot be identified\n",
    "        return barcode_id\n",
    "    else:\n",
    "        # add barcode ID to final ID\n",
    "        final_id = barcode_id\n",
    "\n",
    "    # assign UMI ID (if applicable)\n",
    "    if len(umi_list)>0:\n",
    "        umi_id = assign_umi(umi=umi, umi_list=umi_list, error=1)\n",
    "        if umi_id == 'UNKNOWN':\n",
    "            # UMI cannot be identified\n",
    "            return umi_id\n",
    "        else:\n",
    "            # add UMI ID to final ID\n",
    "            final_id = '-'.join([final_id, str(umi_id)])\n",
    "    \n",
    "    return final_id\n",
    "    \n",
    "\n",
    "def split_barcode_umi_from_seq(seq, barcode_length, umi_length, linker_seq):\n",
    "    '''\n",
    "    Splits seq by linker_seq and returns barcode and UMI sequence\n",
    "    If linker seq cannot be found (e.g. sequencing error) or yields multiple splits,\n",
    "    fall back to splitting by base position.\n",
    "    '''\n",
    "    barcode = ''\n",
    "    umi = ''\n",
    "    \n",
    "    try:\n",
    "        # split by linker\n",
    "        barcode, umi = seq.split(linker_seq) # can try error tolerant regex?\n",
    "    except ValueError:\n",
    "        # split by base position\n",
    "        if barcode_length>0 and umi_length==0:\n",
    "            barcode = seq[:barcode_length]\n",
    "        else:\n",
    "            umi = seq[-umi_length:]\n",
    "            barcode = seq[:-(umi_length+len(linker_seq))] # may return partial barcodes for short sequences\n",
    "\n",
    "    return (barcode, umi)\n",
    "\n",
    "\n",
    "def assign_barcode(barcode, barcode_table, error):\n",
    "    '''\n",
    "    Searches barcode table for barcode sequence and returns unique barcode ID\n",
    "    Tries perfect match first, then error-tolerant regex\n",
    "    '''\n",
    "    if barcode == '':\n",
    "        barcode_id = 'UNKNOWN'\n",
    "    else:\n",
    "        try:\n",
    "            # search for perfect match\n",
    "            barcode_id = barcode_table.loc[barcode, 'Unique_ID']\n",
    "        except KeyError:\n",
    "            # search by error-tolerant regex\n",
    "            pattern = \"(?:\"+barcode+\"){s<=\"+str(error)+\"}\"\n",
    "            search = [bool(regex.search(pattern, x)) for x in barcode_table.index]\n",
    "            if sum(search)==1:\n",
    "                barcode_id = barcode_table.loc[search, 'Unique_ID'][0]\n",
    "            else:\n",
    "                # barcode cannot be identified\n",
    "                barcode_id = 'UNKNOWN'\n",
    "    \n",
    "    return barcode_id\n",
    "\n",
    "\n",
    "def assign_umi(umi, umi_list, error):\n",
    "    '''\n",
    "    Searches umi list for umi sequence and returns 1-index position as umi ID\n",
    "    Tries perfect match first, then error-tolerant regex\n",
    "    '''\n",
    "    if umi == '':\n",
    "        umi_id = 'UNKNOWN'\n",
    "    else:\n",
    "        try:\n",
    "            # search for perfect match\n",
    "            umi_id = umi_list.index(umi)+1\n",
    "        except ValueError:\n",
    "            # search by error-tolerant regex\n",
    "            pattern = \"(?:\"+umi+\"){s<=\"+str(error)+\"}\"\n",
    "            search = [bool(regex.search(pattern, x)) for x in umi_list]\n",
    "            if sum(search) == 1:\n",
    "                umi_id = search.index(True)+1\n",
    "            else:\n",
    "                # UMI cannot be identified\n",
    "                umi_id = 'UNKNOWN'\n",
    "    \n",
    "    return umi_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir=\"/home/users/rang/scratch/yeast/crispey3/pool1_scm_aug2021/fastq/trimmed/merged/\"\n",
    "working_dir=\"/home/users/rang/scratch/yeast/crispey3/gxg_bya_aug2021/fastq/trimmed/merged/\"\n",
    "os.chdir(working_dir)\n",
    "\n",
    "#output directory\n",
    "# output_dir = \"/home/users/rang/scratch/yeast/crispey3/pool1_scm_aug2021/counts/\"\n",
    "output_dir = \"/home/users/rang/scratch/yeast/crispey3/gxg_bya_aug2021/counts/\"\n",
    "\n",
    "# merged reads to count barcodes from\n",
    "fastq_list = sorted(glob.glob(\"*extendedFrags.fastq.gz\")) # check for file name\n",
    "# sample names for each fastq\n",
    "sample_name_list = [fastq_file.split(\"_\")[0] for fastq_file in fastq_list] # adjust accordingly to generate sample name for counts matrix\n",
    "\n",
    "# sequence counts file (before combining)\n",
    "seq_counts_filename = \"seq_counts.txt\"\n",
    "\n",
    "# mapped barcode-UMI counts file\n",
    "barcode_counts_filename = \"barcode_counts.txt\"\n",
    "\n",
    "\n",
    "# open barcode reference file\n",
    "barcode_reference_file = '/home/users/rang/crispey3/library_design/Input/12BP_PBCs_well_grouped.csv'\n",
    "barcode_table = pd.read_csv(barcode_reference_file, index_col=1)\n",
    "\n",
    "# approved list of UMIs used in cloning CRISPEY3 plasmid\n",
    "umi_list = ['ACGCGTGAA',\n",
    "            'ATGTGGCTC',\n",
    "            'CAGAGGATC',\n",
    "            'CTGTGGCAA',\n",
    "            'GTGTGATTC',\n",
    "            'TAGAGGACT',] # only first 6 UMIs were included in cloning CRISPEY3 libaries\n",
    "#             'AAGAGCCTC',\n",
    "#             'AAGAGGAGG',\n",
    "#             'ATGTGCGAA',\n",
    "#             'ATGTGTAGG',\n",
    "#             'CAGAGCCAA',\n",
    "#             'CTGTGATGG',\n",
    "#             'CTGTGTATC',\n",
    "#             'GAGAGGAAA',\n",
    "#             'TCGCGGTAA',\n",
    "#             'TTGTGCGTC']\n",
    "umi_list = sorted(umi_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count sequences in each fastq file\n",
    "fastq_dict = dict(zip(sample_name_list, fastq_list))\n",
    "with mp.Pool(min(len(os.sched_getaffinity(0)), len(fastq_list))) as pool:\n",
    "    seq_counts_df = {sample_name : pool.apply_async(count_seqs, (fastq_file, 20, 27)) for sample_name, fastq_file in fastq_dict.items()}\n",
    "    seq_counts_df = {sample_name : res.get() for sample_name, res in seq_counts_df.items()}\n",
    "    \n",
    "# merge to dataframe\n",
    "seq_counts_df = pd.DataFrame.from_dict(seq_counts_df, orient=\"columns\")\n",
    "seq_counts_df.index.name = 'sequence'\n",
    "\n",
    "# write to file to inspect\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "seq_counts_df.to_csv(output_dir+seq_counts_filename, sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# seq_counts_df = pd.read_csv(output_dir+seq_counts_filename, sep=\"\\t\", index_col='sequence')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>BYA-t1-1</th>\n",
       "      <th>BYA-t1-2</th>\n",
       "      <th>BYA-t1-3</th>\n",
       "      <th>BYA-t10-1</th>\n",
       "      <th>BYA-t10-2</th>\n",
       "      <th>BYA-t10-3</th>\n",
       "      <th>BYA-t13-1</th>\n",
       "      <th>BYA-t13-2</th>\n",
       "      <th>BYA-t13-3</th>\n",
       "      <th>BYA-t4-1</th>\n",
       "      <th>BYA-t4-2</th>\n",
       "      <th>BYA-t4-3</th>\n",
       "      <th>BYA-t7-1</th>\n",
       "      <th>BYA-t7-2</th>\n",
       "      <th>BYA-t7-3</th>\n",
       "      <th>barcode_umi_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TATCGTCTTCATGCATGCACGCGTGAA</td>\n",
       "      <td>4302.0</td>\n",
       "      <td>5068.0</td>\n",
       "      <td>4616.0</td>\n",
       "      <td>4728.0</td>\n",
       "      <td>5121.0</td>\n",
       "      <td>5249.0</td>\n",
       "      <td>5065.0</td>\n",
       "      <td>5002.0</td>\n",
       "      <td>5508.0</td>\n",
       "      <td>4307.0</td>\n",
       "      <td>6222.0</td>\n",
       "      <td>4616.0</td>\n",
       "      <td>4677.0</td>\n",
       "      <td>4721.0</td>\n",
       "      <td>4993.0</td>\n",
       "      <td>054_037-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGCTGGACTCGTGCATGCCTGTGGCAA</td>\n",
       "      <td>12467.0</td>\n",
       "      <td>14109.0</td>\n",
       "      <td>13828.0</td>\n",
       "      <td>13657.0</td>\n",
       "      <td>14528.0</td>\n",
       "      <td>15180.0</td>\n",
       "      <td>13899.0</td>\n",
       "      <td>13459.0</td>\n",
       "      <td>15149.0</td>\n",
       "      <td>13201.0</td>\n",
       "      <td>17873.0</td>\n",
       "      <td>14381.0</td>\n",
       "      <td>13410.0</td>\n",
       "      <td>13931.0</td>\n",
       "      <td>15180.0</td>\n",
       "      <td>065_058-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TATTAAGAGCGCGCATGCCAGAGGATC</td>\n",
       "      <td>9801.0</td>\n",
       "      <td>10695.0</td>\n",
       "      <td>10322.0</td>\n",
       "      <td>11307.0</td>\n",
       "      <td>11981.0</td>\n",
       "      <td>11757.0</td>\n",
       "      <td>11664.0</td>\n",
       "      <td>11432.0</td>\n",
       "      <td>12113.0</td>\n",
       "      <td>10351.0</td>\n",
       "      <td>14210.0</td>\n",
       "      <td>10831.0</td>\n",
       "      <td>10809.0</td>\n",
       "      <td>11083.0</td>\n",
       "      <td>11452.0</td>\n",
       "      <td>076_066-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTCCTTGTTCGCGCATGCGTGTGATTC</td>\n",
       "      <td>216.0</td>\n",
       "      <td>284.0</td>\n",
       "      <td>245.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>143.0</td>\n",
       "      <td>265.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>310.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>048_064-5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCCGAAACTAGAGCATGCCAGAGGATC</td>\n",
       "      <td>12874.0</td>\n",
       "      <td>14597.0</td>\n",
       "      <td>14325.0</td>\n",
       "      <td>13189.0</td>\n",
       "      <td>13620.0</td>\n",
       "      <td>12758.0</td>\n",
       "      <td>12596.0</td>\n",
       "      <td>12369.0</td>\n",
       "      <td>12304.0</td>\n",
       "      <td>13450.0</td>\n",
       "      <td>18026.0</td>\n",
       "      <td>14225.0</td>\n",
       "      <td>13128.0</td>\n",
       "      <td>13413.0</td>\n",
       "      <td>14133.0</td>\n",
       "      <td>035_004-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2424829</th>\n",
       "      <td>TCGCTCACAAACGCACGCATGTGGCTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>040_038-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2425244</th>\n",
       "      <td>TCCGTGTCGGAAATTACCCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>UNKNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2425708</th>\n",
       "      <td>ATCGGTATCATTGGCTTAGA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>UNKNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2426570</th>\n",
       "      <td>GTCGCACGGTGACGGAGGAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>UNKNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2435655</th>\n",
       "      <td>AGTCTGGGAAAATTTAAGAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>UNKNOWN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>641160 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            sequence  BYA-t1-1  BYA-t1-2  BYA-t1-3  BYA-t10-1  \\\n",
       "0        TATCGTCTTCATGCATGCACGCGTGAA    4302.0    5068.0    4616.0     4728.0   \n",
       "1        AGCTGGACTCGTGCATGCCTGTGGCAA   12467.0   14109.0   13828.0    13657.0   \n",
       "2        TATTAAGAGCGCGCATGCCAGAGGATC    9801.0   10695.0   10322.0    11307.0   \n",
       "3        CTCCTTGTTCGCGCATGCGTGTGATTC     216.0     284.0     245.0      176.0   \n",
       "4        GCCGAAACTAGAGCATGCCAGAGGATC   12874.0   14597.0   14325.0    13189.0   \n",
       "...                              ...       ...       ...       ...        ...   \n",
       "2424829  TCGCTCACAAACGCACGCATGTGGCTC       NaN       NaN       NaN        NaN   \n",
       "2425244         TCCGTGTCGGAAATTACCCC       NaN       NaN       NaN        NaN   \n",
       "2425708         ATCGGTATCATTGGCTTAGA       NaN       NaN       NaN        NaN   \n",
       "2426570         GTCGCACGGTGACGGAGGAA       NaN       NaN       NaN        NaN   \n",
       "2435655         AGTCTGGGAAAATTTAAGAG       NaN       NaN       NaN        NaN   \n",
       "\n",
       "         BYA-t10-2  BYA-t10-3  BYA-t13-1  BYA-t13-2  BYA-t13-3  BYA-t4-1  \\\n",
       "0           5121.0     5249.0     5065.0     5002.0     5508.0    4307.0   \n",
       "1          14528.0    15180.0    13899.0    13459.0    15149.0   13201.0   \n",
       "2          11981.0    11757.0    11664.0    11432.0    12113.0   10351.0   \n",
       "3            143.0      265.0      116.0      133.0      250.0     221.0   \n",
       "4          13620.0    12758.0    12596.0    12369.0    12304.0   13450.0   \n",
       "...            ...        ...        ...        ...        ...       ...   \n",
       "2424829        NaN        NaN        NaN        NaN        NaN       NaN   \n",
       "2425244        NaN        NaN        NaN        NaN        NaN       NaN   \n",
       "2425708        NaN        NaN        NaN        NaN        NaN       NaN   \n",
       "2426570        NaN        NaN        NaN        NaN        NaN       NaN   \n",
       "2435655        NaN        NaN        NaN        NaN        NaN       NaN   \n",
       "\n",
       "         BYA-t4-2  BYA-t4-3  BYA-t7-1  BYA-t7-2  BYA-t7-3 barcode_umi_id  \n",
       "0          6222.0    4616.0    4677.0    4721.0    4993.0      054_037-1  \n",
       "1         17873.0   14381.0   13410.0   13931.0   15180.0      065_058-4  \n",
       "2         14210.0   10831.0   10809.0   11083.0   11452.0      076_066-3  \n",
       "3           310.0     230.0     187.0     214.0     301.0      048_064-5  \n",
       "4         18026.0   14225.0   13128.0   13413.0   14133.0      035_004-3  \n",
       "...           ...       ...       ...       ...       ...            ...  \n",
       "2424829       NaN       NaN       NaN       NaN       3.0      040_038-2  \n",
       "2425244       NaN       NaN       NaN       NaN       3.0        UNKNOWN  \n",
       "2425708       NaN       NaN       NaN       NaN       4.0        UNKNOWN  \n",
       "2426570       NaN       NaN       NaN       NaN       3.0        UNKNOWN  \n",
       "2435655       NaN       NaN       NaN       NaN       3.0        UNKNOWN  \n",
       "\n",
       "[641160 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BYA-t1-1</th>\n",
       "      <th>BYA-t1-2</th>\n",
       "      <th>BYA-t1-3</th>\n",
       "      <th>BYA-t10-1</th>\n",
       "      <th>BYA-t10-2</th>\n",
       "      <th>BYA-t10-3</th>\n",
       "      <th>BYA-t13-1</th>\n",
       "      <th>BYA-t13-2</th>\n",
       "      <th>BYA-t13-3</th>\n",
       "      <th>BYA-t4-1</th>\n",
       "      <th>BYA-t4-2</th>\n",
       "      <th>BYA-t4-3</th>\n",
       "      <th>BYA-t7-1</th>\n",
       "      <th>BYA-t7-2</th>\n",
       "      <th>BYA-t7-3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>001_022-6</th>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>34</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>37</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "      <td>13</td>\n",
       "      <td>4</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>001_045-6</th>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>001_119-6</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>002_017-2</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>002_019-6</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ladder_042-3</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ladder_043-1</th>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ladder_043-2</th>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ladder_043-3</th>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ladder_043-4</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32078 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              BYA-t1-1  BYA-t1-2  BYA-t1-3  BYA-t10-1  BYA-t10-2  BYA-t10-3  \\\n",
       "barcode                                                                       \n",
       "001_022-6            5         4         5         34         20         21   \n",
       "001_045-6            7         6         2          0          0          0   \n",
       "001_119-6            7         0         2          0          0          0   \n",
       "002_017-2            1        10         5          0          1          2   \n",
       "002_019-6            2         3         2          0          0          3   \n",
       "...                ...       ...       ...        ...        ...        ...   \n",
       "Ladder_042-3         0         1         2          0          0          0   \n",
       "Ladder_043-1        19         9        11          3         11          3   \n",
       "Ladder_043-2         8        10        11          0          2          1   \n",
       "Ladder_043-3         9        13        10          3          4          9   \n",
       "Ladder_043-4         1         2         3          0          0          2   \n",
       "\n",
       "              BYA-t13-1  BYA-t13-2  BYA-t13-3  BYA-t4-1  BYA-t4-2  BYA-t4-3  \\\n",
       "barcode                                                                       \n",
       "001_022-6            37         18         28        13         4        13   \n",
       "001_045-6             0          0          0         2         2         1   \n",
       "001_119-6             0          0          0         2         2         4   \n",
       "002_017-2             0          0          0         2        11         6   \n",
       "002_019-6             0          0          0         0         0         5   \n",
       "...                 ...        ...        ...       ...       ...       ...   \n",
       "Ladder_042-3          0          0          0         0         0         1   \n",
       "Ladder_043-1          0          7          2        20         1        13   \n",
       "Ladder_043-2          0          3          0         4        12         9   \n",
       "Ladder_043-3          8          3          8         4         8         5   \n",
       "Ladder_043-4          0          0          0         1         8         2   \n",
       "\n",
       "              BYA-t7-1  BYA-t7-2  BYA-t7-3  \n",
       "barcode                                     \n",
       "001_022-6           10        17        20  \n",
       "001_045-6            2         6         0  \n",
       "001_119-6            0         0         2  \n",
       "002_017-2            0         6         3  \n",
       "002_019-6            0         0         5  \n",
       "...                ...       ...       ...  \n",
       "Ladder_042-3         0         0         0  \n",
       "Ladder_043-1         5        14        12  \n",
       "Ladder_043-2         5        13         0  \n",
       "Ladder_043-3        17         5        10  \n",
       "Ladder_043-4         0         0         3  \n",
       "\n",
       "[32078 rows x 15 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# map each sequence in seq_counts_df to barcode-UMI ID\n",
    "mapped_counts_df = seq_counts_df.copy()\n",
    "mapped_counts_df = mapped_counts_df.reset_index()\n",
    "\n",
    "# filter out low count barcodes (computationally expensive to map these rare barcodes, minimal impact to total count (~2%))\n",
    "cutoff = 3\n",
    "mapped_counts_df = mapped_counts_df.loc[mapped_counts_df.sum(axis=1)>=cutoff]\n",
    "\n",
    "# do multiprocessing for sequence mapping\n",
    "num_of_cores = len(os.sched_getaffinity(0))\n",
    "with mp.Pool(num_of_cores) as pool:\n",
    "    barcode_umi_id_list = [pool.apply_async(map_seq_to_barcode_umi, (seq, barcode_table, umi_list, 12, 9, 'GCATGC')) for seq in mapped_counts_df['sequence']]\n",
    "    barcode_umi_id_list = [res.get() for res in barcode_umi_id_list]\n",
    "    \n",
    "# mapped_counts_df['barcode_umi_id'] = [x for res in barcode_umi_id_lists for x in res] #pd.concat(barcode_umi_id_lists)\n",
    "mapped_counts_df['barcode_umi_id'] = barcode_umi_id_list\n",
    "display(mapped_counts_df)\n",
    "\n",
    "# remove unknowns\n",
    "mapped_counts_df = mapped_counts_df.query('barcode_umi_id!=\"UNKNOWN\"')\n",
    "\n",
    "# consolidate counts\n",
    "mapped_counts_df = mapped_counts_df.groupby('barcode_umi_id').sum().fillna(0).astype(int)\n",
    "mapped_counts_df.index.name = 'barcode'\n",
    "display(mapped_counts_df)\n",
    "\n",
    "# write all counts to output file\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "mapped_counts_df.to_csv(output_dir+barcode_counts_filename, sep=\"\\t\")\n",
    "\n",
    "\n",
    "# # one-liner to map sequences to barcode-umi IDs\n",
    "# # warning: single-threaded, expected to be slow\n",
    "# mapped_counts_df['barcode_umi_id'] = mapped_counts_df['sequence'].apply(map_seq_to_barcode_umi, args=(barcode_table, umi_list, 12, 9, 'GCATGC'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (optional) Combine counts across UMIs per barcode\n",
    "The counts of different UMIs of the same barcode can be added together to produce a stacked counts matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine counts from different UMIs of the same barcode\n",
    "stacked_counts_filename = \"stacked_barcode_counts.txt\"\n",
    "\n",
    "stacked_counts_df = mapped_counts_df.groupby(by=lambda x: x.split('-')[0]).sum()\n",
    "stacked_counts_df.index.name = 'barcode'\n",
    "stacked_counts_df.to_csv(output_dir+stacked_counts_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # adapt ladder fastq files by trimming to new seq format (R1: 27bp, R2: 49bp)\n",
    "# r1_len = 27\n",
    "# r2_len = 49\n",
    "\n",
    "# # working directory with fastq files\n",
    "# working_dir=\"/home/users/rang/scratch/yeast/tmp/ladder_pilot_adapted/fastq/\"\n",
    "# os.chdir(working_dir)\n",
    "\n",
    "# # trim R1\n",
    "# fastq_list = sorted([os.path.abspath(x) for x in glob.glob(\"*R1_001.fastq.gz\")])\n",
    "# for fastq_path in fastq_list:\n",
    "#     output_path = fastq_path.replace(\"_R1_001\", \"_R1_adapted\")\n",
    "#     cutadapt_cmd = [\"cutadapt\", \"-l\", str(r1_len),\n",
    "#                     \"-o\", output_path, fastq_path]\n",
    "#     subprocess.run(cutadapt_cmd)\n",
    "    \n",
    "# # trim R2\n",
    "# fastq_list = sorted([os.path.abspath(x) for x in glob.glob(\"*R2_001.fastq.gz\")])\n",
    "# for fastq_path in fastq_list:\n",
    "#     output_path = fastq_path.replace(\"_R2_001\", \"_R2_adapted\")\n",
    "#     cutadapt_cmd = [\"cutadapt\", \"-l\", str(r2_len),\n",
    "#                     \"-o\", output_path, fastq_path]\n",
    "#     subprocess.run(cutadapt_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # one-step count_barcodes function\n",
    "# # warning: less efficient since barcode-UMI mapping is done per sample, rather than a single time after all sequences are counted\n",
    "\n",
    "# def count_barcodes(fastq_file, barcode_table, umi_list, \n",
    "#                    min_seq_length=20, barcode_length=12, umi_length=9, linker_seq='GCATGC'):\n",
    "#     '''\n",
    "#     parses fastq file to count sequences, then extracts barcode-UMI info from sequences and assigns ID based\n",
    "#     on provided reference barcode_table and umi_list. Finally, consolidates counts by assigned ID and returns\n",
    "#     a dict of barcode counts.\n",
    "#     does NOT do UMI mapping if umi_list is empty.\n",
    "#     '''\n",
    "#     # set max_seq_length\n",
    "#     max_seq_length = barcode_length+umi_length+len(linker_seq) # this could be adjusted to allow insertions\n",
    "#     # alphabetical sort umi_list\n",
    "#     umi_list = sorted(umi_list)\n",
    "    \n",
    "#     # count raw sequences\n",
    "#     seq_counts = count_seqs(fastq_file, min_seq_length, max_seq_length)\n",
    "    \n",
    "#     # consolidate barcode counts\n",
    "#     barcode_counts_dict = {}\n",
    "#     for seq, count in seq_counts.items():\n",
    "#         # assign barcode-UMI ID\n",
    "#         assigned_id = map_seq_to_barcode_umi(seq, barcode_table, umi_list, barcode_length, umi_length, linker_seq)\n",
    "#         if assigned_id:\n",
    "#             try:\n",
    "#                 barcode_counts_dict[assigned_id] += count\n",
    "#             except KeyError:\n",
    "#                 barcode_counts_dict[assigned_id] = count\n",
    "    \n",
    "#     return barcode_counts_dict\n",
    "\n",
    "\n",
    "# fastq_dict = dict(zip(sample_name_list, fastq_list))\n",
    "# with mp.Pool(min(len(os.sched_getaffinity(0)), len(fastq_list))) as pool:\n",
    "#     all_counts_df = {sample_name : pool.apply_async(count_barcodes, (fastq_file, barcode_table, umi_list)) for sample_name, fastq_file in fastq_dict.items()}\n",
    "#     all_counts_df = {sample_name : res.get() for sample_name, res in all_counts_df.items()}\n",
    "    \n",
    "# # write all counts to output file\n",
    "# os.makedirs(output_dir, exist_ok=True)\n",
    "# all_counts_df = pd.DataFrame.from_dict(all_counts_df, orient=\"columns\")\n",
    "# all_counts_df.index.name = 'barcode'\n",
    "# all_counts_df.to_csv(output_dir+counts_filename, sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (crispey-bar)",
   "language": "python",
   "name": "crispey-bar"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
