{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize variants selected from select_crispey3_library_variants_2020 notebook into pools\n",
    "The goal is to assign each variant with a pool number, and that number will be used when assigning barcodes to guide-donors and grouping the oligos into pools in the final CRISPEY3 library construction.\n",
    "\n",
    "There are multiple sets going into the final library, and each set will have slightly different grouping and randomization requirements.\n",
    "\n",
    "Broadly speaking, here are some general principles:\n",
    "- Keep oligos targeting the same variant together. This way, we control the pool covariate and can better account for guide-specific differences per variant.\n",
    "- Within each variant set, avoid clustering variants in the same genomic region together. This avoids cloning issues (PCR mispriming, template swaps in Gibson assembly, etc.), and also controls for pool-specific batch effects when comparing fitness effects across variants.\n",
    "    - This negates a gene-based grouping approach, which would make for convenient organization and effective utilization of pools to do gene-specific small growth competition experiments.\n",
    "    - Sets such as the ergosterol pathway set and GxG sets must utilize whole-set variant randomization to avoid pool effects being associated with each gene.\n",
    "- An alternative approach would be to pool variants by gene. The GxE set is already somewhat grouped by gene because although the variants are currently grouped by condition, genes are often condition-specific in their testing and so variants in the same gene will inevitably end up in the same few pools. Some randomization here might still be possible, but not if there are too few pools available, especially for oligos used in multiple conditions\n",
    "- When assigning variants to each pool, take into account the number of oligos generated for each variant so that each pool contains the appropriate number of oligos (118, plus 3 additional technical oligos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of available cores: 20\n"
     ]
    }
   ],
   "source": [
    "import os, random, vcf\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "print(\"Number of available cores:\", len(os.sched_getaffinity(0)))\n",
    "\n",
    "# set working directory\n",
    "working_dir = \"/home/users/rang/crispey3/library_design/Input/\"\n",
    "os.chdir(working_dir)\n",
    "\n",
    "# setting the random seed\n",
    "random.seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_variants_to_pools(vars_df, assignment_order=[], grouping_col = None, num_of_pools_per_group={}, \n",
    "                           spillover_assignment={}, max_pool_size=1, first_pool=1, ):\n",
    "    \n",
    "    # for storing variant ID of spillover variants\n",
    "    spillover_vars = {g : [] for g in assignment_order}\n",
    "#     # add discard list\n",
    "#     spillover_vars['discard'] = []\n",
    "    \n",
    "    # iterate over each group and assign to pools\n",
    "    for group in assignment_order:\n",
    "        # get subset of variants belonging to group\n",
    "        # if column to group by provided, select variant subset by column. Otherwise, filter by variant ID\n",
    "        if grouping_col:\n",
    "            subset = vars_df.loc[vars_df[grouping_col] == group]\n",
    "        else:\n",
    "            subset = vars_df.filter(like=group, axis=0)\n",
    "        # sort variants in descending order of number of oligos, then randomize variants within each group of a certain number of oligos\n",
    "        # this enables even, efficient packing with randomization across pools\n",
    "        subset = subset.groupby('num_of_oligos').apply(lambda x: x.sample(frac=1, random_state=1)).reset_index(level=0, drop=True)[::-1]\n",
    "        \n",
    "        \n",
    "        # get number of pools to use for group\n",
    "        num_of_pools = num_of_pools_per_group[group]\n",
    "        # assign oligos to spillover if no pools set for group\n",
    "        if num_of_pools == 0:\n",
    "#             print(\"Excess oligos for {}: {}\".format(group, subset.num_of_oligos.sum())) # debugging\n",
    "            spillover_vars[spillover_assignment[group]] += subset.index.tolist()\n",
    "            continue\n",
    "\n",
    "\n",
    "        # define current pool sizes\n",
    "        cur_pool_sizes = [0] * num_of_pools\n",
    "        # add spillover variants to first pool(s) in group\n",
    "        # note: packing is done by max-ing out first pool, then moving to next pool, and doing so until all spillover variants are assigned\n",
    "        if len(spillover_vars[group]) > 0:\n",
    "            spillover_pool_num = first_pool\n",
    "            for v, row in vars_df.loc[spillover_vars[group]].iterrows():\n",
    "                # move to next spillover pool number if pool has reached capacity\n",
    "                if cur_pool_sizes[spillover_pool_num-first_pool] + row['num_of_oligos'] > max_pool_size:\n",
    "                    spillover_pool_num += 1\n",
    "                    \n",
    "                # assign each spillover variant to pool number\n",
    "                vars_df.loc[v, 'pool_num'] = spillover_pool_num\n",
    "                # increase pool size by variant's number of oligos\n",
    "                cur_pool_sizes[spillover_pool_num-first_pool] += row['num_of_oligos']\n",
    "\n",
    "                \n",
    "        # add all other variants to pools in group\n",
    "        # variants are assigned to spread oligos out, ensuring similar group sizes across pools\n",
    "        excess = 0\n",
    "        for v, row in subset.iterrows():\n",
    "            # find smallest current pool\n",
    "            i = np.argmin(cur_pool_sizes)\n",
    "            # assign variant to pool if number of oligos in pool is within max pool size\n",
    "            if cur_pool_sizes[i] + row['num_of_oligos'] <= max_pool_size:\n",
    "                vars_df.loc[v, 'pool_num'] = first_pool+i\n",
    "                cur_pool_sizes[i] += row['num_of_oligos']\n",
    "            # if current pools have insufficient space, send to group assigned for spillover\n",
    "            else:\n",
    "                excess +=row['num_of_oligos']\n",
    "                spillover_vars[spillover_assignment[group]].append(v)\n",
    "        \n",
    "#         # (debugging) print spillover variants\n",
    "#         if group in spillover_assignment.keys():\n",
    "#             print(\"Excess oligos for {}: {}\".format(group, excess))\n",
    "        \n",
    "        # once assigned, advance first_pool for next group\n",
    "        first_pool += num_of_pools\n",
    "    \n",
    "#     # remove discarded variants\n",
    "#     vars_df = vars_df.loc[~vars_df.index.isin(spillover_vars['discard'])]\n",
    "#     # identify number of variants discarded\n",
    "#     print(\"Variants discarded:\", spillover_vars['discard'])\n",
    "\n",
    "    vars_df['pool_num'] = vars_df['pool_num'].astype(int)\n",
    "\n",
    "    return vars_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ergosterol set\n",
    "\n",
    "### We should be able to fit all ergosterol variants in 34 pools\n",
    "- ERG: 21 pools, 117-118 oligos each\n",
    "- EGE: 4 pools, 117-118 oligos each\n",
    "- EGD: 1 pool, 118 oligos each\n",
    "- EGC: 2 pools, 110-111 oligos each\n",
    "- EGB: 1 pool, 117 oligos each\n",
    "- EGA: 1 pool, 117 oligos each\n",
    "- EG9: 1 pool, 118 oligos each\n",
    "- EG8: 3 pools, 113-114 oligos each\n",
    "\n",
    "See code for spillover assignment\n",
    "\n",
    "### Variant ID prefix and which genomes they can be edited in:\n",
    "- ERG: BY | RM | YJM | YPS\n",
    "- EGE: BY | RM | YJM | ---\n",
    "- EGD: BY | RM | --- | YPS\n",
    "- EGC: BY | RM | --- | ---\n",
    "- EGB: BY | --- | YJM | YPS\n",
    "- EGA: BY | --- | YJM | ---\n",
    "- EG9: BY | --- | --- | YPS\n",
    "- EG8: BY | --- | --- | ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>3</td>\n",
       "      <td>2382|1631|1338|448|2910</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>448.0</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>17</td>\n",
       "      <td>2396|1645|1324|434|2896</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>434.0</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00008</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TG</td>\n",
       "      <td>661</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.327000</td>\n",
       "      <td>72</td>\n",
       "      <td>2451|1700|1268|378|2840</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>72.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00009</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>72</td>\n",
       "      <td>2451|1700|1268|378|2840</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>72.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00010</th>\n",
       "      <td>V</td>\n",
       "      <td>237196</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>69</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.034000</td>\n",
       "      <td>77</td>\n",
       "      <td>2456|1705|1264|374|2836</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>77.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00039</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499649</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>3868|16</td>\n",
       "      <td>4690|1026|357|1930|4403</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>357.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00040</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499650</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>3867|15</td>\n",
       "      <td>4689|1025|358|1931|4404</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>358.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00041</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00042</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>35</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.017000</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00044</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499659</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>3858|6</td>\n",
       "      <td>4680|1016|367|1940|4413</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>367.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1580 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                       \n",
       "ERG_00003     V  237122   G   A    9  2022  0.004451                     3   \n",
       "ERG_00004     V  237136   T   C   26  2022  0.013000                    17   \n",
       "EGE_00008     V  237191   T  TG  661  2022  0.327000                    72   \n",
       "EGE_00009     V  237191   T  TA    6  2022  0.002967                    72   \n",
       "EGE_00010     V  237196   A   T   69  2022  0.034000                    77   \n",
       "...         ...     ...  ..  ..  ...   ...       ...                   ...   \n",
       "EGA_00039   XVI  499649   G   A    4  2022  0.001978               3868|16   \n",
       "EGA_00040   XVI  499650   G   C    5  2022  0.002473               3867|15   \n",
       "EGA_00041   XVI  499655   G   A    4  2022  0.001978               3862|10   \n",
       "EGA_00042   XVI  499655   G   T   35  2022  0.017000               3862|10   \n",
       "EGA_00044   XVI  499659   G   A    1  2022  0.000495                3858|6   \n",
       "\n",
       "           downstream_distance_str                   region  ...  \\\n",
       "var_id                                                       ...   \n",
       "ERG_00003  2382|1631|1338|448|2910  unidirectional_promoter  ...   \n",
       "ERG_00004  2396|1645|1324|434|2896  unidirectional_promoter  ...   \n",
       "EGE_00008  2451|1700|1268|378|2840  unidirectional_promoter  ...   \n",
       "EGE_00009  2451|1700|1268|378|2840  unidirectional_promoter  ...   \n",
       "EGE_00010  2456|1705|1264|374|2836  unidirectional_promoter  ...   \n",
       "...                            ...                      ...  ...   \n",
       "EGA_00039  4690|1026|357|1930|4403  unidirectional_promoter  ...   \n",
       "EGA_00040  4689|1025|358|1931|4404  unidirectional_promoter  ...   \n",
       "EGA_00041  4684|1020|363|1936|4409  unidirectional_promoter  ...   \n",
       "EGA_00042  4684|1020|363|1936|4409  unidirectional_promoter  ...   \n",
       "EGA_00044  4680|1016|367|1940|4413  unidirectional_promoter  ...   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00003                    SAH1               YER043C   \n",
       "ERG_00004                    SAH1               YER043C   \n",
       "EGE_00008                    SAH1               YER043C   \n",
       "EGE_00009                    SAH1               YER043C   \n",
       "EGE_00010                    SAH1               YER043C   \n",
       "...                           ...                   ...   \n",
       "EGA_00039                   ERG10               YPL028W   \n",
       "EGA_00040                   ERG10               YPL028W   \n",
       "EGA_00041                   ERG10               YPL028W   \n",
       "EGA_00042                   ERG10               YPL028W   \n",
       "EGA_00044                   ERG10               YPL028W   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "ERG_00003    upstream_gene_variant                    3.0   \n",
       "ERG_00004    upstream_gene_variant                   17.0   \n",
       "EGE_00008    upstream_gene_variant                   72.0   \n",
       "EGE_00009    upstream_gene_variant                   72.0   \n",
       "EGE_00010    upstream_gene_variant                   77.0   \n",
       "...                            ...                    ...   \n",
       "EGA_00039  downstream_gene_variant                  357.0   \n",
       "EGA_00040  downstream_gene_variant                  358.0   \n",
       "EGA_00041  downstream_gene_variant                  363.0   \n",
       "EGA_00042  downstream_gene_variant                  363.0   \n",
       "EGA_00044  downstream_gene_variant                  367.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00003                   ERG28               YER044C   \n",
       "ERG_00004                   ERG28               YER044C   \n",
       "EGE_00008                   ERG28               YER044C   \n",
       "EGE_00009                   ERG28               YER044C   \n",
       "EGE_00010                   ERG28               YER044C   \n",
       "...                           ...                   ...   \n",
       "EGA_00039                    SMA1               YPL027W   \n",
       "EGA_00040                    SMA1               YPL027W   \n",
       "EGA_00041                    SMA1               YPL027W   \n",
       "EGA_00042                    SMA1               YPL027W   \n",
       "EGA_00044                    SMA1               YPL027W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \\\n",
       "var_id                                                                    \n",
       "ERG_00003  downstream_gene_variant                  448.0             2   \n",
       "ERG_00004  downstream_gene_variant                  434.0             2   \n",
       "EGE_00008  downstream_gene_variant                  378.0             2   \n",
       "EGE_00009  downstream_gene_variant                  378.0             2   \n",
       "EGE_00010  downstream_gene_variant                  374.0             2   \n",
       "...                            ...                    ...           ...   \n",
       "EGA_00039    upstream_gene_variant                   16.0             2   \n",
       "EGA_00040    upstream_gene_variant                   15.0             2   \n",
       "EGA_00041    upstream_gene_variant                   10.0             2   \n",
       "EGA_00042    upstream_gene_variant                   10.0             2   \n",
       "EGA_00044    upstream_gene_variant                    6.0             2   \n",
       "\n",
       "          pool_num  \n",
       "var_id              \n",
       "ERG_00003       13  \n",
       "ERG_00004       21  \n",
       "EGE_00008       25  \n",
       "EGE_00009       22  \n",
       "EGE_00010       22  \n",
       "...            ...  \n",
       "EGA_00039       30  \n",
       "EGA_00040       30  \n",
       "EGA_00041       30  \n",
       "EGA_00042       32  \n",
       "EGA_00044       30  \n",
       "\n",
       "[1580 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "1     118\n",
       "2     118\n",
       "3     118\n",
       "4     118\n",
       "5     118\n",
       "6     118\n",
       "7     118\n",
       "8     118\n",
       "9     117\n",
       "10    117\n",
       "11    117\n",
       "12    117\n",
       "13    117\n",
       "14    117\n",
       "15    117\n",
       "16    117\n",
       "17    117\n",
       "18    117\n",
       "19    117\n",
       "20    117\n",
       "21    117\n",
       "22    118\n",
       "23    118\n",
       "24    118\n",
       "25    117\n",
       "26    118\n",
       "27    110\n",
       "28    111\n",
       "29    117\n",
       "30    117\n",
       "31    118\n",
       "32    114\n",
       "33    114\n",
       "34    113\n",
       "Name: num_of_oligos, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 3968\n",
      "Total number of variants: 1580\n"
     ]
    }
   ],
   "source": [
    "ergosterol_vars_file = \"/home/users/rang/scratch/yeast/ergosterol/ergosterol_variants_final.txt\"\n",
    "ergosterol_vars = pd.read_csv(ergosterol_vars_file, sep='\\t', index_col=0)\n",
    "ergosterol_vars['num_of_oligos'] = ergosterol_vars['num_of_oligos'].astype(int)\n",
    "\n",
    "# assign ergosterol variants to pools\n",
    "ergosterol_vars = sort_variants_to_pools(vars_df = ergosterol_vars, \n",
    "                                         assignment_order = ['ERG','EGE','EGD','EGC','EGB','EGA','EG9','EG8'],\n",
    "                                         num_of_pools_per_group = {'ERG':21, 'EGE':4, 'EGD':1, 'EGC':2, 'EGB':1, 'EGA':1, 'EG9':1, 'EG8':3},\n",
    "                                         spillover_assignment = {'ERG':'EGE', 'EGE':'EGC', 'EGD':'EGC', 'EGC':'EG8', 'EGB':'EGA', 'EGA':'EG8', 'EG9':'EG8'},\n",
    "                                         max_pool_size = 118, first_pool = 1)\n",
    "\n",
    "display(ergosterol_vars)\n",
    "# check number of oligos in each pool\n",
    "display(ergosterol_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", ergosterol_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(ergosterol_vars))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GXG set\n",
    "\n",
    "### We should be able to fit all GxG variants in 43 pools\n",
    "- GXG: 43 pools, 114-116 oligos each\n",
    "\n",
    "### Variant ID prefix and which genomes they can be edited in:\n",
    "- GXG: BY | RM | YJM | YPS\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>assoc_gene</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00321</th>\n",
       "      <td>I</td>\n",
       "      <td>101530</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>385|1833</td>\n",
       "      <td>1662|4742|35</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>YAL025C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>385.0</td>\n",
       "      <td>LTE1</td>\n",
       "      <td>YAL024C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>35.0</td>\n",
       "      <td>3</td>\n",
       "      <td>['YAL024C', 'YAL025C']</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00329</th>\n",
       "      <td>I</td>\n",
       "      <td>101816</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001980</td>\n",
       "      <td>671|2119</td>\n",
       "      <td>1948|4456</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>['YAL024C']</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00332</th>\n",
       "      <td>I</td>\n",
       "      <td>101901</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>90</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.045000</td>\n",
       "      <td>756|2204</td>\n",
       "      <td>2033|4371</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>['YAL024C']</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00339</th>\n",
       "      <td>I</td>\n",
       "      <td>102261</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>78</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.039000</td>\n",
       "      <td>1116|2564</td>\n",
       "      <td>2393|4011</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>['YAL024C']</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00346</th>\n",
       "      <td>I</td>\n",
       "      <td>102699</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>310</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.154000</td>\n",
       "      <td>1554|3002</td>\n",
       "      <td>2831|3573</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>['YAL024C']</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50437</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899389</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>3625|805|1366|2655</td>\n",
       "      <td>2385</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>['YPR181C']</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50440</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899419</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>196</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.097000</td>\n",
       "      <td>3655|775|1336|2625</td>\n",
       "      <td>2415</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>['YPR181C']</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50446</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001487</td>\n",
       "      <td>3757|673|1234|2523</td>\n",
       "      <td>2517</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>['YPR181C']</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50447</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899679</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>68</td>\n",
       "      <td>2010</td>\n",
       "      <td>0.034000</td>\n",
       "      <td>3915|12|515|1076|2365</td>\n",
       "      <td>2675</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>YPR181C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>12.0</td>\n",
       "      <td>SMX3</td>\n",
       "      <td>YPR182W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>515.0</td>\n",
       "      <td>4</td>\n",
       "      <td>['YPR181C']</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50454</th>\n",
       "      <td>XVI</td>\n",
       "      <td>900065</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001487</td>\n",
       "      <td>4301|398|129|690|1979</td>\n",
       "      <td>3061</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>YPR181C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>398.0</td>\n",
       "      <td>SMX3</td>\n",
       "      <td>YPR182W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>129.0</td>\n",
       "      <td>3</td>\n",
       "      <td>['YPR182W', 'YPR181C']</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF  upstream_distance_str  \\\n",
       "var_id                                                                        \n",
       "GXG_00321     I  101530   G   C    4  2022  0.001978               385|1833   \n",
       "GXG_00329     I  101816   C   T    4  2020  0.001980               671|2119   \n",
       "GXG_00332     I  101901   C   A   90  2022  0.045000               756|2204   \n",
       "GXG_00339     I  102261   C   T   78  2020  0.039000              1116|2564   \n",
       "GXG_00346     I  102699   G   A  310  2016  0.154000              1554|3002   \n",
       "...         ...     ...  ..  ..  ...   ...       ...                    ...   \n",
       "GXG_50437   XVI  899389   G   A    6  2018  0.002973     3625|805|1366|2655   \n",
       "GXG_50440   XVI  899419   A   G  196  2014  0.097000     3655|775|1336|2625   \n",
       "GXG_50446   XVI  899521   G   C    3  2018  0.001487     3757|673|1234|2523   \n",
       "GXG_50447   XVI  899679   A   G   68  2010  0.034000  3915|12|515|1076|2365   \n",
       "GXG_50454   XVI  900065   C   G    3  2018  0.001487  4301|398|129|690|1979   \n",
       "\n",
       "          downstream_distance_str                   region  ...  \\\n",
       "var_id                                                      ...   \n",
       "GXG_00321            1662|4742|35  unidirectional_promoter  ...   \n",
       "GXG_00329               1948|4456         missense_variant  ...   \n",
       "GXG_00332               2033|4371       synonymous_variant  ...   \n",
       "GXG_00339               2393|4011       synonymous_variant  ...   \n",
       "GXG_00346               2831|3573       synonymous_variant  ...   \n",
       "...                           ...                      ...  ...   \n",
       "GXG_50437                    2385       synonymous_variant  ...   \n",
       "GXG_50440                    2415       synonymous_variant  ...   \n",
       "GXG_50446                    2517       synonymous_variant  ...   \n",
       "GXG_50447                    2675   bidirectional_promoter  ...   \n",
       "GXG_50454                    3061   bidirectional_promoter  ...   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00321               YAL025C    upstream_gene_variant   \n",
       "GXG_00329                   NaN                      NaN   \n",
       "GXG_00332                   NaN                      NaN   \n",
       "GXG_00339                   NaN                      NaN   \n",
       "GXG_00346                   NaN                      NaN   \n",
       "...                         ...                      ...   \n",
       "GXG_50437                   NaN                      NaN   \n",
       "GXG_50440                   NaN                      NaN   \n",
       "GXG_50446                   NaN                      NaN   \n",
       "GXG_50447               YPR181C    upstream_gene_variant   \n",
       "GXG_50454               YPR181C    upstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXG_00321                  385.0                    LTE1   \n",
       "GXG_00329                    NaN                     NaN   \n",
       "GXG_00332                    NaN                     NaN   \n",
       "GXG_00339                    NaN                     NaN   \n",
       "GXG_00346                    NaN                     NaN   \n",
       "...                          ...                     ...   \n",
       "GXG_50437                    NaN                     NaN   \n",
       "GXG_50440                    NaN                     NaN   \n",
       "GXG_50446                    NaN                     NaN   \n",
       "GXG_50447                   12.0                    SMX3   \n",
       "GXG_50454                  398.0                    SMX3   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00321               YAL024C  downstream_gene_variant   \n",
       "GXG_00329                   NaN                      NaN   \n",
       "GXG_00332                   NaN                      NaN   \n",
       "GXG_00339                   NaN                      NaN   \n",
       "GXG_00346                   NaN                      NaN   \n",
       "...                         ...                      ...   \n",
       "GXG_50437                   NaN                      NaN   \n",
       "GXG_50440                   NaN                      NaN   \n",
       "GXG_50446                   NaN                      NaN   \n",
       "GXG_50447               YPR182W    upstream_gene_variant   \n",
       "GXG_50454               YPR182W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance num_of_oligos              assoc_gene  \\\n",
       "var_id                                                                   \n",
       "GXG_00321                   35.0             3  ['YAL024C', 'YAL025C']   \n",
       "GXG_00329                    NaN             4             ['YAL024C']   \n",
       "GXG_00332                    NaN             2             ['YAL024C']   \n",
       "GXG_00339                    NaN             2             ['YAL024C']   \n",
       "GXG_00346                    NaN             3             ['YAL024C']   \n",
       "...                          ...           ...                     ...   \n",
       "GXG_50437                    NaN             2             ['YPR181C']   \n",
       "GXG_50440                    NaN             3             ['YPR181C']   \n",
       "GXG_50446                    NaN             2             ['YPR181C']   \n",
       "GXG_50447                  515.0             4             ['YPR181C']   \n",
       "GXG_50454                  129.0             3  ['YPR182W', 'YPR181C']   \n",
       "\n",
       "          pool_num  \n",
       "var_id              \n",
       "GXG_00321       68  \n",
       "GXG_00329       37  \n",
       "GXG_00332       50  \n",
       "GXG_00339       76  \n",
       "GXG_00346       41  \n",
       "...            ...  \n",
       "GXG_50437       72  \n",
       "GXG_50440       66  \n",
       "GXG_50446       43  \n",
       "GXG_50447       47  \n",
       "GXG_50454       60  \n",
       "\n",
       "[2000 rows x 44 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "35    115\n",
       "36    115\n",
       "37    115\n",
       "38    115\n",
       "39    115\n",
       "40    115\n",
       "41    115\n",
       "42    115\n",
       "43    115\n",
       "44    115\n",
       "45    115\n",
       "46    116\n",
       "47    116\n",
       "48    116\n",
       "49    116\n",
       "50    116\n",
       "51    116\n",
       "52    116\n",
       "53    116\n",
       "54    116\n",
       "55    116\n",
       "56    116\n",
       "57    116\n",
       "58    116\n",
       "59    116\n",
       "60    116\n",
       "61    116\n",
       "62    116\n",
       "63    116\n",
       "64    116\n",
       "65    116\n",
       "66    116\n",
       "67    116\n",
       "68    114\n",
       "69    114\n",
       "70    114\n",
       "71    114\n",
       "72    114\n",
       "73    114\n",
       "74    114\n",
       "75    114\n",
       "76    114\n",
       "77    114\n",
       "Name: num_of_oligos, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 4957\n",
      "Total number of variants: 2000\n"
     ]
    }
   ],
   "source": [
    "gxg_vars_file = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/gxg_variants_final.txt\"\n",
    "gxg_vars = pd.read_csv(gxg_vars_file, sep='\\t', index_col=0)\n",
    "gxg_vars['num_of_oligos'] = gxg_vars['num_of_oligos'].astype(int)\n",
    "\n",
    "# assign gxg variants to pools\n",
    "gxg_vars = sort_variants_to_pools(vars_df = gxg_vars, \n",
    "                                  assignment_order = ['GXG'],\n",
    "                                  num_of_pools_per_group = {'GXG':43},\n",
    "                                  spillover_assignment = {},\n",
    "                                  max_pool_size = 118, first_pool = 35)\n",
    "\n",
    "display(gxg_vars)\n",
    "# check number of oligos in each pool\n",
    "display(gxg_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", gxg_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(gxg_vars))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Epistasis validation set\n",
    "\n",
    "### We should be able to fit known effects variants in 6 pools\n",
    "- TDH, VAL: 6 pools, 115-116 oligos each\n",
    "\n",
    "### Variant ID prefix and which genomes they can be edited in:\n",
    "- TDH: BY | --- | --- | ---\n",
    "- VAL: BY | RM | YJM | YPS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TDH_00001</th>\n",
       "      <td>VII</td>\n",
       "      <td>884068</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TDH_00002</th>\n",
       "      <td>VII</td>\n",
       "      <td>884073</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TDH_00003</th>\n",
       "      <td>VII</td>\n",
       "      <td>884105</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TDH_00004</th>\n",
       "      <td>VII</td>\n",
       "      <td>884107</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TDH_00005</th>\n",
       "      <td>VII</td>\n",
       "      <td>884108</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00239</th>\n",
       "      <td>XVI</td>\n",
       "      <td>501421</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00240</th>\n",
       "      <td>XVI</td>\n",
       "      <td>563805</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00241</th>\n",
       "      <td>XVI</td>\n",
       "      <td>798797</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00242</th>\n",
       "      <td>XVI</td>\n",
       "      <td>820797</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00243</th>\n",
       "      <td>XVI</td>\n",
       "      <td>868939</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>5</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>257 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT  num_of_oligos  pool_num\n",
       "var_id                                                  \n",
       "TDH_00001   VII  884068   C   T              2        78\n",
       "TDH_00002   VII  884073   G   A              2        78\n",
       "TDH_00003   VII  884105   C   T              3        78\n",
       "TDH_00004   VII  884107   G   A              3        78\n",
       "TDH_00005   VII  884108   G   A              3        78\n",
       "...         ...     ...  ..  ..            ...       ...\n",
       "VAL_00239   XVI  501421   T   C              3        78\n",
       "VAL_00240   XVI  563805   G   A              2        80\n",
       "VAL_00241   XVI  798797   A   C              2        79\n",
       "VAL_00242   XVI  820797   C   T              4        81\n",
       "VAL_00243   XVI  868939   A   G              5        83\n",
       "\n",
       "[257 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "78    116\n",
       "79    115\n",
       "80    115\n",
       "81    115\n",
       "82    115\n",
       "83    116\n",
       "Name: num_of_oligos, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 692\n",
      "Total number of variants: 257\n"
     ]
    }
   ],
   "source": [
    "epival_vars_file = \"/home/users/rang/scratch/yeast/epistasis/epival_variants_final.txt\" # also VCF file epival_variants_final.vcf\n",
    "epival_vars = pd.read_csv(epival_vars_file, sep='\\t')\n",
    "\n",
    "# calculate number of oligos per variant, add info to column and adjust to reflect one variant per row\n",
    "epival_vars['num_of_oligos'] = epival_vars['var_id'].map(epival_vars.groupby('var_id').size())\n",
    "# convert epival_vars to simplified dataframe aligned with the other sets\n",
    "epival_vars = epival_vars[['var_id', 'chrom', 'SNP_chr_pos', 'REF', 'ALT', 'num_of_oligos']].rename(columns={'chrom':'CHROM', 'SNP_chr_pos':'POS'}).set_index('var_id').drop_duplicates()\n",
    "epival_vars['ALT'] = epival_vars['ALT'].apply(lambda x: x[1:-1])\n",
    "\n",
    "# assign epival variants to pools\n",
    "epival_vars = sort_variants_to_pools(vars_df = epival_vars, \n",
    "                                  assignment_order = ['TDH', 'VAL'],\n",
    "                                  num_of_pools_per_group = {'TDH':0, 'VAL':6},\n",
    "                                  spillover_assignment = {'TDH':'VAL'},\n",
    "                                  max_pool_size = 118, first_pool = 78)\n",
    "\n",
    "display(epival_vars)\n",
    "# check number of oligos in each pool\n",
    "display(epival_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", epival_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(epival_vars))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GXE set\n",
    "\n",
    "### We should be able to fit all GxE variants in 94 pools\n",
    "Oligos will be grouped according to condition (see assoc_traits_str)\n",
    "- GXE: 94 pools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>ACS</th>\n",
       "      <th>ADE</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADG</th>\n",
       "      <th>ADI</th>\n",
       "      <th>ADR</th>\n",
       "      <th>SACE_MAA</th>\n",
       "      <th>assoc_traits</th>\n",
       "      <th>assoc_traits_str</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00027</th>\n",
       "      <td>I</td>\n",
       "      <td>192429</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1506</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.754000</td>\n",
       "      <td>12|190</td>\n",
       "      <td>4697|3618|173</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00033</th>\n",
       "      <td>I</td>\n",
       "      <td>192466</td>\n",
       "      <td>AAAGGG</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001485</td>\n",
       "      <td>50|148</td>\n",
       "      <td>4735|3656|211</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17166</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375077</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001484</td>\n",
       "      <td>425|4414</td>\n",
       "      <td>4043|2214|2156</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17167</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375078</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>424|4415</td>\n",
       "      <td>4042|2213|2157</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17168</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375080</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>422|4417</td>\n",
       "      <td>4040|2211|2159</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17188</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375331</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>171|4668</td>\n",
       "      <td>3789|1960|162|2410</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17189</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375333</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>168|4670</td>\n",
       "      <td>3786|1957|164|2412</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4384 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT    AC    AN  \\\n",
       "var_id                                                                  \n",
       "GXE_00023     I  192419                C                A     6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT    39  2014   \n",
       "GXE_00026     I  192427                T                C   902  2000   \n",
       "GXE_00027     I  192429                G                C  1506  1998   \n",
       "GXE_00033     I  192466           AAAGGG                A     3  2020   \n",
       "...         ...     ...              ...              ...   ...   ...   \n",
       "GXE_17166   XVI  375077                T                C     3  2022   \n",
       "GXE_17167   XVI  375078                C                A     8  2020   \n",
       "GXE_17168   XVI  375080                G                A    31  2022   \n",
       "GXE_17188   XVI  375331                C                A     8  2020   \n",
       "GXE_17189   XVI  375333                A               AT     8  2022   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "GXE_00027  0.754000                12|190           4697|3618|173   \n",
       "GXE_00033  0.001485                50|148           4735|3656|211   \n",
       "...             ...                   ...                     ...   \n",
       "GXE_17166  0.001484              425|4414          4043|2214|2156   \n",
       "GXE_17167  0.003960              424|4415          4042|2213|2157   \n",
       "GXE_17168  0.015000              422|4417          4040|2211|2159   \n",
       "GXE_17188  0.003960              171|4668      3789|1960|162|2410   \n",
       "GXE_17189  0.003956              168|4670      3786|1957|164|2412   \n",
       "\n",
       "                            region  ...  ACS  ADE  ADF  ADG  ADI  ADR  \\\n",
       "var_id                              ...                                 \n",
       "GXE_00023   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_00024   bidirectional_promoter  ...  0/0  NaN  0/0  0/0  0/0  0/0   \n",
       "GXE_00026   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  1/1  1/1   \n",
       "GXE_00027   bidirectional_promoter  ...  1/1  NaN  1/1  1/1  1/1  1/1   \n",
       "GXE_00033   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "...                            ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "GXE_17166         missense_variant  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17167              stop_gained  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17168         missense_variant  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17188  unidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17189  unidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "\n",
       "          SACE_MAA                                       assoc_traits  \\\n",
       "var_id                                                                  \n",
       "GXE_00023      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "GXE_00024      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "GXE_00026      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "GXE_00027      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "GXE_00033      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "...            ...                                                ...   \n",
       "GXE_17166      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17167      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17168      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17188      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17189      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "\n",
       "                                            assoc_traits_str pool_num  \n",
       "var_id                                                                 \n",
       "GXE_00023  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       85  \n",
       "GXE_00024  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       84  \n",
       "GXE_00026  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       85  \n",
       "GXE_00027  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       86  \n",
       "GXE_00033  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       86  \n",
       "...                                                      ...      ...  \n",
       "GXE_17166  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...      109  \n",
       "GXE_17167  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...      109  \n",
       "GXE_17168  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17188  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17189  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "\n",
       "[4384 rows x 61 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "84     118\n",
       "85     118\n",
       "86     117\n",
       "87     118\n",
       "88     118\n",
       "      ... \n",
       "173    117\n",
       "174    117\n",
       "175    117\n",
       "176    118\n",
       "177    118\n",
       "Name: num_of_oligos, Length: 94, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 10983\n",
      "Total number of variants: 4384\n"
     ]
    }
   ],
   "source": [
    "gxe_vars_file = \"/home/users/rang/scratch/yeast/GxE/gxe_variants_final.txt\"\n",
    "gxe_vars = pd.read_csv(gxe_vars_file, sep='\\t', index_col=0)\n",
    "gxe_vars['num_of_oligos'] = gxe_vars['num_of_oligos'].astype(int)\n",
    "\n",
    "# assign gxe variants to pools\n",
    "gxe_vars = sort_variants_to_pools(vars_df = gxe_vars, \n",
    "                                  assignment_order = ['Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1',\n",
    "                                                      'Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2',\n",
    "                                                      'Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2',\n",
    "                                                      'Caffeine;15mM;2_Cobalt_Chloride;2mM;2',\n",
    "                                                      'Caffeine;15mM;2_Neomycin;5mg/mL;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2_Fructose;;1',\n",
    "                                                      'Fluconazole;100uM;2_Fructose;;1',\n",
    "                                                      'Caffeine;15mM;2_Fructose;;1',\n",
    "                                                      'Lithium_Chloride;100mM;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2',\n",
    "                                                      'Fluconazole;100uM;2',\n",
    "                                                      'Neomycin;5mg/mL;2',\n",
    "                                                      'Fructose;;1',\n",
    "                                                      'Caffeine;15mM;2'],\n",
    "                                  \n",
    "                                  grouping_col = 'assoc_traits_str',\n",
    "                                  \n",
    "                                  num_of_pools_per_group = {'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2':3,\n",
    "                                                            'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1':0,\n",
    "                                                            'Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2':3,\n",
    "                                                            'Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2':1,\n",
    "                                                            'Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2':1,\n",
    "                                                            'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2':1,\n",
    "                                                            'Caffeine;15mM;2_Cobalt_Chloride;2mM;2':1,\n",
    "                                                            'Caffeine;15mM;2_Neomycin;5mg/mL;2':0,\n",
    "                                                            'Cobalt_Chloride;2mM;2_Fructose;;1':1,\n",
    "                                                            'Fluconazole;100uM;2_Fructose;;1':0,\n",
    "                                                            'Caffeine;15mM;2_Fructose;;1':0,\n",
    "                                                            'Lithium_Chloride;100mM;2':14,\n",
    "                                                            'Cobalt_Chloride;2mM;2':14,\n",
    "                                                            'Fluconazole;100uM;2':13,\n",
    "                                                            'Neomycin;5mg/mL;2':10,\n",
    "                                                            'Fructose;;1':13,\n",
    "                                                            'Caffeine;15mM;2':19},\n",
    "                                                            \n",
    "                                  spillover_assignment = {'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2':'Neomycin;5mg/mL;2',\n",
    "                                                          'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1':'Cobalt_Chloride;2mM;2',\n",
    "                                                          'Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2':'Neomycin;5mg/mL;2',\n",
    "                                                          'Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2':'Cobalt_Chloride;2mM;2',\n",
    "                                                          'Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2':'Lithium_Chloride;100mM;2',\n",
    "                                                          'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2':'Fluconazole;100uM;2',\n",
    "                                                          'Caffeine;15mM;2_Neomycin;5mg/mL;2':'Neomycin;5mg/mL;2',\n",
    "                                                          'Cobalt_Chloride;2mM;2_Fructose;;1':'Cobalt_Chloride;2mM;2',\n",
    "                                                          'Fluconazole;100uM;2_Fructose;;1':'Fluconazole;100uM;2',\n",
    "                                                          'Caffeine;15mM;2_Fructose;;1':'Caffeine;15mM;2',\n",
    "                                                          'Lithium_Chloride;100mM;2':'Caffeine;15mM;2',\n",
    "                                                          'Fructose;;1':'Caffeine;15mM;2'},\n",
    "                                  max_pool_size = 118, first_pool = 84)\n",
    "\n",
    "display(gxe_vars)\n",
    "# check number of oligos in each pool\n",
    "display(gxe_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", gxe_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(gxe_vars))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>ACS</th>\n",
       "      <th>ADE</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADG</th>\n",
       "      <th>ADI</th>\n",
       "      <th>ADR</th>\n",
       "      <th>SACE_MAA</th>\n",
       "      <th>assoc_traits</th>\n",
       "      <th>assoc_traits_str</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_17015</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373309</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>2193|484|2646|3868</td>\n",
       "      <td>3982|388</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17016</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>2189|480|2650|3872</td>\n",
       "      <td>3978|392</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17017</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373314</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>2188|479|2651|3873</td>\n",
       "      <td>3977|393</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17018</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373318</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>18</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.008902</td>\n",
       "      <td>2184|475|2655|3877</td>\n",
       "      <td>3973|397</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17033</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373506</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>10</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004946</td>\n",
       "      <td>1996|287|2843|4065</td>\n",
       "      <td>3785|585</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17034</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373512</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>35</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.017000</td>\n",
       "      <td>1990|281|2849|4071</td>\n",
       "      <td>3779|591</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17042</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373604</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>1898|189|2941|4163</td>\n",
       "      <td>3687|683</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17053</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373707</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>98</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.049000</td>\n",
       "      <td>1795|86|3044|4266</td>\n",
       "      <td>3584|786</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17055</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373710</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>2010</td>\n",
       "      <td>0.004975</td>\n",
       "      <td>1792|83|3047|4269</td>\n",
       "      <td>3581|789</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17058</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373805</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>1697|3142|4364</td>\n",
       "      <td>3486|884</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17065</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373838</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>20</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.009891</td>\n",
       "      <td>1664|3175|4397</td>\n",
       "      <td>3453|917</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17066</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373839</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>15</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.007418</td>\n",
       "      <td>1663|3176|4398</td>\n",
       "      <td>3452|918</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17071</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373899</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>1603|3236|4458</td>\n",
       "      <td>3392|978</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17072</th>\n",
       "      <td>XVI</td>\n",
       "      <td>373909</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>22</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.011000</td>\n",
       "      <td>1593|3246|4468</td>\n",
       "      <td>3382|988</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17081</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374063</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>1439|3400|4622</td>\n",
       "      <td>3228|1142</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17082</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374065</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>30</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>1437|3402|4624</td>\n",
       "      <td>3226|1144</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17084</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374089</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.007913</td>\n",
       "      <td>1413|3426|4648</td>\n",
       "      <td>3202|1168</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17086</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374149</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>108</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.053000</td>\n",
       "      <td>1353|3486|4708</td>\n",
       "      <td>4971|3142|1228</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17087</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374156</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>876</td>\n",
       "      <td>2006</td>\n",
       "      <td>0.437000</td>\n",
       "      <td>1346|3493|4715</td>\n",
       "      <td>4964|3135|1235</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17089</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374183</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>1319|3520|4742</td>\n",
       "      <td>4937|3108|1262</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17090</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374220</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>1282|3557|4779</td>\n",
       "      <td>4900|3071|1299</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17091</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374224</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.001488</td>\n",
       "      <td>1278|3561|4783</td>\n",
       "      <td>4896|3067|1303</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17097</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374261</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1649</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.817000</td>\n",
       "      <td>1241|3598|4820</td>\n",
       "      <td>4859|3030|1340</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17098</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374268</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001484</td>\n",
       "      <td>1234|3605|4827</td>\n",
       "      <td>4852|3023|1347</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17099</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374270</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>1232|3607|4829</td>\n",
       "      <td>4850|3021|1349</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17101</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374271</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>14</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.006924</td>\n",
       "      <td>1231|3608|4830</td>\n",
       "      <td>4849|3020|1350</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17103</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374272</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001484</td>\n",
       "      <td>1230|3609|4831</td>\n",
       "      <td>4848|3019|1351</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17118</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374535</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>967|3872</td>\n",
       "      <td>4585|2756|1614</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17120</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374541</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>961|3878</td>\n",
       "      <td>4579|2750|1620</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17125</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374570</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>932|3907</td>\n",
       "      <td>4550|2721|1649</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17127</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374633</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>869|3970</td>\n",
       "      <td>4487|2658|1712</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17132</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374711</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>29</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>791|4048</td>\n",
       "      <td>4409|2580|1790</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17144</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374855</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>67</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.033000</td>\n",
       "      <td>647|4192</td>\n",
       "      <td>4265|2436|1934</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17145</th>\n",
       "      <td>XVI</td>\n",
       "      <td>374857</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>136</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.067000</td>\n",
       "      <td>645|4194</td>\n",
       "      <td>4263|2434|1936</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17168</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375080</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>422|4417</td>\n",
       "      <td>4040|2211|2159</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17188</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375331</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>171|4668</td>\n",
       "      <td>3789|1960|162|2410</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17189</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375333</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>168|4670</td>\n",
       "      <td>3786|1957|164|2412</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT    AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                        \n",
       "GXE_17015   XVI  373309   C   T     5  2022  0.002473    2193|484|2646|3868   \n",
       "GXE_17016   XVI  373313   C   T     6  2022  0.002967    2189|480|2650|3872   \n",
       "GXE_17017   XVI  373314   G   C     6  2018  0.002973    2188|479|2651|3873   \n",
       "GXE_17018   XVI  373318   T   A    18  2022  0.008902    2184|475|2655|3877   \n",
       "GXE_17033   XVI  373506   T   C    10  2022  0.004946    1996|287|2843|4065   \n",
       "GXE_17034   XVI  373512   A   G    35  2020  0.017000    1990|281|2849|4071   \n",
       "GXE_17042   XVI  373604   C   G     6  2022  0.002967    1898|189|2941|4163   \n",
       "GXE_17053   XVI  373707   C   T    98  2020  0.049000     1795|86|3044|4266   \n",
       "GXE_17055   XVI  373710   G   A    10  2010  0.004975     1792|83|3047|4269   \n",
       "GXE_17058   XVI  373805   T   C     5  2022  0.002473        1697|3142|4364   \n",
       "GXE_17065   XVI  373838   T   A    20  2022  0.009891        1664|3175|4397   \n",
       "GXE_17066   XVI  373839   T   C    15  2022  0.007418        1663|3176|4398   \n",
       "GXE_17071   XVI  373899   A   T     6  2022  0.002967        1603|3236|4458   \n",
       "GXE_17072   XVI  373909   G   A    22  2022  0.011000        1593|3246|4468   \n",
       "GXE_17081   XVI  374063   T   C     4  2022  0.001978        1439|3400|4622   \n",
       "GXE_17082   XVI  374065   A   G    30  2020  0.015000        1437|3402|4624   \n",
       "GXE_17084   XVI  374089   G   A    16  2022  0.007913        1413|3426|4648   \n",
       "GXE_17086   XVI  374149   C   T   108  2022  0.053000        1353|3486|4708   \n",
       "GXE_17087   XVI  374156   G   T   876  2006  0.437000        1346|3493|4715   \n",
       "GXE_17089   XVI  374183   G   C     4  2022  0.001978        1319|3520|4742   \n",
       "GXE_17090   XVI  374220   T   C     4  2022  0.001978        1282|3557|4779   \n",
       "GXE_17091   XVI  374224   G   C     3  2016  0.001488        1278|3561|4783   \n",
       "GXE_17097   XVI  374261   C   T  1649  2018  0.817000        1241|3598|4820   \n",
       "GXE_17098   XVI  374268   A   G     3  2022  0.001484        1234|3605|4827   \n",
       "GXE_17099   XVI  374270   G   T    26  2022  0.013000        1232|3607|4829   \n",
       "GXE_17101   XVI  374271   G   T    14  2022  0.006924        1231|3608|4830   \n",
       "GXE_17103   XVI  374272   G   A     3  2022  0.001484        1230|3609|4831   \n",
       "GXE_17118   XVI  374535   C   A     8  2022  0.003956              967|3872   \n",
       "GXE_17120   XVI  374541   C   T     4  2022  0.001978              961|3878   \n",
       "GXE_17125   XVI  374570   G   A     5  2022  0.002473              932|3907   \n",
       "GXE_17127   XVI  374633   G   A     4  2022  0.001978              869|3970   \n",
       "GXE_17132   XVI  374711   C   A    29  2022  0.014000              791|4048   \n",
       "GXE_17144   XVI  374855   G   A    67  2020  0.033000              647|4192   \n",
       "GXE_17145   XVI  374857   T   C   136  2022  0.067000              645|4194   \n",
       "GXE_17168   XVI  375080   G   A    31  2022  0.015000              422|4417   \n",
       "GXE_17188   XVI  375331   C   A     8  2020  0.003960              171|4668   \n",
       "GXE_17189   XVI  375333   A  AT     8  2022  0.003956              168|4670   \n",
       "\n",
       "          downstream_distance_str                   region  ...  ACS  ADE  \\\n",
       "var_id                                                      ...             \n",
       "GXE_17015                3982|388  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17016                3978|392  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17017                3977|393  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17018                3973|397  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17033                3785|585  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17034                3779|591  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17042                3687|683  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17053                3584|786  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17055                3581|789  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17058                3486|884         missense_variant  ...  0/0  0/0   \n",
       "GXE_17065                3453|917         missense_variant  ...  0/0  0/0   \n",
       "GXE_17066                3452|918         missense_variant  ...  0/0  0/0   \n",
       "GXE_17071                3392|978         missense_variant  ...  0/0  0/0   \n",
       "GXE_17072                3382|988       synonymous_variant  ...  0/0  0/0   \n",
       "GXE_17081               3228|1142       synonymous_variant  ...  0/0  0/0   \n",
       "GXE_17082               3226|1144       synonymous_variant  ...  0/0  0/0   \n",
       "GXE_17084               3202|1168       synonymous_variant  ...  0/0  0/0   \n",
       "GXE_17086          4971|3142|1228       synonymous_variant  ...  0/0  0/0   \n",
       "GXE_17087          4964|3135|1235         missense_variant  ...  1/1  0/0   \n",
       "GXE_17089          4937|3108|1262         missense_variant  ...  0/0  0/0   \n",
       "GXE_17090          4900|3071|1299         missense_variant  ...  0/0  0/0   \n",
       "GXE_17091          4896|3067|1303       synonymous_variant  ...  0/0  0/0   \n",
       "GXE_17097          4859|3030|1340         missense_variant  ...  1/1  1/1   \n",
       "GXE_17098          4852|3023|1347         missense_variant  ...  0/0  0/0   \n",
       "GXE_17099          4850|3021|1349         missense_variant  ...  0/0  0/0   \n",
       "GXE_17101          4849|3020|1350         missense_variant  ...  0/0  0/0   \n",
       "GXE_17103          4848|3019|1351       synonymous_variant  ...  0/0  0/0   \n",
       "GXE_17118          4585|2756|1614         missense_variant  ...  0/0  0/0   \n",
       "GXE_17120          4579|2750|1620         missense_variant  ...  0/0  0/0   \n",
       "GXE_17125          4550|2721|1649         missense_variant  ...  0/0  0/0   \n",
       "GXE_17127          4487|2658|1712         missense_variant  ...  0/0  0/0   \n",
       "GXE_17132          4409|2580|1790         missense_variant  ...  0/0  0/0   \n",
       "GXE_17144          4265|2436|1934         missense_variant  ...  0/0  0/0   \n",
       "GXE_17145          4263|2434|1936       synonymous_variant  ...  0/0  0/0   \n",
       "GXE_17168          4040|2211|2159         missense_variant  ...  0/0  0/0   \n",
       "GXE_17188      3789|1960|162|2410  unidirectional_promoter  ...  0/0  0/0   \n",
       "GXE_17189      3786|1957|164|2412  unidirectional_promoter  ...  0/0  0/0   \n",
       "\n",
       "           ADF  ADG  ADI  ADR SACE_MAA  \\\n",
       "var_id                                   \n",
       "GXE_17015  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17016  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17017  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17018  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17033  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17034  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17042  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17053  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17055  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17058  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17065  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17066  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17071  0/0  0/0  0/0  0/0      1/1   \n",
       "GXE_17072  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17081  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17082  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17084  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17086  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17087  0/0  0/0  0/0  0/0      1/1   \n",
       "GXE_17089  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17090  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17091  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17097  1/1  1/1  1/1  1/1      1/1   \n",
       "GXE_17098  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17099  0/0  0/0  0/0  0/0      1/1   \n",
       "GXE_17101  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17103  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17118  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17120  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17125  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17127  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17132  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17144  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17145  1/1  0/0  0/0  0/0      0/0   \n",
       "GXE_17168  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17188  0/0  0/0  0/0  0/0      0/0   \n",
       "GXE_17189  0/0  0/0  0/0  0/0      0/0   \n",
       "\n",
       "                                                assoc_traits  \\\n",
       "var_id                                                         \n",
       "GXE_17015  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17016  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17017  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17018  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17033  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17034  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17042  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17053  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17055  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17058  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17065  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17066  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17071  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17072  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17081  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17082  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17084  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17086  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17087  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17089  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17090  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17091  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17097  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17098  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17099  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17101  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17103  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17118  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17120  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17125  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17127  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17132  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17144  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17145  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17168  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17188  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17189  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "\n",
       "                                            assoc_traits_str pool_num  \n",
       "var_id                                                                 \n",
       "GXE_17015  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17016  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17017  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17018  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17033  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17034  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17042  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17053  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17055  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17058  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17065  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17066  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17071  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17072  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17081  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17082  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17084  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17086  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17087  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17089  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17090  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17091  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17097  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17098  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17099  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17101  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17103  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17118  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17120  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17125  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17127  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17132  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17144  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17145  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17168  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17188  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17189  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "\n",
       "[37 rows x 61 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gxe_vars.query('pool_num==90')\n",
    "g"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HSP90 set\n",
    "\n",
    "### We should be able to fit all HSP90 variants in 7 pools\n",
    "- HSP+HSX: 7 pools, 111-113 oligos each\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HSP_00004</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631977</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>56</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>4169|378</td>\n",
       "      <td>3192|7|2712|4314</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00005</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631978</td>\n",
       "      <td>CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>4171|259</td>\n",
       "      <td>3194|9|2593|4195</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>259.0</td>\n",
       "      <td>2</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00006</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631979</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.005946</td>\n",
       "      <td>4171|376</td>\n",
       "      <td>3194|9|2710|4312</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>376.0</td>\n",
       "      <td>2</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00007</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>53</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.026000</td>\n",
       "      <td>4172|375</td>\n",
       "      <td>3195|10|2709|4311</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>375.0</td>\n",
       "      <td>2</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00008</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631981</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>4173|374</td>\n",
       "      <td>3196|11|2708|4310</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>11.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "      <td>2</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00327</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98320</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>1996</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4864|2128|1116|2109|3258</td>\n",
       "      <td>3240</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00329</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98335</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4819|2083|1071|2155|3304</td>\n",
       "      <td>3195</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00330</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98335</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4818|2082|1070|2155|3304</td>\n",
       "      <td>3194</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00334</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98367</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>0</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4729|1993|981|2245|3394</td>\n",
       "      <td>3105</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00336</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98413</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>2004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4728|1992|980|2245|3394</td>\n",
       "      <td>3104</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>332 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS                                                REF  \\\n",
       "var_id                                                                       \n",
       "HSP_00004  XIII  631977                                                  A   \n",
       "HSP_00005  XIII  631978  CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...   \n",
       "HSP_00006  XIII  631979                                                  G   \n",
       "HSP_00007  XIII  631980                                                  C   \n",
       "HSP_00008  XIII  631981                                                  A   \n",
       "...         ...     ...                                                ...   \n",
       "HSX_00327   XVI   98320                                                  T   \n",
       "HSX_00329   XVI   98335                                                  G   \n",
       "HSX_00330   XVI   98335                                                  G   \n",
       "HSX_00334   XVI   98367                                                  C   \n",
       "HSX_00336   XVI   98413                                                  C   \n",
       "\n",
       "          ALT  AC    AN        AF     upstream_distance_str  \\\n",
       "var_id                                                        \n",
       "HSP_00004   G  56  2022  0.028000                  4169|378   \n",
       "HSP_00005   C   6  2022  0.002967                  4171|259   \n",
       "HSP_00006   A  12  2018  0.005946                  4171|376   \n",
       "HSP_00007   T  53  2018  0.026000                  4172|375   \n",
       "HSP_00008   G  25  2020  0.012000                  4173|374   \n",
       "...        ..  ..   ...       ...                       ...   \n",
       "HSX_00327   G   0  1996  0.000000  4864|2128|1116|2109|3258   \n",
       "HSX_00329   A   0  2012  0.000000  4819|2083|1071|2155|3304   \n",
       "HSX_00330   C   0  2012  0.000000  4818|2082|1070|2155|3304   \n",
       "HSX_00334  CT   0  2016  0.000000   4729|1993|981|2245|3394   \n",
       "HSX_00336   G   0  2004  0.000000   4728|1992|980|2245|3394   \n",
       "\n",
       "          downstream_distance_str                   region  ...  \\\n",
       "var_id                                                      ...   \n",
       "HSP_00004        3192|7|2712|4314  unidirectional_promoter  ...   \n",
       "HSP_00005        3194|9|2593|4195  unidirectional_promoter  ...   \n",
       "HSP_00006        3194|9|2710|4312  unidirectional_promoter  ...   \n",
       "HSP_00007       3195|10|2709|4311  unidirectional_promoter  ...   \n",
       "HSP_00008       3196|11|2708|4310  unidirectional_promoter  ...   \n",
       "...                           ...                      ...  ...   \n",
       "HSX_00327                    3240       frameshift_variant  ...   \n",
       "HSX_00329                    3195         missense_variant  ...   \n",
       "HSX_00330                    3194       frameshift_variant  ...   \n",
       "HSX_00334                    3105       synonymous_variant  ...   \n",
       "HSX_00336                    3104       frameshift_variant  ...   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "HSP_00004                    RTP1               YMR185W   \n",
       "HSP_00005                    RTP1               YMR185W   \n",
       "HSP_00006                    RTP1               YMR185W   \n",
       "HSP_00007                    RTP1               YMR185W   \n",
       "HSP_00008                    RTP1               YMR185W   \n",
       "...                           ...                   ...   \n",
       "HSX_00327                     NaN                   NaN   \n",
       "HSX_00329                     NaN                   NaN   \n",
       "HSX_00330                     NaN                   NaN   \n",
       "HSX_00334                     NaN                   NaN   \n",
       "HSX_00336                     NaN                   NaN   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "HSP_00004  downstream_gene_variant                    7.0   \n",
       "HSP_00005  downstream_gene_variant                    9.0   \n",
       "HSP_00006  downstream_gene_variant                    9.0   \n",
       "HSP_00007  downstream_gene_variant                   10.0   \n",
       "HSP_00008  downstream_gene_variant                   11.0   \n",
       "...                            ...                    ...   \n",
       "HSX_00327                      NaN                    NaN   \n",
       "HSX_00329                      NaN                    NaN   \n",
       "HSX_00330                      NaN                    NaN   \n",
       "HSX_00334                      NaN                    NaN   \n",
       "HSX_00336                      NaN                    NaN   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "HSP_00004                   HSC82               YMR186W   \n",
       "HSP_00005                   HSC82               YMR186W   \n",
       "HSP_00006                   HSC82               YMR186W   \n",
       "HSP_00007                   HSC82               YMR186W   \n",
       "HSP_00008                   HSC82               YMR186W   \n",
       "...                           ...                   ...   \n",
       "HSX_00327                     NaN                   NaN   \n",
       "HSX_00329                     NaN                   NaN   \n",
       "HSX_00330                     NaN                   NaN   \n",
       "HSX_00334                     NaN                   NaN   \n",
       "HSX_00336                     NaN                   NaN   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \\\n",
       "var_id                                                                    \n",
       "HSP_00004    upstream_gene_variant                  378.0             2   \n",
       "HSP_00005    upstream_gene_variant                  259.0             2   \n",
       "HSP_00006    upstream_gene_variant                  376.0             2   \n",
       "HSP_00007    upstream_gene_variant                  375.0             2   \n",
       "HSP_00008    upstream_gene_variant                  374.0             2   \n",
       "...                            ...                    ...           ...   \n",
       "HSX_00327                      NaN                    NaN             2   \n",
       "HSX_00329                      NaN                    NaN             3   \n",
       "HSX_00330                      NaN                    NaN             3   \n",
       "HSX_00334                      NaN                    NaN             2   \n",
       "HSX_00336                      NaN                    NaN             2   \n",
       "\n",
       "          pool_num  \n",
       "var_id              \n",
       "HSP_00004      179  \n",
       "HSP_00005      179  \n",
       "HSP_00006      182  \n",
       "HSP_00007      182  \n",
       "HSP_00008      180  \n",
       "...            ...  \n",
       "HSX_00327      178  \n",
       "HSX_00329      180  \n",
       "HSX_00330      179  \n",
       "HSX_00334      180  \n",
       "HSX_00336      182  \n",
       "\n",
       "[332 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "178    113\n",
       "179    111\n",
       "180    111\n",
       "181    111\n",
       "182    111\n",
       "183    112\n",
       "184    112\n",
       "Name: num_of_oligos, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 781\n",
      "Total number of variants: 332\n"
     ]
    }
   ],
   "source": [
    "hsp90_vars_file = \"/home/users/rang/scratch/yeast/hsp90/hsp90_variants_final.txt\"\n",
    "hsp90_vars = pd.read_csv(hsp90_vars_file, sep='\\t', index_col=0)\n",
    "hsp90_vars['num_of_oligos'] = hsp90_vars['num_of_oligos'].astype(int)\n",
    "\n",
    "# assign hsp90 variants to pools\n",
    "hsp90_vars = sort_variants_to_pools(vars_df = hsp90_vars, \n",
    "                                    assignment_order = ['HS'],\n",
    "                                    num_of_pools_per_group = {'HS':7},\n",
    "                                    spillover_assignment = {},\n",
    "                                    max_pool_size = 118, first_pool = 178)\n",
    "\n",
    "display(hsp90_vars)\n",
    "# check number of oligos in each pool\n",
    "display(hsp90_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", hsp90_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(hsp90_vars))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Humanized yeast pilot\n",
    "\n",
    "### We should be able to fit humanized yeast variants in 6 pools\n",
    "Assign one pool per human gene to study\n",
    "- DHFR: 185\n",
    "- DPAGT1: 186\n",
    "- NSDHL: 187\n",
    "- PGK1: 188\n",
    "- PKLR: 189\n",
    "- UROS: 190"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DHFR_10007</th>\n",
       "      <td>DHFR</td>\n",
       "      <td>457</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHFR_10008</th>\n",
       "      <td>DHFR</td>\n",
       "      <td>237</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHFR_10104</th>\n",
       "      <td>DHFR</td>\n",
       "      <td>186</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHFR_00019</th>\n",
       "      <td>chr5</td>\n",
       "      <td>79929796</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHFR_00059</th>\n",
       "      <td>chr5</td>\n",
       "      <td>79950223</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00009</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477478</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00010</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477480</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00012</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477484</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00020</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477526</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00029</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477556</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>223 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            CHROM        POS REF ALT  num_of_oligos  pool_num\n",
       "var_id                                                       \n",
       "DHFR_10007   DHFR        457   A   T              2       185\n",
       "DHFR_10008   DHFR        237   C   T              1       185\n",
       "DHFR_10104   DHFR        186   G   C              1       185\n",
       "DHFR_00019   chr5   79929796   G   C              3       185\n",
       "DHFR_00059   chr5   79950223   C   T              4       185\n",
       "...           ...        ...  ..  ..            ...       ...\n",
       "UROS_00009  chr10  127477478   T   C              4       190\n",
       "UROS_00010  chr10  127477480   G   C              4       190\n",
       "UROS_00012  chr10  127477484   G   C              3       190\n",
       "UROS_00020  chr10  127477526   G   A              4       190\n",
       "UROS_00029  chr10  127477556   T   C              3       190\n",
       "\n",
       "[223 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "185     70\n",
       "186    118\n",
       "187    118\n",
       "188    118\n",
       "189    118\n",
       "190    118\n",
       "Name: num_of_oligos, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 660\n",
      "Total number of variants: 223\n"
     ]
    }
   ],
   "source": [
    "gene_to_pool = {'DHFR':185,\n",
    "                'DPAGT1':186,\n",
    "                'NSDHL':187,\n",
    "                'PGK1':188,\n",
    "                'PKLR':189,\n",
    "                'UROS':190}\n",
    "\n",
    "# read in all_SNPs_humanized_GG_9bp_OLIGO.tab, subset to variants of interest, count oligos and assign to pools\n",
    "\n",
    "humanized_vars_file = \"/home/users/rang/crispey3/humanized/Output/all_SNPs_humanized_combined_GG_9bp_OLIGO.tab\"\n",
    "humanized_vars = pd.read_csv(humanized_vars_file, sep='\\t')\n",
    "\n",
    "# calculate number of oligos per variant, add info to column and adjust to reflect one variant per row\n",
    "humanized_vars['num_of_oligos'] = humanized_vars['var_id'].map(humanized_vars.groupby('var_id').size())\n",
    "# convert humanized_vars to simplified dataframe aligned with the other sets\n",
    "humanized_vars = humanized_vars[['var_id', 'chrom', 'SNP_chr_pos', 'REF', 'ALT', 'num_of_oligos']].rename(columns={'chrom':'CHROM', 'SNP_chr_pos':'POS'}).set_index('var_id').drop_duplicates()\n",
    "humanized_vars['ALT'] = humanized_vars['ALT'].apply(lambda x: x[1:-1])\n",
    "\n",
    "# assign humanized variants to pools by variant prefix, which indicates the gene of interest\n",
    "humanized_vars['pool_num'] = [gene_to_pool[v.split('_')[0]] for v in humanized_vars.index]\n",
    "\n",
    "display(humanized_vars)\n",
    "# check number of oligos in each pool\n",
    "display(humanized_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", humanized_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(humanized_vars))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This leaves the following sets to assign to pools:\n",
    "- Fitness ladder (ladder oligo, synonymous donor+guide, scramble donor+guide): 191\n",
    "- Neutrals (CRISPEY1 neutral, donor+sgGFP, ladder scrambled donor+sgGFP): 192"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write pool assignment of all variants to single file for reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00008</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TG</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00009</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00010</th>\n",
       "      <td>V</td>\n",
       "      <td>237196</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00009</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477478</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00010</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477480</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00012</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477484</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00020</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477526</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UROS_00029</th>\n",
       "      <td>chr10</td>\n",
       "      <td>127477556</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8776 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            CHROM        POS REF ALT  num_of_oligos  pool_num\n",
       "var_id                                                       \n",
       "ERG_00003       V     237122   G   A              2        13\n",
       "ERG_00004       V     237136   T   C              2        21\n",
       "EGE_00008       V     237191   T  TG              2        25\n",
       "EGE_00009       V     237191   T  TA              2        22\n",
       "EGE_00010       V     237196   A   T              2        22\n",
       "...           ...        ...  ..  ..            ...       ...\n",
       "UROS_00009  chr10  127477478   T   C              4       190\n",
       "UROS_00010  chr10  127477480   G   C              4       190\n",
       "UROS_00012  chr10  127477484   G   C              3       190\n",
       "UROS_00020  chr10  127477526   G   A              4       190\n",
       "UROS_00029  chr10  127477556   T   C              3       190\n",
       "\n",
       "[8776 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output_columns = ['CHROM', 'POS', 'REF', 'ALT', 'num_of_oligos', 'pool_num']\n",
    "out_df = pd.concat([ergosterol_vars[output_columns], \n",
    "                    gxg_vars[output_columns], \n",
    "                    epival_vars[output_columns],\n",
    "                    gxe_vars[output_columns],\n",
    "                    hsp90_vars[output_columns],\n",
    "                    humanized_vars[output_columns]])\n",
    "display(out_df)\n",
    "\n",
    "out_df.to_csv('crispey3_vars_pool_assignment.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "117    63\n",
       "116    40\n",
       "118    39\n",
       "115    25\n",
       "114    12\n",
       "111     5\n",
       "113     2\n",
       "112     2\n",
       "110     1\n",
       "70      1\n",
       "Name: num_of_oligos, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df.groupby('pool_num').num_of_oligos.sum().value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assemble all variants (except humanized yeast set) from separate VCF files into single VCF for oligo design pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500 variants written to file\n",
      "1000 variants written to file\n",
      "1500 variants written to file\n",
      "2000 variants written to file\n",
      "2500 variants written to file\n",
      "3000 variants written to file\n",
      "3500 variants written to file\n",
      "4000 variants written to file\n",
      "4500 variants written to file\n",
      "5000 variants written to file\n",
      "5500 variants written to file\n",
      "6000 variants written to file\n",
      "6500 variants written to file\n",
      "7000 variants written to file\n",
      "7500 variants written to file\n",
      "8000 variants written to file\n",
      "8500 variants written to file\n",
      "Total number of variants in design VCF: 8553\n"
     ]
    }
   ],
   "source": [
    "lib_design_vcf_file = 'all_variants_design_oligos.vcf'\n",
    "template_vcf_file = '/home/users/rang/crispey3/initial_design/Input/gxg_variants_design_oligos_initial.vcf'\n",
    "template_vcf = vcf.Reader(filename=template_vcf_file)\n",
    "\n",
    "# open output file\n",
    "lib_design_vcf = vcf.Writer(open(lib_design_vcf_file, 'w'), template_vcf)\n",
    "\n",
    "vcf_paths_list = ['/home/users/rang/crispey3/initial_design/Input/ergosterol_variants_design_oligos_initial.vcf',\n",
    "                  '/home/users/rang/crispey3/initial_design/Input/gxg_variants_design_oligos_initial.vcf',\n",
    "                  '/home/users/rang/crispey3/initial_design/Input/epival_variants_final.vcf',\n",
    "                  '/home/users/rang/crispey3/initial_design/Input/gxe_variants_design_oligos_initial.vcf',\n",
    "                  '/home/users/rang/scratch/yeast/hsp90/hsp90_variants_final.vcf']\n",
    "\n",
    "var_num=0\n",
    "for variants_vcf_path in vcf_paths_list:\n",
    "    variants = vcf.Reader(filename=variants_vcf_path)\n",
    "    for record in variants:\n",
    "        if record.ID in out_df.index:\n",
    "            var_num+=1\n",
    "            lib_design_vcf.write_record(record)\n",
    "            \n",
    "            if var_num % 500 == 0:\n",
    "                print(var_num, 'variants written to file')\n",
    "lib_design_vcf.close()\n",
    "\n",
    "print('Total number of variants in design VCF:', var_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(84, \"['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 118\n",
      "(85, \"['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 118\n",
      "(86, \"['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 117\n",
      "(87, \"['Caffeine;15mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 118\n",
      "(88, \"['Caffeine;15mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 118\n",
      "(89, \"['Caffeine;15mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 117\n",
      "(90, \"['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neomycin;5mg/mL;2']\") 118\n",
      "(91, \"['Lithium_Chloride;100mM;2', 'Neomycin;5mg/mL;2']\") 118\n",
      "(92, \"['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2']\") 117\n",
      "(93, \"['Caffeine;15mM;2', 'Cobalt_Chloride;2mM;2']\") 117\n",
      "(94, \"['Cobalt_Chloride;2mM;2', 'Fructose;;1']\") 118\n",
      "(95, \"['Lithium_Chloride;100mM;2', 'Neomycin;5mg/mL;2']\") 28\n",
      "(95, \"['Lithium_Chloride;100mM;2']\") 90\n",
      "(96, \"['Lithium_Chloride;100mM;2']\") 118\n",
      "(97, \"['Lithium_Chloride;100mM;2']\") 118\n",
      "(98, \"['Lithium_Chloride;100mM;2']\") 118\n",
      "(99, \"['Lithium_Chloride;100mM;2']\") 118\n",
      "(100, \"['Lithium_Chloride;100mM;2']\") 118\n",
      "(101, \"['Lithium_Chloride;100mM;2']\") 117\n",
      "(102, \"['Lithium_Chloride;100mM;2']\") 117\n",
      "(103, \"['Lithium_Chloride;100mM;2']\") 117\n",
      "(104, \"['Lithium_Chloride;100mM;2']\") 117\n",
      "(105, \"['Lithium_Chloride;100mM;2']\") 117\n",
      "(106, \"['Lithium_Chloride;100mM;2']\") 117\n",
      "(107, \"['Lithium_Chloride;100mM;2']\") 117\n",
      "(108, \"['Lithium_Chloride;100mM;2']\") 117\n",
      "(109, \"['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2', 'Fructose;;1']\") 10\n",
      "(109, \"['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neomycin;5mg/mL;2']\") 22\n",
      "(109, \"['Cobalt_Chloride;2mM;2', 'Fructose;;1']\") 52\n",
      "(109, \"['Cobalt_Chloride;2mM;2']\") 32\n",
      "(110, \"['Cobalt_Chloride;2mM;2']\") 117\n",
      "(111, \"['Cobalt_Chloride;2mM;2']\") 117\n",
      "(112, \"['Cobalt_Chloride;2mM;2']\") 115\n",
      "(113, \"['Cobalt_Chloride;2mM;2']\") 115\n",
      "(114, \"['Cobalt_Chloride;2mM;2']\") 115\n",
      "(115, \"['Cobalt_Chloride;2mM;2']\") 116\n",
      "(116, \"['Cobalt_Chloride;2mM;2']\") 116\n",
      "(117, \"['Cobalt_Chloride;2mM;2']\") 116\n",
      "(118, \"['Cobalt_Chloride;2mM;2']\") 116\n",
      "(119, \"['Cobalt_Chloride;2mM;2']\") 116\n",
      "(120, \"['Cobalt_Chloride;2mM;2']\") 116\n",
      "(121, \"['Cobalt_Chloride;2mM;2']\") 116\n",
      "(122, \"['Cobalt_Chloride;2mM;2']\") 116\n",
      "(123, \"['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2']\") 58\n",
      "(123, \"['Fluconazole;100uM;2', 'Fructose;;1']\") 59\n",
      "(124, \"['Fluconazole;100uM;2', 'Fructose;;1']\") 14\n",
      "(124, \"['Fluconazole;100uM;2']\") 103\n",
      "(125, \"['Fluconazole;100uM;2']\") 117\n",
      "(126, \"['Fluconazole;100uM;2']\") 117\n",
      "(127, \"['Fluconazole;100uM;2']\") 117\n",
      "(128, \"['Fluconazole;100uM;2']\") 115\n",
      "(129, \"['Fluconazole;100uM;2']\") 115\n",
      "(130, \"['Fluconazole;100uM;2']\") 115\n",
      "(131, \"['Fluconazole;100uM;2']\") 115\n",
      "(132, \"['Fluconazole;100uM;2']\") 115\n",
      "(133, \"['Fluconazole;100uM;2']\") 115\n",
      "(134, \"['Fluconazole;100uM;2']\") 116\n",
      "(135, \"['Fluconazole;100uM;2']\") 116\n",
      "(136, \"['Caffeine;15mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 14\n",
      "(136, \"['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 104\n",
      "(137, \"['Caffeine;15mM;2', 'Fluconazole;100uM;2', 'Neomycin;5mg/mL;2']\") 6\n",
      "(137, \"['Caffeine;15mM;2', 'Neomycin;5mg/mL;2']\") 12\n",
      "(137, \"['Neomycin;5mg/mL;2']\") 99\n",
      "(138, \"['Neomycin;5mg/mL;2']\") 117\n",
      "(139, \"['Neomycin;5mg/mL;2']\") 117\n",
      "(140, \"['Neomycin;5mg/mL;2']\") 115\n",
      "(141, \"['Neomycin;5mg/mL;2']\") 116\n",
      "(142, \"['Neomycin;5mg/mL;2']\") 116\n",
      "(143, \"['Neomycin;5mg/mL;2']\") 116\n",
      "(144, \"['Neomycin;5mg/mL;2']\") 116\n",
      "(145, \"['Neomycin;5mg/mL;2']\") 116\n",
      "(146, \"['Fructose;;1']\") 118\n",
      "(147, \"['Fructose;;1']\") 117\n",
      "(148, \"['Fructose;;1']\") 117\n",
      "(149, \"['Fructose;;1']\") 117\n",
      "(150, \"['Fructose;;1']\") 117\n",
      "(151, \"['Fructose;;1']\") 117\n",
      "(152, \"['Fructose;;1']\") 117\n",
      "(153, \"['Fructose;;1']\") 117\n",
      "(154, \"['Fructose;;1']\") 117\n",
      "(155, \"['Fructose;;1']\") 118\n",
      "(156, \"['Fructose;;1']\") 118\n",
      "(157, \"['Fructose;;1']\") 118\n",
      "(158, \"['Fructose;;1']\") 118\n",
      "(159, \"['Caffeine;15mM;2', 'Fructose;;1']\") 2\n",
      "(159, \"['Caffeine;15mM;2']\") 63\n",
      "(159, \"['Fructose;;1']\") 8\n",
      "(159, \"['Lithium_Chloride;100mM;2']\") 44\n",
      "(160, \"['Caffeine;15mM;2']\") 117\n",
      "(161, \"['Caffeine;15mM;2']\") 117\n",
      "(162, \"['Caffeine;15mM;2']\") 117\n",
      "(163, \"['Caffeine;15mM;2']\") 117\n",
      "(164, \"['Caffeine;15mM;2']\") 117\n",
      "(165, \"['Caffeine;15mM;2']\") 117\n",
      "(166, \"['Caffeine;15mM;2']\") 117\n",
      "(167, \"['Caffeine;15mM;2']\") 117\n",
      "(168, \"['Caffeine;15mM;2']\") 117\n",
      "(169, \"['Caffeine;15mM;2']\") 117\n",
      "(170, \"['Caffeine;15mM;2']\") 117\n",
      "(171, \"['Caffeine;15mM;2']\") 117\n",
      "(172, \"['Caffeine;15mM;2']\") 117\n",
      "(173, \"['Caffeine;15mM;2']\") 117\n",
      "(174, \"['Caffeine;15mM;2']\") 117\n",
      "(175, \"['Caffeine;15mM;2']\") 117\n",
      "(176, \"['Caffeine;15mM;2']\") 118\n",
      "(177, \"['Caffeine;15mM;2']\") 118\n"
     ]
    }
   ],
   "source": [
    "for i, j in gxe_vars.groupby(['pool_num', 'assoc_traits']).num_of_oligos.sum().items():\n",
    "    print(i, j)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (crispey-bar)",
   "language": "python",
   "name": "crispey-bar"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
