{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize variants selected from select_crispey3_library_variants notebook into pools\n",
    "The goal is to assign each variant with a pool number, and that number will be used when assigning barcodes to guide-donors and grouping the oligos into pools in the final CRISPEY3 library construction.\n",
    "\n",
    "There are multiple sets going into the final library, and each set will have slightly different grouping and randomization requirements.\n",
    "\n",
    "Broadly speaking, here are some general principles:\n",
    "- Keep oligos targeting the same variant together. This way, we control the pool covariate and can better account for guide-specific differences per variant.\n",
    "- Within each variant set, avoid clustering variants in the same genomic region together. This avoids cloning issues (PCR mispriming, template swaps in Gibson assembly, etc.), and also controls for pool-specific batch effects when comparing fitness effects across variants.\n",
    "    - This negates a gene-based grouping approach, which would make for convenient organization and effective utilization of pools to do gene-specific small growth competition experiments.\n",
    "    - Sets such as the ergosterol pathway set and GxG sets must utilize whole-set variant randomization to avoid pool effects being associated with each gene.\n",
    "- An alternative approach would be to pool variants by gene. The GxE set is already somewhat grouped by gene because although the variants are currently grouped by condition, genes are often condition-specific in their testing and so variants in the same gene will inevitably end up in the same few pools. Some randomization here might still be possible, but not if there are too few pools available, especially for oligos used in multiple conditions\n",
    "- When assigning variants to each pool, take into account the number of oligos generated for each variant so that each pool contains the appropriate number of oligos (118, plus 3 additional technical oligos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, random, vcf\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_variants_to_pools(vars_df, assignment_order=[], grouping_col = None, num_of_pools_per_group={}, \n",
    "                           spillover_assignment={}, max_pool_size=1, first_pool=1, ):\n",
    "    \n",
    "    # for storing variant ID of spillover variants\n",
    "    spillover_vars = {g : [] for g in assignment_order}\n",
    "#     # add discard list\n",
    "#     spillover_vars['discard'] = []\n",
    "    \n",
    "    # iterate over each group and assign to pools\n",
    "    for group in assignment_order:\n",
    "        # get subset of variants belonging to group\n",
    "        # if column to group by provided, select variant subset by column. Otherwise, filter by variant ID\n",
    "        if grouping_col:\n",
    "            subset = vars_df.loc[vars_df[grouping_col] == group]\n",
    "        else:\n",
    "            subset = vars_df.filter(like=group, axis=0)\n",
    "        # sort variants in descending order of number of oligos, then randomize variants within each group of a certain number of oligos\n",
    "        # this enables even, efficient packing with randomization across pools\n",
    "        subset = subset.groupby('num_of_oligos').apply(lambda x: x.sample(frac=1, random_state=1)).reset_index(level=0, drop=True)[::-1]\n",
    "        \n",
    "        # get number of pools to use for group\n",
    "        num_of_pools = num_of_pools_per_group[group]\n",
    "        # assign oligos to spillover if no pools set for group\n",
    "        if num_of_pools == 0:\n",
    "#             print(\"Excess oligos for {}: {}\".format(group, subset.num_of_oligos.sum())) # debugging\n",
    "            spillover_vars[spillover_assignment[group]] += subset.index.tolist()\n",
    "            continue\n",
    "\n",
    "        # define current pool sizes\n",
    "        cur_pool_sizes = [0] * num_of_pools\n",
    "        # add spillover variants to first pool(s) in group\n",
    "        # note: packing is done by max-ing out first pool, then moving to next pool, and doing so until all spillover variants are assigned\n",
    "        if len(spillover_vars[group]) > 0:\n",
    "            spillover_pool_num = first_pool\n",
    "            for v, row in vars_df.loc[spillover_vars[group]].iterrows():\n",
    "                # move to next spillover pool number if pool has reached capacity\n",
    "                if cur_pool_sizes[spillover_pool_num-first_pool] + row['num_of_oligos'] > max_pool_size:\n",
    "                    spillover_pool_num += 1\n",
    "                    \n",
    "                # assign each spillover variant to pool number\n",
    "                vars_df.loc[v, 'pool_num'] = spillover_pool_num\n",
    "                # increase pool size by variant's number of oligos\n",
    "                cur_pool_sizes[spillover_pool_num-first_pool] += row['num_of_oligos']\n",
    "       \n",
    "        # add all other variants to pools in group\n",
    "        # variants are assigned to spread oligos out, ensuring similar group sizes across pools\n",
    "        excess = 0\n",
    "        for v, row in subset.iterrows():\n",
    "            # find smallest current pool\n",
    "            i = np.argmin(cur_pool_sizes)\n",
    "            # assign variant to pool if number of oligos in pool is within max pool size\n",
    "            if cur_pool_sizes[i] + row['num_of_oligos'] <= max_pool_size:\n",
    "                vars_df.loc[v, 'pool_num'] = first_pool+i\n",
    "                cur_pool_sizes[i] += row['num_of_oligos']\n",
    "            # if current pools have insufficient space, send to group assigned for spillover\n",
    "            else:\n",
    "                excess +=row['num_of_oligos']\n",
    "                spillover_vars[spillover_assignment[group]].append(v)\n",
    "        \n",
    "#         # (debugging) print spillover variants\n",
    "#         if group in spillover_assignment.keys():\n",
    "#             print(\"Excess oligos for {}: {}\".format(group, excess))\n",
    "        \n",
    "        # once assigned, advance first_pool for next group\n",
    "        first_pool += num_of_pools\n",
    "    \n",
    "#     # remove discarded variants\n",
    "#     vars_df = vars_df.loc[~vars_df.index.isin(spillover_vars['discard'])]\n",
    "#     # identify number of variants discarded\n",
    "#     print(\"Variants discarded:\", spillover_vars['discard'])\n",
    "\n",
    "    vars_df['pool_num'] = vars_df['pool_num'].astype(int)\n",
    "\n",
    "    return vars_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/crispey3/library_design/Input/\"\n",
    "os.chdir(working_dir)\n",
    "\n",
    "# setting the random seed\n",
    "random.seed(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ergosterol set\n",
    "- ERG: 21 pools, 117-118 oligos each\n",
    "- EGE: 4 pools, 117-118 oligos each\n",
    "- EGD: 1 pool, 118 oligos each\n",
    "- EGC: 2 pools, 110-111 oligos each\n",
    "- EGB: 1 pool, 117 oligos each\n",
    "- EGA: 1 pool, 117 oligos each\n",
    "- EG9: 1 pool, 118 oligos each\n",
    "- EG8: 3 pools, 113-114 oligos each\n",
    "\n",
    "See code for spillover assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>3</td>\n",
       "      <td>2382|1631|1338|448|2910</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>448.0</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>17</td>\n",
       "      <td>2396|1645|1324|434|2896</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>434.0</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00008</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TG</td>\n",
       "      <td>661</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.327000</td>\n",
       "      <td>72</td>\n",
       "      <td>2451|1700|1268|378|2840</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>72.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00009</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>72</td>\n",
       "      <td>2451|1700|1268|378|2840</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>72.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00010</th>\n",
       "      <td>V</td>\n",
       "      <td>237196</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>69</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.034000</td>\n",
       "      <td>77</td>\n",
       "      <td>2456|1705|1264|374|2836</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>77.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00039</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499649</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>3868|16</td>\n",
       "      <td>4690|1026|357|1930|4403</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>357.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00040</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499650</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>3867|15</td>\n",
       "      <td>4689|1025|358|1931|4404</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>358.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00041</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00042</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>35</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.017000</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00044</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499659</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>3858|6</td>\n",
       "      <td>4680|1016|367|1940|4413</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>367.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1580 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                       \n",
       "ERG_00003     V  237122   G   A    9  2022  0.004451                     3   \n",
       "ERG_00004     V  237136   T   C   26  2022  0.013000                    17   \n",
       "EGE_00008     V  237191   T  TG  661  2022  0.327000                    72   \n",
       "EGE_00009     V  237191   T  TA    6  2022  0.002967                    72   \n",
       "EGE_00010     V  237196   A   T   69  2022  0.034000                    77   \n",
       "...         ...     ...  ..  ..  ...   ...       ...                   ...   \n",
       "EGA_00039   XVI  499649   G   A    4  2022  0.001978               3868|16   \n",
       "EGA_00040   XVI  499650   G   C    5  2022  0.002473               3867|15   \n",
       "EGA_00041   XVI  499655   G   A    4  2022  0.001978               3862|10   \n",
       "EGA_00042   XVI  499655   G   T   35  2022  0.017000               3862|10   \n",
       "EGA_00044   XVI  499659   G   A    1  2022  0.000495                3858|6   \n",
       "\n",
       "           downstream_distance_str                   region  ...  \\\n",
       "var_id                                                       ...   \n",
       "ERG_00003  2382|1631|1338|448|2910  unidirectional_promoter  ...   \n",
       "ERG_00004  2396|1645|1324|434|2896  unidirectional_promoter  ...   \n",
       "EGE_00008  2451|1700|1268|378|2840  unidirectional_promoter  ...   \n",
       "EGE_00009  2451|1700|1268|378|2840  unidirectional_promoter  ...   \n",
       "EGE_00010  2456|1705|1264|374|2836  unidirectional_promoter  ...   \n",
       "...                            ...                      ...  ...   \n",
       "EGA_00039  4690|1026|357|1930|4403  unidirectional_promoter  ...   \n",
       "EGA_00040  4689|1025|358|1931|4404  unidirectional_promoter  ...   \n",
       "EGA_00041  4684|1020|363|1936|4409  unidirectional_promoter  ...   \n",
       "EGA_00042  4684|1020|363|1936|4409  unidirectional_promoter  ...   \n",
       "EGA_00044  4680|1016|367|1940|4413  unidirectional_promoter  ...   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00003                    SAH1               YER043C   \n",
       "ERG_00004                    SAH1               YER043C   \n",
       "EGE_00008                    SAH1               YER043C   \n",
       "EGE_00009                    SAH1               YER043C   \n",
       "EGE_00010                    SAH1               YER043C   \n",
       "...                           ...                   ...   \n",
       "EGA_00039                   ERG10               YPL028W   \n",
       "EGA_00040                   ERG10               YPL028W   \n",
       "EGA_00041                   ERG10               YPL028W   \n",
       "EGA_00042                   ERG10               YPL028W   \n",
       "EGA_00044                   ERG10               YPL028W   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "ERG_00003    upstream_gene_variant                    3.0   \n",
       "ERG_00004    upstream_gene_variant                   17.0   \n",
       "EGE_00008    upstream_gene_variant                   72.0   \n",
       "EGE_00009    upstream_gene_variant                   72.0   \n",
       "EGE_00010    upstream_gene_variant                   77.0   \n",
       "...                            ...                    ...   \n",
       "EGA_00039  downstream_gene_variant                  357.0   \n",
       "EGA_00040  downstream_gene_variant                  358.0   \n",
       "EGA_00041  downstream_gene_variant                  363.0   \n",
       "EGA_00042  downstream_gene_variant                  363.0   \n",
       "EGA_00044  downstream_gene_variant                  367.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00003                   ERG28               YER044C   \n",
       "ERG_00004                   ERG28               YER044C   \n",
       "EGE_00008                   ERG28               YER044C   \n",
       "EGE_00009                   ERG28               YER044C   \n",
       "EGE_00010                   ERG28               YER044C   \n",
       "...                           ...                   ...   \n",
       "EGA_00039                    SMA1               YPL027W   \n",
       "EGA_00040                    SMA1               YPL027W   \n",
       "EGA_00041                    SMA1               YPL027W   \n",
       "EGA_00042                    SMA1               YPL027W   \n",
       "EGA_00044                    SMA1               YPL027W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \\\n",
       "var_id                                                                    \n",
       "ERG_00003  downstream_gene_variant                  448.0             2   \n",
       "ERG_00004  downstream_gene_variant                  434.0             2   \n",
       "EGE_00008  downstream_gene_variant                  378.0             2   \n",
       "EGE_00009  downstream_gene_variant                  378.0             2   \n",
       "EGE_00010  downstream_gene_variant                  374.0             2   \n",
       "...                            ...                    ...           ...   \n",
       "EGA_00039    upstream_gene_variant                   16.0             2   \n",
       "EGA_00040    upstream_gene_variant                   15.0             2   \n",
       "EGA_00041    upstream_gene_variant                   10.0             2   \n",
       "EGA_00042    upstream_gene_variant                   10.0             2   \n",
       "EGA_00044    upstream_gene_variant                    6.0             2   \n",
       "\n",
       "          pool_num  \n",
       "var_id              \n",
       "ERG_00003       13  \n",
       "ERG_00004       21  \n",
       "EGE_00008       25  \n",
       "EGE_00009       22  \n",
       "EGE_00010       22  \n",
       "...            ...  \n",
       "EGA_00039       30  \n",
       "EGA_00040       30  \n",
       "EGA_00041       30  \n",
       "EGA_00042       32  \n",
       "EGA_00044       30  \n",
       "\n",
       "[1580 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "1     118\n",
       "2     118\n",
       "3     118\n",
       "4     118\n",
       "5     118\n",
       "6     118\n",
       "7     118\n",
       "8     118\n",
       "9     117\n",
       "10    117\n",
       "11    117\n",
       "12    117\n",
       "13    117\n",
       "14    117\n",
       "15    117\n",
       "16    117\n",
       "17    117\n",
       "18    117\n",
       "19    117\n",
       "20    117\n",
       "21    117\n",
       "22    118\n",
       "23    118\n",
       "24    118\n",
       "25    117\n",
       "26    118\n",
       "27    110\n",
       "28    111\n",
       "29    117\n",
       "30    117\n",
       "31    118\n",
       "32    114\n",
       "33    114\n",
       "34    113\n",
       "Name: num_of_oligos, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 3968\n",
      "Total number of variants: 1580\n"
     ]
    }
   ],
   "source": [
    "ergosterol_vars_file = \"/home/users/rang/scratch/yeast/ergosterol/ergosterol_variants_final.txt\"\n",
    "ergosterol_vars = pd.read_csv(ergosterol_vars_file, sep='\\t', index_col=0)\n",
    "ergosterol_vars['num_of_oligos'] = ergosterol_vars['num_of_oligos'].astype(int)\n",
    "\n",
    "# assign ergosterol variants to pools\n",
    "ergosterol_vars = sort_variants_to_pools(vars_df = ergosterol_vars, \n",
    "                                         assignment_order = ['ERG','EGE','EGD','EGC','EGB','EGA','EG9','EG8'],\n",
    "                                         num_of_pools_per_group = {'ERG':21, 'EGE':4, 'EGD':1, 'EGC':2, 'EGB':1, 'EGA':1, 'EG9':1, 'EG8':3},\n",
    "                                         spillover_assignment = {'ERG':'EGE', 'EGE':'EGC', 'EGD':'EGC', 'EGC':'EG8', 'EGB':'EGA', 'EGA':'EG8', 'EG9':'EG8'},\n",
    "                                         max_pool_size = 118, first_pool = 1)\n",
    "\n",
    "display(ergosterol_vars)\n",
    "# check number of oligos in each pool\n",
    "display(ergosterol_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", ergosterol_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(ergosterol_vars))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GXG set\n",
    "- GXG: 43 pools, 114-116 oligos each"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>assoc_gene</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00321</th>\n",
       "      <td>I</td>\n",
       "      <td>101530</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>385|1833</td>\n",
       "      <td>1662|4742|35</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>YAL025C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>385.0</td>\n",
       "      <td>LTE1</td>\n",
       "      <td>YAL024C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>35.0</td>\n",
       "      <td>3</td>\n",
       "      <td>['YAL024C', 'YAL025C']</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00329</th>\n",
       "      <td>I</td>\n",
       "      <td>101816</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001980</td>\n",
       "      <td>671|2119</td>\n",
       "      <td>1948|4456</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>['YAL024C']</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00332</th>\n",
       "      <td>I</td>\n",
       "      <td>101901</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>90</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.045000</td>\n",
       "      <td>756|2204</td>\n",
       "      <td>2033|4371</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>['YAL024C']</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00339</th>\n",
       "      <td>I</td>\n",
       "      <td>102261</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>78</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.039000</td>\n",
       "      <td>1116|2564</td>\n",
       "      <td>2393|4011</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>['YAL024C']</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00346</th>\n",
       "      <td>I</td>\n",
       "      <td>102699</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>310</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.154000</td>\n",
       "      <td>1554|3002</td>\n",
       "      <td>2831|3573</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>['YAL024C']</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50437</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899389</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>3625|805|1366|2655</td>\n",
       "      <td>2385</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>['YPR181C']</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50440</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899419</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>196</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.097000</td>\n",
       "      <td>3655|775|1336|2625</td>\n",
       "      <td>2415</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>['YPR181C']</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50446</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001487</td>\n",
       "      <td>3757|673|1234|2523</td>\n",
       "      <td>2517</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>['YPR181C']</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50447</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899679</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>68</td>\n",
       "      <td>2010</td>\n",
       "      <td>0.034000</td>\n",
       "      <td>3915|12|515|1076|2365</td>\n",
       "      <td>2675</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>YPR181C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>12.0</td>\n",
       "      <td>SMX3</td>\n",
       "      <td>YPR182W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>515.0</td>\n",
       "      <td>4</td>\n",
       "      <td>['YPR181C']</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50454</th>\n",
       "      <td>XVI</td>\n",
       "      <td>900065</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001487</td>\n",
       "      <td>4301|398|129|690|1979</td>\n",
       "      <td>3061</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>YPR181C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>398.0</td>\n",
       "      <td>SMX3</td>\n",
       "      <td>YPR182W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>129.0</td>\n",
       "      <td>3</td>\n",
       "      <td>['YPR182W', 'YPR181C']</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF  upstream_distance_str  \\\n",
       "var_id                                                                        \n",
       "GXG_00321     I  101530   G   C    4  2022  0.001978               385|1833   \n",
       "GXG_00329     I  101816   C   T    4  2020  0.001980               671|2119   \n",
       "GXG_00332     I  101901   C   A   90  2022  0.045000               756|2204   \n",
       "GXG_00339     I  102261   C   T   78  2020  0.039000              1116|2564   \n",
       "GXG_00346     I  102699   G   A  310  2016  0.154000              1554|3002   \n",
       "...         ...     ...  ..  ..  ...   ...       ...                    ...   \n",
       "GXG_50437   XVI  899389   G   A    6  2018  0.002973     3625|805|1366|2655   \n",
       "GXG_50440   XVI  899419   A   G  196  2014  0.097000     3655|775|1336|2625   \n",
       "GXG_50446   XVI  899521   G   C    3  2018  0.001487     3757|673|1234|2523   \n",
       "GXG_50447   XVI  899679   A   G   68  2010  0.034000  3915|12|515|1076|2365   \n",
       "GXG_50454   XVI  900065   C   G    3  2018  0.001487  4301|398|129|690|1979   \n",
       "\n",
       "          downstream_distance_str                   region  ...  \\\n",
       "var_id                                                      ...   \n",
       "GXG_00321            1662|4742|35  unidirectional_promoter  ...   \n",
       "GXG_00329               1948|4456         missense_variant  ...   \n",
       "GXG_00332               2033|4371       synonymous_variant  ...   \n",
       "GXG_00339               2393|4011       synonymous_variant  ...   \n",
       "GXG_00346               2831|3573       synonymous_variant  ...   \n",
       "...                           ...                      ...  ...   \n",
       "GXG_50437                    2385       synonymous_variant  ...   \n",
       "GXG_50440                    2415       synonymous_variant  ...   \n",
       "GXG_50446                    2517       synonymous_variant  ...   \n",
       "GXG_50447                    2675   bidirectional_promoter  ...   \n",
       "GXG_50454                    3061   bidirectional_promoter  ...   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00321               YAL025C    upstream_gene_variant   \n",
       "GXG_00329                   NaN                      NaN   \n",
       "GXG_00332                   NaN                      NaN   \n",
       "GXG_00339                   NaN                      NaN   \n",
       "GXG_00346                   NaN                      NaN   \n",
       "...                         ...                      ...   \n",
       "GXG_50437                   NaN                      NaN   \n",
       "GXG_50440                   NaN                      NaN   \n",
       "GXG_50446                   NaN                      NaN   \n",
       "GXG_50447               YPR181C    upstream_gene_variant   \n",
       "GXG_50454               YPR181C    upstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXG_00321                  385.0                    LTE1   \n",
       "GXG_00329                    NaN                     NaN   \n",
       "GXG_00332                    NaN                     NaN   \n",
       "GXG_00339                    NaN                     NaN   \n",
       "GXG_00346                    NaN                     NaN   \n",
       "...                          ...                     ...   \n",
       "GXG_50437                    NaN                     NaN   \n",
       "GXG_50440                    NaN                     NaN   \n",
       "GXG_50446                    NaN                     NaN   \n",
       "GXG_50447                   12.0                    SMX3   \n",
       "GXG_50454                  398.0                    SMX3   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00321               YAL024C  downstream_gene_variant   \n",
       "GXG_00329                   NaN                      NaN   \n",
       "GXG_00332                   NaN                      NaN   \n",
       "GXG_00339                   NaN                      NaN   \n",
       "GXG_00346                   NaN                      NaN   \n",
       "...                         ...                      ...   \n",
       "GXG_50437                   NaN                      NaN   \n",
       "GXG_50440                   NaN                      NaN   \n",
       "GXG_50446                   NaN                      NaN   \n",
       "GXG_50447               YPR182W    upstream_gene_variant   \n",
       "GXG_50454               YPR182W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance num_of_oligos              assoc_gene  \\\n",
       "var_id                                                                   \n",
       "GXG_00321                   35.0             3  ['YAL024C', 'YAL025C']   \n",
       "GXG_00329                    NaN             4             ['YAL024C']   \n",
       "GXG_00332                    NaN             2             ['YAL024C']   \n",
       "GXG_00339                    NaN             2             ['YAL024C']   \n",
       "GXG_00346                    NaN             3             ['YAL024C']   \n",
       "...                          ...           ...                     ...   \n",
       "GXG_50437                    NaN             2             ['YPR181C']   \n",
       "GXG_50440                    NaN             3             ['YPR181C']   \n",
       "GXG_50446                    NaN             2             ['YPR181C']   \n",
       "GXG_50447                  515.0             4             ['YPR181C']   \n",
       "GXG_50454                  129.0             3  ['YPR182W', 'YPR181C']   \n",
       "\n",
       "          pool_num  \n",
       "var_id              \n",
       "GXG_00321       68  \n",
       "GXG_00329       37  \n",
       "GXG_00332       50  \n",
       "GXG_00339       76  \n",
       "GXG_00346       41  \n",
       "...            ...  \n",
       "GXG_50437       72  \n",
       "GXG_50440       66  \n",
       "GXG_50446       43  \n",
       "GXG_50447       47  \n",
       "GXG_50454       60  \n",
       "\n",
       "[2000 rows x 44 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "35    115\n",
       "36    115\n",
       "37    115\n",
       "38    115\n",
       "39    115\n",
       "40    115\n",
       "41    115\n",
       "42    115\n",
       "43    115\n",
       "44    115\n",
       "45    115\n",
       "46    116\n",
       "47    116\n",
       "48    116\n",
       "49    116\n",
       "50    116\n",
       "51    116\n",
       "52    116\n",
       "53    116\n",
       "54    116\n",
       "55    116\n",
       "56    116\n",
       "57    116\n",
       "58    116\n",
       "59    116\n",
       "60    116\n",
       "61    116\n",
       "62    116\n",
       "63    116\n",
       "64    116\n",
       "65    116\n",
       "66    116\n",
       "67    116\n",
       "68    114\n",
       "69    114\n",
       "70    114\n",
       "71    114\n",
       "72    114\n",
       "73    114\n",
       "74    114\n",
       "75    114\n",
       "76    114\n",
       "77    114\n",
       "Name: num_of_oligos, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 4957\n",
      "Total number of variants: 2000\n"
     ]
    }
   ],
   "source": [
    "gxg_vars_file = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/gxg_variants_final.txt\"\n",
    "gxg_vars = pd.read_csv(gxg_vars_file, sep='\\t', index_col=0)\n",
    "gxg_vars['num_of_oligos'] = gxg_vars['num_of_oligos'].astype(int)\n",
    "\n",
    "# assign gxg variants to pools\n",
    "gxg_vars = sort_variants_to_pools(vars_df = gxg_vars, \n",
    "                                  assignment_order = ['GXG'],\n",
    "                                  num_of_pools_per_group = {'GXG':43},\n",
    "                                  spillover_assignment = {},\n",
    "                                  max_pool_size = 118, first_pool = 35)\n",
    "\n",
    "display(gxg_vars)\n",
    "# check number of oligos in each pool\n",
    "display(gxg_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", gxg_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(gxg_vars))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GXE set\n",
    "Oligos will be grouped according to condition (see assoc_traits_str)\n",
    "- GXE: 94 pools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>ACS</th>\n",
       "      <th>ADE</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADG</th>\n",
       "      <th>ADI</th>\n",
       "      <th>ADR</th>\n",
       "      <th>SACE_MAA</th>\n",
       "      <th>assoc_traits</th>\n",
       "      <th>assoc_traits_str</th>\n",
       "      <th>pool_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00027</th>\n",
       "      <td>I</td>\n",
       "      <td>192429</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1506</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.754000</td>\n",
       "      <td>12|190</td>\n",
       "      <td>4697|3618|173</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00033</th>\n",
       "      <td>I</td>\n",
       "      <td>192466</td>\n",
       "      <td>AAAGGG</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001485</td>\n",
       "      <td>50|148</td>\n",
       "      <td>4735|3656|211</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17166</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375077</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001484</td>\n",
       "      <td>425|4414</td>\n",
       "      <td>4043|2214|2156</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17167</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375078</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>424|4415</td>\n",
       "      <td>4042|2213|2157</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17168</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375080</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>422|4417</td>\n",
       "      <td>4040|2211|2159</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17188</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375331</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>171|4668</td>\n",
       "      <td>3789|1960|162|2410</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17189</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375333</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>168|4670</td>\n",
       "      <td>3786|1957|164|2412</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4384 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT    AC    AN  \\\n",
       "var_id                                                                  \n",
       "GXE_00023     I  192419                C                A     6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT    39  2014   \n",
       "GXE_00026     I  192427                T                C   902  2000   \n",
       "GXE_00027     I  192429                G                C  1506  1998   \n",
       "GXE_00033     I  192466           AAAGGG                A     3  2020   \n",
       "...         ...     ...              ...              ...   ...   ...   \n",
       "GXE_17166   XVI  375077                T                C     3  2022   \n",
       "GXE_17167   XVI  375078                C                A     8  2020   \n",
       "GXE_17168   XVI  375080                G                A    31  2022   \n",
       "GXE_17188   XVI  375331                C                A     8  2020   \n",
       "GXE_17189   XVI  375333                A               AT     8  2022   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "GXE_00027  0.754000                12|190           4697|3618|173   \n",
       "GXE_00033  0.001485                50|148           4735|3656|211   \n",
       "...             ...                   ...                     ...   \n",
       "GXE_17166  0.001484              425|4414          4043|2214|2156   \n",
       "GXE_17167  0.003960              424|4415          4042|2213|2157   \n",
       "GXE_17168  0.015000              422|4417          4040|2211|2159   \n",
       "GXE_17188  0.003960              171|4668      3789|1960|162|2410   \n",
       "GXE_17189  0.003956              168|4670      3786|1957|164|2412   \n",
       "\n",
       "                            region  ...  ACS  ADE  ADF  ADG  ADI  ADR  \\\n",
       "var_id                              ...                                 \n",
       "GXE_00023   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_00024   bidirectional_promoter  ...  0/0  NaN  0/0  0/0  0/0  0/0   \n",
       "GXE_00026   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  1/1  1/1   \n",
       "GXE_00027   bidirectional_promoter  ...  1/1  NaN  1/1  1/1  1/1  1/1   \n",
       "GXE_00033   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "...                            ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "GXE_17166         missense_variant  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17167              stop_gained  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17168         missense_variant  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17188  unidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17189  unidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "\n",
       "          SACE_MAA                                       assoc_traits  \\\n",
       "var_id                                                                  \n",
       "GXE_00023      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "GXE_00024      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "GXE_00026      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "GXE_00027      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "GXE_00033      0/0  ['Cobalt_Chloride;2mM;2', 'Fluconazole;100uM;2...   \n",
       "...            ...                                                ...   \n",
       "GXE_17166      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17167      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17168      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17188      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "GXE_17189      0/0  ['Cobalt_Chloride;2mM;2', 'Fructose;;1', 'Neom...   \n",
       "\n",
       "                                            assoc_traits_str pool_num  \n",
       "var_id                                                                 \n",
       "GXE_00023  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       85  \n",
       "GXE_00024  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       84  \n",
       "GXE_00026  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       85  \n",
       "GXE_00027  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       86  \n",
       "GXE_00033  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...       86  \n",
       "...                                                      ...      ...  \n",
       "GXE_17166  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...      109  \n",
       "GXE_17167  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...      109  \n",
       "GXE_17168  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17188  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "GXE_17189  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...       90  \n",
       "\n",
       "[4384 rows x 61 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "pool_num\n",
       "84     118\n",
       "85     118\n",
       "86     117\n",
       "87     118\n",
       "88     118\n",
       "      ... \n",
       "173    117\n",
       "174    117\n",
       "175    117\n",
       "176    118\n",
       "177    118\n",
       "Name: num_of_oligos, Length: 94, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos: 10983\n",
      "Total number of variants: 4384\n"
     ]
    }
   ],
   "source": [
    "gxe_vars_file = \"/home/users/rang/scratch/yeast/GxE/gxe_variants_final.txt\"\n",
    "gxe_vars = pd.read_csv(gxe_vars_file, sep='\\t', index_col=0)\n",
    "gxe_vars['num_of_oligos'] = gxe_vars['num_of_oligos'].astype(int)\n",
    "\n",
    "# assign gxe variants to pools\n",
    "gxe_vars = sort_variants_to_pools(vars_df = gxe_vars, \n",
    "                                  assignment_order = ['Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1',\n",
    "                                                      'Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2',\n",
    "                                                      'Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2',\n",
    "                                                      'Caffeine;15mM;2_Cobalt_Chloride;2mM;2',\n",
    "                                                      'Caffeine;15mM;2_Neomycin;5mg/mL;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2_Fructose;;1',\n",
    "                                                      'Fluconazole;100uM;2_Fructose;;1',\n",
    "                                                      'Caffeine;15mM;2_Fructose;;1',\n",
    "                                                      'Lithium_Chloride;100mM;2',\n",
    "                                                      'Cobalt_Chloride;2mM;2',\n",
    "                                                      'Fluconazole;100uM;2',\n",
    "                                                      'Neomycin;5mg/mL;2',\n",
    "                                                      'Fructose;;1',\n",
    "                                                      'Caffeine;15mM;2'],\n",
    "                                  \n",
    "                                  grouping_col = 'assoc_traits_str',\n",
    "                                  \n",
    "                                  num_of_pools_per_group = {'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2':3,\n",
    "                                                            'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1':0,\n",
    "                                                            'Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2':3,\n",
    "                                                            'Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2':1,\n",
    "                                                            'Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2':1,\n",
    "                                                            'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2':1,\n",
    "                                                            'Caffeine;15mM;2_Cobalt_Chloride;2mM;2':1,\n",
    "                                                            'Caffeine;15mM;2_Neomycin;5mg/mL;2':0,\n",
    "                                                            'Cobalt_Chloride;2mM;2_Fructose;;1':1,\n",
    "                                                            'Fluconazole;100uM;2_Fructose;;1':0,\n",
    "                                                            'Caffeine;15mM;2_Fructose;;1':0,\n",
    "                                                            'Lithium_Chloride;100mM;2':14,\n",
    "                                                            'Cobalt_Chloride;2mM;2':14,\n",
    "                                                            'Fluconazole;100uM;2':13,\n",
    "                                                            'Neomycin;5mg/mL;2':10,\n",
    "                                                            'Fructose;;1':13,\n",
    "                                                            'Caffeine;15mM;2':19},\n",
    "                                                            \n",
    "                                  spillover_assignment = {'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2':'Neomycin;5mg/mL;2',\n",
    "                                                          'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1':'Cobalt_Chloride;2mM;2',\n",
    "                                                          'Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2':'Neomycin;5mg/mL;2',\n",
    "                                                          'Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2':'Cobalt_Chloride;2mM;2',\n",
    "                                                          'Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2':'Lithium_Chloride;100mM;2',\n",
    "                                                          'Cobalt_Chloride;2mM;2_Fluconazole;100uM;2':'Fluconazole;100uM;2',\n",
    "                                                          'Caffeine;15mM;2_Neomycin;5mg/mL;2':'Neomycin;5mg/mL;2',\n",
    "                                                          'Cobalt_Chloride;2mM;2_Fructose;;1':'Cobalt_Chloride;2mM;2',\n",
    "                                                          'Fluconazole;100uM;2_Fructose;;1':'Fluconazole;100uM;2',\n",
    "                                                          'Caffeine;15mM;2_Fructose;;1':'Caffeine;15mM;2',\n",
    "                                                          'Lithium_Chloride;100mM;2':'Caffeine;15mM;2',\n",
    "                                                          'Fructose;;1':'Caffeine;15mM;2'},\n",
    "                                  max_pool_size = 118, first_pool = 84)\n",
    "\n",
    "display(gxe_vars)\n",
    "# check number of oligos in each pool\n",
    "display(gxe_vars.groupby('pool_num').num_of_oligos.sum())\n",
    "print(\"Total number of oligos:\", gxe_vars.num_of_oligos.sum())\n",
    "print(\"Total number of variants:\", len(gxe_vars))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (crispey-bar)",
   "language": "python",
   "name": "crispey-bar"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
