{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant selection for CRISPEY-BAR library\n",
    "This notebook details the variant selection process for several CRISPEY-BAR projects, including:\n",
    "1. Epistasis\n",
    "2. GxE (QTL pools)\n",
    "3. GxE (ergosterol)\n",
    "4. Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, vcf, re\n",
    "import multiprocessing as mp\n",
    "from Bio import SeqIO\n",
    "from collections import OrderedDict, namedtuple\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_variants(genes_df, vcf_file, chrom_dict, upstream_length=500, downstream_length=500):\n",
    "    '''\n",
    "    given a dataframe of genes of interest, searches vcf for variants in and around that region\n",
    "    can be parallelized by splitting genes and vcf by chromosome\n",
    "    '''\n",
    "    genes_df = genes_df.sort_values('start')\n",
    "    # define start and end points to search depending on gene strand, and the upstream/downstream search parameters\n",
    "    genes_df.loc[genes_df['strand']=='+', 'start_upstream'] = genes_df['start']-upstream_length\n",
    "    genes_df.loc[genes_df['strand']=='+', 'end_downstream'] = genes_df['end']+downstream_length\n",
    "    genes_df.loc[genes_df['strand']=='-', 'start_upstream'] = genes_df['start']-downstream_length\n",
    "    genes_df.loc[genes_df['strand']=='-', 'end_downstream'] = genes_df['end']+upstream_length\n",
    "    genes_df.loc[:,'start_upstream'] = genes_df['start_upstream'].astype(int)\n",
    "    genes_df.loc[:,'end_downstream'] = genes_df['end_downstream'].astype(int)\n",
    "\n",
    "    # read in vcf file to parse for variants\n",
    "    allVars = vcf.Reader(filename=vcf_file)\n",
    "    candVars = []\n",
    "    for i, row in genes_df.iterrows():\n",
    "        retrieve = allVars.fetch(chrom=chrom_dict[row['chrom']], \n",
    "                                 start=row['start_upstream']-1,\n",
    "                                 end=row['end_downstream'])\n",
    "        for rec in retrieve:\n",
    "            rec.samples=[]\n",
    "            rec.FORMAT=None\n",
    "            candVars.append(rec)\n",
    "    candVars = pd.Series(candVars).drop_duplicates().reset_index(drop=True)\n",
    "    return candVars\n",
    "\n",
    "def ngg_check(sequence, threshold):\n",
    "    '''\n",
    "    searches a Bio.seq object for NGG sequence around the center of the sequence\n",
    "    set threshold on size of window to search for NGG (recommended: 10)\n",
    "    returns a boolean of whether one or more NGG found\n",
    "    '''\n",
    "    hits=0\n",
    "    midpt = (len(sequence)-1)//2\n",
    "    for s in [sequence[midpt-2:midpt+threshold], sequence.reverse_complement()[midpt-2:midpt+threshold]]:\n",
    "        matches = re.finditer('.GG', str(s), re.I)\n",
    "        index_list = [m.start() for m in matches]\n",
    "        hits+=len(index_list)\n",
    "    if hits > 0:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def search_genome(sequence, genome):\n",
    "    '''\n",
    "    searches a genome (stored as dictionary of SeqRecords) for a Bio.Seq sequence (exact string match)\n",
    "    returns number of occurences of the sequence in the genome\n",
    "    '''\n",
    "    hits = 0\n",
    "    for scaffold in genome.values():\n",
    "        # search both sequence and its reverse complement to find hits on forward and reverse strands\n",
    "        for s in [sequence, sequence.reverse_complement()]:\n",
    "            matches = re.finditer(str(s), str(scaffold.seq), re.I)\n",
    "            index_list = [m.start() for m in matches]\n",
    "            hits+=len(index_list)\n",
    "    return hits\n",
    "\n",
    "def filter_targetable_variants(candVars, genomes_list, threshold, window):\n",
    "    '''\n",
    "    Takes a list of VCF records, a list of genomes, a threshold for searching NGGs, and a window size for \n",
    "    flanking sequences, filters and sorts records into two lists, whether they are targetable in all genomes\n",
    "    or just in the reference genome.\n",
    "    '''\n",
    "    ref_gen = genomes_list[0]\n",
    "    candVars_allStrains=[]\n",
    "    candVars_refOnly=[]\n",
    "    \n",
    "    for record in candVars:\n",
    "        edits_in_ref_strain = False\n",
    "        edits_in_nonref_strain = False\n",
    "        ##### FILTERING #####\n",
    "        # 1. Skip variant if near chromosome edge\n",
    "        seq_window = ref_gen[record.CHROM][record.POS-window//2-1:record.POS+window//2]\n",
    "        if len(seq_window) != window:\n",
    "            continue\n",
    "\n",
    "        # 2. is there an 'NGG' sequence around the center of the seq_window?\n",
    "        if not ngg_check(sequence=seq_window.seq, threshold=threshold):\n",
    "            continue\n",
    "\n",
    "        # 3. check genomes if sequence flanking variant (seq_window) is unique\n",
    "        if search_genome(seq_window.seq, ref_gen)==1: # reference genome\n",
    "            edits_in_ref_strain = True\n",
    "            # check remaining genomes to ensure seq_window is unique hit in all of them\n",
    "            edits_in_nonref_strain = True\n",
    "            for gen in genomes_list[1:]:\n",
    "                if search_genome(seq_window.seq, gen)!=1:\n",
    "                    edits_in_nonref_strain=False\n",
    "                    break\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        # 4. store variant info according to whether it can be edited in all strains, or just reference strain\n",
    "        if edits_in_ref_strain:\n",
    "            if edits_in_nonref_strain:\n",
    "                candVars_allStrains.append(record)\n",
    "            else:\n",
    "                candVars_refOnly.append(record)\n",
    "        else:\n",
    "            print(\"Variant skipped: Cannot be edited in reference strain\") # sanity check step: should not appear in output as filters in previous steps should remove all such instances\n",
    "    \n",
    "    return (candVars_allStrains, candVars_refOnly)\n",
    "\n",
    "def find_targetable_variants(candVars, genomes_list, threshold, window):\n",
    "    \"\"\"\n",
    "    UPGRADED FROM filter_targetable_variants\n",
    "    Takes a list of VCF records, a list of genomes (at least 1), a threshold for searching NGGs, and a window size for search sequence.\n",
    "    Filters records for NGG availability and targetability in each genome.\n",
    "    Adds a list of integers (0/1) under \"TGT\" field in VCF INFO section. Number of digits corresponds to genomes provided\n",
    "    For each genome, 1 indicates unique match for sequence is found, otherwise 0.\n",
    "    NOTE: REMEMBER TO ADD \"TGT\" INFO TO VCF HEADER BEFORE WRITING RECORDS TO FILE\n",
    "    header_info = collections.namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "    header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "    \"\"\"\n",
    "    ref_gen = genomes_list[0]\n",
    "\n",
    "    record_list = []\n",
    "    for record in candVars:\n",
    "        tgt_code = [0]*len(genomes_list)\n",
    "        ##### FILTERING #####\n",
    "        # 1. Skip variant if near chromosome edge\n",
    "        seq_window = ref_gen[record.CHROM][record.POS-window//2-1:record.POS+window//2]\n",
    "        if len(seq_window) != window:\n",
    "            continue\n",
    "\n",
    "        # 2. is there an 'NGG' sequence around the center of the seq_window?\n",
    "        if not ngg_check(sequence=seq_window.seq, threshold=threshold):\n",
    "            continue\n",
    "\n",
    "        # 3. check sequence flanking variant (seq_window) for unique match in genome\n",
    "        for i in range(len(genomes_list)):\n",
    "            gen = genomes_list[i]\n",
    "            num_of_matches = search_genome(seq_window.seq, gen)\n",
    "            if num_of_matches==1: # unique match\n",
    "                tgt_code[i]=1\n",
    "        # sequence window must match at least one genome in order to be written\n",
    "        if sum(tgt_code)==0:\n",
    "#             print(\"Variant at chrom {}, pos {} skipped: No unique matches found in tested genome(s)\".format(record.CHROM, record.POS))\n",
    "            continue\n",
    "        # 4. add TGT code to variant record, add to record list\n",
    "        else:\n",
    "            record.INFO['TGT'] = tgt_code\n",
    "            record_list.append(record)\n",
    "        \n",
    "    return record_list\n",
    "\n",
    "##### FUNCTIONS TO ANNOTATE VARIANTS, BASED ON EILON'S OLIGO DESIGN PIPELINE #####\n",
    "def get_genes_len(input_gff_filename):\n",
    "    with open(input_gff_filename) as f:\n",
    "        gff_lines = f.readlines()\n",
    "    gff_lines = [l for l in gff_lines if len(l.split('\\t'))>=5 ]    \n",
    "    gff_lines = [l for l in gff_lines if l.split('\\t')[2] == 'gene' ]\n",
    "    genes_ids = [l.split('ID=')[1].split(';Name=')[0] for l in gff_lines]\n",
    "    genes_lens = [ int(l.split('\\t')[4]) - int(l.split('\\t')[3]) + 1 for l in gff_lines]\n",
    "\n",
    "    gene_le_df = pd.DataFrame({'Gene' : genes_ids, 'len_bp' : genes_lens})\n",
    "    gene_le_df.set_index(keys='Gene', drop=False, inplace=True)\n",
    "    gene_le_df.index.name=''\n",
    "    gene_le_df.drop_duplicates(inplace=True)\n",
    "\n",
    "    return(gene_le_df)\n",
    "\n",
    "def get_cdss_len(input_gff_filename):\n",
    "    with open(input_gff_filename) as f:\n",
    "        gff_lines = f.readlines()\n",
    "    gff_lines = [l for l in gff_lines if len(l.split('\\t'))>=5 ]\n",
    "    gff_lines = [l for l in gff_lines if l.split('\\t')[2] == 'CDS' ]\n",
    "    genes_ids = [l.split('Parent=')[1].split(';Name=')[0] for l in gff_lines]\n",
    "    genes_lens = [ int(l.split('\\t')[4]) - int(l.split('\\t')[3]) + 1 for l in gff_lines]\n",
    "\n",
    "    gene_le_df = pd.DataFrame({'Gene' : genes_ids, 'len_bp' : genes_lens})\n",
    "    gene_le_df.set_index(keys='Gene', drop=False, inplace=True)\n",
    "    gene_le_df.index.name=''\n",
    "    gene_le_df = gene_le_df.groupby('Gene').sum()\n",
    "    gene_le_df['Gene'] = gene_le_df.index\n",
    "    gene_le_df.index.name=''\n",
    "    gene_le_df.drop_duplicates(inplace=True)\n",
    "\n",
    "    return(gene_le_df)\n",
    "\n",
    "def get_dubious_genes(input_gff_filename):\n",
    "    with open(input_gff_filename) as f:\n",
    "        gff_lines = f.readlines()\n",
    "    gff_lines = [l for l in gff_lines if len(l.split('\\t'))>=5 ]\n",
    "    gff_lines = [l for l in gff_lines if (l.split('\\t')[2] == 'gene') & (len(re.findall('dubious', l, flags=re.I))>0)]\n",
    "    genes_ids = [l.split('ID=')[1].split(';Name=')[0] for l in gff_lines]\n",
    "    \n",
    "    return genes_ids\n",
    "\n",
    "def annotate_variants_by_VEPoutput(vep_input_vcf_filename, vep_output_filename, input_gff_filename, annotated_res_output_filename, id_colname = 'var_id'):\n",
    "    cdss_len_df = get_cdss_len(input_gff_filename)\n",
    "    genes_len_df = get_genes_len(input_gff_filename)\n",
    "    dubious_genes = get_dubious_genes(input_gff_filename)\n",
    "    \n",
    "    vep_df = pd.read_table(vep_output_filename, sep='\\t', na_values = \"-\", low_memory=False)\n",
    "    vep_df = vep_df.loc[~vep_df.Gene.isin(dubious_genes),]\n",
    "    vep_df.rename(columns={\"#Uploaded_variation\" : id_colname}, inplace=True)\n",
    "    \n",
    "    # prepare annotated variants df using VCF supplied to VEP\n",
    "    variants_annotated_df = pd.Series([rec for rec in vcf.Reader(filename=vep_input_vcf_filename)])\n",
    "    var_id_list = []\n",
    "    assemble = []\n",
    "    columns = ['CHROM', 'POS', 'REF', 'ALT', 'AC', 'AN' ,'AF']\n",
    "    for rec in variants_annotated_df:\n",
    "        var_id_list.append(rec.ID)\n",
    "        assemble.append((rec.CHROM, rec.POS, rec.REF, rec.ALT[0], rec.INFO['AC'][0], rec.INFO['AN'], rec.INFO['AF'][0]))\n",
    "    variants_annotated_df = pd.DataFrame.from_records(assemble, index=var_id_list, columns=columns)\n",
    "    variants_annotated_df.index.name = id_colname\n",
    "    \n",
    "    # parse VEP output text file by ID for annotations\n",
    "    entry_cnt = 0\n",
    "    for cur_id, anno in vep_df.groupby(id_colname):\n",
    "        entry_cnt +=1\n",
    "        if (entry_cnt % 5000 == 0):\n",
    "            print(\"parsing entry: %d\"% (entry_cnt))\n",
    "        \n",
    "        # all up/down stream genes (VEP default search range is 5kb from variant position)\n",
    "        cur_upstream_all_str =  '|'.join([str(int(d)) for d in anno.DISTANCE[anno.Consequence.isin(['upstream_gene_variant' ])].values if not np.isnan(d)])\n",
    "        cur_downstream_all_str =  '|'.join([str(int(d)) for d in anno.DISTANCE[anno.Consequence.isin(['downstream_gene_variant' ])].values if not np.isnan(d)])\n",
    "        \n",
    "        variants_annotated_df.loc[cur_id, 'upstream_distance_str'] = cur_upstream_all_str\n",
    "        variants_annotated_df.loc[cur_id, 'downstream_distance_str'] = cur_downstream_all_str\n",
    "\n",
    "        is_up_down_stream = anno.Consequence.isin(['upstream_gene_variant' ,'downstream_gene_variant'])\n",
    "        \n",
    "        # identify intergenic variants\n",
    "        if np.all(is_up_down_stream):\n",
    "            left_genes = ( ((anno.Consequence =='upstream_gene_variant') & (anno.STRAND<0)) |\n",
    "                          ((anno.Consequence =='downstream_gene_variant') & (anno.STRAND>0)) )\n",
    "\n",
    "            right_genes = ( ((anno.Consequence =='upstream_gene_variant') & (anno.STRAND>0)) |\n",
    "                              ((anno.Consequence =='downstream_gene_variant') & (anno.STRAND<0)) )\n",
    "\n",
    "            if (left_genes.sum() > 0):\n",
    "                ind_of_closest_left = anno.loc[left_genes,'DISTANCE'].idxmin()\n",
    "                \n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene1_Gene_Name'] = anno['SYMBOL'][ind_of_closest_left]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene1_Gene_ID'] = anno['Gene'][ind_of_closest_left]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene1_Annotation'] = anno['Consequence'][ind_of_closest_left]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene1_Distance'] = anno['DISTANCE'][ind_of_closest_left]\n",
    "\n",
    "            if (right_genes.sum() > 0):\n",
    "                ind_of_closest_right = anno.loc[right_genes,'DISTANCE'].idxmin()\n",
    "                \n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene2_Gene_Name'] = anno['SYMBOL'][ind_of_closest_right]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene2_Gene_ID'] = anno['Gene'][ind_of_closest_right]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene2_Annotation'] = anno['Consequence'][ind_of_closest_right]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene2_Distance'] = anno['DISTANCE'][ind_of_closest_right]\n",
    "\n",
    "            # determine type of noncoding variant\n",
    "            # Caution: variants without genes on one side will be labelled as \"intergenic\"\n",
    "            cur_region = 'intergenic'\n",
    "            if ((left_genes.sum() > 0) and (right_genes.sum() > 0)): \n",
    "                if (anno['Consequence'][ind_of_closest_left] == 'upstream_gene_variant'):\n",
    "                    if (anno['Consequence'][ind_of_closest_right] == 'upstream_gene_variant'):\n",
    "                        cur_region = 'bidirectional_promoter'\n",
    "                    else:\n",
    "                        cur_region = 'unidirectional_promoter'\n",
    "                else:\n",
    "                    if (anno['Consequence'][ind_of_closest_right] == 'upstream_gene_variant'):\n",
    "                        cur_region = 'unidirectional_promoter'\n",
    "\n",
    "            variants_annotated_df.loc[cur_id, 'region'] = cur_region\n",
    "\n",
    "            # index of closest gene\n",
    "            sel_row = anno.DISTANCE.idxmin()\n",
    "        \n",
    "        # for variants that land in a gene, select annotation with highest impact\n",
    "        else:\n",
    "            anno = anno.loc[~is_up_down_stream,:]\n",
    "            if (anno.IMPACT == 'HIGH').sum()>0:\n",
    "                anno = anno.loc[anno.IMPACT == 'HIGH',:]\n",
    "            elif (anno.IMPACT == 'MODERATE').sum()>0:\n",
    "                anno = anno.loc[anno.IMPACT == 'MODERATE',:]\n",
    "            elif (anno.IMPACT == 'LOW').sum()>0:\n",
    "                anno = anno.loc[anno.IMPACT == 'LOW',:]\n",
    "            elif (anno.IMPACT == 'MODIFIER').sum()>0:\n",
    "                anno = anno.loc[anno.IMPACT == 'MODIFIER',:] #NOTE: includes introns, noncoding transcripts within length of gene\n",
    "            \n",
    "            else:\n",
    "                raise ValueError(anno)\n",
    "\n",
    "            sel_row = anno.index[0]\n",
    "            variants_annotated_df.loc[cur_id, 'region'] = anno['Consequence'][sel_row]\n",
    "\n",
    "        # closest annotation recorded\n",
    "        variants_annotated_df.loc[cur_id, 'Annotation'] = anno['Consequence'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Annotation_Impact'] = anno['IMPACT'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Gene_Name'] = anno['SYMBOL'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Gene_ID'] = anno['Gene'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Transcript_BioType'] = anno['BIOTYPE'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Existing_variation'] = anno['Existing_variation'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'HGVSc'] = anno['HGVSc'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'HGVSp'] = anno['HGVSp'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'SWISSPROT'] = anno['SWISSPROT'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'UNIPARC'] = anno['UNIPARC'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'BLOSUM62'] = anno['BLOSUM62'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'HGVSp'] = anno['HGVSp'][sel_row]\n",
    "#         variants_annotated_df.set_value(index=cur_id,col='VEP_CSN',value=anno['CSN'][sel_row])\n",
    "        \n",
    "        # store cDNA and CDS positions\n",
    "        cdna_pos = anno['cDNA_position'][sel_row]\n",
    "        cds_pos = anno['CDS_position'][sel_row]\n",
    "    \n",
    "        # add additional info for missense and synonymous variants\n",
    "        if (anno['Consequence'][sel_row] in ['synonymous_variant','missense_variant']):\n",
    "            # because variants were extracted from 1011 genomes GVCF, some SNPs have trailing bases.\n",
    "            # Calculate the correct cDNA and CDS positions if so \n",
    "            if '-' in cdna_pos:\n",
    "                cdna_pos = cdna_pos.split('-')[0]\n",
    "            if '-' in cds_pos:\n",
    "                cds_pos = cds_pos.split('-')[0]\n",
    "                \n",
    "            \n",
    "            if anno['Gene'][sel_row] in genes_len_df.index:\n",
    "                cur_gene_len = int(genes_len_df.loc[anno['Gene'][sel_row],'len_bp'])\n",
    "                variants_annotated_df.loc[cur_id, 'cDNA_len'] = cur_gene_len\n",
    "                variants_annotated_df.loc[cur_id, 'cDNA_frac'] = int(cdna_pos)/cur_gene_len\n",
    "            \n",
    "            if anno['Gene'][sel_row] in cdss_len_df.index:\n",
    "                cur_cds_len = int(cdss_len_df.loc[anno['Gene'][sel_row],'len_bp'])\n",
    "                variants_annotated_df.loc[cur_id, 'CDS_len'] = cur_cds_len\n",
    "                variants_annotated_df.loc[cur_id, 'CDS_frac'] = int(cds_pos)/cur_cds_len\n",
    "                variants_annotated_df.loc[cur_id, 'AA_len'] = cur_cds_len/3\n",
    "                variants_annotated_df.loc[cur_id, 'AA_frac'] = int(cds_pos)/cur_cds_len\n",
    "                \n",
    "            variants_annotated_df.loc[cur_id, 'cDNA_pos'] = cdna_pos\n",
    "            variants_annotated_df.loc[cur_id, 'CDS_pos'] = cds_pos\n",
    "            variants_annotated_df.loc[cur_id, 'AA_pos'] = anno['Protein_position'][sel_row]\n",
    "            variants_annotated_df.loc[cur_id, 'Amino_acids'] = anno['Amino_acids'][sel_row]\n",
    "            variants_annotated_df.loc[cur_id, 'Codons'] = anno['Codons'][sel_row]\n",
    "            variants_annotated_df.loc[cur_id, 'DOMAINS'] = anno['DOMAINS'][sel_row]\n",
    "\n",
    "    variants_annotated_df.to_csv(annotated_res_output_filename, sep='\\t', header=True, index=True)\n",
    "    \n",
    "    return(variants_annotated_df)\n",
    "\n",
    "###############################################################################################################\n",
    "def find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size=121):\n",
    "    \"\"\"\n",
    "    helper function to find number of pools required to fit a certain number of oligos,\n",
    "    as well as the average size of the pool\n",
    "    \"\"\"\n",
    "    print('Total number of oligos to fit:', num_of_oligos)\n",
    "\n",
    "    # num of pools to use\n",
    "    num_of_pools_lower = num_of_oligos//max_pool_size\n",
    "    num_of_pools_upper = num_of_oligos//min_pool_size\n",
    "    display((num_of_pools_lower, num_of_pools_upper))\n",
    "\n",
    "    # explore pooling options for all possible pool configurations\n",
    "    for num_of_pools in range(num_of_pools_lower, num_of_pools_upper+1):\n",
    "        if num_of_pools == 0:\n",
    "            continue\n",
    "        print(\"Use {} pools:\".format(num_of_pools))\n",
    "        oligos_per_pool = min((num_of_oligos // num_of_pools), max_pool_size)\n",
    "        print(\"Number of oligos per pool:\", oligos_per_pool)\n",
    "        print(\"Number of oligos leftover:\", num_of_oligos - num_of_pools*oligos_per_pool)\n",
    "        print(\"Number of technical oligos:\", (complete_pool_size - oligos_per_pool) * num_of_pools)\n",
    "        print()\n",
    "    \n",
    "    # return list of possible pool sizes\n",
    "    return np.arange(num_of_pools_lower, num_of_pools_upper+1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count num. of genetic interactions per gene in Costanzo et al. (2016) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/\"\n",
    "# list of files to parse for genetic interactions\n",
    "file_list = ['SGA_ExE.txt', 'SGA_ExN_NxE.txt', 'SGA_NxN.txt']\n",
    "# summary of genetic interactions output file\n",
    "out_file = working_dir+'genetic_interactions_summary_by_gene.txt'\n",
    "os.chdir(working_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(890107, 17)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_strain</th>\n",
       "      <th>query_allele</th>\n",
       "      <th>array_strain</th>\n",
       "      <th>array_allele</th>\n",
       "      <th>arraytype</th>\n",
       "      <th>genetic_interaction_score</th>\n",
       "      <th>pval</th>\n",
       "      <th>query_smf</th>\n",
       "      <th>array_smf</th>\n",
       "      <th>dmf</th>\n",
       "      <th>dmf_std</th>\n",
       "      <th>genetic_interaction_score_abs</th>\n",
       "      <th>query_gene_id</th>\n",
       "      <th>query_gene_name</th>\n",
       "      <th>array_gene_id</th>\n",
       "      <th>array_gene_name</th>\n",
       "      <th>gene_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL026W_tsa1065</td>\n",
       "      <td>lsm2-5001</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.3529</td>\n",
       "      <td>3.591000e-06</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9408</td>\n",
       "      <td>0.4266</td>\n",
       "      <td>0.0790</td>\n",
       "      <td>0.3529</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL026W</td>\n",
       "      <td>lsm2</td>\n",
       "      <td>YAL001C_YBL026W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL034C_tsa950</td>\n",
       "      <td>stu1-7</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.1294</td>\n",
       "      <td>1.931000e-02</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.6690</td>\n",
       "      <td>0.4249</td>\n",
       "      <td>0.0482</td>\n",
       "      <td>0.1294</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL034C</td>\n",
       "      <td>stu1</td>\n",
       "      <td>YAL001C_YBL034C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL097W_tsa510</td>\n",
       "      <td>brn1-9</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.0808</td>\n",
       "      <td>5.582000e-15</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.5464</td>\n",
       "      <td>0.3719</td>\n",
       "      <td>0.0077</td>\n",
       "      <td>0.0808</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL097W</td>\n",
       "      <td>brn1</td>\n",
       "      <td>YAL001C_YBL097W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBR029C_tsa1063</td>\n",
       "      <td>cds1-5001</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.1173</td>\n",
       "      <td>8.243000e-05</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9007</td>\n",
       "      <td>0.6289</td>\n",
       "      <td>0.0226</td>\n",
       "      <td>0.1173</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBR029C</td>\n",
       "      <td>cds1</td>\n",
       "      <td>YAL001C_YBR029C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBR060C_tsa311</td>\n",
       "      <td>orc2-2</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>0.2516</td>\n",
       "      <td>4.729000e-12</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.7384</td>\n",
       "      <td>0.8634</td>\n",
       "      <td>0.0362</td>\n",
       "      <td>0.2516</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBR060C</td>\n",
       "      <td>orc2</td>\n",
       "      <td>YAL001C_YBR060C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      query_strain query_allele     array_strain array_allele arraytype  \\\n",
       "1   YAL001C_tsq508   tfc3-g349e  YBL026W_tsa1065    lsm2-5001     TSA30   \n",
       "6   YAL001C_tsq508   tfc3-g349e   YBL034C_tsa950       stu1-7     TSA30   \n",
       "15  YAL001C_tsq508   tfc3-g349e   YBL097W_tsa510       brn1-9     TSA30   \n",
       "24  YAL001C_tsq508   tfc3-g349e  YBR029C_tsa1063    cds1-5001     TSA30   \n",
       "30  YAL001C_tsq508   tfc3-g349e   YBR060C_tsa311       orc2-2     TSA30   \n",
       "\n",
       "    genetic_interaction_score          pval  query_smf  array_smf     dmf  \\\n",
       "1                     -0.3529  3.591000e-06     0.8285     0.9408  0.4266   \n",
       "6                     -0.1294  1.931000e-02     0.8285     0.6690  0.4249   \n",
       "15                    -0.0808  5.582000e-15     0.8285     0.5464  0.3719   \n",
       "24                    -0.1173  8.243000e-05     0.8285     0.9007  0.6289   \n",
       "30                     0.2516  4.729000e-12     0.8285     0.7384  0.8634   \n",
       "\n",
       "    dmf_std  genetic_interaction_score_abs query_gene_id query_gene_name  \\\n",
       "1    0.0790                         0.3529       YAL001C            tfc3   \n",
       "6    0.0482                         0.1294       YAL001C            tfc3   \n",
       "15   0.0077                         0.0808       YAL001C            tfc3   \n",
       "24   0.0226                         0.1173       YAL001C            tfc3   \n",
       "30   0.0362                         0.2516       YAL001C            tfc3   \n",
       "\n",
       "   array_gene_id array_gene_name        gene_pair  \n",
       "1        YBL026W            lsm2  YAL001C_YBL026W  \n",
       "6        YBL034C            stu1  YAL001C_YBL034C  \n",
       "15       YBL097W            brn1  YAL001C_YBL097W  \n",
       "24       YBR029C            cds1  YAL001C_YBR029C  \n",
       "30       YBR060C            orc2  YAL001C_YBR060C  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = []\n",
    "for file in file_list:\n",
    "    df.append(pd.read_csv(working_dir+file, sep='\\t', header = 0))\n",
    "df = pd.concat(df, axis=0, join='inner')\n",
    "df.columns = ['query_strain', 'query_allele', 'array_strain', 'array_allele', 'arraytype', 'genetic_interaction_score', 'pval', 'query_smf', 'array_smf', 'dmf', 'dmf_std']\n",
    "df['genetic_interaction_score_abs'] = df['genetic_interaction_score'].abs()\n",
    "df['query_gene_id'] = [x.split('_')[0] for x in df['query_strain']]\n",
    "df['query_gene_name'] = [x.split('-')[0] for x in df['query_allele']]\n",
    "df['array_gene_id'] = [x.split('_')[0] for x in df['array_strain']]\n",
    "df['array_gene_name'] = [x.split('-')[0] for x in df['array_allele']]\n",
    "df['gene_pair'] = df.apply(lambda x: '_'.join(sorted([x['query_gene_id'], x['array_gene_id']])), axis=1)\n",
    "\n",
    "# get list of genes\n",
    "all_genes = set(df.query_gene_id).union(set(df.array_gene_id))\n",
    "\n",
    "# filter down to significant interactions only\n",
    "pval_threshold=0.05\n",
    "interaction_threshold=0.08\n",
    "df = df.query('genetic_interaction_score_abs>@interaction_threshold & pval<@pval_threshold')\n",
    "display(df.shape)\n",
    "display(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(780653, 17)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_strain</th>\n",
       "      <th>query_allele</th>\n",
       "      <th>array_strain</th>\n",
       "      <th>array_allele</th>\n",
       "      <th>arraytype</th>\n",
       "      <th>genetic_interaction_score</th>\n",
       "      <th>pval</th>\n",
       "      <th>query_smf</th>\n",
       "      <th>array_smf</th>\n",
       "      <th>dmf</th>\n",
       "      <th>dmf_std</th>\n",
       "      <th>genetic_interaction_score_abs</th>\n",
       "      <th>query_gene_id</th>\n",
       "      <th>query_gene_name</th>\n",
       "      <th>array_gene_id</th>\n",
       "      <th>array_gene_name</th>\n",
       "      <th>gene_pair</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_pair</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBL010C</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL010C_dma88</td>\n",
       "      <td>ybl010c</td>\n",
       "      <td>DMA30</td>\n",
       "      <td>-0.0818</td>\n",
       "      <td>3.430000e-02</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9870</td>\n",
       "      <td>0.7360</td>\n",
       "      <td>0.0361</td>\n",
       "      <td>0.0818</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL010C</td>\n",
       "      <td>ybl010c</td>\n",
       "      <td>YAL001C_YBL010C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBL026W</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL026W_tsa1065</td>\n",
       "      <td>lsm2-5001</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.3529</td>\n",
       "      <td>3.591000e-06</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9408</td>\n",
       "      <td>0.4266</td>\n",
       "      <td>0.0790</td>\n",
       "      <td>0.3529</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL026W</td>\n",
       "      <td>lsm2</td>\n",
       "      <td>YAL001C_YBL026W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBL034C</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL034C_tsa950</td>\n",
       "      <td>stu1-7</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.1294</td>\n",
       "      <td>1.931000e-02</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.6690</td>\n",
       "      <td>0.4249</td>\n",
       "      <td>0.0482</td>\n",
       "      <td>0.1294</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL034C</td>\n",
       "      <td>stu1</td>\n",
       "      <td>YAL001C_YBL034C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBL097W</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL097W_tsa510</td>\n",
       "      <td>brn1-9</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.0808</td>\n",
       "      <td>5.582000e-15</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.5464</td>\n",
       "      <td>0.3719</td>\n",
       "      <td>0.0077</td>\n",
       "      <td>0.0808</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL097W</td>\n",
       "      <td>brn1</td>\n",
       "      <td>YAL001C_YBL097W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBR029C</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBR029C_tsa1063</td>\n",
       "      <td>cds1-5001</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.1173</td>\n",
       "      <td>8.243000e-05</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9007</td>\n",
       "      <td>0.6289</td>\n",
       "      <td>0.0226</td>\n",
       "      <td>0.1173</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBR029C</td>\n",
       "      <td>cds1</td>\n",
       "      <td>YAL001C_YBR029C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   query_strain query_allele     array_strain array_allele  \\\n",
       "gene_pair                                                                    \n",
       "YAL001C_YBL010C  YAL001C_tsq508   tfc3-g349e    YBL010C_dma88      ybl010c   \n",
       "YAL001C_YBL026W  YAL001C_tsq508   tfc3-g349e  YBL026W_tsa1065    lsm2-5001   \n",
       "YAL001C_YBL034C  YAL001C_tsq508   tfc3-g349e   YBL034C_tsa950       stu1-7   \n",
       "YAL001C_YBL097W  YAL001C_tsq508   tfc3-g349e   YBL097W_tsa510       brn1-9   \n",
       "YAL001C_YBR029C  YAL001C_tsq508   tfc3-g349e  YBR029C_tsa1063    cds1-5001   \n",
       "\n",
       "                arraytype  genetic_interaction_score          pval  query_smf  \\\n",
       "gene_pair                                                                       \n",
       "YAL001C_YBL010C     DMA30                    -0.0818  3.430000e-02     0.8285   \n",
       "YAL001C_YBL026W     TSA30                    -0.3529  3.591000e-06     0.8285   \n",
       "YAL001C_YBL034C     TSA30                    -0.1294  1.931000e-02     0.8285   \n",
       "YAL001C_YBL097W     TSA30                    -0.0808  5.582000e-15     0.8285   \n",
       "YAL001C_YBR029C     TSA30                    -0.1173  8.243000e-05     0.8285   \n",
       "\n",
       "                 array_smf     dmf  dmf_std  genetic_interaction_score_abs  \\\n",
       "gene_pair                                                                    \n",
       "YAL001C_YBL010C     0.9870  0.7360   0.0361                         0.0818   \n",
       "YAL001C_YBL026W     0.9408  0.4266   0.0790                         0.3529   \n",
       "YAL001C_YBL034C     0.6690  0.4249   0.0482                         0.1294   \n",
       "YAL001C_YBL097W     0.5464  0.3719   0.0077                         0.0808   \n",
       "YAL001C_YBR029C     0.9007  0.6289   0.0226                         0.1173   \n",
       "\n",
       "                query_gene_id query_gene_name array_gene_id array_gene_name  \\\n",
       "gene_pair                                                                     \n",
       "YAL001C_YBL010C       YAL001C            tfc3       YBL010C         ybl010c   \n",
       "YAL001C_YBL026W       YAL001C            tfc3       YBL026W            lsm2   \n",
       "YAL001C_YBL034C       YAL001C            tfc3       YBL034C            stu1   \n",
       "YAL001C_YBL097W       YAL001C            tfc3       YBL097W            brn1   \n",
       "YAL001C_YBR029C       YAL001C            tfc3       YBR029C            cds1   \n",
       "\n",
       "                       gene_pair  \n",
       "gene_pair                         \n",
       "YAL001C_YBL010C  YAL001C_YBL010C  \n",
       "YAL001C_YBL026W  YAL001C_YBL026W  \n",
       "YAL001C_YBL034C  YAL001C_YBL034C  \n",
       "YAL001C_YBL097W  YAL001C_YBL097W  \n",
       "YAL001C_YBR029C  YAL001C_YBR029C  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# find strongest interaction in each pair of genes\n",
    "gene_pairs_df = df.groupby('gene_pair').apply(lambda x: x.sort_values('genetic_interaction_score_abs', ascending=False).iloc[0])\n",
    "display(gene_pairs_df.shape)\n",
    "display(gene_pairs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5707, 2)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFL039C</th>\n",
       "      <td>1324</td>\n",
       "      <td>2220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL034C-B</th>\n",
       "      <td>1150</td>\n",
       "      <td>1897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJR076C</th>\n",
       "      <td>749</td>\n",
       "      <td>1342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBL105C</th>\n",
       "      <td>841</td>\n",
       "      <td>1139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YER157W</th>\n",
       "      <td>925</td>\n",
       "      <td>967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           positive_interactions  negative_interactions\n",
       "gene_id                                                \n",
       "YFL039C                     1324                   2220\n",
       "YFL034C-B                   1150                   1897\n",
       "YJR076C                      749                   1342\n",
       "YBL105C                      841                   1139\n",
       "YER157W                      925                    967"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# for each gene, record number of positive and negative interactions that meet interaction and pvalue thresholds\n",
    "gene_list = []\n",
    "positive_interactions = []\n",
    "negative_interactions = []\n",
    "for gene in all_genes:\n",
    "    sub = gene_pairs_df.query('query_gene_id==@gene | array_gene_id==@gene')\n",
    "    gene_list.append(gene)\n",
    "    positive_interactions.append(len(sub[sub['genetic_interaction_score']>interaction_threshold]))\n",
    "    negative_interactions.append(len(sub[sub['genetic_interaction_score']<interaction_threshold*-1]))\n",
    "\n",
    "num_of_interactions_df = pd.DataFrame({'positive_interactions':positive_interactions, 'negative_interactions':negative_interactions}, index=gene_list)\n",
    "num_of_interactions_df = num_of_interactions_df.loc[num_of_interactions_df.sum(axis=1).sort_values(ascending=False).index]\n",
    "num_of_interactions_df.index.name = 'gene_id'\n",
    "display(num_of_interactions_df.shape)\n",
    "display(num_of_interactions_df.head())\n",
    "\n",
    "num_of_interactions_df.to_csv(out_file, sep='\\t', header=True, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read S288C GFF file and extract gene information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info_df = {}\n",
    "gff_file = '/home/users/rang/yeast/genomes/saccharomyces_cerevisiae_R64-1-1_20110208.gff'\n",
    "char_map = {'%20' : ' ',\n",
    "            '%2C' : ',',\n",
    "            '%2F' : '/',\n",
    "            '%3B' : ';'}\n",
    "\n",
    "with open(gff_file, 'r') as gff:\n",
    "    line = gff.readline()\n",
    "    while line.rstrip() != '###':\n",
    "        while line[0] == \"#\":\n",
    "            line = gff.readline()\n",
    "            \n",
    "        chrom, _, feature, start, end, _, strand, _, info = line.rstrip().split('\\t')\n",
    "        if feature=='gene':\n",
    "            info = {x.split('=')[0] : x.split('=')[1] for x in info.split(';')}\n",
    "            gene_id = info['ID']\n",
    "            if 'gene' in info:\n",
    "                gene_name = info['gene']\n",
    "            else:\n",
    "                gene_name = np.nan\n",
    "                \n",
    "            if 'Note' in info:\n",
    "                desc = info['Note']\n",
    "                for c, r in char_map.items():\n",
    "                    desc = desc.replace(c, r)\n",
    "            else:\n",
    "                desc = np.nan\n",
    "            # store gene info\n",
    "            gene_info_df[gene_id] = [chrom, int(start), int(end), strand, gene_name, desc]\n",
    "        line = gff.readline()\n",
    "        \n",
    "gene_info_df = pd.DataFrame.from_dict(gene_info_df, orient='index', columns=['chrom', 'start', 'end', 'strand', 'gene_name', 'description'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFL039C</th>\n",
       "      <td>1324</td>\n",
       "      <td>2220</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>53260.0</td>\n",
       "      <td>54696.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ACT1</td>\n",
       "      <td>Actin, structural protein involved in cell pol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL034C-B</th>\n",
       "      <td>1150</td>\n",
       "      <td>1897</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>63016.0</td>\n",
       "      <td>63993.0</td>\n",
       "      <td>-</td>\n",
       "      <td>MOB2</td>\n",
       "      <td>Component of the RAM signaling network, that a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJR076C</th>\n",
       "      <td>749</td>\n",
       "      <td>1342</td>\n",
       "      <td>chrX</td>\n",
       "      <td>575354.0</td>\n",
       "      <td>576601.0</td>\n",
       "      <td>-</td>\n",
       "      <td>CDC11</td>\n",
       "      <td>Component of the septin ring of the mother-bud...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBL105C</th>\n",
       "      <td>841</td>\n",
       "      <td>1139</td>\n",
       "      <td>chrII</td>\n",
       "      <td>14241.0</td>\n",
       "      <td>17696.0</td>\n",
       "      <td>-</td>\n",
       "      <td>PKC1</td>\n",
       "      <td>Protein serine/threonine kinase essential for ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YER157W</th>\n",
       "      <td>925</td>\n",
       "      <td>967</td>\n",
       "      <td>chrV</td>\n",
       "      <td>484788.0</td>\n",
       "      <td>487193.0</td>\n",
       "      <td>+</td>\n",
       "      <td>COG3</td>\n",
       "      <td>Essential component of the conserved oligomeri...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           positive_interactions  negative_interactions  chrom     start  \\\n",
       "gene_id                                                                    \n",
       "YFL039C                     1324                   2220  chrVI   53260.0   \n",
       "YFL034C-B                   1150                   1897  chrVI   63016.0   \n",
       "YJR076C                      749                   1342   chrX  575354.0   \n",
       "YBL105C                      841                   1139  chrII   14241.0   \n",
       "YER157W                      925                    967   chrV  484788.0   \n",
       "\n",
       "                end strand gene_name  \\\n",
       "gene_id                                \n",
       "YFL039C     54696.0      -      ACT1   \n",
       "YFL034C-B   63993.0      -      MOB2   \n",
       "YJR076C    576601.0      -     CDC11   \n",
       "YBL105C     17696.0      -      PKC1   \n",
       "YER157W    487193.0      +      COG3   \n",
       "\n",
       "                                                 description  \n",
       "gene_id                                                       \n",
       "YFL039C    Actin, structural protein involved in cell pol...  \n",
       "YFL034C-B  Component of the RAM signaling network, that a...  \n",
       "YJR076C    Component of the septin ring of the mother-bud...  \n",
       "YBL105C    Protein serine/threonine kinase essential for ...  \n",
       "YER157W    Essential component of the conserved oligomeri...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# merge GFF info with genetic interactions df\n",
    "num_of_interactions_file = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/genetic_interactions_summary_by_gene.txt\"\n",
    "num_of_interactions_df = pd.read_csv(num_of_interactions_file, sep='\\t', header = 0, index_col=0)\n",
    "num_of_interactions_df = num_of_interactions_df.join(gene_info_df).sort_values(['chrom', 'start'])\n",
    "num_of_interactions_df = num_of_interactions_df.loc[num_of_interactions_df[['positive_interactions', 'negative_interactions']].sum(axis=1).sort_values(ascending=False).index]\n",
    "display(num_of_interactions_df.head())\n",
    "\n",
    "num_of_interactions_df.to_csv(out_file.replace('.txt', '_annotated.txt'), sep='\\t', header=True, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selecting natural variants for yeast epistasis project \n",
    "Screen natural variation in \"hub\" genes enriched for gene-gene interactions (Costanzo et al., 2016) to identify genetic background effects on growth fitness."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/\"\n",
    "# summary of genetic interactions output file\n",
    "genetic_interactions_summary_file = working_dir+'genetic_interactions_summary_by_gene.txt'\n",
    "os.chdir(working_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select top genes with most genetic interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(800, 11)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>essential</th>\n",
       "      <th>paralog</th>\n",
       "      <th>KA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFL039C</th>\n",
       "      <td>1324</td>\n",
       "      <td>2220</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>53260</td>\n",
       "      <td>54696</td>\n",
       "      <td>-</td>\n",
       "      <td>ACT1</td>\n",
       "      <td>Actin, structural protein involved in cell pol...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL034C-B</th>\n",
       "      <td>1150</td>\n",
       "      <td>1897</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>63016</td>\n",
       "      <td>63993</td>\n",
       "      <td>-</td>\n",
       "      <td>MOB2</td>\n",
       "      <td>Component of the RAM signaling network, that a...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJR076C</th>\n",
       "      <td>749</td>\n",
       "      <td>1342</td>\n",
       "      <td>chrX</td>\n",
       "      <td>575354</td>\n",
       "      <td>576601</td>\n",
       "      <td>-</td>\n",
       "      <td>CDC11</td>\n",
       "      <td>Component of the septin ring of the mother-bud...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBL105C</th>\n",
       "      <td>841</td>\n",
       "      <td>1139</td>\n",
       "      <td>chrII</td>\n",
       "      <td>14241</td>\n",
       "      <td>17696</td>\n",
       "      <td>-</td>\n",
       "      <td>PKC1</td>\n",
       "      <td>Protein serine/threonine kinase essential for ...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YER157W</th>\n",
       "      <td>925</td>\n",
       "      <td>967</td>\n",
       "      <td>chrV</td>\n",
       "      <td>484788</td>\n",
       "      <td>487193</td>\n",
       "      <td>+</td>\n",
       "      <td>COG3</td>\n",
       "      <td>Essential component of the conserved oligomeri...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           positive_interactions  negative_interactions  chrom   start  \\\n",
       "gene_id                                                                  \n",
       "YFL039C                     1324                   2220  chrVI   53260   \n",
       "YFL034C-B                   1150                   1897  chrVI   63016   \n",
       "YJR076C                      749                   1342   chrX  575354   \n",
       "YBL105C                      841                   1139  chrII   14241   \n",
       "YER157W                      925                    967   chrV  484788   \n",
       "\n",
       "              end strand gene_name  \\\n",
       "gene_id                              \n",
       "YFL039C     54696      -      ACT1   \n",
       "YFL034C-B   63993      -      MOB2   \n",
       "YJR076C    576601      -     CDC11   \n",
       "YBL105C     17696      -      PKC1   \n",
       "YER157W    487193      +      COG3   \n",
       "\n",
       "                                                 description  essential  \\\n",
       "gene_id                                                                   \n",
       "YFL039C    Actin, structural protein involved in cell pol...       True   \n",
       "YFL034C-B  Component of the RAM signaling network, that a...       True   \n",
       "YJR076C    Component of the septin ring of the mother-bud...       True   \n",
       "YBL105C    Protein serine/threonine kinase essential for ...       True   \n",
       "YER157W    Essential component of the conserved oligomeri...       True   \n",
       "\n",
       "           paralog  KA  \n",
       "gene_id                 \n",
       "YFL039C      False NaN  \n",
       "YFL034C-B    False NaN  \n",
       "YJR076C      False NaN  \n",
       "YBL105C      False NaN  \n",
       "YER157W      False NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_cutoff = 800\n",
    "\n",
    "num_of_interactions_df = pd.read_csv(genetic_interactions_summary_file.replace('.txt', '_annotated.txt'), sep='\\t', header = 0, index_col=0)\n",
    "num_of_interactions_df = num_of_interactions_df.dropna().astype({'start':'int', 'end':'int'})\n",
    "\n",
    "# remove dubious orfs\n",
    "dubious_orfs_file = working_dir+'dubious_orfs.tsv'\n",
    "dubious_orfs = pd.read_csv(dubious_orfs_file, sep='\\t', header=None, names=['gene_id', 'description'])\n",
    "num_of_interactions_df = num_of_interactions_df.query('~gene_id.isin(@dubious_orfs.gene_id)')\n",
    "\n",
    "# genes in excluded list\n",
    "genes_to_exclude = ['YDL227C', # HO\n",
    "                    'YOR202W', # HIS3\n",
    "                    'YEL021W', # URA3\n",
    "                    'YCL018W', # LEU2\n",
    "                    'YBR115C', # LYS2\n",
    "                    'YBR020W', # GAL1\n",
    "                    'YDR009W', # GAL3\n",
    "                    'YPL248C', # GAL4\n",
    "                    'YBR018C', # GAL7\n",
    "                    'YBR019C', # GAL10\n",
    "                    'YML051W', # GAL80\n",
    "                    'YLR256W'] # HAP1\n",
    "\n",
    "# exclude ergosterol pathway genes that are being used in GxE project\n",
    "ergosterol_genes_file = \"/home/users/rang/scratch/yeast/ergosterol/ergosterol_biosynthetic_process_genes.txt\"\n",
    "ergosterol_genes = pd.read_csv(ergosterol_genes_file, sep='\\t')\n",
    "ergosterol_genes = ergosterol_genes.rename(columns={'Gene Systematic Name':'gene_id', 'Gene':'gene_name'}).set_index('gene_id')\n",
    "\n",
    "genes_to_exclude = genes_to_exclude + ergosterol_genes.index.tolist() \n",
    "num_of_interactions_df = num_of_interactions_df.query('~gene_id.isin(@genes_to_exclude)')\n",
    "\n",
    "# find top genes with most genetic interactions\n",
    "interaction_rich_genes = num_of_interactions_df.loc[num_of_interactions_df[['positive_interactions', 'negative_interactions']].sum(axis=1).sort_values(ascending=False).index].head(genes_cutoff)\n",
    "\n",
    "# annotate genes by essentiality\n",
    "essential_genes_file = working_dir+\"S288C_essential_genes.txt\"\n",
    "essential_genes_df = pd.read_csv(essential_genes_file, sep='\\t')\n",
    "interaction_rich_genes.loc[:,'essential'] = interaction_rich_genes.index.isin(essential_genes_df['Gene Systematic Name'])\n",
    "\n",
    "# annotate genes by paralog status - data from supp. table 2 in Byrne & Wolfe 2005\n",
    "paralogs_file = working_dir+\"S288C_paralogs.txt\"\n",
    "paralogs_df = pd.read_csv(paralogs_file, sep='\\t', index_col='Byrne')\n",
    "paralogs_ka_list = pd.concat([paralogs_df[['Gene 1', 'KA']].rename(columns={'Gene 1':'Gene'}), paralogs_df[['Gene 2', 'KA']].rename(columns={'Gene 2':'Gene'})], axis=0, ignore_index=True).set_index('Gene')\n",
    "interaction_rich_genes.loc[:,'paralog'] = interaction_rich_genes.index.isin(paralogs_ka_list.index)\n",
    "interaction_rich_genes.loc[interaction_rich_genes['paralog'],'KA'] = paralogs_ka_list.loc[interaction_rich_genes.index[interaction_rich_genes['paralog']], 'KA']\n",
    "\n",
    "display(interaction_rich_genes.shape)\n",
    "display(interaction_rich_genes.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scan 1,011 genomes VCF for variants in and around gene of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search hub genes for variants in each gvcf\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16',}\n",
    "\n",
    "genes_to_process = []\n",
    "for chrom in chromosome_dict.keys():\n",
    "    genes_df = interaction_rich_genes.query('chrom==@chrom').sort_values('start')\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict[chrom])\n",
    "    genes_to_process.append((genes_df, gvcf_file, chromosome_dict))\n",
    "\n",
    "# consolidate candidate variants\n",
    "num_of_cores = min(len(os.sched_getaffinity(0)), len(genes_to_process))\n",
    "with mp.Pool(num_of_cores) as pool:\n",
    "    results = pool.starmap(select_variants, genes_to_process)\n",
    "results = pd.concat(results).reset_index(drop=True)\n",
    "\n",
    "# write to new vcf file\n",
    "gxg_variants_file = \"gxg_variants.vcf\"\n",
    "template = vcf.Reader(filename=genes_to_process[0][1])\n",
    "# change chromosome naming convention to a single Roman numeral (follows ref genome, VEP formatting)\n",
    "chromosome_dict_rev = {v:k for k, v in chromosome_dict.items()} \n",
    "template.contigs = OrderedDict((chromosome_dict_rev[k][3:], v._replace(id=chromosome_dict_rev[k][3:])) for k, v in template.contigs.items())\n",
    "# remove 1011 genomes samples data\n",
    "template.samples=[]\n",
    "\n",
    "output = vcf.Writer(open(gxg_variants_file, 'w'), template)\n",
    "for record in results:\n",
    "    # update chromosome naming\n",
    "    record.CHROM = chromosome_dict_rev[record.CHROM][3:]\n",
    "    # write to file\n",
    "    output.write_record(record)\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter variants for those that can be targeted for REF to ALT allele editing\n",
    "Use a 60 bp window to ensure the same sequence (ref allele +- 30bp) can be found in four yeast strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to genome fasta files\n",
    "ref_genome_fasta = \"/home/users/rang/yeast/genomes/Saccharomyces_cerevisiae.R64-1-1.dna.chromosome.I.fa\"\n",
    "rm_genome_fasta = \"/home/users/rang/yeast/genomes/RM11-1A_SGD_2015_JRIP00000000.fsa\"\n",
    "yjm_genome_fasta = \"/home/users/rang/yeast/genomes/YJM789_Stanford_2007_AAFW02000000_highQuality31.fsa\"\n",
    "yps_genome_fasta = \"/home/users/rang/yeast/genomes/YPS128.genome.fa\"\n",
    "\n",
    "# load genome files\n",
    "ref_genome = SeqIO.to_dict(SeqIO.parse(ref_genome_fasta, \"fasta\"))\n",
    "rm_genome = SeqIO.to_dict(SeqIO.parse(rm_genome_fasta, \"fasta\"))\n",
    "yjm_genome = SeqIO.to_dict(SeqIO.parse(yjm_genome_fasta, \"fasta\"))\n",
    "yps_genome = SeqIO.to_dict(SeqIO.parse(yps_genome_fasta, \"fasta\"))\n",
    "genomes_list = [ref_genome, rm_genome, yjm_genome, yps_genome] # IMPORTANT: SET REFERENCE GENOME AS FIRST IN LIST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# path to input and output VCF files\n",
    "gxg_variants_file = \"gxg_variants.vcf\"\n",
    "gxg_variants_select_file = gxg_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "threshold = 10 # threshold value for searching NGGs\n",
    "window = 61 # size of flanking sequences to search for unique hits\n",
    "\n",
    "# read in all candidate variants to filter\n",
    "gxg_variants = vcf.Reader(filename=gxg_variants_file)\n",
    "gxg_variants = [rec for rec in gxg_variants]\n",
    "# split variants into chunks for multiprocessing\n",
    "n = len(gxg_variants)//len(os.sched_getaffinity(0)) + 1\n",
    "gxg_variants = [gxg_variants[i:i+n] for i in range(0, len(gxg_variants), n)]\n",
    "\n",
    "# find targetable variants (TGT info field indicates which genomes variant is targetable in)\n",
    "to_process = [(rec_list, genomes_list, threshold, window) for rec_list in gxg_variants]\n",
    "with mp.Pool(len(os.sched_getaffinity(0))) as pool:\n",
    "    result = pool.starmap(find_targetable_variants, to_process) \n",
    "\n",
    "# write targetable variants to file\n",
    "header = vcf.Reader(filename=gxg_variants_file)\n",
    "# add \"TGT\" info field into VCF header\n",
    "header_info = namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "header.infos['TGT'] = header_info_tgt\n",
    "# write to file\n",
    "gxg_variants_select = vcf.Writer(open(gxg_variants_select_file, 'w'), header)\n",
    "to_write = [rec for rec_list in result for rec in rec_list]\n",
    "for rec in to_write:\n",
    "    gxg_variants_select.write_record(rec)\n",
    "gxg_variants_select.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter variants by allele frequency, and editability in four strains\n",
    "1. Exclude singletons and doubletons\n",
    "2. Exclude variants that cannot be edited in all four strains\n",
    "3. For variants with multiple alternate alleles, select most common alternate allele\n",
    "\n",
    "Variants are assigned an ID prefix \"GXG\" to denote their use in the CRISPEY-BAR epistasis project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# candidate variants editable in all strains\n",
    "gxg_variants_select = vcf.Reader(filename=gxg_variants_select_file)\n",
    "\n",
    "# open output file\n",
    "gxg_variants_select_AFfilter_file = gxg_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "gxg_variants_select_AFfilter = vcf.Writer(open(gxg_variants_select_AFfilter_file, 'w'), gxg_variants_select)\n",
    "\n",
    "var_counter = {}\n",
    "for record in gxg_variants_select:        \n",
    "    # skip variant if all possible alternate allele count is less than 3\n",
    "    if all([ac<3 for ac in record.INFO['AC']]):\n",
    "        continue\n",
    "    # if at least one alternate allele meets minimum count, select most common alternate allele to study\n",
    "    else:\n",
    "        i = np.argmax(record.INFO['AC']) # most common alt allele\n",
    "        # update record entry\n",
    "        record.ALT = [record.ALT[i]]\n",
    "        record.INFO['AC'] = [record.INFO['AC'][i]]\n",
    "        record.INFO['AF'] = [record.INFO['AF'][i]]\n",
    "        record.INFO['MLEAC'] = [record.INFO['MLEAC'][i]]\n",
    "        record.INFO['MLEAF'] = [record.INFO['MLEAF'][i]]\n",
    "        \n",
    "        # inspect TGT field and assign variant ID prefix\n",
    "        ### if tgt field contains all 1 (i.e. targetable by all queried genomes), assign \"GXG\" prefix to variant ID\n",
    "        ### otherwise, skip variant\n",
    "        if all([i==1 for i in record.INFO['TGT']]):\n",
    "            var_id_prefix = \"GXG\"\n",
    "        else:\n",
    "#             print('variant does not pass TGT check: skipping')\n",
    "            continue\n",
    "            \n",
    "        # add to counter\n",
    "        if var_id_prefix not in var_counter:\n",
    "            var_counter[var_id_prefix] = 1\n",
    "        else:\n",
    "            var_counter[var_id_prefix] += 1\n",
    "        \n",
    "        # add unique variant ID (follow format XXX_XXXXX)\n",
    "        record.ID='{}_{:05d}'.format(var_id_prefix, var_counter[var_id_prefix])\n",
    "        \n",
    "        # write to output\n",
    "        gxg_variants_select_AFfilter.write_record(record)\n",
    "gxg_variants_select_AFfilter.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit gxg_variants_select_AFfilter VCF to VEP to get annotations, and download VEP output in TXT format (vep_output_filename).\n",
    "http://uswest.ensembl.org/Saccharomyces_cerevisiae/Tools/VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parsing entry: 5000\n",
      "parsing entry: 10000\n",
      "parsing entry: 15000\n",
      "parsing entry: 20000\n",
      "parsing entry: 25000\n",
      "parsing entry: 30000\n",
      "parsing entry: 35000\n",
      "parsing entry: 40000\n",
      "parsing entry: 45000\n",
      "parsing entry: 50000\n"
     ]
    }
   ],
   "source": [
    "# add annotations to variants \n",
    "gff_file = '/home/users/rang/yeast/genomes/saccharomyces_cerevisiae_R64-1-1_20110208.gff'\n",
    "gxg_variants_file = \"gxg_variants.vcf\"\n",
    "gxg_variants_select_file = gxg_variants_file.replace('.vcf', '_select.vcf')\n",
    "gxg_variants_select_AFfilter_file = gxg_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "vep_output_filename = gxg_variants_select_AFfilter_file.replace('.vcf', '_VEPoutput.txt')\n",
    "\n",
    "variants_annotated_filename = vep_output_filename.replace('_VEPoutput.txt', '_annotated.txt')\n",
    "variants_annotated_df = annotate_variants_by_VEPoutput(gxg_variants_select_AFfilter_file, \n",
    "                                                       vep_output_filename, \n",
    "                                                       gff_file, \n",
    "                                                       variants_annotated_filename, \n",
    "                                                       id_colname='var_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(44835, 41)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00005</th>\n",
       "      <td>I</td>\n",
       "      <td>42019</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>61</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3880|158</td>\n",
       "      <td>862|118|3047|4872</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>118.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00006</th>\n",
       "      <td>I</td>\n",
       "      <td>42023</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.005941</td>\n",
       "      <td>3876|154</td>\n",
       "      <td>858|122|3051|4876</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>122.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00007</th>\n",
       "      <td>I</td>\n",
       "      <td>42033</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>60</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3865|143</td>\n",
       "      <td>847|132|3061|4886</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>132.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00008</th>\n",
       "      <td>I</td>\n",
       "      <td>42036</td>\n",
       "      <td>TTGG</td>\n",
       "      <td>CTGG</td>\n",
       "      <td>6</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.002976</td>\n",
       "      <td>3860|138</td>\n",
       "      <td>842|135|3064|4889</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>135.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>138.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00009</th>\n",
       "      <td>I</td>\n",
       "      <td>42037</td>\n",
       "      <td>TGG</td>\n",
       "      <td>T</td>\n",
       "      <td>30</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>3860|138</td>\n",
       "      <td>842|137|3066|4891</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>137.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>138.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM    POS   REF   ALT  AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                         \n",
       "GXG_00005     I  42019     A     C  61  2022  0.030000              3880|158   \n",
       "GXG_00006     I  42023     C     T  12  2020  0.005941              3876|154   \n",
       "GXG_00007     I  42033     A    AT  60  2012  0.030000              3865|143   \n",
       "GXG_00008     I  42036  TTGG  CTGG   6  2016  0.002976              3860|138   \n",
       "GXG_00009     I  42037   TGG     T  30  2022  0.015000              3860|138   \n",
       "\n",
       "          downstream_distance_str                   region  ... Codons  \\\n",
       "var_id                                                      ...          \n",
       "GXG_00005       862|118|3047|4872  unidirectional_promoter  ...    NaN   \n",
       "GXG_00006       858|122|3051|4876  unidirectional_promoter  ...    NaN   \n",
       "GXG_00007       847|132|3061|4886  unidirectional_promoter  ...    NaN   \n",
       "GXG_00008       842|135|3064|4889  unidirectional_promoter  ...    NaN   \n",
       "GXG_00009       842|137|3066|4891  unidirectional_promoter  ...    NaN   \n",
       "\n",
       "          DOMAINS closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                            \n",
       "GXG_00005     NaN                    GPB2               YAL056W   \n",
       "GXG_00006     NaN                    GPB2               YAL056W   \n",
       "GXG_00007     NaN                    GPB2               YAL056W   \n",
       "GXG_00008     NaN                    GPB2               YAL056W   \n",
       "GXG_00009     NaN                    GPB2               YAL056W   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "GXG_00005  downstream_gene_variant                  118.0   \n",
       "GXG_00006  downstream_gene_variant                  122.0   \n",
       "GXG_00007  downstream_gene_variant                  132.0   \n",
       "GXG_00008  downstream_gene_variant                  135.0   \n",
       "GXG_00009  downstream_gene_variant                  137.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXG_00005                   PEX22               YAL055W   \n",
       "GXG_00006                   PEX22               YAL055W   \n",
       "GXG_00007                   PEX22               YAL055W   \n",
       "GXG_00008                   PEX22               YAL055W   \n",
       "GXG_00009                   PEX22               YAL055W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "GXG_00005    upstream_gene_variant                  158.0  \n",
       "GXG_00006    upstream_gene_variant                  154.0  \n",
       "GXG_00007    upstream_gene_variant                  143.0  \n",
       "GXG_00008    upstream_gene_variant                  138.0  \n",
       "GXG_00009    upstream_gene_variant                  138.0  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read in annotated variant file\n",
    "variants_annotated_df = pd.read_csv(variants_annotated_filename, sep='\\t', index_col=0)\n",
    "\n",
    "# remove variants from genes not listed in interaction_rich_genes\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "gxg_design_oligos = variants_annotated_df.query('(Gene_ID.isin(@interaction_rich_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(gxg_design_oligos.shape)\n",
    "display(gxg_design_oligos.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write variants to gxg_design_oligos VCF as a preliminary list to design oligos for\n",
    "gxg_design_oligos_vcf_file = gxg_variants_file.replace('.vcf', '_design_oligos_initial.vcf')\n",
    "gxg_variants_select_AFfilter = vcf.Reader(filename=gxg_variants_select_AFfilter_file)\n",
    "gxg_design_oligos_vcf = vcf.Writer(open(gxg_design_oligos_vcf_file, 'w'), gxg_variants_select_AFfilter)\n",
    "for record in gxg_variants_select_AFfilter:\n",
    "    if record.ID in gxg_design_oligos.index:\n",
    "        gxg_design_oligos_vcf.write_record(record)\n",
    "gxg_design_oligos_vcf.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run oligo design script (crispey3_design_library.ipynb) with these variants to identify which can be targeted by at least two unique guides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18255, 42)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00006</th>\n",
       "      <td>I</td>\n",
       "      <td>42023</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.005941</td>\n",
       "      <td>3876|154</td>\n",
       "      <td>858|122|3051|4876</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>122.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>154.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00007</th>\n",
       "      <td>I</td>\n",
       "      <td>42033</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>60</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3865|143</td>\n",
       "      <td>847|132|3061|4886</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>132.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>143.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00012</th>\n",
       "      <td>I</td>\n",
       "      <td>42184</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>3715</td>\n",
       "      <td>697|283|3212</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00013</th>\n",
       "      <td>I</td>\n",
       "      <td>42186</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>59</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.029000</td>\n",
       "      <td>3713</td>\n",
       "      <td>695|285|3214</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00014</th>\n",
       "      <td>I</td>\n",
       "      <td>42452</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>3447</td>\n",
       "      <td>429|551|3480</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Pfam:PF12827,Gene3D:3.40.50.11730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM    POS REF ALT  AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                     \n",
       "GXG_00006     I  42023   C   T  12  2020  0.005941              3876|154   \n",
       "GXG_00007     I  42033   A  AT  60  2012  0.030000              3865|143   \n",
       "GXG_00012     I  42184   C   T   5  2022  0.002473                  3715   \n",
       "GXG_00013     I  42186   C   T  59  2022  0.029000                  3713   \n",
       "GXG_00014     I  42452   C   T   8  2022  0.003956                  3447   \n",
       "\n",
       "          downstream_distance_str                   region  ...  \\\n",
       "var_id                                                      ...   \n",
       "GXG_00006       858|122|3051|4876  unidirectional_promoter  ...   \n",
       "GXG_00007       847|132|3061|4886  unidirectional_promoter  ...   \n",
       "GXG_00012            697|283|3212         missense_variant  ...   \n",
       "GXG_00013            695|285|3214         missense_variant  ...   \n",
       "GXG_00014            429|551|3480       synonymous_variant  ...   \n",
       "\n",
       "                                     DOMAINS closest_gene1_Gene_Name  \\\n",
       "var_id                                                                 \n",
       "GXG_00006                                NaN                    GPB2   \n",
       "GXG_00007                                NaN                    GPB2   \n",
       "GXG_00012                                NaN                     NaN   \n",
       "GXG_00013                                NaN                     NaN   \n",
       "GXG_00014  Pfam:PF12827,Gene3D:3.40.50.11730                     NaN   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00006               YAL056W  downstream_gene_variant   \n",
       "GXG_00007               YAL056W  downstream_gene_variant   \n",
       "GXG_00012                   NaN                      NaN   \n",
       "GXG_00013                   NaN                      NaN   \n",
       "GXG_00014                   NaN                      NaN   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXG_00006                  122.0                   PEX22   \n",
       "GXG_00007                  132.0                   PEX22   \n",
       "GXG_00012                    NaN                     NaN   \n",
       "GXG_00013                    NaN                     NaN   \n",
       "GXG_00014                    NaN                     NaN   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00006               YAL055W    upstream_gene_variant   \n",
       "GXG_00007               YAL055W    upstream_gene_variant   \n",
       "GXG_00012                   NaN                      NaN   \n",
       "GXG_00013                   NaN                      NaN   \n",
       "GXG_00014                   NaN                      NaN   \n",
       "\n",
       "          closest_gene2_Distance num_of_oligos  \n",
       "var_id                                          \n",
       "GXG_00006                  154.0           2.0  \n",
       "GXG_00007                  143.0           2.0  \n",
       "GXG_00012                    NaN           2.0  \n",
       "GXG_00013                    NaN           3.0  \n",
       "GXG_00014                    NaN           3.0  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gxg_oligo_file = os.path.expanduser(\"~/crispey3/initial_design/Output/all_SNPs_gxg_initial_GG_9bp_OLIGO.tab\")\n",
    "gxg_oligo_df = pd.read_csv(gxg_oligo_file, sep='\\t')\n",
    "\n",
    "gxg_design_oligos = pd.read_csv(\"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/gxg_variants_select_AFfilter_annotated.txt\", sep='\\t', index_col=0)\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "gxg_design_oligos = gxg_design_oligos.query('(Gene_ID.isin(@interaction_rich_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "\n",
    "# select oligos that have at least 2 oligos designed\n",
    "gxg_design_oligos['num_of_oligos'] = gxg_oligo_df.groupby('var_id').apply(len)\n",
    "gxg_oligos_final = gxg_design_oligos.query('num_of_oligos>1').copy()\n",
    "display(gxg_oligos_final.shape)\n",
    "display(gxg_oligos_final.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Of the 18,255 natural variants eligible for the CRISPEY-BAR epistasis library, select ~2,000 variants - yields a final library of ~5k oligos\n",
    "1. Two groups of 1,000 variants are selected. Group 1 consists of variants with allele frequency (AF) greater than or equal to 2%. Group 2 consists of \"ultra-rare\" variants with allele frequency less than or equal to 0.2%\n",
    "2. Genes are ranked, in descending order, by the number of variants with AF>=2%, followed by number of variants with AF<=0.2%.\n",
    "3. For each gene, select all available group 1 variants, then select a matching number of group 2 variants.\n",
    "4. Go through each gene and select variants till at least 2,000 variants are selected for library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(800, 18)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>essential</th>\n",
       "      <th>paralog</th>\n",
       "      <th>KA</th>\n",
       "      <th>left_limit</th>\n",
       "      <th>right_limit</th>\n",
       "      <th>region_length</th>\n",
       "      <th>num_of_vars</th>\n",
       "      <th>num_of_vars_common</th>\n",
       "      <th>num_of_vars_rare</th>\n",
       "      <th>var_density</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YNL262W</th>\n",
       "      <td>252</td>\n",
       "      <td>319</td>\n",
       "      <td>XIV</td>\n",
       "      <td>148212</td>\n",
       "      <td>154880</td>\n",
       "      <td>+</td>\n",
       "      <td>POL2</td>\n",
       "      <td>Catalytic subunit of DNA polymerase (II) epsil...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147712</td>\n",
       "      <td>155380</td>\n",
       "      <td>7669</td>\n",
       "      <td>99</td>\n",
       "      <td>24</td>\n",
       "      <td>15</td>\n",
       "      <td>1.290911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPL231W</th>\n",
       "      <td>348</td>\n",
       "      <td>548</td>\n",
       "      <td>XVI</td>\n",
       "      <td>108652</td>\n",
       "      <td>114315</td>\n",
       "      <td>+</td>\n",
       "      <td>FAS2</td>\n",
       "      <td>Alpha subunit of fatty acid synthetase, which ...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>108152</td>\n",
       "      <td>114815</td>\n",
       "      <td>6664</td>\n",
       "      <td>99</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.485594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR207C</th>\n",
       "      <td>242</td>\n",
       "      <td>298</td>\n",
       "      <td>XIII</td>\n",
       "      <td>677193</td>\n",
       "      <td>683564</td>\n",
       "      <td>-</td>\n",
       "      <td>HFA1</td>\n",
       "      <td>Mitochondrial acetyl-coenzyme A carboxylase, c...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.366</td>\n",
       "      <td>676693</td>\n",
       "      <td>684064</td>\n",
       "      <td>7372</td>\n",
       "      <td>81</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>1.098752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGR097W</th>\n",
       "      <td>159</td>\n",
       "      <td>370</td>\n",
       "      <td>VII</td>\n",
       "      <td>678695</td>\n",
       "      <td>682135</td>\n",
       "      <td>+</td>\n",
       "      <td>ASK10</td>\n",
       "      <td>Component of RNA polymerase II holoenzyme, pho...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.561</td>\n",
       "      <td>678195</td>\n",
       "      <td>682635</td>\n",
       "      <td>4441</td>\n",
       "      <td>52</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>1.170907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL033C</th>\n",
       "      <td>276</td>\n",
       "      <td>396</td>\n",
       "      <td>VI</td>\n",
       "      <td>69115</td>\n",
       "      <td>74427</td>\n",
       "      <td>-</td>\n",
       "      <td>RIM15</td>\n",
       "      <td>Glucose-repressible protein kinase involved in...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68615</td>\n",
       "      <td>74927</td>\n",
       "      <td>6313</td>\n",
       "      <td>61</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>0.966260</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         positive_interactions  negative_interactions chrom   start     end  \\\n",
       "gene_id                                                                       \n",
       "YNL262W                    252                    319   XIV  148212  154880   \n",
       "YPL231W                    348                    548   XVI  108652  114315   \n",
       "YMR207C                    242                    298  XIII  677193  683564   \n",
       "YGR097W                    159                    370   VII  678695  682135   \n",
       "YFL033C                    276                    396    VI   69115   74427   \n",
       "\n",
       "        strand gene_name                                        description  \\\n",
       "gene_id                                                                       \n",
       "YNL262W      +      POL2  Catalytic subunit of DNA polymerase (II) epsil...   \n",
       "YPL231W      +      FAS2  Alpha subunit of fatty acid synthetase, which ...   \n",
       "YMR207C      -      HFA1  Mitochondrial acetyl-coenzyme A carboxylase, c...   \n",
       "YGR097W      +     ASK10  Component of RNA polymerase II holoenzyme, pho...   \n",
       "YFL033C      -     RIM15  Glucose-repressible protein kinase involved in...   \n",
       "\n",
       "         essential  paralog     KA  left_limit  right_limit  region_length  \\\n",
       "gene_id                                                                      \n",
       "YNL262W       True    False    NaN      147712       155380           7669   \n",
       "YPL231W       True    False    NaN      108152       114815           6664   \n",
       "YMR207C      False     True  0.366      676693       684064           7372   \n",
       "YGR097W      False     True  0.561      678195       682635           4441   \n",
       "YFL033C      False    False    NaN       68615        74927           6313   \n",
       "\n",
       "         num_of_vars  num_of_vars_common  num_of_vars_rare  var_density  \n",
       "gene_id                                                                  \n",
       "YNL262W           99                  24                15     1.290911  \n",
       "YPL231W           99                  21                21     1.485594  \n",
       "YMR207C           81                  19                22     1.098752  \n",
       "YGR097W           52                  19                12     1.170907  \n",
       "YFL033C           61                  18                12     0.966260  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create a dataframe of hub genes with annotations and number of variants in each group\n",
    "left_limit=500 # distance from left of gene to search\n",
    "right_limit=500 # distance from right of gene to search\n",
    "\n",
    "genes_df = interaction_rich_genes.copy()\n",
    "genes_df.loc[:,'chrom'] = genes_df['chrom'].str.replace('chr', '')\n",
    "genes_df.loc[:,'left_limit'] = genes_df['start'] - left_limit\n",
    "genes_df.loc[:,'right_limit'] = genes_df['end'] + right_limit\n",
    "genes_df.loc[:, 'region_length'] = genes_df['right_limit'] - genes_df['left_limit'] + 1\n",
    "\n",
    "# associate each variant with hub gene of interest if it falls within gene limits\n",
    "gxg_oligos_final.loc[:,'assoc_gene'] = gxg_oligos_final.apply(lambda x: genes_df.query('chrom==@x.CHROM & left_limit<=@x.POS & right_limit>=@x.POS').index.tolist(), axis=1)\n",
    "# remove variants without gene associations\n",
    "gxg_oligos_final = gxg_oligos_final[gxg_oligos_final.assoc_gene.apply(len)>0]\n",
    "# for coding variants with multiple associations (i.e. genes of interest adjacent to each other),\n",
    "# adjust to associate only gene whose coding region contains variant\n",
    "selection = (gxg_oligos_final.assoc_gene.apply(len)>1) & (~gxg_oligos_final.region.isin(intergenic_regions_list))\n",
    "gxg_oligos_final.loc[selection, 'assoc_gene'] = gxg_oligos_final.loc[selection, 'Gene_ID'].apply(lambda x: [x])\n",
    "\n",
    "# # finally, exclude remaining variants that are associated with more than one gene (should be only noncoding variants)\n",
    "# gxg_oligos_final = gxg_oligos_final[gxg_oligos_final['assoc_gene'].apply(lambda x: len(x)==1)]\n",
    "\n",
    "# count the number of variants associated with each gene\n",
    "af_cutoff_common=0.02\n",
    "af_cutoff_rare=0.002\n",
    "genes_df.loc[:, 'num_of_vars'] = [gxg_oligos_final['assoc_gene'].apply(lambda x: gene in x).sum() for gene in genes_df.index]\n",
    "genes_df.loc[:, 'num_of_vars_common'] = [gxg_oligos_final.query('AF>=@af_cutoff_common')['assoc_gene'].apply(lambda x: gene in x).sum() for gene in genes_df.index]\n",
    "genes_df.loc[:, 'num_of_vars_rare'] = [gxg_oligos_final.query('AF<=@af_cutoff_rare')['assoc_gene'].apply(lambda x: gene in x).sum() for gene in genes_df.index]\n",
    "# estimate density of variants (num of variants per 100bp) surveyed in each gene_region\n",
    "genes_df.loc[:, 'var_density'] = genes_df['num_of_vars'] / genes_df['region_length'] * 100\n",
    "\n",
    "genes_df = genes_df.sort_values(['num_of_vars_common', 'num_of_vars_rare'], ascending=False)\n",
    "display(genes_df.shape)\n",
    "display(genes_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of hub genes selected: 103\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2000, 43)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>assoc_gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00321</th>\n",
       "      <td>I</td>\n",
       "      <td>101530</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>385|1833</td>\n",
       "      <td>1662|4742|35</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>MAK16</td>\n",
       "      <td>YAL025C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>385.0</td>\n",
       "      <td>LTE1</td>\n",
       "      <td>YAL024C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>35.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[YAL024C, YAL025C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00329</th>\n",
       "      <td>I</td>\n",
       "      <td>101816</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001980</td>\n",
       "      <td>671|2119</td>\n",
       "      <td>1948|4456</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[YAL024C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00332</th>\n",
       "      <td>I</td>\n",
       "      <td>101901</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>90</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.045000</td>\n",
       "      <td>756|2204</td>\n",
       "      <td>2033|4371</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[YAL024C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00339</th>\n",
       "      <td>I</td>\n",
       "      <td>102261</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>78</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.039000</td>\n",
       "      <td>1116|2564</td>\n",
       "      <td>2393|4011</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[YAL024C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00346</th>\n",
       "      <td>I</td>\n",
       "      <td>102699</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>310</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.154000</td>\n",
       "      <td>1554|3002</td>\n",
       "      <td>2831|3573</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[YAL024C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                       \n",
       "GXG_00321     I  101530   G   C    4  2022  0.001978              385|1833   \n",
       "GXG_00329     I  101816   C   T    4  2020  0.001980              671|2119   \n",
       "GXG_00332     I  101901   C   A   90  2022  0.045000              756|2204   \n",
       "GXG_00339     I  102261   C   T   78  2020  0.039000             1116|2564   \n",
       "GXG_00346     I  102699   G   A  310  2016  0.154000             1554|3002   \n",
       "\n",
       "          downstream_distance_str                   region  ...  \\\n",
       "var_id                                                      ...   \n",
       "GXG_00321            1662|4742|35  unidirectional_promoter  ...   \n",
       "GXG_00329               1948|4456         missense_variant  ...   \n",
       "GXG_00332               2033|4371       synonymous_variant  ...   \n",
       "GXG_00339               2393|4011       synonymous_variant  ...   \n",
       "GXG_00346               2831|3573       synonymous_variant  ...   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXG_00321                   MAK16               YAL025C   \n",
       "GXG_00329                     NaN                   NaN   \n",
       "GXG_00332                     NaN                   NaN   \n",
       "GXG_00339                     NaN                   NaN   \n",
       "GXG_00346                     NaN                   NaN   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "GXG_00321    upstream_gene_variant                  385.0   \n",
       "GXG_00329                      NaN                    NaN   \n",
       "GXG_00332                      NaN                    NaN   \n",
       "GXG_00339                      NaN                    NaN   \n",
       "GXG_00346                      NaN                    NaN   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXG_00321                    LTE1               YAL024C   \n",
       "GXG_00329                     NaN                   NaN   \n",
       "GXG_00332                     NaN                   NaN   \n",
       "GXG_00339                     NaN                   NaN   \n",
       "GXG_00346                     NaN                   NaN   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \\\n",
       "var_id                                                                    \n",
       "GXG_00321  downstream_gene_variant                   35.0           3.0   \n",
       "GXG_00329                      NaN                    NaN           4.0   \n",
       "GXG_00332                      NaN                    NaN           2.0   \n",
       "GXG_00339                      NaN                    NaN           2.0   \n",
       "GXG_00346                      NaN                    NaN           3.0   \n",
       "\n",
       "                   assoc_gene  \n",
       "var_id                         \n",
       "GXG_00321  [YAL024C, YAL025C]  \n",
       "GXG_00329           [YAL024C]  \n",
       "GXG_00332           [YAL024C]  \n",
       "GXG_00339           [YAL024C]  \n",
       "GXG_00346           [YAL024C]  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# select 2000 variants, in as few genes as possible\n",
    "cutoff = 2000\n",
    "\n",
    "var_list = []\n",
    "gene_list = []\n",
    "exclude = []\n",
    "for gene in genes_df.index:\n",
    "    # skip blacklisted genes in exclusion list\n",
    "    if gene in exclude:\n",
    "#         print(gene, 'is excluded')\n",
    "        continue\n",
    "        \n",
    "    df = gxg_oligos_final[gxg_oligos_final['assoc_gene'].apply(lambda x: gene in x)].sort_values(['AF', 'AC'], ascending=False)\n",
    "    com_var = df.query('AF>=@af_cutoff_common').index.tolist()\n",
    "    if len(com_var) > len(df)//2:\n",
    "        com_var = com_var[:len(df)//2]        \n",
    "    rar_var = df.tail(len(com_var)).index.tolist()\n",
    "    com_rar_var = com_var + rar_var\n",
    "    \n",
    "    # if selected variants are associated with genes other than the one being parsed (e.g. two hub genes adjacent to each other),\n",
    "    # add those other genes to the exclude list to avoid resampling the same variants\n",
    "    other_genes = set([g for l in df.loc[com_rar_var, 'assoc_gene'] for g in l if g != gene])\n",
    "    if len(other_genes)>0:\n",
    "#         print(other_genes, \"added to exclude list\")\n",
    "        exclude += list(other_genes)\n",
    "    \n",
    "    # add selected variants to var_list\n",
    "    var_list += com_rar_var\n",
    "    # add gene to gene_list\n",
    "    gene_list.append(gene)\n",
    "    \n",
    "    if len(var_list)>= cutoff:\n",
    "        break\n",
    "\n",
    "# number of genes selected\n",
    "hub_genes = genes_df.loc[gene_list]\n",
    "print('Number of hub genes selected:', len(hub_genes))\n",
    "\n",
    "# select final list of variants that will be targeted\n",
    "gxg_oligos_final = gxg_oligos_final.query('var_id.isin(@var_list)')\n",
    "display(gxg_oligos_final.shape)\n",
    "display(gxg_oligos_final.head())\n",
    "\n",
    "# write gxg_oligos_final table to file\n",
    "gxg_oligos_final.to_csv(\"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/gxg_variants_final.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of GXG variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants: 2000\n",
      "Number of oligos: 4957\n",
      "\n",
      "Distribution of (common) variants among hub genes\n",
      "count    103.000000\n",
      "mean       9.708738\n",
      "std        3.285831\n",
      "min        7.000000\n",
      "25%        8.000000\n",
      "50%        9.000000\n",
      "75%       11.000000\n",
      "max       24.000000\n",
      "Name: num_of_vars_common, dtype: float64\n",
      "\n",
      "Distribution of oligos among GxG variants\n",
      "count    2000.000000\n",
      "mean        2.478500\n",
      "std         0.740143\n",
      "min         2.000000\n",
      "25%         2.000000\n",
      "50%         2.000000\n",
      "75%         3.000000\n",
      "max         7.000000\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Proportion of variant annotation classes\n",
      "synonymous_variant         0.4395\n",
      "missense_variant           0.3470\n",
      "unidirectional_promoter    0.1020\n",
      "bidirectional_promoter     0.0835\n",
      "intergenic                 0.0175\n",
      "intron_variant             0.0030\n",
      "inframe_insertion          0.0030\n",
      "inframe_deletion           0.0025\n",
      "frameshift_variant         0.0010\n",
      "stop_gained                0.0010\n",
      "Name: region, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of variants:\", len(gxg_oligos_final))\n",
    "print(\"Number of oligos:\", int(sum(gxg_oligos_final.num_of_oligos)))\n",
    "print()\n",
    "print(\"Distribution of (common) variants among hub genes\")\n",
    "# print(gxg_oligos_final.Gene_ID.value_counts().filter(hub_genes.index).describe())\n",
    "print(hub_genes.num_of_vars_common.describe())\n",
    "print()\n",
    "print(\"Distribution of oligos among GxG variants\")\n",
    "print(gxg_oligos_final.num_of_oligos.describe())\n",
    "print()\n",
    "print(\"Proportion of variant annotation classes\")\n",
    "print(gxg_oligos_final.region.value_counts()/len(gxg_oligos_final))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine number of wells required in matrixed oligo synthesis array\n",
    "Twist Biosciences synthesizes 121 oligos per well, up to 384 wells per plate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GXG    4957.0\n",
       "Name: num_of_oligos, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos to fit: 4957\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(42, 45)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use 42 pools:\n",
      "Number of oligos per pool: 118\n",
      "Number of oligos leftover: 1\n",
      "Number of technical oligos: 126\n",
      "\n",
      "Use 43 pools:\n",
      "Number of oligos per pool: 115\n",
      "Number of oligos leftover: 12\n",
      "Number of technical oligos: 258\n",
      "\n",
      "Use 44 pools:\n",
      "Number of oligos per pool: 112\n",
      "Number of oligos leftover: 29\n",
      "Number of technical oligos: 396\n",
      "\n",
      "Use 45 pools:\n",
      "Number of oligos per pool: 110\n",
      "Number of oligos leftover: 7\n",
      "Number of technical oligos: 495\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([42, 43, 44, 45])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count OLIGOS in each variant prefix set\n",
    "display(gxg_oligos_final.groupby(lambda x: x[:3]).num_of_oligos.sum().sort_values(ascending=False))\n",
    "\n",
    "# determine number of pools to use\n",
    "num_of_oligos = int(gxg_oligos_final.num_of_oligos.sum())\n",
    "min_pool_size = 109 # allows 12 technical oligos\n",
    "max_pool_size = 118 # allows 3 technical oligos\n",
    "complete_pool_size = 121 # maximum oligos synthesized per pool\n",
    "\n",
    "# find number of pools to use\n",
    "find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selecting natural variants in yeast QTLs for GxE project\n",
    "Screen natural variants in select list of QTLs from Bloom et al. (2019) tested for environment-specific fitness effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_variants_from_qtls(qtls_df, vcf_file, chrom_dict, upstream_length=500, downstream_length=500):\n",
    "    '''\n",
    "    modified version of select_variants function that accepts a dataframe of QTLs\n",
    "    (requires qtl_left and qtl_right columns defined, and assumes each QTL is associated with one gene)\n",
    "    defines search boundaries for each QTL and then searches vcf for variants within it\n",
    "    can be parallelized by splitting genes and vcf by chromosome\n",
    "    '''\n",
    "    # find gene upstream and downstream boundaries\n",
    "    qtls_df.loc[qtls_df['strand']=='+', 'gene_left'] = qtls_df['start']-upstream_length\n",
    "    qtls_df.loc[qtls_df['strand']=='+', 'gene_right'] = qtls_df['end']+downstream_length\n",
    "    qtls_df.loc[qtls_df['strand']=='-', 'gene_left'] = qtls_df['start']-downstream_length\n",
    "    qtls_df.loc[qtls_df['strand']=='-', 'gene_right'] = qtls_df['end']+upstream_length\n",
    "    \n",
    "    # define search boundaries by either gene or QTL boundary, whichever is larger\n",
    "    qtls_df.loc[:,'search_left'] = qtls_df[['gene_left', 'qtl_left']].min(axis=1)\n",
    "    qtls_df.loc[:,'search_right'] = qtls_df[['gene_right', 'qtl_right']].max(axis=1)\n",
    "    qtls_df.loc[:,'search_left'] = qtls_df['search_left'].astype(int)\n",
    "    qtls_df.loc[:,'search_right'] = qtls_df['search_right'].astype(int)\n",
    "    \n",
    "    # from these search boundaries, remove duplicates\n",
    "    search_boundaries_df = qtls_df[['chrom', 'search_left', 'search_right']].drop_duplicates()\n",
    "\n",
    "    # read in vcf file to parse for variants\n",
    "    allVars = vcf.Reader(filename=vcf_file)\n",
    "    candVars = []\n",
    "    for i, row in search_boundaries_df.iterrows():\n",
    "        retrieve = allVars.fetch(chrom=chrom_dict[row['chrom']], \n",
    "                                 start=row['search_left']-1,\n",
    "                                 end=row['search_right'])\n",
    "        for rec in retrieve:\n",
    "            rec.samples = []\n",
    "            rec.FORMAT=None\n",
    "            candVars.append(rec)\n",
    "    candVars = pd.Series(candVars).drop_duplicates().reset_index(drop=True)\n",
    "    return candVars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/GxE/\"\n",
    "# list of QTLs to parse for GxE variants\n",
    "gxe_qtls_file = working_dir+\"GxE_gene_selection_20201130.tsv\"\n",
    "os.chdir(working_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(92, 33)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>cross</th>\n",
       "      <th>forward_scan_marker</th>\n",
       "      <th>index</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>LOD</th>\n",
       "      <th>relocalized_peak</th>\n",
       "      <th>...</th>\n",
       "      <th>Beta_abs</th>\n",
       "      <th>qtl_genes</th>\n",
       "      <th>num_qtl_genes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Caffeine;15mM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>624258</td>\n",
       "      <td>B</td>\n",
       "      <td>chrXV_620334_A_G_61508</td>\n",
       "      <td>18833</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>74.281303</td>\n",
       "      <td>chrXV_624258_G_A_61750</td>\n",
       "      <td>...</td>\n",
       "      <td>0.909699</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fluconazole;100uM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>621444</td>\n",
       "      <td>B</td>\n",
       "      <td>chrXV_621444_A_G_61540</td>\n",
       "      <td>18839</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>117.931452</td>\n",
       "      <td>chrXV_621444_A_G_61540</td>\n",
       "      <td>...</td>\n",
       "      <td>0.920556</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fluconazole;100uM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619941</td>\n",
       "      <td>377</td>\n",
       "      <td>chrXV_619941_G_T_47212</td>\n",
       "      <td>23734</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>168.252715</td>\n",
       "      <td>chrXV_619941_G_T_47212</td>\n",
       "      <td>...</td>\n",
       "      <td>0.711397</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fluconazole;100uM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>622413</td>\n",
       "      <td>3049</td>\n",
       "      <td>chrXV_622413_A_G_72894</td>\n",
       "      <td>25133</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>120.009983</td>\n",
       "      <td>chrXV_622413_A_G_72894</td>\n",
       "      <td>...</td>\n",
       "      <td>0.607320</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fluconazole;100uM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>624320</td>\n",
       "      <td>377</td>\n",
       "      <td>chrXV_474477_C_A_46484</td>\n",
       "      <td>23322</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>7.932991</td>\n",
       "      <td>chrXV_624320_TTTTC_T_47455</td>\n",
       "      <td>...</td>\n",
       "      <td>0.107987</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 trait    chr     pos cross     forward_scan_marker  index  \\\n",
       "0      Caffeine;15mM;2  chrXV  624258     B  chrXV_620334_A_G_61508  18833   \n",
       "1  Fluconazole;100uM;2  chrXV  621444     B  chrXV_621444_A_G_61540  18839   \n",
       "2  Fluconazole;100uM;2  chrXV  619941   377  chrXV_619941_G_T_47212  23734   \n",
       "3  Fluconazole;100uM;2  chrXV  622413  3049  chrXV_622413_A_G_72894  25133   \n",
       "4  Fluconazole;100uM;2  chrXV  624320   377  chrXV_474477_C_A_46484  23322   \n",
       "\n",
       "   p-value  q-value         LOD            relocalized_peak  ...  Beta_abs  \\\n",
       "0   0.0001   0.0001   74.281303      chrXV_624258_G_A_61750  ...  0.909699   \n",
       "1   0.0001   0.0001  117.931452      chrXV_621444_A_G_61540  ...  0.920556   \n",
       "2   0.0001   0.0001  168.252715      chrXV_619941_G_T_47212  ...  0.711397   \n",
       "3   0.0001   0.0001  120.009983      chrXV_622413_A_G_72894  ...  0.607320   \n",
       "4   0.0009   0.0003    7.932991  chrXV_624320_TTTTC_T_47455  ...  0.107987   \n",
       "\n",
       "     qtl_genes  num_qtl_genes  gene_id  gene_name  chrom   start     end  \\\n",
       "0  ['YOR153W']              1  YOR153W       PDR5  chrXV  619840  624375   \n",
       "1  ['YOR153W']              1  YOR153W       PDR5  chrXV  619840  624375   \n",
       "2  ['YOR153W']              1  YOR153W       PDR5  chrXV  619840  624375   \n",
       "3  ['YOR153W']              1  YOR153W       PDR5  chrXV  619840  624375   \n",
       "4  ['YOR153W']              1  YOR153W       PDR5  chrXV  619840  624375   \n",
       "\n",
       "  strand                                        description  \n",
       "0      +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "1      +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "2      +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "3      +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "4      +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read in GxE QTL table\n",
    "gxe_qtls = pd.read_csv(gxe_qtls_file, sep='\\t', index_col=0)\n",
    "gxe_qtls = gxe_qtls.rename(columns={'Systematic Name':'gene_id', 'Gene ID':'gene_name', 'Chr':'chrom'})#.set_index('gene_id')\n",
    "# add gene strand and description annotations to gxe QTLs\n",
    "gxe_qtls = gxe_qtls.merge(gene_info_df[['strand', 'description']], how='left', left_on='gene_id', right_index=True)\n",
    "\n",
    "# remove QTLs containing blacklisted genes\n",
    "genes_to_exclude = ['YDL227C', # HO\n",
    "                    'YOR202W', # HIS3\n",
    "                    'YEL021W', # URA3\n",
    "                    'YCL018W', # LEU2\n",
    "                    'YBR115C', # LYS2\n",
    "                    'YBR020W', # GAL1\n",
    "                    'YDR009W', # GAL3\n",
    "                    'YPL248C', # GAL4\n",
    "                    'YBR018C', # GAL7\n",
    "                    'YBR019C', # GAL10\n",
    "                    'YML051W', # GAL80\n",
    "                    'YLR256W'] # HAP1\n",
    "                   \n",
    "gxe_qtls = gxe_qtls.query('~gene_id.isin(@genes_to_exclude)')\n",
    "\n",
    "# get list of 1,011 yeast strains used in QTL mapping\n",
    "yeast_strains_id = gxe_qtls['Strain ID of Parent 1 in Peter et al. 2018'].dropna().unique()\n",
    "yeast_strains_name_to_id = dict(zip(gxe_qtls['Parent 1'].apply(lambda x: x.split(' ')[0]),\n",
    "                                    gxe_qtls['Strain ID of Parent 1 in Peter et al. 2018']))\n",
    "display(gxe_qtls.shape)\n",
    "display(gxe_qtls.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scan 1,011 genomes VCF for variants in and around QTLs of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search GxE QTLs for variants in each gvcf\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16'}\n",
    "\n",
    "qtls_to_process = []\n",
    "for chrom in chromosome_dict.keys():\n",
    "    qtls_df = gxe_qtls.query('chrom==@chrom').sort_values('pos')\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict[chrom])\n",
    "    qtls_to_process.append((qtls_df, gvcf_file, chromosome_dict))\n",
    "    \n",
    "# consolidate candidate variants\n",
    "num_of_cores = min(len(os.sched_getaffinity(0)), len(qtls_to_process))\n",
    "with mp.Pool(num_of_cores) as pool:\n",
    "    results = pool.starmap(select_variants_from_qtls, qtls_to_process)\n",
    "results = pd.concat(results).reset_index(drop=True)\n",
    "\n",
    "# write to new vcf file\n",
    "gxe_variants_file = \"gxe_variants.vcf\"\n",
    "template = vcf.Reader(filename=qtls_to_process[0][1])\n",
    "# change chromosome naming convention to a single Roman numeral (follows ref genome, VEP formatting)\n",
    "chromosome_dict_rev = {v:k for k, v in chromosome_dict.items()} \n",
    "template.contigs = OrderedDict((chromosome_dict_rev[k][3:], v._replace(id=chromosome_dict_rev[k][3:])) for k, v in template.contigs.items())\n",
    "# remove 1011 genomes strain allele info\n",
    "template.samples = []\n",
    "\n",
    "output = vcf.Writer(open(gxe_variants_file, 'w'), template)\n",
    "for record in results:\n",
    "    # update chromosome naming\n",
    "    record.CHROM = chromosome_dict_rev[record.CHROM][3:]\n",
    "    # write to file\n",
    "    output.write_record(record)\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter variants for those that can be targeted for editing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to genome fasta files\n",
    "ref_genome_fasta = \"/home/users/rang/yeast/genomes/Saccharomyces_cerevisiae.R64-1-1.dna.chromosome.I.fa\"\n",
    "\n",
    "# load genome files\n",
    "ref_genome = SeqIO.to_dict(SeqIO.parse(ref_genome_fasta, \"fasta\"))\n",
    "genomes_list = [ref_genome]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# path to input and output VCF files\n",
    "gxe_variants_file = \"gxe_variants.vcf\"\n",
    "gxe_variants_select_file = gxe_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "threshold = 10 # threshold value for searching NGGs\n",
    "window = 61 # size of flanking sequences to search for unique hits\n",
    "\n",
    "# read in all candidate variants to filter\n",
    "gxe_variants = vcf.Reader(filename=gxe_variants_file)\n",
    "gxe_variants = [rec for rec in gxe_variants]\n",
    "# split variants into chunks for multiprocessing\n",
    "n = len(gxe_variants)//len(os.sched_getaffinity(0)) + 1\n",
    "gxe_variants = [gxe_variants[i:i+n] for i in range(0, len(gxe_variants), n)]\n",
    "\n",
    "# find targetable variants (TGT info field indicates which genomes variant is targetable in)\n",
    "to_process = [(rec_list, genomes_list, threshold, window) for rec_list in gxe_variants]\n",
    "with mp.Pool(len(os.sched_getaffinity(0))) as pool:\n",
    "    result = pool.starmap(find_targetable_variants, to_process) \n",
    "\n",
    "# write targetable variants to file\n",
    "header = vcf.Reader(filename=gxe_variants_file)\n",
    "# add \"TGT\" info field into VCF header\n",
    "header_info = namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "header.infos['TGT'] = header_info_tgt\n",
    "# write to file\n",
    "gxe_variants_select = vcf.Writer(open(gxe_variants_select_file, 'w'), header)\n",
    "to_write = [rec for rec_list in result for rec in rec_list]\n",
    "for rec in to_write:\n",
    "    gxe_variants_select.write_record(rec)\n",
    "gxe_variants_select.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter variants by allele frequency\n",
    "1. For variants with multiple alternate alleles, split variant entry into separate lines in VCF for library design - allows oligo design pipeline to synthesize separate oligos for each type of REF-to-ALT edit\n",
    "\n",
    "Variants are assigned an ID prefix \"GXE\" to denote their use in the CRISPEY-BAR GxE project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select variants by allele frequency (currently disabled singleton/doubleton filter)\n",
    "### if multiple alternate alleles, SPLIT VARIANT ENTRY INTO SEPARATE ENTRIES in VCF (useful for oligo design pipeline)\n",
    "### assign unitl variant ID prefix and number (use format VAR_XXXXX)\n",
    "\n",
    "# candidate variants editable in all strains\n",
    "gxe_variants_select = vcf.Reader(filename=gxe_variants_select_file)\n",
    "\n",
    "# open output file\n",
    "gxe_variants_select_AFfilter_file = gxe_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "gxe_variants_select_AFfilter = vcf.Writer(open(gxe_variants_select_AFfilter_file, 'w'), gxe_variants_select)\n",
    "\n",
    "var_counter = {}\n",
    "for record in gxe_variants_select:        \n",
    "    # (optional step to skip singletons and doubletons -- DISABLED by setting ac<1)\n",
    "    if all([ac<1 for ac in record.INFO['AC']]):\n",
    "        continue\n",
    "    # if at least one alternate allele meets minimum count, write to VCF\n",
    "    else:\n",
    "        alt = record.ALT[:]\n",
    "        ac = record.INFO['AC'][:]\n",
    "        af = record.INFO['AF'][:]\n",
    "        mleac = record.INFO['MLEAC'][:]\n",
    "        mleaf = record.INFO['MLEAF'][:]\n",
    "        \n",
    "        # split each alt allele into its own variant entry (useful if a locus has multiple possible alleles)\n",
    "        for i in range(len(alt)):\n",
    "            record.ALT = [alt[i]]\n",
    "            record.INFO['AC'] = [ac[i]]\n",
    "            record.INFO['AF'] = [af[i]]\n",
    "            record.INFO['MLEAC'] = [mleac[i]]\n",
    "            record.INFO['MLEAF'] = [mleaf[i]]\n",
    "\n",
    "            # inspect TGT field and assign variant ID prefix\n",
    "            ### if tgt field contains all 1 (i.e. targetable by all queried genomes), assign \"GXE\"\n",
    "            ### otherwise, skip variant\n",
    "            if all([i==1 for i in record.INFO['TGT']]):\n",
    "                var_id_prefix = \"GXE\"\n",
    "            else:\n",
    "#                 print('variant does not pass TGT check: skipping')\n",
    "                continue\n",
    "\n",
    "            # add to counter\n",
    "            if var_id_prefix not in var_counter:\n",
    "                var_counter[var_id_prefix] = 1\n",
    "            else:\n",
    "                var_counter[var_id_prefix] += 1\n",
    "\n",
    "            # add unique variant ID (follow format VAR_XXXXX)\n",
    "            record.ID='{}_{:05d}'.format(var_id_prefix, var_counter[var_id_prefix])\n",
    "\n",
    "            # write to output\n",
    "            gxe_variants_select_AFfilter.write_record(record)\n",
    "            \n",
    "gxe_variants_select_AFfilter.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit gxe_variants_select_AFfilter VCF to VEP to get annotations, and download VEP output in TXT format (vep_output_filename).\n",
    "http://uswest.ensembl.org/Saccharomyces_cerevisiae/Tools/VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parsing entry: 5000\n",
      "parsing entry: 10000\n",
      "parsing entry: 15000\n"
     ]
    }
   ],
   "source": [
    "# add annotations to variants \n",
    "gff_file = '/home/users/rang/yeast/genomes/saccharomyces_cerevisiae_R64-1-1_20110208.gff'\n",
    "gxe_variants_file = \"gxe_variants.vcf\"\n",
    "gxe_variants_select_file = gxe_variants_file.replace('.vcf', '_select.vcf')\n",
    "gxe_variants_select_AFfilter_file = gxe_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "vep_output_filename = gxe_variants_select_AFfilter_file.replace('.vcf', '_VEPoutput.txt')\n",
    "\n",
    "variants_annotated_filename = vep_output_filename.replace('_VEPoutput.txt', '_annotated.txt')\n",
    "variants_annotated_df = annotate_variants_by_VEPoutput(gxe_variants_select_AFfilter_file, \n",
    "                                                       vep_output_filename, \n",
    "                                                       gff_file, \n",
    "                                                       variants_annotated_filename, \n",
    "                                                       id_colname='var_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15390, 41)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00017</th>\n",
       "      <td>I</td>\n",
       "      <td>192257</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2006</td>\n",
       "      <td>0.003988</td>\n",
       "      <td>362</td>\n",
       "      <td>4525|3446|80|1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAT1</td>\n",
       "      <td>YAR035W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>200.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00025</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.000993</td>\n",
       "      <td>8|181</td>\n",
       "      <td>4693|3614|169</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>8.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT   AC    AN  \\\n",
       "var_id                                                                 \n",
       "GXE_00017     I  192257                G                A    8  2006   \n",
       "GXE_00023     I  192419                C                A    6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT   39  2014   \n",
       "GXE_00025     I  192424  GTTTGGATTACCTCT                G    2  2014   \n",
       "GXE_00026     I  192427                T                C  902  2000   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00017  0.003988                   362          4525|3446|80|1   \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00025  0.000993                 8|181           4693|3614|169   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "\n",
       "                           region  ... Codons DOMAINS closest_gene1_Gene_Name  \\\n",
       "var_id                             ...                                          \n",
       "GXE_00017              intergenic  ...    NaN     NaN                    YAT1   \n",
       "GXE_00023  bidirectional_promoter  ...    NaN     NaN                     NaN   \n",
       "GXE_00024  bidirectional_promoter  ...    NaN     NaN                     NaN   \n",
       "GXE_00025  bidirectional_promoter  ...    NaN     NaN                     NaN   \n",
       "GXE_00026  bidirectional_promoter  ...    NaN     NaN                     NaN   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXE_00017               YAR035W  downstream_gene_variant   \n",
       "GXE_00023             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00024             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00025             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00026             YAR035C-A    upstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXE_00017                    1.0                     NaN   \n",
       "GXE_00023                    2.0                    SWH1   \n",
       "GXE_00024                    7.0                    SWH1   \n",
       "GXE_00025                    8.0                    SWH1   \n",
       "GXE_00026                   10.0                    SWH1   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXE_00017             YAR035C-A  downstream_gene_variant   \n",
       "GXE_00023               YAR042W    upstream_gene_variant   \n",
       "GXE_00024               YAR042W    upstream_gene_variant   \n",
       "GXE_00025               YAR042W    upstream_gene_variant   \n",
       "GXE_00026               YAR042W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance  \n",
       "var_id                            \n",
       "GXE_00017                   80.0  \n",
       "GXE_00023                  200.0  \n",
       "GXE_00024                  181.0  \n",
       "GXE_00025                  181.0  \n",
       "GXE_00026                  192.0  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read in annotated file\n",
    "variants_annotated_df = pd.read_csv(variants_annotated_filename, sep='\\t', index_col=0)\n",
    "\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "gxe_design_oligos = variants_annotated_df.query('(Gene_ID.isin(@gxe_qtls.gene_id.unique()) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(gxe_design_oligos.shape)\n",
    "display(gxe_design_oligos.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write variants to gxe_design_oligos VCF as a preliminary list to design oligos for\n",
    "gxe_design_oligos_vcf_file = gxe_variants_file.replace('.vcf', '_design_oligos_initial.vcf')\n",
    "\n",
    "gxe_variants_select_AFfilter = vcf.Reader(filename=gxe_variants_select_AFfilter_file)\n",
    "gxe_design_oligos_vcf = vcf.Writer(open(gxe_design_oligos_vcf_file, 'w'), gxe_variants_select_AFfilter)\n",
    "for record in gxe_variants_select_AFfilter:\n",
    "    if record.ID in gxe_design_oligos.index:\n",
    "        gxe_design_oligos_vcf.write_record(record)\n",
    "gxe_design_oligos_vcf.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run oligo design script (crispey3_design_library.ipynb) with these variants to identify which can be targeted by at least two unique guides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6480, 42)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>200.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00025</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.000993</td>\n",
       "      <td>8|181</td>\n",
       "      <td>4693|3614|169</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>8.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>192.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00027</th>\n",
       "      <td>I</td>\n",
       "      <td>192429</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1506</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.754000</td>\n",
       "      <td>12|190</td>\n",
       "      <td>4697|3618|173</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>12.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>190.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT    AC    AN  \\\n",
       "var_id                                                                  \n",
       "GXE_00023     I  192419                C                A     6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT    39  2014   \n",
       "GXE_00025     I  192424  GTTTGGATTACCTCT                G     2  2014   \n",
       "GXE_00026     I  192427                T                C   902  2000   \n",
       "GXE_00027     I  192429                G                C  1506  1998   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00025  0.000993                 8|181           4693|3614|169   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "GXE_00027  0.754000                12|190           4697|3618|173   \n",
       "\n",
       "                           region  ... DOMAINS closest_gene1_Gene_Name  \\\n",
       "var_id                             ...                                   \n",
       "GXE_00023  bidirectional_promoter  ...     NaN                     NaN   \n",
       "GXE_00024  bidirectional_promoter  ...     NaN                     NaN   \n",
       "GXE_00025  bidirectional_promoter  ...     NaN                     NaN   \n",
       "GXE_00026  bidirectional_promoter  ...     NaN                     NaN   \n",
       "GXE_00027  bidirectional_promoter  ...     NaN                     NaN   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXE_00023             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00024             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00025             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00026             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00027             YAR035C-A    upstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXE_00023                    2.0                    SWH1   \n",
       "GXE_00024                    7.0                    SWH1   \n",
       "GXE_00025                    8.0                    SWH1   \n",
       "GXE_00026                   10.0                    SWH1   \n",
       "GXE_00027                   12.0                    SWH1   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXE_00023               YAR042W    upstream_gene_variant   \n",
       "GXE_00024               YAR042W    upstream_gene_variant   \n",
       "GXE_00025               YAR042W    upstream_gene_variant   \n",
       "GXE_00026               YAR042W    upstream_gene_variant   \n",
       "GXE_00027               YAR042W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance num_of_oligos  \n",
       "var_id                                          \n",
       "GXE_00023                  200.0           3.0  \n",
       "GXE_00024                  181.0           3.0  \n",
       "GXE_00025                  181.0           3.0  \n",
       "GXE_00026                  192.0           2.0  \n",
       "GXE_00027                  190.0           3.0  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gxe_oligo_file = os.path.expanduser(\"~/crispey3/initial_design/Output/all_SNPs_gxe_initial_GG_9bp_OLIGO.tab\")\n",
    "gxe_oligo_df = pd.read_csv(gxe_oligo_file, sep='\\t')\n",
    "\n",
    "gxe_design_oligos = pd.read_csv(\"/home/users/rang/scratch/yeast/GxE/gxe_variants_select_AFfilter_annotated.txt\", sep='\\t', index_col=0)\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "gxe_design_oligos = gxe_design_oligos.query('(Gene_ID.isin(@gxe_qtls.gene_id.unique()) | region.isin(@intergenic_regions_list))').copy()\n",
    "\n",
    "# select oligos that have at least 2 oligos designed\n",
    "gxe_design_oligos['num_of_oligos'] = gxe_oligo_df.groupby('var_id').apply(len)\n",
    "gxe_oligos_final = gxe_design_oligos.query('num_of_oligos>1').copy()\n",
    "display(gxe_oligos_final.shape)\n",
    "display(gxe_oligos_final.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove singletons & doubletons, but keep those that appear in at least one of the 15 QTL parental strains (Bloom et al., 2019)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500 variants processed\n",
      "1000 variants processed\n",
      "1500 variants processed\n",
      "2000 variants processed\n",
      "2500 variants processed\n",
      "3000 variants processed\n",
      "3500 variants processed\n",
      "4000 variants processed\n",
      "4500 variants processed\n",
      "5000 variants processed\n",
      "5500 variants processed\n",
      "6000 variants processed\n"
     ]
    }
   ],
   "source": [
    "# search GVCF and add allele information for parent strains in QTL crosses\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16'}\n",
    "\n",
    "gxe_variants_file = \"gxe_variants.vcf\"\n",
    "i=0\n",
    "for chromosome, df in gxe_oligos_final.groupby('CHROM'):\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict['chr'+chromosome])\n",
    "    allVars = vcf.Reader(filename=gvcf_file)\n",
    "    \n",
    "    for v, row in df.iterrows():\n",
    "        # find variant in 1011 genomes GVCF\n",
    "        retrieve = allVars.fetch(chrom=chromosome_dict['chr'+row.CHROM], \n",
    "                                 start=row.POS-1,\n",
    "                                 end=row.POS)\n",
    "        # select record that matches \n",
    "        for r in retrieve:\n",
    "            if (r.POS==row.POS) & (r.REF==row.REF):\n",
    "                record = r\n",
    "                break\n",
    "                \n",
    "        # get all alternate alleles at this position\n",
    "        gxe_oligos_final.loc[v, 'all_alt_alleles'] = ','.join([str(a) for a in record.ALT])\n",
    "        # get allele calls for strains in yeast_strains_id\n",
    "        for s in record.samples:\n",
    "            if (s.sample in yeast_strains_id) and (None not in s.gt_alleles):\n",
    "                gxe_oligos_final.loc[v, s.sample] = '/'.join(s.gt_alleles)\n",
    "        \n",
    "        i+=1\n",
    "        if i % 500 == 0:\n",
    "            print(i, \"variants processed\")\n",
    "        \n",
    "# write updated gxe_oligos_final table\n",
    "gxe_oligos_final.to_csv(gxe_variants_file.replace('.vcf', '_with_parents.txt'), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4384, 58)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACI</th>\n",
       "      <th>ACQ</th>\n",
       "      <th>ACS</th>\n",
       "      <th>ADE</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADG</th>\n",
       "      <th>ADI</th>\n",
       "      <th>ADR</th>\n",
       "      <th>SACE_MAA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00027</th>\n",
       "      <td>I</td>\n",
       "      <td>192429</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1506</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.754000</td>\n",
       "      <td>12|190</td>\n",
       "      <td>4697|3618|173</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00033</th>\n",
       "      <td>I</td>\n",
       "      <td>192466</td>\n",
       "      <td>AAAGGG</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001485</td>\n",
       "      <td>50|148</td>\n",
       "      <td>4735|3656|211</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT    AC    AN  \\\n",
       "var_id                                                                  \n",
       "GXE_00023     I  192419                C                A     6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT    39  2014   \n",
       "GXE_00026     I  192427                T                C   902  2000   \n",
       "GXE_00027     I  192429                G                C  1506  1998   \n",
       "GXE_00033     I  192466           AAAGGG                A     3  2020   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "GXE_00027  0.754000                12|190           4697|3618|173   \n",
       "GXE_00033  0.001485                50|148           4735|3656|211   \n",
       "\n",
       "                           region  ...  ACG  ACI  ACQ  ACS  ADE  ADF  ADG  \\\n",
       "var_id                             ...                                      \n",
       "GXE_00023  bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_00024  bidirectional_promoter  ...  0/0  0/0  0/0  0/0  NaN  0/0  0/0   \n",
       "GXE_00026  bidirectional_promoter  ...  1/1  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_00027  bidirectional_promoter  ...  1/1  1/1  1/1  1/1  NaN  1/1  1/1   \n",
       "GXE_00033  bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "\n",
       "           ADI  ADR SACE_MAA  \n",
       "var_id                        \n",
       "GXE_00023  0/0  0/0      0/0  \n",
       "GXE_00024  0/0  0/0      0/0  \n",
       "GXE_00026  1/1  1/1      0/0  \n",
       "GXE_00027  1/1  1/1      0/0  \n",
       "GXE_00033  0/0  0/0      0/0  \n",
       "\n",
       "[5 rows x 58 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# find singletons and doubletons where at least 1 alt allele found in 15 parental strains used to map QTLs\n",
    "df = gxe_oligos_final[~( (gxe_oligos_final.loc[:,yeast_strains_id] == '0/0') | (gxe_oligos_final.loc[:,yeast_strains_id].isna()) ).all(axis=1)].query('AC<3')\n",
    "\n",
    "# check if parental strains carry alt allele that matches variant entry\n",
    "singletons_doubletons_include_list = []\n",
    "for v, row in df.iterrows():\n",
    "    alt_allele_num = row.all_alt_alleles.split(',').index(row.ALT)+1\n",
    "    if any([str(alt_allele_num) in gt for gt in row[yeast_strains_id].fillna('unknown')]):\n",
    "        singletons_doubletons_include_list.append(v)\n",
    "        \n",
    "# finally, subset gxe_oligos_final to remove singletons and doubletons, unless variant is in singletons_doubletons_include_list\n",
    "gxe_oligos_final_no_singletons = gxe_oligos_final.query('AC>=3 | var_id.isin(@singletons_doubletons_include_list)')\n",
    "display(gxe_oligos_final_no_singletons.shape)\n",
    "display(gxe_oligos_final_no_singletons.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/groups/hbfraser/modules/packages/conda/4.6.14/envs/crispey-bar/lib/python3.7/site-packages/pandas/core/indexing.py:376: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/home/groups/hbfraser/modules/packages/conda/4.6.14/envs/crispey-bar/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item] = s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(4384, 60)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>ACQ</th>\n",
       "      <th>ACS</th>\n",
       "      <th>ADE</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADG</th>\n",
       "      <th>ADI</th>\n",
       "      <th>ADR</th>\n",
       "      <th>SACE_MAA</th>\n",
       "      <th>assoc_traits</th>\n",
       "      <th>assoc_traits_str</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00027</th>\n",
       "      <td>I</td>\n",
       "      <td>192429</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1506</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.754000</td>\n",
       "      <td>12|190</td>\n",
       "      <td>4697|3618|173</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00033</th>\n",
       "      <td>I</td>\n",
       "      <td>192466</td>\n",
       "      <td>AAAGGG</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001485</td>\n",
       "      <td>50|148</td>\n",
       "      <td>4735|3656|211</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT    AC    AN  \\\n",
       "var_id                                                                  \n",
       "GXE_00023     I  192419                C                A     6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT    39  2014   \n",
       "GXE_00026     I  192427                T                C   902  2000   \n",
       "GXE_00027     I  192429                G                C  1506  1998   \n",
       "GXE_00033     I  192466           AAAGGG                A     3  2020   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "GXE_00027  0.754000                12|190           4697|3618|173   \n",
       "GXE_00033  0.001485                50|148           4735|3656|211   \n",
       "\n",
       "                           region  ...  ACQ  ACS  ADE  ADF  ADG  ADI  ADR  \\\n",
       "var_id                             ...                                      \n",
       "GXE_00023  bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_00024  bidirectional_promoter  ...  0/0  0/0  NaN  0/0  0/0  0/0  0/0   \n",
       "GXE_00026  bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  1/1  1/1   \n",
       "GXE_00027  bidirectional_promoter  ...  1/1  1/1  NaN  1/1  1/1  1/1  1/1   \n",
       "GXE_00033  bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "\n",
       "          SACE_MAA                                       assoc_traits  \\\n",
       "var_id                                                                  \n",
       "GXE_00023      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "GXE_00024      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "GXE_00026      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "GXE_00027      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "GXE_00033      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "\n",
       "                                            assoc_traits_str  \n",
       "var_id                                                        \n",
       "GXE_00023  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "GXE_00024  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "GXE_00026  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "GXE_00027  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "GXE_00033  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "\n",
       "[5 rows x 60 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# use QTL/gene coordinates to associate each variant with traits in gxe_qtls\n",
    "qtls_df = gxe_qtls[['trait', 'chrom', 'start', 'end', 'qtl_left', 'qtl_right']].copy()\n",
    "qtls_df.loc[:,'chrom'] = qtls_df['chrom'].str.replace('chr', '')\n",
    "# find gene upstream and downstream boundaries\n",
    "qtls_df.loc[:, 'gene_left'] = qtls_df['start']-500\n",
    "qtls_df.loc[:, 'gene_right'] = qtls_df['end']+500\n",
    "# define search boundaries by either gene or QTL boundary, whichever is larger\n",
    "qtls_df.loc[:,'left_limit'] = qtls_df[['gene_left', 'qtl_left']].min(axis=1)\n",
    "qtls_df.loc[:,'right_limit'] = qtls_df[['gene_right', 'qtl_right']].max(axis=1)\n",
    "qtls_df.loc[:,'left_limit'] = qtls_df['left_limit'].astype(int)\n",
    "qtls_df.loc[:,'right_limit'] = qtls_df['right_limit'].astype(int)\n",
    "qtls_df.loc[:, 'region_length'] = qtls_df['right_limit'] - qtls_df['left_limit'] + 1\n",
    "# remove search boundaries duplicates\n",
    "qtls_df = qtls_df[['trait', 'chrom', 'left_limit', 'right_limit']].drop_duplicates()\n",
    "\n",
    "# associate each variant with traits of interest if it falls within QTL/gene search boundaries\n",
    "gxe_oligos_final_no_singletons.loc[:,'assoc_traits'] = gxe_oligos_final_no_singletons.apply(lambda x: qtls_df.query('chrom==@x.CHROM & left_limit<=@x.POS & right_limit>=@x.POS').trait.unique().tolist(), axis=1)\n",
    "gxe_oligos_final_no_singletons.loc[:,'assoc_traits_str'] = gxe_oligos_final_no_singletons['assoc_traits'].apply(lambda x: '_'.join(x))\n",
    "\n",
    "display(gxe_oligos_final_no_singletons.shape)\n",
    "display(gxe_oligos_final_no_singletons.head())\n",
    "\n",
    "# write gxe_oligos_final_no_singletons table to file\n",
    "gxe_oligos_final_no_singletons.to_csv(gxe_variants_file.replace('.vcf', '_final.txt'), sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of GxE variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants: 4384\n",
      "Number of oligos: 10983\n",
      "\n",
      "Distribution of variants among gxe genes\n",
      "count     53.000000\n",
      "mean      73.245283\n",
      "std       44.006951\n",
      "min       10.000000\n",
      "25%       39.000000\n",
      "50%       68.000000\n",
      "75%       98.000000\n",
      "max      189.000000\n",
      "Name: Gene_ID, dtype: float64\n",
      "\n",
      "Distribution of oligos among gxe variants\n",
      "count    4384.000000\n",
      "mean        2.505246\n",
      "std         0.808900\n",
      "min         2.000000\n",
      "25%         2.000000\n",
      "50%         2.000000\n",
      "75%         3.000000\n",
      "max         7.000000\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Proportion of variant annotation classes\n",
      "synonymous_variant         0.377965\n",
      "missense_variant           0.288549\n",
      "bidirectional_promoter     0.140511\n",
      "unidirectional_promoter    0.123859\n",
      "intergenic                 0.055201\n",
      "frameshift_variant         0.008212\n",
      "stop_gained                0.003650\n",
      "inframe_insertion          0.001369\n",
      "inframe_deletion           0.000456\n",
      "stop_lost                  0.000228\n",
      "Name: region, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of variants:\", len(gxe_oligos_final_no_singletons))\n",
    "print(\"Number of oligos:\", int(sum(gxe_oligos_final_no_singletons.num_of_oligos)))\n",
    "print()\n",
    "print(\"Distribution of variants among gxe genes\")\n",
    "print(gxe_oligos_final_no_singletons.Gene_ID.value_counts().filter(gxe_qtls.gene_id.unique()).describe())\n",
    "print()\n",
    "print(\"Distribution of oligos among gxe variants\")\n",
    "print(gxe_oligos_final_no_singletons.num_of_oligos.describe())\n",
    "print()\n",
    "print(\"Proportion of variant annotation classes\")\n",
    "print(gxe_oligos_final_no_singletons.region.value_counts()/len(gxe_oligos_final_no_singletons))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine number of wells required in matrixed oligo synthesis array\n",
    "Twist Biosciences synthesizes 121 oligos per well, up to 384 wells per plate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants per condition(s)\n",
      "assoc_traits_str\n",
      "Caffeine;15mM;2                                                879\n",
      "Caffeine;15mM;2_Cobalt_Chloride;2mM;2                           41\n",
      "Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2          163\n",
      "Caffeine;15mM;2_Fructose;;1                                      1\n",
      "Caffeine;15mM;2_Neomycin;5mg/mL;2                                5\n",
      "Cobalt_Chloride;2mM;2                                          611\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2                       74\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1            4\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2    172\n",
      "Cobalt_Chloride;2mM;2_Fructose;;1                               70\n",
      "Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2             48\n",
      "Fluconazole;100uM;2                                            531\n",
      "Fluconazole;100uM;2_Fructose;;1                                 32\n",
      "Fructose;;1                                                    632\n",
      "Lithium_Chloride;100mM;2                                       659\n",
      "Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2                      53\n",
      "Neomycin;5mg/mL;2                                              409\n",
      "dtype: int64\n",
      "\n",
      "Number of oligos per condition(s)\n",
      "assoc_traits_str\n",
      "Caffeine;15mM;2                                                2171.0\n",
      "Caffeine;15mM;2_Cobalt_Chloride;2mM;2                           117.0\n",
      "Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2           373.0\n",
      "Caffeine;15mM;2_Fructose;;1                                       2.0\n",
      "Caffeine;15mM;2_Neomycin;5mg/mL;2                                12.0\n",
      "Cobalt_Chloride;2mM;2                                          1539.0\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2                       175.0\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1            10.0\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2     457.0\n",
      "Cobalt_Chloride;2mM;2_Fructose;;1                               170.0\n",
      "Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2             140.0\n",
      "Fluconazole;100uM;2                                            1376.0\n",
      "Fluconazole;100uM;2_Fructose;;1                                  73.0\n",
      "Fructose;;1                                                    1534.0\n",
      "Lithium_Chloride;100mM;2                                       1660.0\n",
      "Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2                      146.0\n",
      "Neomycin;5mg/mL;2                                              1028.0\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Trait: Caffeine;15mM;2\n",
      "Number of variants: 1089\n",
      "Number of oligos: 2675\n",
      "\n",
      "Trait: Fluconazole;100uM;2\n",
      "Number of variants: 976\n",
      "Number of oligos: 2464\n",
      "\n",
      "Trait: Neomycin;5mg/mL;2\n",
      "Number of variants: 850\n",
      "Number of oligos: 2156\n",
      "\n",
      "Trait: Cobalt_Chloride;2mM;2\n",
      "Number of variants: 1020\n",
      "Number of oligos: 2608\n",
      "\n",
      "Trait: Fructose;;1\n",
      "Number of variants: 787\n",
      "Number of oligos: 1929\n",
      "\n",
      "Trait: Lithium_Chloride;100mM;2\n",
      "Number of variants: 712\n",
      "Number of oligos: 1806\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('Number of variants per condition(s)')\n",
    "print(gxe_oligos_final_no_singletons.groupby('assoc_traits_str').size())\n",
    "print()\n",
    "print('Number of oligos per condition(s)')\n",
    "print(gxe_oligos_final_no_singletons.groupby('assoc_traits_str').num_of_oligos.sum())\n",
    "print()\n",
    "\n",
    "# count variants and oligos per trait\n",
    "for trait in gxe_qtls.trait.unique():\n",
    "    print(\"Trait:\", trait)\n",
    "    print('Number of variants:', len(gxe_oligos_final_no_singletons[gxe_oligos_final_no_singletons.assoc_traits.apply(lambda x: trait in x)]) )\n",
    "    print('Number of oligos:', int(gxe_oligos_final_no_singletons[gxe_oligos_final_no_singletons.assoc_traits.apply(lambda x: trait in x)]['num_of_oligos'].sum()) )\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GXE    10983.0\n",
       "Name: num_of_oligos, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos to fit: 10983\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(93, 100)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use 93 pools:\n",
      "Number of oligos per pool: 118\n",
      "Number of oligos leftover: 9\n",
      "Number of technical oligos: 279\n",
      "\n",
      "Use 94 pools:\n",
      "Number of oligos per pool: 116\n",
      "Number of oligos leftover: 79\n",
      "Number of technical oligos: 470\n",
      "\n",
      "Use 95 pools:\n",
      "Number of oligos per pool: 115\n",
      "Number of oligos leftover: 58\n",
      "Number of technical oligos: 570\n",
      "\n",
      "Use 96 pools:\n",
      "Number of oligos per pool: 114\n",
      "Number of oligos leftover: 39\n",
      "Number of technical oligos: 672\n",
      "\n",
      "Use 97 pools:\n",
      "Number of oligos per pool: 113\n",
      "Number of oligos leftover: 22\n",
      "Number of technical oligos: 776\n",
      "\n",
      "Use 98 pools:\n",
      "Number of oligos per pool: 112\n",
      "Number of oligos leftover: 7\n",
      "Number of technical oligos: 882\n",
      "\n",
      "Use 99 pools:\n",
      "Number of oligos per pool: 110\n",
      "Number of oligos leftover: 93\n",
      "Number of technical oligos: 1089\n",
      "\n",
      "Use 100 pools:\n",
      "Number of oligos per pool: 109\n",
      "Number of oligos leftover: 83\n",
      "Number of technical oligos: 1200\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([ 93,  94,  95,  96,  97,  98,  99, 100])"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count OLIGOS in each variant prefix set\n",
    "display(gxe_oligos_final_no_singletons.groupby(lambda x: x[:3]).num_of_oligos.sum().sort_values(ascending=False))\n",
    "\n",
    "# determine number of pools to use\n",
    "num_of_oligos = int(gxe_oligos_final_no_singletons.num_of_oligos.sum())\n",
    "min_pool_size = 109 # allows 12 technical oligos\n",
    "max_pool_size = 118 # allows 3 technical oligos\n",
    "complete_pool_size = 121 # maximum oligos synthesized per pool\n",
    "\n",
    "# find number of pools to use\n",
    "find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selecting variants in ergosterol pathway genes for GxE project\n",
    "Screen natural variation in ergosterol genes and test for GxE effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/ergosterol/\"\n",
    "# list of ergosterol genes\n",
    "ergosterol_genes_file = working_dir+\"ergosterol_biosynthetic_process_genes.txt\"\n",
    "# summary of genetic interactions output file\n",
    "genetic_interactions_summary_file = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/genetic_interactions_summary_by_gene.txt\"\n",
    "os.chdir(working_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26, 12)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutant Information</th>\n",
       "      <th>Breslow fitness</th>\n",
       "      <th>pathway</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>description</th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YPL028W</th>\n",
       "      <td>ERG10</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.935</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXVI</td>\n",
       "      <td>498096</td>\n",
       "      <td>499292</td>\n",
       "      <td>+</td>\n",
       "      <td>Acetyl-CoA C-acetyltransferase (acetoacetyl-Co...</td>\n",
       "      <td>152.0</td>\n",
       "      <td>263.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR208W</th>\n",
       "      <td>ERG12</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>1.016</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>684467</td>\n",
       "      <td>685798</td>\n",
       "      <td>+</td>\n",
       "      <td>Mevalonate kinase, acts in the biosynthesis of...</td>\n",
       "      <td>27.0</td>\n",
       "      <td>107.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YML126C</th>\n",
       "      <td>ERG13</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.929</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>19060</td>\n",
       "      <td>20535</td>\n",
       "      <td>-</td>\n",
       "      <td>3-hydroxy-3-methylglutaryl-CoA (HMG-CoA) synth...</td>\n",
       "      <td>206.0</td>\n",
       "      <td>316.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJL167W</th>\n",
       "      <td>ERG20</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.995</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrX</td>\n",
       "      <td>105014</td>\n",
       "      <td>106072</td>\n",
       "      <td>+</td>\n",
       "      <td>Farnesyl pyrophosphate synthetase, has both di...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR220W</th>\n",
       "      <td>ERG8</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.953</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>712316</td>\n",
       "      <td>713671</td>\n",
       "      <td>+</td>\n",
       "      <td>Phosphomevalonate kinase, an essential cytosol...</td>\n",
       "      <td>102.0</td>\n",
       "      <td>150.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name                                            Details  \\\n",
       "gene_id                                                                \n",
       "YPL028W     ERG10  Media: minimal medium Details: Relative fitnes...   \n",
       "YMR208W     ERG12  Media: minimal medium Details: Relative fitnes...   \n",
       "YML126C     ERG13  Media: minimal medium Details: Relative fitnes...   \n",
       "YJL167W     ERG20  Media: minimal medium Details: Relative fitnes...   \n",
       "YMR220W      ERG8  Media: minimal medium Details: Relative fitnes...   \n",
       "\n",
       "            Mutant Information  Breslow fitness   pathway    chrom   start  \\\n",
       "gene_id                                                                      \n",
       "YPL028W  reduction of function            0.935  upstream   chrXVI  498096   \n",
       "YMR208W  reduction of function            1.016  upstream  chrXIII  684467   \n",
       "YML126C  reduction of function            0.929  upstream  chrXIII   19060   \n",
       "YJL167W  reduction of function            0.995  upstream     chrX  105014   \n",
       "YMR220W  reduction of function            0.953  upstream  chrXIII  712316   \n",
       "\n",
       "            end strand                                        description  \\\n",
       "gene_id                                                                     \n",
       "YPL028W  499292      +  Acetyl-CoA C-acetyltransferase (acetoacetyl-Co...   \n",
       "YMR208W  685798      +  Mevalonate kinase, acts in the biosynthesis of...   \n",
       "YML126C   20535      -  3-hydroxy-3-methylglutaryl-CoA (HMG-CoA) synth...   \n",
       "YJL167W  106072      +  Farnesyl pyrophosphate synthetase, has both di...   \n",
       "YMR220W  713671      +  Phosphomevalonate kinase, an essential cytosol...   \n",
       "\n",
       "         positive_interactions  negative_interactions  \n",
       "gene_id                                                \n",
       "YPL028W                  152.0                  263.0  \n",
       "YMR208W                   27.0                  107.0  \n",
       "YML126C                  206.0                  316.0  \n",
       "YJL167W                    NaN                    NaN  \n",
       "YMR220W                  102.0                  150.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read in ergosterol genes table\n",
    "ergosterol_genes = pd.read_csv(ergosterol_genes_file, sep='\\t')\n",
    "ergosterol_genes = ergosterol_genes.rename(columns={'Gene Systematic Name':'gene_id', 'Gene':'gene_name'}).set_index('gene_id')\n",
    "\n",
    "# annotate ergosterol genes\n",
    "ergosterol_genes = ergosterol_genes.merge(gene_info_df.drop('gene_name', axis=1), how='left', left_index=True, right_index=True)\n",
    "\n",
    "# add genetic interaction data\n",
    "genetic_interactions = pd.read_csv(genetic_interactions_summary_file, sep='\\t', index_col=0)\n",
    "ergosterol_genes = ergosterol_genes.merge(genetic_interactions, how='left', left_index=True, right_index=True)\n",
    "display(ergosterol_genes.shape)\n",
    "display(ergosterol_genes.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scan 1,011 genomes VCF for variants in and around gene of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search ergosterol genes for variants in each gvcf\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16',}\n",
    "\n",
    "genes_to_process = []\n",
    "for chrom in chromosome_dict.keys():\n",
    "    genes_df = ergosterol_genes.query('chrom==@chrom').sort_values('start')\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict[chrom])\n",
    "    genes_to_process.append((genes_df, gvcf_file, chromosome_dict))\n",
    "    \n",
    "# consolidate candidate variants\n",
    "num_of_cores = min(len(os.sched_getaffinity(0)), len(genes_to_process))\n",
    "with mp.Pool(num_of_cores) as pool:\n",
    "    results = pool.starmap(select_variants, genes_to_process)\n",
    "results = pd.concat(results).reset_index(drop=True)\n",
    "\n",
    "# write to new vcf file\n",
    "ergosterol_variants_file = \"ergosterol_variants.vcf\"\n",
    "template = vcf.Reader(filename=genes_to_process[0][1])\n",
    "# change chromosome naming convention to a single Roman numeral (follows ref genome, VEP formatting)\n",
    "chromosome_dict_rev = {v:k for k, v in chromosome_dict.items()} \n",
    "template.contigs = OrderedDict((chromosome_dict_rev[k][3:], v._replace(id=chromosome_dict_rev[k][3:])) for k, v in template.contigs.items())\n",
    "# remove 1011 genomes samples data\n",
    "template.samples=[]\n",
    "\n",
    "output = vcf.Writer(open(ergosterol_variants_file, 'w'), template)\n",
    "for record in results:\n",
    "    # update chromosome naming\n",
    "    record.CHROM = chromosome_dict_rev[record.CHROM][3:]\n",
    "    # write to file\n",
    "    output.write_record(record)\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter variants for those that can be targeted for REF to ALT allele editing\n",
    "Use a 60 bp window to ensure the same sequence (ref allele +- 30bp) can be found in four yeast strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to genome fasta files\n",
    "ref_genome_fasta = \"/home/users/rang/yeast/genomes/Saccharomyces_cerevisiae.R64-1-1.dna.chromosome.I.fa\"\n",
    "rm_genome_fasta = \"/home/users/rang/yeast/genomes/RM11-1A_SGD_2015_JRIP00000000.fsa\"\n",
    "yjm_genome_fasta = \"/home/users/rang/yeast/genomes/YJM789_Stanford_2007_AAFW02000000_highQuality31.fsa\"\n",
    "yps_genome_fasta = \"/home/users/rang/yeast/genomes/YPS128.genome.fa\"\n",
    "\n",
    "# load genome files\n",
    "ref_genome = SeqIO.to_dict(SeqIO.parse(ref_genome_fasta, \"fasta\"))\n",
    "rm_genome = SeqIO.to_dict(SeqIO.parse(rm_genome_fasta, \"fasta\"))\n",
    "yjm_genome = SeqIO.to_dict(SeqIO.parse(yjm_genome_fasta, \"fasta\"))\n",
    "yps_genome = SeqIO.to_dict(SeqIO.parse(yps_genome_fasta, \"fasta\"))\n",
    "genomes_list = [ref_genome, rm_genome, yjm_genome, yps_genome] # IMPORTANT: SET REFERENCE GENOME AS FIRST IN LIST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to input and output VCF files\n",
    "ergosterol_variants_file = \"ergosterol_variants.vcf\"\n",
    "ergosterol_variants_select_file = ergosterol_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "threshold = 10 # threshold value for searching NGGs\n",
    "window = 61 # size of flanking sequences to search for unique hits\n",
    "\n",
    "# read in all candidate variants to filter\n",
    "ergosterol_variants = vcf.Reader(filename=ergosterol_variants_file)\n",
    "ergosterol_variants = [rec for rec in ergosterol_variants]\n",
    "# split variants into chunks for multiprocessing\n",
    "n = len(ergosterol_variants)//len(os.sched_getaffinity(0)) + 1\n",
    "ergosterol_variants = [ergosterol_variants[i:i+n] for i in range(0, len(ergosterol_variants), n)]\n",
    "\n",
    "# find targetable variants (TGT info field indicates which genomes variant is targetable in)\n",
    "to_process = [(rec_list, genomes_list, threshold, window) for rec_list in ergosterol_variants]\n",
    "with mp.Pool(len(os.sched_getaffinity(0))) as pool:\n",
    "    result = pool.starmap(find_targetable_variants, to_process) \n",
    "\n",
    "# write targetable variants to file\n",
    "header = vcf.Reader(filename=ergosterol_variants_file)\n",
    "# add \"TGT\" info field into VCF header\n",
    "header_info = namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "header.infos['TGT'] = header_info_tgt\n",
    "# write to file\n",
    "ergosterol_variants_select = vcf.Writer(open(ergosterol_variants_select_file, 'w'), header)\n",
    "to_write = [rec for rec_list in result for rec in rec_list]\n",
    "for rec in to_write:\n",
    "    ergosterol_variants_select.write_record(rec)\n",
    "ergosterol_variants_select.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter variants by allele frequency\n",
    "1. For variants with multiple alternate alleles, split variant entry into separate lines in VCF for library design - allows oligo design pipeline to synthesize separate oligos for each type of REF-to-ALT edit\n",
    "\n",
    "Variants are assigned an ID prefix \"ERG\" or \"EG_\" to denote their use in the CRISPEY-BAR GxE project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select variants by allele frequency (currently disabled singleton/doubleton filter)\n",
    "### if multiple alternate alleles, SPLIT VARIANT ENTRY INTO SEPARATE ENTRIES in VCF (useful for oligo design pipeline)\n",
    "### assign unitl variant ID prefix and number (use format VAR_XXXXX)\n",
    "\n",
    "# candidate variants editable in all strains\n",
    "ergosterol_variants_select = vcf.Reader(filename=ergosterol_variants_select_file)\n",
    "\n",
    "# open output file\n",
    "ergosterol_variants_select_AFfilter_file = ergosterol_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "ergosterol_variants_select_AFfilter = vcf.Writer(open(ergosterol_variants_select_AFfilter_file, 'w'), ergosterol_variants_select)\n",
    "\n",
    "var_counter = {}\n",
    "for record in ergosterol_variants_select:        \n",
    "    # (optional step to skip singletons and doubletons -- DISABLED by setting ac<1)\n",
    "    if all([ac<1 for ac in record.INFO['AC']]):\n",
    "        continue\n",
    "    # if at least one alternate allele meets minimum count, write to VCF\n",
    "    else:\n",
    "        alt = record.ALT[:]\n",
    "        ac = record.INFO['AC'][:]\n",
    "        af = record.INFO['AF'][:]\n",
    "        mleac = record.INFO['MLEAC'][:]\n",
    "        mleaf = record.INFO['MLEAF'][:]\n",
    "        \n",
    "        # split each alt allele into its own variant entry (useful if a locus has multiple possible alleles)\n",
    "        for i in range(len(alt)):\n",
    "            record.ALT = [alt[i]]\n",
    "            record.INFO['AC'] = [ac[i]]\n",
    "            record.INFO['AF'] = [af[i]]\n",
    "            record.INFO['MLEAC'] = [mleac[i]]\n",
    "            record.INFO['MLEAF'] = [mleaf[i]]\n",
    "\n",
    "            # inspect TGT field and assign variant ID prefix\n",
    "            ### if tgt field contains all 1 (i.e. targetable by all queried genomes), assign \"ERG\"\n",
    "            ### otherwise, assign unique prefix \"EG-\" with the last digit determined by converting binary to hexadecimal (works well for up to 4 genomes)\n",
    "            if all([i==1 for i in record.INFO['TGT']]):\n",
    "                var_id_prefix = \"ERG\"\n",
    "            else:\n",
    "                binary_str = ''.join([str(x) for x in record.INFO['TGT']])\n",
    "                hex_digit = hex(int(binary_str, 2))[2:].upper()\n",
    "                var_id_prefix = \"EG{}\".format(hex_digit)\n",
    "\n",
    "            # add to counter\n",
    "            if var_id_prefix not in var_counter:\n",
    "                var_counter[var_id_prefix] = 1\n",
    "            else:\n",
    "                var_counter[var_id_prefix] += 1\n",
    "\n",
    "            # add unique variant ID (follow format VAR_XXXXX)\n",
    "            record.ID='{}_{:05d}'.format(var_id_prefix, var_counter[var_id_prefix])\n",
    "\n",
    "            # write to output\n",
    "            ergosterol_variants_select_AFfilter.write_record(record)\n",
    "            \n",
    "ergosterol_variants_select_AFfilter.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit ergosterol_variants_select_AFfilter VCF to VEP to get annotations, and download VEP output in TXT format (vep_output_filename).\n",
    "http://uswest.ensembl.org/Saccharomyces_cerevisiae/Tools/VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add annotations to variants and remove variants that fall in non-ergosterol genes.\n",
    "ergosterol_variants_file = \"ergosterol_variants.vcf\"\n",
    "ergosterol_variants_select_file = ergosterol_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# get VEP annotations and remove variants that fall in non-ergosterol pathway genes\n",
    "ergosterol_variants_select_AFfilter_file = ergosterol_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "vep_output_filename = ergosterol_variants_select_AFfilter_file.replace('.vcf', '_VEPoutput.txt')\n",
    "variants_annotated_filename = vep_output_filename.replace('_VEPoutput.txt', '_annotated.txt')\n",
    "variants_annotated_df = annotate_variants_by_VEPoutput(ergosterol_variants_select_AFfilter_file, \n",
    "                                                       vep_output_filename, \n",
    "                                                       gff_file, \n",
    "                                                       variants_annotated_filename, \n",
    "                                                       id_colname='var_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4082, 41)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00002</th>\n",
       "      <td>V</td>\n",
       "      <td>237121</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>2</td>\n",
       "      <td>2381|1630|1339|449|2911</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>449.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>3</td>\n",
       "      <td>2382|1631|1338|448|2910</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>448.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>17</td>\n",
       "      <td>2396|1645|1324|434|2896</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00001</th>\n",
       "      <td>V</td>\n",
       "      <td>237149</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>43</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>30</td>\n",
       "      <td>2409|1658|1311|421|2883</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>30.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>421.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00002</th>\n",
       "      <td>V</td>\n",
       "      <td>237151</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>32</td>\n",
       "      <td>2411|1660|1309|419|2881</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>32.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>419.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT  AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                      \n",
       "ERG_00002     V  237121   G   A   9  2022  0.004451                     2   \n",
       "ERG_00003     V  237122   G   A   9  2022  0.004451                     3   \n",
       "ERG_00004     V  237136   T   C  26  2022  0.013000                    17   \n",
       "EGE_00001     V  237149   A   C  43  2022  0.021000                    30   \n",
       "EGE_00002     V  237151   T   C   2  2022  0.000989                    32   \n",
       "\n",
       "           downstream_distance_str                   region  ... Codons  \\\n",
       "var_id                                                       ...          \n",
       "ERG_00002  2381|1630|1339|449|2911  unidirectional_promoter  ...    NaN   \n",
       "ERG_00003  2382|1631|1338|448|2910  unidirectional_promoter  ...    NaN   \n",
       "ERG_00004  2396|1645|1324|434|2896  unidirectional_promoter  ...    NaN   \n",
       "EGE_00001  2409|1658|1311|421|2883  unidirectional_promoter  ...    NaN   \n",
       "EGE_00002  2411|1660|1309|419|2881  unidirectional_promoter  ...    NaN   \n",
       "\n",
       "          DOMAINS closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                            \n",
       "ERG_00002     NaN                    SAH1               YER043C   \n",
       "ERG_00003     NaN                    SAH1               YER043C   \n",
       "ERG_00004     NaN                    SAH1               YER043C   \n",
       "EGE_00001     NaN                    SAH1               YER043C   \n",
       "EGE_00002     NaN                    SAH1               YER043C   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "ERG_00002    upstream_gene_variant                    2.0   \n",
       "ERG_00003    upstream_gene_variant                    3.0   \n",
       "ERG_00004    upstream_gene_variant                   17.0   \n",
       "EGE_00001    upstream_gene_variant                   30.0   \n",
       "EGE_00002    upstream_gene_variant                   32.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00002                   ERG28               YER044C   \n",
       "ERG_00003                   ERG28               YER044C   \n",
       "ERG_00004                   ERG28               YER044C   \n",
       "EGE_00001                   ERG28               YER044C   \n",
       "EGE_00002                   ERG28               YER044C   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "ERG_00002  downstream_gene_variant                  449.0  \n",
       "ERG_00003  downstream_gene_variant                  448.0  \n",
       "ERG_00004  downstream_gene_variant                  434.0  \n",
       "EGE_00001  downstream_gene_variant                  421.0  \n",
       "EGE_00002  downstream_gene_variant                  419.0  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read in annotated file\n",
    "variants_annotated_df = pd.read_csv(variants_annotated_filename, sep='\\t', index_col=0)\n",
    "\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "ergosterol_design_oligos = variants_annotated_df.query('(Gene_ID.isin(@ergosterol_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(ergosterol_design_oligos.shape)\n",
    "display(ergosterol_design_oligos.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# write variants to ergosterol_design_oligos VCF as a preliminary list to design oligos for\n",
    "ergosterol_design_oligos_vcf_file = ergosterol_variants_file.replace('.vcf', '_design_oligos_initial.vcf')\n",
    "\n",
    "ergosterol_variants_select_AFfilter = vcf.Reader(filename=ergosterol_variants_select_AFfilter_file)\n",
    "ergosterol_design_oligos_vcf = vcf.Writer(open(ergosterol_design_oligos_vcf_file, 'w'), ergosterol_variants_select_AFfilter)\n",
    "for record in ergosterol_variants_select_AFfilter:\n",
    "    if record.ID in ergosterol_design_oligos.index:\n",
    "        ergosterol_design_oligos_vcf.write_record(record)\n",
    "ergosterol_design_oligos_vcf.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run oligo design script (crispey3_design_library.ipynb) with these variants to identify which can be targeted by at least two unique guides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1580, 42)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>3</td>\n",
       "      <td>2382|1631|1338|448|2910</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>448.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>17</td>\n",
       "      <td>2396|1645|1324|434|2896</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>434.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00008</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TG</td>\n",
       "      <td>661</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.327000</td>\n",
       "      <td>72</td>\n",
       "      <td>2451|1700|1268|378|2840</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>72.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00009</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>72</td>\n",
       "      <td>2451|1700|1268|378|2840</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>72.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00010</th>\n",
       "      <td>V</td>\n",
       "      <td>237196</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>69</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.034000</td>\n",
       "      <td>77</td>\n",
       "      <td>2456|1705|1264|374|2836</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>77.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                       \n",
       "ERG_00003     V  237122   G   A    9  2022  0.004451                     3   \n",
       "ERG_00004     V  237136   T   C   26  2022  0.013000                    17   \n",
       "EGE_00008     V  237191   T  TG  661  2022  0.327000                    72   \n",
       "EGE_00009     V  237191   T  TA    6  2022  0.002967                    72   \n",
       "EGE_00010     V  237196   A   T   69  2022  0.034000                    77   \n",
       "\n",
       "           downstream_distance_str                   region  ... DOMAINS  \\\n",
       "var_id                                                       ...           \n",
       "ERG_00003  2382|1631|1338|448|2910  unidirectional_promoter  ...     NaN   \n",
       "ERG_00004  2396|1645|1324|434|2896  unidirectional_promoter  ...     NaN   \n",
       "EGE_00008  2451|1700|1268|378|2840  unidirectional_promoter  ...     NaN   \n",
       "EGE_00009  2451|1700|1268|378|2840  unidirectional_promoter  ...     NaN   \n",
       "EGE_00010  2456|1705|1264|374|2836  unidirectional_promoter  ...     NaN   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00003                    SAH1               YER043C   \n",
       "ERG_00004                    SAH1               YER043C   \n",
       "EGE_00008                    SAH1               YER043C   \n",
       "EGE_00009                    SAH1               YER043C   \n",
       "EGE_00010                    SAH1               YER043C   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "ERG_00003    upstream_gene_variant                    3.0   \n",
       "ERG_00004    upstream_gene_variant                   17.0   \n",
       "EGE_00008    upstream_gene_variant                   72.0   \n",
       "EGE_00009    upstream_gene_variant                   72.0   \n",
       "EGE_00010    upstream_gene_variant                   77.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00003                   ERG28               YER044C   \n",
       "ERG_00004                   ERG28               YER044C   \n",
       "EGE_00008                   ERG28               YER044C   \n",
       "EGE_00009                   ERG28               YER044C   \n",
       "EGE_00010                   ERG28               YER044C   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \n",
       "var_id                                                                   \n",
       "ERG_00003  downstream_gene_variant                  448.0           2.0  \n",
       "ERG_00004  downstream_gene_variant                  434.0           2.0  \n",
       "EGE_00008  downstream_gene_variant                  378.0           2.0  \n",
       "EGE_00009  downstream_gene_variant                  378.0           2.0  \n",
       "EGE_00010  downstream_gene_variant                  374.0           2.0  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "oligo_file = os.path.expanduser(\"~/crispey3/initial_design/Output/all_SNPs_ergosterol_initial_GG_9bp_OLIGO.tab\")\n",
    "ergosterol_oligo_df = pd.read_csv(oligo_file, sep='\\t')\n",
    "\n",
    "ergosterol_design_oligos = pd.read_csv(\"/home/users/rang/scratch/yeast/ergosterol/ergosterol_variants_select_AFfilter_annotated.txt\", sep='\\t', index_col=0)\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "ergosterol_design_oligos = ergosterol_design_oligos.query('(Gene_ID.isin(@ergosterol_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "\n",
    "ergosterol_design_oligos['num_of_oligos'] = ergosterol_oligo_df.groupby('var_id').apply(len)\n",
    "ergosterol_oligos_final = ergosterol_design_oligos.query('num_of_oligos>1').copy()\n",
    "display(ergosterol_oligos_final.shape)\n",
    "display(ergosterol_oligos_final.head())\n",
    "\n",
    "# write ergosterol_oligos_final table to file\n",
    "ergosterol_oligos_final.to_csv(\"/home/users/rang/scratch/yeast/ergosterol/ergosterol_variants_final.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of ergosterol variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants: 1580\n",
      "Number of oligos: 3968\n",
      "\n",
      "Distribution of variants among ergosterol genes\n",
      "count     26.000000\n",
      "mean      49.384615\n",
      "std       26.023185\n",
      "min       19.000000\n",
      "25%       32.250000\n",
      "50%       42.000000\n",
      "75%       57.000000\n",
      "max      114.000000\n",
      "Name: Gene_ID, dtype: float64\n",
      "\n",
      "Distribution of oligos among ergosterol variants\n",
      "count    1580.000000\n",
      "mean        2.511392\n",
      "std         0.858882\n",
      "min         2.000000\n",
      "25%         2.000000\n",
      "50%         2.000000\n",
      "75%         3.000000\n",
      "max         7.000000\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Proportion of variant annotation classes\n",
      "synonymous_variant         0.372152\n",
      "missense_variant           0.236076\n",
      "unidirectional_promoter    0.183544\n",
      "bidirectional_promoter     0.149367\n",
      "intergenic                 0.053165\n",
      "frameshift_variant         0.002532\n",
      "stop_gained                0.001899\n",
      "inframe_deletion           0.001266\n",
      "Name: region, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of variants:\", len(ergosterol_oligos_final))\n",
    "print(\"Number of oligos:\", int(sum(ergosterol_oligos_final.num_of_oligos)))\n",
    "print()\n",
    "print(\"Distribution of variants among ergosterol genes\")\n",
    "print(ergosterol_oligos_final.Gene_ID.value_counts().filter(ergosterol_genes.index).describe())\n",
    "print()\n",
    "print(\"Distribution of oligos among ergosterol variants\")\n",
    "print(ergosterol_oligos_final.num_of_oligos.describe())\n",
    "print()\n",
    "print(\"Proportion of variant annotation classes\")\n",
    "print(ergosterol_oligos_final.region.value_counts()/len(ergosterol_oligos_final))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine number of wells required in matrixed oligo synthesis array\n",
    "Twist Biosciences synthesizes 121 oligos per well, up to 384 wells per plate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ERG    2479.0\n",
       "EGE     469.0\n",
       "EG8     253.0\n",
       "EGB     209.0\n",
       "EG9     196.0\n",
       "EGC     193.0\n",
       "EGD     134.0\n",
       "EGA      35.0\n",
       "Name: num_of_oligos, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos to fit: 3968\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(33, 36)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use 33 pools:\n",
      "Number of oligos per pool: 118\n",
      "Number of oligos leftover: 74\n",
      "Number of technical oligos: 99\n",
      "\n",
      "Use 34 pools:\n",
      "Number of oligos per pool: 116\n",
      "Number of oligos leftover: 24\n",
      "Number of technical oligos: 170\n",
      "\n",
      "Use 35 pools:\n",
      "Number of oligos per pool: 113\n",
      "Number of oligos leftover: 13\n",
      "Number of technical oligos: 280\n",
      "\n",
      "Use 36 pools:\n",
      "Number of oligos per pool: 110\n",
      "Number of oligos leftover: 8\n",
      "Number of technical oligos: 396\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([33, 34, 35, 36])"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count OLIGOS in each variant prefix set\n",
    "display(ergosterol_oligos_final.groupby(lambda x: x[:3]).num_of_oligos.sum().sort_values(ascending=False))\n",
    "\n",
    "# determine number of pools to use\n",
    "num_of_oligos = int(ergosterol_oligos_final.num_of_oligos.sum())\n",
    "min_pool_size = 109 # allows 12 technical oligos\n",
    "max_pool_size = 118 # allows 3 technical oligos\n",
    "complete_pool_size = 121 # maximum oligos synthesized per pool\n",
    "\n",
    "# find number of pools to use\n",
    "find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (crispey-bar)",
   "language": "python",
   "name": "crispey-bar"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
