{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant selection for CRISPEY3 library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of available cores: 20\n"
     ]
    }
   ],
   "source": [
    "import os, vcf, re\n",
    "import multiprocessing as mp\n",
    "from Bio import SeqIO\n",
    "from collections import OrderedDict, namedtuple\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "print(\"Number of available cores:\", len(os.sched_getaffinity(0)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_variants(genes_df, vcf_file, chrom_dict, upstream_length=500, downstream_length=500):\n",
    "    '''\n",
    "    given a dataframe of genes of interest, searches vcf for variants in and around that region\n",
    "    can be parallelized by splitting genes and vcf by chromosome\n",
    "    '''\n",
    "    genes_df = genes_df.sort_values('start')\n",
    "    # define start and end points to search depending on gene strand, and the upstream/downstream search parameters\n",
    "    genes_df.loc[genes_df['strand']=='+', 'start_upstream'] = genes_df['start']-upstream_length\n",
    "    genes_df.loc[genes_df['strand']=='+', 'end_downstream'] = genes_df['end']+downstream_length\n",
    "    genes_df.loc[genes_df['strand']=='-', 'start_upstream'] = genes_df['start']-downstream_length\n",
    "    genes_df.loc[genes_df['strand']=='-', 'end_downstream'] = genes_df['end']+upstream_length\n",
    "    genes_df.loc[:,'start_upstream'] = genes_df['start_upstream'].astype(int)\n",
    "    genes_df.loc[:,'end_downstream'] = genes_df['end_downstream'].astype(int)\n",
    "\n",
    "    # read in vcf file to parse for variants\n",
    "    allVars = vcf.Reader(filename=vcf_file)\n",
    "    candVars = []\n",
    "    for i, row in genes_df.iterrows():\n",
    "        retrieve = allVars.fetch(chrom=chrom_dict[row['chrom']], \n",
    "                                 start=row['start_upstream']-1,\n",
    "                                 end=row['end_downstream'])\n",
    "        for rec in retrieve:\n",
    "            rec.samples=[]\n",
    "            rec.FORMAT=None\n",
    "            candVars.append(rec)\n",
    "    candVars = pd.Series(candVars).drop_duplicates().reset_index(drop=True)\n",
    "    return candVars\n",
    "\n",
    "def ngg_check(sequence, threshold):\n",
    "    '''\n",
    "    searches a Bio.seq object for NGG sequence around the center of the sequence\n",
    "    set threshold on size of window to search for NGG (recommended: 10)\n",
    "    returns a boolean of whether one or more NGG found\n",
    "    '''\n",
    "    hits=0\n",
    "    midpt = (len(sequence)-1)//2\n",
    "    for s in [sequence[midpt-2:midpt+threshold], sequence.reverse_complement()[midpt-2:midpt+threshold]]:\n",
    "        matches = re.finditer('.GG', str(s), re.I)\n",
    "        index_list = [m.start() for m in matches]\n",
    "        hits+=len(index_list)\n",
    "    if hits > 0:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def search_genome(sequence, genome):\n",
    "    '''\n",
    "    searches a genome (stored as dictionary of SeqRecords) for a Bio.Seq sequence (exact string match)\n",
    "    returns number of occurences of the sequence in the genome\n",
    "    '''\n",
    "    hits = 0\n",
    "    for scaffold in genome.values():\n",
    "        # search both sequence and its reverse complement to find hits on forward and reverse strands\n",
    "        for s in [sequence, sequence.reverse_complement()]:\n",
    "            matches = re.finditer(str(s), str(scaffold.seq), re.I)\n",
    "            index_list = [m.start() for m in matches]\n",
    "            hits+=len(index_list)\n",
    "    return hits\n",
    "\n",
    "def filter_targetable_variants(candVars, genomes_list, threshold, window):\n",
    "    '''\n",
    "    Takes a list of VCF records, a list of genomes, a threshold for searching NGGs, and a window size for \n",
    "    flanking sequences, filters and sorts records into two lists, whether they are targetable in all genomes\n",
    "    or just in the reference genome.\n",
    "    '''\n",
    "    ref_gen = genomes_list[0]\n",
    "    candVars_allStrains=[]\n",
    "    candVars_refOnly=[]\n",
    "    \n",
    "    for record in candVars:\n",
    "        edits_in_ref_strain = False\n",
    "        edits_in_nonref_strain = False\n",
    "        ##### FILTERING #####\n",
    "        # 1. Skip variant if near chromosome edge\n",
    "        seq_window = ref_gen[record.CHROM][record.POS-window//2-1:record.POS+window//2]\n",
    "        if len(seq_window) != window:\n",
    "            continue\n",
    "\n",
    "        # 2. is there an 'NGG' sequence around the center of the seq_window?\n",
    "        if not ngg_check(sequence=seq_window.seq, threshold=threshold):\n",
    "            continue\n",
    "\n",
    "        # 3. check genomes if sequence flanking variant (seq_window) is unique\n",
    "        if search_genome(seq_window.seq, ref_gen)==1: # reference genome\n",
    "            edits_in_ref_strain = True\n",
    "            # check remaining genomes to ensure seq_window is unique hit in all of them\n",
    "            edits_in_nonref_strain = True\n",
    "            for gen in genomes_list[1:]:\n",
    "                if search_genome(seq_window.seq, gen)!=1:\n",
    "                    edits_in_nonref_strain=False\n",
    "                    break\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        # 4. store variant info according to whether it can be edited in all strains, or just reference strain\n",
    "        if edits_in_ref_strain:\n",
    "            if edits_in_nonref_strain:\n",
    "                candVars_allStrains.append(record)\n",
    "            else:\n",
    "                candVars_refOnly.append(record)\n",
    "        else:\n",
    "            print(\"Variant skipped: Cannot be edited in reference strain\") # sanity check step: should not appear in output as filters in previous steps should remove all such instances\n",
    "    \n",
    "    return (candVars_allStrains, candVars_refOnly)\n",
    "\n",
    "def find_targetable_variants(candVars, genomes_list, threshold, window):\n",
    "    \"\"\"\n",
    "    UPGRADED FROM filter_targetable_variants\n",
    "    Takes a list of VCF records, a list of genomes (at least 1), a threshold for searching NGGs, and a window size for search sequence.\n",
    "    Filters records for NGG availability and targetability in each genome.\n",
    "    Adds a list of integers (0/1) under \"TGT\" field in VCF INFO section. Number of digits corresponds to genomes provided\n",
    "    For each genome, 1 indicates unique match for sequence is found, otherwise 0.\n",
    "    NOTE: REMEMBER TO ADD \"TGT\" INFO TO VCF HEADER BEFORE WRITING RECORDS TO FILE\n",
    "    header_info = collections.namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "    header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "    \"\"\"\n",
    "    ref_gen = genomes_list[0]\n",
    "\n",
    "    record_list = []\n",
    "    for record in candVars:\n",
    "        tgt_code = [0]*len(genomes_list)\n",
    "        ##### FILTERING #####\n",
    "        # 1. Skip variant if near chromosome edge\n",
    "        seq_window = ref_gen[record.CHROM][record.POS-window//2-1:record.POS+window//2]\n",
    "        if len(seq_window) != window:\n",
    "            continue\n",
    "\n",
    "        # 2. is there an 'NGG' sequence around the center of the seq_window?\n",
    "        if not ngg_check(sequence=seq_window.seq, threshold=threshold):\n",
    "            continue\n",
    "\n",
    "        # 3. check sequence flanking variant (seq_window) for unique match in genome\n",
    "        for i in range(len(genomes_list)):\n",
    "            gen = genomes_list[i]\n",
    "            num_of_matches = search_genome(seq_window.seq, gen)\n",
    "            if num_of_matches==1: # unique match\n",
    "                tgt_code[i]=1\n",
    "        # sequence window must match at least one genome in order to be written\n",
    "        if sum(tgt_code)==0:\n",
    "            print(\"Variant at chrom {}, pos {} skipped: No unique matches found in tested genome(s)\".format(record.CHROM, record.POS))\n",
    "            continue\n",
    "        # 4. add TGT code to variant record, add to record list\n",
    "        else:\n",
    "            record.INFO['TGT'] = tgt_code\n",
    "            record_list.append(record)\n",
    "        \n",
    "    return record_list\n",
    "\n",
    "##### FUNCTIONS TO ANNOTATE VARIANTS, BASED ON EILON'S OLIGO DESIGN PIPELINE #####\n",
    "def get_genes_len(input_gff_filename):\n",
    "    with open(input_gff_filename) as f:\n",
    "        gff_lines = f.readlines()\n",
    "    gff_lines = [l for l in gff_lines if len(l.split('\\t'))>=5 ]    \n",
    "    gff_lines = [l for l in gff_lines if l.split('\\t')[2] == 'gene' ]\n",
    "    genes_ids = [l.split('ID=')[1].split(';Name=')[0] for l in gff_lines]\n",
    "    genes_lens = [ int(l.split('\\t')[4]) - int(l.split('\\t')[3]) + 1 for l in gff_lines]\n",
    "\n",
    "    gene_le_df = pd.DataFrame({'Gene' : genes_ids, 'len_bp' : genes_lens})\n",
    "    gene_le_df.set_index(keys='Gene', drop=False, inplace=True)\n",
    "    gene_le_df.index.name=''\n",
    "    gene_le_df.drop_duplicates(inplace=True)\n",
    "\n",
    "    return(gene_le_df)\n",
    "\n",
    "def get_cdss_len(input_gff_filename):\n",
    "    with open(input_gff_filename) as f:\n",
    "        gff_lines = f.readlines()\n",
    "    gff_lines = [l for l in gff_lines if len(l.split('\\t'))>=5 ]\n",
    "    gff_lines = [l for l in gff_lines if l.split('\\t')[2] == 'CDS' ]\n",
    "    genes_ids = [l.split('Parent=')[1].split(';Name=')[0] for l in gff_lines]\n",
    "    genes_lens = [ int(l.split('\\t')[4]) - int(l.split('\\t')[3]) + 1 for l in gff_lines]\n",
    "\n",
    "    gene_le_df = pd.DataFrame({'Gene' : genes_ids, 'len_bp' : genes_lens})\n",
    "    gene_le_df.set_index(keys='Gene', drop=False, inplace=True)\n",
    "    gene_le_df.index.name=''\n",
    "    gene_le_df = gene_le_df.groupby('Gene').sum()\n",
    "    gene_le_df['Gene'] = gene_le_df.index\n",
    "    gene_le_df.index.name=''\n",
    "    gene_le_df.drop_duplicates(inplace=True)\n",
    "\n",
    "    return(gene_le_df)\n",
    "\n",
    "def get_dubious_genes(input_gff_filename):\n",
    "    with open(input_gff_filename) as f:\n",
    "        gff_lines = f.readlines()\n",
    "    gff_lines = [l for l in gff_lines if len(l.split('\\t'))>=5 ]\n",
    "    gff_lines = [l for l in gff_lines if (l.split('\\t')[2] == 'gene') & (len(re.findall('dubious', l, flags=re.I))>0)]\n",
    "    genes_ids = [l.split('ID=')[1].split(';Name=')[0] for l in gff_lines]\n",
    "    \n",
    "    return genes_ids\n",
    "\n",
    "def annotate_variants_by_VEPoutput(vep_input_vcf_filename, vep_output_filename, input_gff_filename, annotated_res_output_filename, id_colname = 'var_id'):\n",
    "    cdss_len_df = get_cdss_len(input_gff_filename)\n",
    "    genes_len_df = get_genes_len(input_gff_filename)\n",
    "    dubious_genes = get_dubious_genes(input_gff_filename)\n",
    "    \n",
    "    vep_df = pd.read_table(vep_output_filename, sep='\\t', na_values = \"-\", low_memory=False)\n",
    "    vep_df = vep_df.loc[~vep_df.Gene.isin(dubious_genes),]\n",
    "    vep_df.rename(columns={\"#Uploaded_variation\" : id_colname}, inplace=True)\n",
    "    \n",
    "    # prepare annotated variants df using VCF supplied to VEP\n",
    "    variants_annotated_df = pd.Series([rec for rec in vcf.Reader(filename=vep_input_vcf_filename)])\n",
    "    var_id_list = []\n",
    "    assemble = []\n",
    "    columns = ['CHROM', 'POS', 'REF', 'ALT', 'AC', 'AN' ,'AF']\n",
    "    for rec in variants_annotated_df:\n",
    "        var_id_list.append(rec.ID)\n",
    "        assemble.append((rec.CHROM, rec.POS, rec.REF, rec.ALT[0], rec.INFO['AC'][0], rec.INFO['AN'], rec.INFO['AF'][0]))\n",
    "    variants_annotated_df = pd.DataFrame.from_records(assemble, index=var_id_list, columns=columns)\n",
    "    variants_annotated_df.index.name = id_colname\n",
    "    \n",
    "    # parse VEP output text file by ID for annotations\n",
    "    entry_cnt = 0\n",
    "    for cur_id, anno in vep_df.groupby(id_colname):\n",
    "        entry_cnt +=1\n",
    "        if (entry_cnt % 5000 == 0):\n",
    "            print(\"parsing entry: %d\"% (entry_cnt))\n",
    "        \n",
    "        # all up/down stream genes (VEP default search range is 5kb from variant position)\n",
    "        cur_upstream_all_str =  '|'.join([str(int(d)) for d in anno.DISTANCE[anno.Consequence.isin(['upstream_gene_variant' ])].values if not np.isnan(d)])\n",
    "        cur_downstream_all_str =  '|'.join([str(int(d)) for d in anno.DISTANCE[anno.Consequence.isin(['downstream_gene_variant' ])].values if not np.isnan(d)])\n",
    "        \n",
    "        variants_annotated_df.loc[cur_id, 'upstream_distance_str'] = cur_upstream_all_str\n",
    "        variants_annotated_df.loc[cur_id, 'downstream_distance_str'] = cur_downstream_all_str\n",
    "\n",
    "        is_up_down_stream = anno.Consequence.isin(['upstream_gene_variant' ,'downstream_gene_variant'])\n",
    "        \n",
    "        # identify intergenic variants\n",
    "        if np.all(is_up_down_stream):\n",
    "            left_genes = ( ((anno.Consequence =='upstream_gene_variant') & (anno.STRAND<0)) |\n",
    "                          ((anno.Consequence =='downstream_gene_variant') & (anno.STRAND>0)) )\n",
    "\n",
    "            right_genes = ( ((anno.Consequence =='upstream_gene_variant') & (anno.STRAND>0)) |\n",
    "                              ((anno.Consequence =='downstream_gene_variant') & (anno.STRAND<0)) )\n",
    "\n",
    "            if (left_genes.sum() > 0):\n",
    "                ind_of_closest_left = anno.loc[left_genes,'DISTANCE'].idxmin()\n",
    "                \n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene1_Gene_Name'] = anno['SYMBOL'][ind_of_closest_left]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene1_Gene_ID'] = anno['Gene'][ind_of_closest_left]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene1_Annotation'] = anno['Consequence'][ind_of_closest_left]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene1_Distance'] = anno['DISTANCE'][ind_of_closest_left]\n",
    "\n",
    "            if (right_genes.sum() > 0):\n",
    "                ind_of_closest_right = anno.loc[right_genes,'DISTANCE'].idxmin()\n",
    "                \n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene2_Gene_Name'] = anno['SYMBOL'][ind_of_closest_right]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene2_Gene_ID'] = anno['Gene'][ind_of_closest_right]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene2_Annotation'] = anno['Consequence'][ind_of_closest_right]\n",
    "                variants_annotated_df.loc[cur_id, 'closest_gene2_Distance'] = anno['DISTANCE'][ind_of_closest_right]\n",
    "\n",
    "            # determine type of noncoding variant\n",
    "            # Caution: variants without genes on one side will be labelled as \"intergenic\"\n",
    "            cur_region = 'intergenic'\n",
    "            if ((left_genes.sum() > 0) and (right_genes.sum() > 0)): \n",
    "                if (anno['Consequence'][ind_of_closest_left] == 'upstream_gene_variant'):\n",
    "                    if (anno['Consequence'][ind_of_closest_right] == 'upstream_gene_variant'):\n",
    "                        cur_region = 'bidirectional_promoter'\n",
    "                    else:\n",
    "                        cur_region = 'unidirectional_promoter'\n",
    "                else:\n",
    "                    if (anno['Consequence'][ind_of_closest_right] == 'upstream_gene_variant'):\n",
    "                        cur_region = 'unidirectional_promoter'\n",
    "\n",
    "            variants_annotated_df.loc[cur_id, 'region'] = cur_region\n",
    "\n",
    "            # index of closest gene\n",
    "            sel_row = anno.DISTANCE.idxmin()\n",
    "        \n",
    "        # for variants that land in a gene, select annotation with highest impact\n",
    "        else:\n",
    "            anno = anno.loc[~is_up_down_stream,:]\n",
    "            if (anno.IMPACT == 'HIGH').sum()>0:\n",
    "                anno = anno.loc[anno.IMPACT == 'HIGH',:]\n",
    "            elif (anno.IMPACT == 'MODERATE').sum()>0:\n",
    "                anno = anno.loc[anno.IMPACT == 'MODERATE',:]\n",
    "            elif (anno.IMPACT == 'LOW').sum()>0:\n",
    "                anno = anno.loc[anno.IMPACT == 'LOW',:]\n",
    "            elif (anno.IMPACT == 'MODIFIER').sum()>0:\n",
    "                anno = anno.loc[anno.IMPACT == 'MODIFIER',:] #NOTE: includes introns, noncoding transcripts within length of gene\n",
    "            \n",
    "            else:\n",
    "                raise ValueError(anno)\n",
    "\n",
    "            sel_row = anno.index[0]\n",
    "            variants_annotated_df.loc[cur_id, 'region'] = anno['Consequence'][sel_row]\n",
    "\n",
    "        # closest annotation recorded\n",
    "        variants_annotated_df.loc[cur_id, 'Annotation'] = anno['Consequence'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Annotation_Impact'] = anno['IMPACT'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Gene_Name'] = anno['SYMBOL'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Gene_ID'] = anno['Gene'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Transcript_BioType'] = anno['BIOTYPE'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'Existing_variation'] = anno['Existing_variation'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'HGVSc'] = anno['HGVSc'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'HGVSp'] = anno['HGVSp'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'SWISSPROT'] = anno['SWISSPROT'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'UNIPARC'] = anno['UNIPARC'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'BLOSUM62'] = anno['BLOSUM62'][sel_row]\n",
    "        variants_annotated_df.loc[cur_id, 'HGVSp'] = anno['HGVSp'][sel_row]\n",
    "#         variants_annotated_df.set_value(index=cur_id,col='VEP_CSN',value=anno['CSN'][sel_row])\n",
    "        \n",
    "        # store cDNA and CDS positions\n",
    "        cdna_pos = anno['cDNA_position'][sel_row]\n",
    "        cds_pos = anno['CDS_position'][sel_row]\n",
    "    \n",
    "        # add additional info for missense and synonymous variants\n",
    "        if (anno['Consequence'][sel_row] in ['synonymous_variant','missense_variant']):\n",
    "            # because variants were extracted from 1011 genomes GVCF, some SNPs have trailing bases.\n",
    "            # Calculate the correct cDNA and CDS positions if so \n",
    "            if '-' in cdna_pos:\n",
    "                cdna_pos = cdna_pos.split('-')[0]\n",
    "            if '-' in cds_pos:\n",
    "                cds_pos = cds_pos.split('-')[0]\n",
    "                \n",
    "            \n",
    "            if anno['Gene'][sel_row] in genes_len_df.index:\n",
    "                cur_gene_len = int(genes_len_df.loc[anno['Gene'][sel_row],'len_bp'])\n",
    "                variants_annotated_df.loc[cur_id, 'cDNA_len'] = cur_gene_len\n",
    "                variants_annotated_df.loc[cur_id, 'cDNA_frac'] = int(cdna_pos)/cur_gene_len\n",
    "            \n",
    "            if anno['Gene'][sel_row] in cdss_len_df.index:\n",
    "                cur_cds_len = int(cdss_len_df.loc[anno['Gene'][sel_row],'len_bp'])\n",
    "                variants_annotated_df.loc[cur_id, 'CDS_len'] = cur_cds_len\n",
    "                variants_annotated_df.loc[cur_id, 'CDS_frac'] = int(cds_pos)/cur_cds_len\n",
    "                variants_annotated_df.loc[cur_id, 'AA_len'] = cur_cds_len/3\n",
    "                variants_annotated_df.loc[cur_id, 'AA_frac'] = int(cds_pos)/cur_cds_len\n",
    "                \n",
    "            variants_annotated_df.loc[cur_id, 'cDNA_pos'] = cdna_pos\n",
    "            variants_annotated_df.loc[cur_id, 'CDS_pos'] = cds_pos\n",
    "            variants_annotated_df.loc[cur_id, 'AA_pos'] = anno['Protein_position'][sel_row]\n",
    "            variants_annotated_df.loc[cur_id, 'Amino_acids'] = anno['Amino_acids'][sel_row]\n",
    "            variants_annotated_df.loc[cur_id, 'Codons'] = anno['Codons'][sel_row]\n",
    "            variants_annotated_df.loc[cur_id, 'DOMAINS'] = anno['DOMAINS'][sel_row]\n",
    "\n",
    "    variants_annotated_df.to_csv(annotated_res_output_filename, sep='\\t', header=True, index=True)\n",
    "    \n",
    "    return(variants_annotated_df)\n",
    "\n",
    "###############################################################################################################\n",
    "def find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size=121):\n",
    "    \"\"\"\n",
    "    helper function to find number of pools required to fit a certain number of oligos,\n",
    "    as well as the average size of the pool\n",
    "    \"\"\"\n",
    "    print('Total number of oligos to fit:', num_of_oligos)\n",
    "\n",
    "    # num of pools to use\n",
    "    num_of_pools_lower = num_of_oligos//max_pool_size\n",
    "    num_of_pools_upper = num_of_oligos//min_pool_size\n",
    "    display((num_of_pools_lower, num_of_pools_upper))\n",
    "\n",
    "    # explore pooling options for all possible pool configurations\n",
    "    for num_of_pools in range(num_of_pools_lower, num_of_pools_upper+1):\n",
    "        if num_of_pools == 0:\n",
    "            continue\n",
    "        print(\"Use {} pools:\".format(num_of_pools))\n",
    "        oligos_per_pool = min((num_of_oligos // num_of_pools), max_pool_size)\n",
    "        print(\"Number of oligos per pool:\", oligos_per_pool)\n",
    "        print(\"Number of oligos leftover:\", num_of_oligos - num_of_pools*oligos_per_pool)\n",
    "        print(\"Number of technical oligos:\", (complete_pool_size - oligos_per_pool) * num_of_pools)\n",
    "        print()\n",
    "    \n",
    "    # return list of possible pool sizes\n",
    "    return np.arange(num_of_pools_lower, num_of_pools_upper+1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse Costanzo dataset to count number of genetic interactions per gene and assemble summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/\"\n",
    "# list of files to parse for genetic interactions\n",
    "file_list = ['SGA_ExE.txt', 'SGA_ExN_NxE.txt', 'SGA_NxN.txt']\n",
    "# summary of genetic interactions output file\n",
    "out_file = working_dir+'genetic_interactions_summary_by_gene.txt'\n",
    "os.chdir(working_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(890107, 17)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_strain</th>\n",
       "      <th>query_allele</th>\n",
       "      <th>array_strain</th>\n",
       "      <th>array_allele</th>\n",
       "      <th>arraytype</th>\n",
       "      <th>genetic_interaction_score</th>\n",
       "      <th>pval</th>\n",
       "      <th>query_smf</th>\n",
       "      <th>array_smf</th>\n",
       "      <th>dmf</th>\n",
       "      <th>dmf_std</th>\n",
       "      <th>genetic_interaction_score_abs</th>\n",
       "      <th>query_gene_id</th>\n",
       "      <th>query_gene_name</th>\n",
       "      <th>array_gene_id</th>\n",
       "      <th>array_gene_name</th>\n",
       "      <th>gene_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL026W_tsa1065</td>\n",
       "      <td>lsm2-5001</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.3529</td>\n",
       "      <td>3.591000e-06</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9408</td>\n",
       "      <td>0.4266</td>\n",
       "      <td>0.0790</td>\n",
       "      <td>0.3529</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL026W</td>\n",
       "      <td>lsm2</td>\n",
       "      <td>YAL001C_YBL026W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL034C_tsa950</td>\n",
       "      <td>stu1-7</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.1294</td>\n",
       "      <td>1.931000e-02</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.6690</td>\n",
       "      <td>0.4249</td>\n",
       "      <td>0.0482</td>\n",
       "      <td>0.1294</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL034C</td>\n",
       "      <td>stu1</td>\n",
       "      <td>YAL001C_YBL034C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL097W_tsa510</td>\n",
       "      <td>brn1-9</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.0808</td>\n",
       "      <td>5.582000e-15</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.5464</td>\n",
       "      <td>0.3719</td>\n",
       "      <td>0.0077</td>\n",
       "      <td>0.0808</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL097W</td>\n",
       "      <td>brn1</td>\n",
       "      <td>YAL001C_YBL097W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBR029C_tsa1063</td>\n",
       "      <td>cds1-5001</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.1173</td>\n",
       "      <td>8.243000e-05</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9007</td>\n",
       "      <td>0.6289</td>\n",
       "      <td>0.0226</td>\n",
       "      <td>0.1173</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBR029C</td>\n",
       "      <td>cds1</td>\n",
       "      <td>YAL001C_YBR029C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBR060C_tsa311</td>\n",
       "      <td>orc2-2</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>0.2516</td>\n",
       "      <td>4.729000e-12</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.7384</td>\n",
       "      <td>0.8634</td>\n",
       "      <td>0.0362</td>\n",
       "      <td>0.2516</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBR060C</td>\n",
       "      <td>orc2</td>\n",
       "      <td>YAL001C_YBR060C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      query_strain query_allele     array_strain array_allele arraytype  \\\n",
       "1   YAL001C_tsq508   tfc3-g349e  YBL026W_tsa1065    lsm2-5001     TSA30   \n",
       "6   YAL001C_tsq508   tfc3-g349e   YBL034C_tsa950       stu1-7     TSA30   \n",
       "15  YAL001C_tsq508   tfc3-g349e   YBL097W_tsa510       brn1-9     TSA30   \n",
       "24  YAL001C_tsq508   tfc3-g349e  YBR029C_tsa1063    cds1-5001     TSA30   \n",
       "30  YAL001C_tsq508   tfc3-g349e   YBR060C_tsa311       orc2-2     TSA30   \n",
       "\n",
       "    genetic_interaction_score          pval  query_smf  array_smf     dmf  \\\n",
       "1                     -0.3529  3.591000e-06     0.8285     0.9408  0.4266   \n",
       "6                     -0.1294  1.931000e-02     0.8285     0.6690  0.4249   \n",
       "15                    -0.0808  5.582000e-15     0.8285     0.5464  0.3719   \n",
       "24                    -0.1173  8.243000e-05     0.8285     0.9007  0.6289   \n",
       "30                     0.2516  4.729000e-12     0.8285     0.7384  0.8634   \n",
       "\n",
       "    dmf_std  genetic_interaction_score_abs query_gene_id query_gene_name  \\\n",
       "1    0.0790                         0.3529       YAL001C            tfc3   \n",
       "6    0.0482                         0.1294       YAL001C            tfc3   \n",
       "15   0.0077                         0.0808       YAL001C            tfc3   \n",
       "24   0.0226                         0.1173       YAL001C            tfc3   \n",
       "30   0.0362                         0.2516       YAL001C            tfc3   \n",
       "\n",
       "   array_gene_id array_gene_name        gene_pair  \n",
       "1        YBL026W            lsm2  YAL001C_YBL026W  \n",
       "6        YBL034C            stu1  YAL001C_YBL034C  \n",
       "15       YBL097W            brn1  YAL001C_YBL097W  \n",
       "24       YBR029C            cds1  YAL001C_YBR029C  \n",
       "30       YBR060C            orc2  YAL001C_YBR060C  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = []\n",
    "for file in file_list:\n",
    "    df.append(pd.read_csv(working_dir+file, sep='\\t', header = 0))\n",
    "df = pd.concat(df, axis=0, join='inner')\n",
    "df.columns = ['query_strain', 'query_allele', 'array_strain', 'array_allele', 'arraytype', 'genetic_interaction_score', 'pval', 'query_smf', 'array_smf', 'dmf', 'dmf_std']\n",
    "df['genetic_interaction_score_abs'] = df['genetic_interaction_score'].abs()\n",
    "df['query_gene_id'] = [x.split('_')[0] for x in df['query_strain']]\n",
    "df['query_gene_name'] = [x.split('-')[0] for x in df['query_allele']]\n",
    "df['array_gene_id'] = [x.split('_')[0] for x in df['array_strain']]\n",
    "df['array_gene_name'] = [x.split('-')[0] for x in df['array_allele']]\n",
    "df['gene_pair'] = df.apply(lambda x: '_'.join(sorted([x['query_gene_id'], x['array_gene_id']])), axis=1)\n",
    "\n",
    "# get list of genes\n",
    "all_genes = set(df.query_gene_id).union(set(df.array_gene_id))\n",
    "\n",
    "# filter down to significant interactions only\n",
    "pval_threshold=0.05\n",
    "interaction_threshold=0.08\n",
    "df = df.query('genetic_interaction_score_abs>@interaction_threshold & pval<@pval_threshold')\n",
    "display(df.shape)\n",
    "display(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(780653, 17)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_strain</th>\n",
       "      <th>query_allele</th>\n",
       "      <th>array_strain</th>\n",
       "      <th>array_allele</th>\n",
       "      <th>arraytype</th>\n",
       "      <th>genetic_interaction_score</th>\n",
       "      <th>pval</th>\n",
       "      <th>query_smf</th>\n",
       "      <th>array_smf</th>\n",
       "      <th>dmf</th>\n",
       "      <th>dmf_std</th>\n",
       "      <th>genetic_interaction_score_abs</th>\n",
       "      <th>query_gene_id</th>\n",
       "      <th>query_gene_name</th>\n",
       "      <th>array_gene_id</th>\n",
       "      <th>array_gene_name</th>\n",
       "      <th>gene_pair</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_pair</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBL010C</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL010C_dma88</td>\n",
       "      <td>ybl010c</td>\n",
       "      <td>DMA30</td>\n",
       "      <td>-0.0818</td>\n",
       "      <td>3.430000e-02</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9870</td>\n",
       "      <td>0.7360</td>\n",
       "      <td>0.0361</td>\n",
       "      <td>0.0818</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL010C</td>\n",
       "      <td>ybl010c</td>\n",
       "      <td>YAL001C_YBL010C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBL026W</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL026W_tsa1065</td>\n",
       "      <td>lsm2-5001</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.3529</td>\n",
       "      <td>3.591000e-06</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9408</td>\n",
       "      <td>0.4266</td>\n",
       "      <td>0.0790</td>\n",
       "      <td>0.3529</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL026W</td>\n",
       "      <td>lsm2</td>\n",
       "      <td>YAL001C_YBL026W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBL034C</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL034C_tsa950</td>\n",
       "      <td>stu1-7</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.1294</td>\n",
       "      <td>1.931000e-02</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.6690</td>\n",
       "      <td>0.4249</td>\n",
       "      <td>0.0482</td>\n",
       "      <td>0.1294</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL034C</td>\n",
       "      <td>stu1</td>\n",
       "      <td>YAL001C_YBL034C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBL097W</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBL097W_tsa510</td>\n",
       "      <td>brn1-9</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.0808</td>\n",
       "      <td>5.582000e-15</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.5464</td>\n",
       "      <td>0.3719</td>\n",
       "      <td>0.0077</td>\n",
       "      <td>0.0808</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBL097W</td>\n",
       "      <td>brn1</td>\n",
       "      <td>YAL001C_YBL097W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL001C_YBR029C</th>\n",
       "      <td>YAL001C_tsq508</td>\n",
       "      <td>tfc3-g349e</td>\n",
       "      <td>YBR029C_tsa1063</td>\n",
       "      <td>cds1-5001</td>\n",
       "      <td>TSA30</td>\n",
       "      <td>-0.1173</td>\n",
       "      <td>8.243000e-05</td>\n",
       "      <td>0.8285</td>\n",
       "      <td>0.9007</td>\n",
       "      <td>0.6289</td>\n",
       "      <td>0.0226</td>\n",
       "      <td>0.1173</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>tfc3</td>\n",
       "      <td>YBR029C</td>\n",
       "      <td>cds1</td>\n",
       "      <td>YAL001C_YBR029C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   query_strain query_allele     array_strain array_allele  \\\n",
       "gene_pair                                                                    \n",
       "YAL001C_YBL010C  YAL001C_tsq508   tfc3-g349e    YBL010C_dma88      ybl010c   \n",
       "YAL001C_YBL026W  YAL001C_tsq508   tfc3-g349e  YBL026W_tsa1065    lsm2-5001   \n",
       "YAL001C_YBL034C  YAL001C_tsq508   tfc3-g349e   YBL034C_tsa950       stu1-7   \n",
       "YAL001C_YBL097W  YAL001C_tsq508   tfc3-g349e   YBL097W_tsa510       brn1-9   \n",
       "YAL001C_YBR029C  YAL001C_tsq508   tfc3-g349e  YBR029C_tsa1063    cds1-5001   \n",
       "\n",
       "                arraytype  genetic_interaction_score          pval  query_smf  \\\n",
       "gene_pair                                                                       \n",
       "YAL001C_YBL010C     DMA30                    -0.0818  3.430000e-02     0.8285   \n",
       "YAL001C_YBL026W     TSA30                    -0.3529  3.591000e-06     0.8285   \n",
       "YAL001C_YBL034C     TSA30                    -0.1294  1.931000e-02     0.8285   \n",
       "YAL001C_YBL097W     TSA30                    -0.0808  5.582000e-15     0.8285   \n",
       "YAL001C_YBR029C     TSA30                    -0.1173  8.243000e-05     0.8285   \n",
       "\n",
       "                 array_smf     dmf  dmf_std  genetic_interaction_score_abs  \\\n",
       "gene_pair                                                                    \n",
       "YAL001C_YBL010C     0.9870  0.7360   0.0361                         0.0818   \n",
       "YAL001C_YBL026W     0.9408  0.4266   0.0790                         0.3529   \n",
       "YAL001C_YBL034C     0.6690  0.4249   0.0482                         0.1294   \n",
       "YAL001C_YBL097W     0.5464  0.3719   0.0077                         0.0808   \n",
       "YAL001C_YBR029C     0.9007  0.6289   0.0226                         0.1173   \n",
       "\n",
       "                query_gene_id query_gene_name array_gene_id array_gene_name  \\\n",
       "gene_pair                                                                     \n",
       "YAL001C_YBL010C       YAL001C            tfc3       YBL010C         ybl010c   \n",
       "YAL001C_YBL026W       YAL001C            tfc3       YBL026W            lsm2   \n",
       "YAL001C_YBL034C       YAL001C            tfc3       YBL034C            stu1   \n",
       "YAL001C_YBL097W       YAL001C            tfc3       YBL097W            brn1   \n",
       "YAL001C_YBR029C       YAL001C            tfc3       YBR029C            cds1   \n",
       "\n",
       "                       gene_pair  \n",
       "gene_pair                         \n",
       "YAL001C_YBL010C  YAL001C_YBL010C  \n",
       "YAL001C_YBL026W  YAL001C_YBL026W  \n",
       "YAL001C_YBL034C  YAL001C_YBL034C  \n",
       "YAL001C_YBL097W  YAL001C_YBL097W  \n",
       "YAL001C_YBR029C  YAL001C_YBR029C  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# find strongest interaction in each pair of genes\n",
    "gene_pairs_df = df.groupby('gene_pair').apply(lambda x: x.sort_values('genetic_interaction_score_abs', ascending=False).iloc[0])\n",
    "display(gene_pairs_df.shape)\n",
    "display(gene_pairs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5707, 2)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFL039C</th>\n",
       "      <td>1324</td>\n",
       "      <td>2220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL034C-B</th>\n",
       "      <td>1150</td>\n",
       "      <td>1897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJR076C</th>\n",
       "      <td>749</td>\n",
       "      <td>1342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBL105C</th>\n",
       "      <td>841</td>\n",
       "      <td>1139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YER157W</th>\n",
       "      <td>925</td>\n",
       "      <td>967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLR148W</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJL189W</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL067W-A</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJL056C</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR014W</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5707 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           positive_interactions  negative_interactions\n",
       "gene_id                                                \n",
       "YFL039C                     1324                   2220\n",
       "YFL034C-B                   1150                   1897\n",
       "YJR076C                      749                   1342\n",
       "YBL105C                      841                   1139\n",
       "YER157W                      925                    967\n",
       "...                          ...                    ...\n",
       "YLR148W                        2                      2\n",
       "YJL189W                        0                      3\n",
       "YAL067W-A                      0                      3\n",
       "YJL056C                        0                      2\n",
       "YMR014W                        1                      0\n",
       "\n",
       "[5707 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# for each gene, record number of positive and negative interactions that meet interaction and pvalue thresholds\n",
    "gene_list = []\n",
    "positive_interactions = []\n",
    "negative_interactions = []\n",
    "for gene in all_genes:\n",
    "    sub = gene_pairs_df.query('query_gene_id==@gene | array_gene_id==@gene')\n",
    "    gene_list.append(gene)\n",
    "    positive_interactions.append(len(sub[sub['genetic_interaction_score']>interaction_threshold]))\n",
    "    negative_interactions.append(len(sub[sub['genetic_interaction_score']<interaction_threshold*-1]))\n",
    "\n",
    "num_of_interactions_df = pd.DataFrame({'positive_interactions':positive_interactions, 'negative_interactions':negative_interactions}, index=gene_list)\n",
    "num_of_interactions_df = num_of_interactions_df.loc[num_of_interactions_df.sum(axis=1).sort_values(ascending=False).index]\n",
    "num_of_interactions_df.index.name = 'gene_id'\n",
    "display(num_of_interactions_df.shape)\n",
    "display(num_of_interactions_df)\n",
    "\n",
    "num_of_interactions_df.to_csv(out_file, sep='\\t', header=True, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read S288C GFF file and extract gene information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info_df = {}\n",
    "gff_file = '/home/users/rang/yeast/genomes/saccharomyces_cerevisiae_R64-1-1_20110208.gff'\n",
    "char_map = {'%20' : ' ',\n",
    "            '%2C' : ',',\n",
    "            '%2F' : '/',\n",
    "            '%3B' : ';'}\n",
    "\n",
    "with open(gff_file, 'r') as gff:\n",
    "    line = gff.readline()\n",
    "    while line.rstrip() != '###':\n",
    "        while line[0] == \"#\":\n",
    "            line = gff.readline()\n",
    "            \n",
    "        chrom, _, feature, start, end, _, strand, _, info = line.rstrip().split('\\t')\n",
    "        if feature=='gene':\n",
    "            info = {x.split('=')[0] : x.split('=')[1] for x in info.split(';')}\n",
    "            gene_id = info['ID']\n",
    "            if 'gene' in info:\n",
    "                gene_name = info['gene']\n",
    "            else:\n",
    "                gene_name = np.nan\n",
    "                \n",
    "            if 'Note' in info:\n",
    "                desc = info['Note']\n",
    "                for c, r in char_map.items():\n",
    "                    desc = desc.replace(c, r)\n",
    "            else:\n",
    "                desc = np.nan\n",
    "            # store gene info\n",
    "            gene_info_df[gene_id] = [chrom, int(start), int(end), strand, gene_name, desc]\n",
    "        line = gff.readline()\n",
    "        \n",
    "gene_info_df = pd.DataFrame.from_dict(gene_info_df, orient='index', columns=['chrom', 'start', 'end', 'strand', 'gene_name', 'description'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFL039C</th>\n",
       "      <td>1324</td>\n",
       "      <td>2220</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>53260.0</td>\n",
       "      <td>54696.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ACT1</td>\n",
       "      <td>Actin, structural protein involved in cell pol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL034C-B</th>\n",
       "      <td>1150</td>\n",
       "      <td>1897</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>63016.0</td>\n",
       "      <td>63993.0</td>\n",
       "      <td>-</td>\n",
       "      <td>MOB2</td>\n",
       "      <td>Component of the RAM signaling network, that a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJR076C</th>\n",
       "      <td>749</td>\n",
       "      <td>1342</td>\n",
       "      <td>chrX</td>\n",
       "      <td>575354.0</td>\n",
       "      <td>576601.0</td>\n",
       "      <td>-</td>\n",
       "      <td>CDC11</td>\n",
       "      <td>Component of the septin ring of the mother-bud...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBL105C</th>\n",
       "      <td>841</td>\n",
       "      <td>1139</td>\n",
       "      <td>chrII</td>\n",
       "      <td>14241.0</td>\n",
       "      <td>17696.0</td>\n",
       "      <td>-</td>\n",
       "      <td>PKC1</td>\n",
       "      <td>Protein serine/threonine kinase essential for ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YER157W</th>\n",
       "      <td>925</td>\n",
       "      <td>967</td>\n",
       "      <td>chrV</td>\n",
       "      <td>484788.0</td>\n",
       "      <td>487193.0</td>\n",
       "      <td>+</td>\n",
       "      <td>COG3</td>\n",
       "      <td>Essential component of the conserved oligomeri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL060C</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>10301.0</td>\n",
       "      <td>10969.0</td>\n",
       "      <td>-</td>\n",
       "      <td>SNO3</td>\n",
       "      <td>Protein of unknown function, nearly identical ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJL189W</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chrX</td>\n",
       "      <td>75933.0</td>\n",
       "      <td>76474.0</td>\n",
       "      <td>+</td>\n",
       "      <td>RPL39</td>\n",
       "      <td>Protein component of the large (60S) ribosomal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL067W-A</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chrI</td>\n",
       "      <td>2480.0</td>\n",
       "      <td>2707.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Putative protein of unknown function; identifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJL056C</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>330431.0</td>\n",
       "      <td>333073.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ZAP1</td>\n",
       "      <td>Zinc-regulated transcription factor; binds to ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR014W</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>298868.0</td>\n",
       "      <td>300427.0</td>\n",
       "      <td>+</td>\n",
       "      <td>BUD22</td>\n",
       "      <td>Protein required for 18S rRNA maturation and s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5707 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           positive_interactions  negative_interactions    chrom     start  \\\n",
       "gene_id                                                                      \n",
       "YFL039C                     1324                   2220    chrVI   53260.0   \n",
       "YFL034C-B                   1150                   1897    chrVI   63016.0   \n",
       "YJR076C                      749                   1342     chrX  575354.0   \n",
       "YBL105C                      841                   1139    chrII   14241.0   \n",
       "YER157W                      925                    967     chrV  484788.0   \n",
       "...                          ...                    ...      ...       ...   \n",
       "YFL060C                        0                      4    chrVI   10301.0   \n",
       "YJL189W                        0                      3     chrX   75933.0   \n",
       "YAL067W-A                      0                      3     chrI    2480.0   \n",
       "YJL056C                        0                      2     chrX  330431.0   \n",
       "YMR014W                        1                      0  chrXIII  298868.0   \n",
       "\n",
       "                end strand gene_name  \\\n",
       "gene_id                                \n",
       "YFL039C     54696.0      -      ACT1   \n",
       "YFL034C-B   63993.0      -      MOB2   \n",
       "YJR076C    576601.0      -     CDC11   \n",
       "YBL105C     17696.0      -      PKC1   \n",
       "YER157W    487193.0      +      COG3   \n",
       "...             ...    ...       ...   \n",
       "YFL060C     10969.0      -      SNO3   \n",
       "YJL189W     76474.0      +     RPL39   \n",
       "YAL067W-A    2707.0      +       NaN   \n",
       "YJL056C    333073.0      -      ZAP1   \n",
       "YMR014W    300427.0      +     BUD22   \n",
       "\n",
       "                                                 description  \n",
       "gene_id                                                       \n",
       "YFL039C    Actin, structural protein involved in cell pol...  \n",
       "YFL034C-B  Component of the RAM signaling network, that a...  \n",
       "YJR076C    Component of the septin ring of the mother-bud...  \n",
       "YBL105C    Protein serine/threonine kinase essential for ...  \n",
       "YER157W    Essential component of the conserved oligomeri...  \n",
       "...                                                      ...  \n",
       "YFL060C    Protein of unknown function, nearly identical ...  \n",
       "YJL189W    Protein component of the large (60S) ribosomal...  \n",
       "YAL067W-A  Putative protein of unknown function; identifi...  \n",
       "YJL056C    Zinc-regulated transcription factor; binds to ...  \n",
       "YMR014W    Protein required for 18S rRNA maturation and s...  \n",
       "\n",
       "[5707 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# merge GFF info with genetic interactions df\n",
    "num_of_interactions_file = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/genetic_interactions_summary_by_gene.txt\"\n",
    "num_of_interactions_df = pd.read_csv(num_of_interactions_file, sep='\\t', header = 0, index_col=0)\n",
    "num_of_interactions_df = num_of_interactions_df.join(gene_info_df).sort_values(['chrom', 'start'])\n",
    "num_of_interactions_df = num_of_interactions_df.loc[num_of_interactions_df[['positive_interactions', 'negative_interactions']].sum(axis=1).sort_values(ascending=False).index]\n",
    "display(num_of_interactions_df)\n",
    "\n",
    "num_of_interactions_df.to_csv(out_file.replace('.txt', '_annotated.txt'), sep='\\t', header=True, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selecting variants in ergosterol pathway genes\n",
    "Survey natural variation in ergosterol genes and test for GxE/GxG effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/ergosterol/\"\n",
    "# list of ergosterol genes\n",
    "ergosterol_genes_file = working_dir+\"ergosterol_biosynthetic_process_genes.txt\"\n",
    "# summary of genetic interactions output file\n",
    "genetic_interactions_summary_file = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/genetic_interactions_summary_by_gene.txt\"\n",
    "\n",
    "os.chdir(working_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutant Information</th>\n",
       "      <th>Breslow fitness</th>\n",
       "      <th>pathway</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>description</th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YPL028W</th>\n",
       "      <td>ERG10</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.935</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXVI</td>\n",
       "      <td>498096</td>\n",
       "      <td>499292</td>\n",
       "      <td>+</td>\n",
       "      <td>Acetyl-CoA C-acetyltransferase (acetoacetyl-Co...</td>\n",
       "      <td>152.0</td>\n",
       "      <td>263.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR208W</th>\n",
       "      <td>ERG12</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>1.016</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>684467</td>\n",
       "      <td>685798</td>\n",
       "      <td>+</td>\n",
       "      <td>Mevalonate kinase, acts in the biosynthesis of...</td>\n",
       "      <td>27.0</td>\n",
       "      <td>107.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YML126C</th>\n",
       "      <td>ERG13</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.929</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>19060</td>\n",
       "      <td>20535</td>\n",
       "      <td>-</td>\n",
       "      <td>3-hydroxy-3-methylglutaryl-CoA (HMG-CoA) synth...</td>\n",
       "      <td>206.0</td>\n",
       "      <td>316.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJL167W</th>\n",
       "      <td>ERG20</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.995</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrX</td>\n",
       "      <td>105014</td>\n",
       "      <td>106072</td>\n",
       "      <td>+</td>\n",
       "      <td>Farnesyl pyrophosphate synthetase, has both di...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR220W</th>\n",
       "      <td>ERG8</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.953</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>712316</td>\n",
       "      <td>713671</td>\n",
       "      <td>+</td>\n",
       "      <td>Phosphomevalonate kinase, an essential cytosol...</td>\n",
       "      <td>102.0</td>\n",
       "      <td>150.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YHR190W</th>\n",
       "      <td>ERG9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrVIII</td>\n",
       "      <td>484845</td>\n",
       "      <td>486179</td>\n",
       "      <td>+</td>\n",
       "      <td>Farnesyl-diphosphate farnesyl transferase (squ...</td>\n",
       "      <td>284.0</td>\n",
       "      <td>423.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YML075C</th>\n",
       "      <td>HMG1</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.959</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>115734</td>\n",
       "      <td>118898</td>\n",
       "      <td>-</td>\n",
       "      <td>One of two isozymes of HMG-CoA reductase that ...</td>\n",
       "      <td>108.0</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLR450W</th>\n",
       "      <td>HMG2</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.929</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXII</td>\n",
       "      <td>1032627</td>\n",
       "      <td>1035764</td>\n",
       "      <td>+</td>\n",
       "      <td>One of two isozymes of HMG-CoA reductase that ...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>113.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPL117C</th>\n",
       "      <td>IDI1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXVI</td>\n",
       "      <td>327864</td>\n",
       "      <td>328730</td>\n",
       "      <td>-</td>\n",
       "      <td>Isopentenyl diphosphate%3Adimethylallyl diphos...</td>\n",
       "      <td>78.0</td>\n",
       "      <td>168.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YNR043W</th>\n",
       "      <td>MVD1</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.909</td>\n",
       "      <td>upstream</td>\n",
       "      <td>chrXIV</td>\n",
       "      <td>701895</td>\n",
       "      <td>703085</td>\n",
       "      <td>+</td>\n",
       "      <td>Mevalonate pyrophosphate decarboxylase, essent...</td>\n",
       "      <td>176.0</td>\n",
       "      <td>364.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGR175C</th>\n",
       "      <td>ERG1</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>1.008</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrVII</td>\n",
       "      <td>846933</td>\n",
       "      <td>848423</td>\n",
       "      <td>-</td>\n",
       "      <td>Squalene epoxidase, catalyzes the epoxidation ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YHR007C</th>\n",
       "      <td>ERG11</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.999</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrVIII</td>\n",
       "      <td>120091</td>\n",
       "      <td>121683</td>\n",
       "      <td>-</td>\n",
       "      <td>Lanosterol 14-alpha-demethylase, catalyzes the...</td>\n",
       "      <td>226.0</td>\n",
       "      <td>388.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR202W</th>\n",
       "      <td>ERG2</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.780</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>667537</td>\n",
       "      <td>668205</td>\n",
       "      <td>+</td>\n",
       "      <td>C-8 sterol isomerase, catalyzes the isomerizat...</td>\n",
       "      <td>57.0</td>\n",
       "      <td>155.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YNL280C</th>\n",
       "      <td>ERG24</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.001</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrXIV</td>\n",
       "      <td>109095</td>\n",
       "      <td>110411</td>\n",
       "      <td>-</td>\n",
       "      <td>C-14 sterol reductase, acts in ergosterol bios...</td>\n",
       "      <td>224.0</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGR060W</th>\n",
       "      <td>ERG25</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.921</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrVII</td>\n",
       "      <td>610564</td>\n",
       "      <td>611493</td>\n",
       "      <td>+</td>\n",
       "      <td>C-4 methyl sterol oxidase, catalyzes the first...</td>\n",
       "      <td>60.0</td>\n",
       "      <td>157.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGL001C</th>\n",
       "      <td>ERG26</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.976</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrVII</td>\n",
       "      <td>495453</td>\n",
       "      <td>496502</td>\n",
       "      <td>-</td>\n",
       "      <td>C-3 sterol dehydrogenase, catalyzes the second...</td>\n",
       "      <td>106.0</td>\n",
       "      <td>224.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLR100W</th>\n",
       "      <td>ERG27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrXII</td>\n",
       "      <td>341810</td>\n",
       "      <td>342853</td>\n",
       "      <td>+</td>\n",
       "      <td>3-keto sterol reductase, catalyzes the last of...</td>\n",
       "      <td>307.0</td>\n",
       "      <td>418.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YER044C</th>\n",
       "      <td>ERG28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrV</td>\n",
       "      <td>237570</td>\n",
       "      <td>238016</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoplasmic reticulum membrane protein, may fa...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR134W</th>\n",
       "      <td>ERG29</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>0.995</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>537838</td>\n",
       "      <td>538551</td>\n",
       "      <td>+</td>\n",
       "      <td>Protein of unknown function that may be involv...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLR056W</th>\n",
       "      <td>ERG3</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.899</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrXII</td>\n",
       "      <td>253861</td>\n",
       "      <td>254958</td>\n",
       "      <td>+</td>\n",
       "      <td>C-5 sterol desaturase, catalyzes the introduct...</td>\n",
       "      <td>335.0</td>\n",
       "      <td>654.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGL012W</th>\n",
       "      <td>ERG4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrVII</td>\n",
       "      <td>472855</td>\n",
       "      <td>474276</td>\n",
       "      <td>+</td>\n",
       "      <td>C-24(28) sterol reductase, catalyzes the final...</td>\n",
       "      <td>158.0</td>\n",
       "      <td>228.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR015C</th>\n",
       "      <td>ERG5</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.019</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>300869</td>\n",
       "      <td>302485</td>\n",
       "      <td>-</td>\n",
       "      <td>C-22 sterol desaturase, a cytochrome P450 enzy...</td>\n",
       "      <td>164.0</td>\n",
       "      <td>184.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YML008C</th>\n",
       "      <td>ERG6</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.911</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrXIII</td>\n",
       "      <td>251839</td>\n",
       "      <td>252990</td>\n",
       "      <td>-</td>\n",
       "      <td>Delta(24)-sterol C-methyltransferase, converts...</td>\n",
       "      <td>303.0</td>\n",
       "      <td>492.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YHR072W</th>\n",
       "      <td>ERG7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrVIII</td>\n",
       "      <td>239098</td>\n",
       "      <td>241293</td>\n",
       "      <td>+</td>\n",
       "      <td>Lanosterol synthase, an essential enzyme that ...</td>\n",
       "      <td>214.0</td>\n",
       "      <td>386.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YHR042W</th>\n",
       "      <td>NCP1</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>reduction of function</td>\n",
       "      <td>1.006</td>\n",
       "      <td>downstream</td>\n",
       "      <td>chrVIII</td>\n",
       "      <td>190543</td>\n",
       "      <td>192618</td>\n",
       "      <td>+</td>\n",
       "      <td>NADP-cytochrome P450 reductase; involved in er...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YKL150W</th>\n",
       "      <td>MCR1</td>\n",
       "      <td>Media: minimal medium Details: Relative fitnes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrXI</td>\n",
       "      <td>166400</td>\n",
       "      <td>167452</td>\n",
       "      <td>+</td>\n",
       "      <td>Mitochondrial NADH-cytochrome b5 reductase, in...</td>\n",
       "      <td>113.0</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name                                            Details  \\\n",
       "gene_id                                                                \n",
       "YPL028W     ERG10  Media: minimal medium Details: Relative fitnes...   \n",
       "YMR208W     ERG12  Media: minimal medium Details: Relative fitnes...   \n",
       "YML126C     ERG13  Media: minimal medium Details: Relative fitnes...   \n",
       "YJL167W     ERG20  Media: minimal medium Details: Relative fitnes...   \n",
       "YMR220W      ERG8  Media: minimal medium Details: Relative fitnes...   \n",
       "YHR190W      ERG9                                                NaN   \n",
       "YML075C      HMG1  Media: minimal medium Details: Relative fitnes...   \n",
       "YLR450W      HMG2  Media: minimal medium Details: Relative fitnes...   \n",
       "YPL117C      IDI1                                                NaN   \n",
       "YNR043W      MVD1  Media: minimal medium Details: Relative fitnes...   \n",
       "YGR175C      ERG1  Media: minimal medium Details: Relative fitnes...   \n",
       "YHR007C     ERG11  Media: minimal medium Details: Relative fitnes...   \n",
       "YMR202W      ERG2  Media: minimal medium Details: Relative fitnes...   \n",
       "YNL280C     ERG24  Media: minimal medium Details: Relative fitnes...   \n",
       "YGR060W     ERG25  Media: minimal medium Details: Relative fitnes...   \n",
       "YGL001C     ERG26  Media: minimal medium Details: Relative fitnes...   \n",
       "YLR100W     ERG27                                                NaN   \n",
       "YER044C     ERG28                                                NaN   \n",
       "YMR134W     ERG29  Media: minimal medium Details: Relative fitnes...   \n",
       "YLR056W      ERG3  Media: minimal medium Details: Relative fitnes...   \n",
       "YGL012W      ERG4                                                NaN   \n",
       "YMR015C      ERG5  Media: minimal medium Details: Relative fitnes...   \n",
       "YML008C      ERG6  Media: minimal medium Details: Relative fitnes...   \n",
       "YHR072W      ERG7                                                NaN   \n",
       "YHR042W      NCP1  Media: minimal medium Details: Relative fitnes...   \n",
       "YKL150W      MCR1  Media: minimal medium Details: Relative fitnes...   \n",
       "\n",
       "            Mutant Information  Breslow fitness     pathway    chrom    start  \\\n",
       "gene_id                                                                         \n",
       "YPL028W  reduction of function            0.935    upstream   chrXVI   498096   \n",
       "YMR208W  reduction of function            1.016    upstream  chrXIII   684467   \n",
       "YML126C  reduction of function            0.929    upstream  chrXIII    19060   \n",
       "YJL167W  reduction of function            0.995    upstream     chrX   105014   \n",
       "YMR220W  reduction of function            0.953    upstream  chrXIII   712316   \n",
       "YHR190W                    NaN              NaN    upstream  chrVIII   484845   \n",
       "YML075C                    NaN            0.959    upstream  chrXIII   115734   \n",
       "YLR450W                    NaN            0.929    upstream   chrXII  1032627   \n",
       "YPL117C                    NaN              NaN    upstream   chrXVI   327864   \n",
       "YNR043W  reduction of function            0.909    upstream   chrXIV   701895   \n",
       "YGR175C  reduction of function            1.008  downstream   chrVII   846933   \n",
       "YHR007C  reduction of function            0.999  downstream  chrVIII   120091   \n",
       "YMR202W                    NaN            0.780  downstream  chrXIII   667537   \n",
       "YNL280C                    NaN            1.001  downstream   chrXIV   109095   \n",
       "YGR060W  reduction of function            0.921  downstream   chrVII   610564   \n",
       "YGL001C  reduction of function            0.976  downstream   chrVII   495453   \n",
       "YLR100W                    NaN              NaN  downstream   chrXII   341810   \n",
       "YER044C                    NaN              NaN  downstream     chrV   237570   \n",
       "YMR134W  reduction of function            0.995  downstream  chrXIII   537838   \n",
       "YLR056W                    NaN            0.899  downstream   chrXII   253861   \n",
       "YGL012W                    NaN              NaN  downstream   chrVII   472855   \n",
       "YMR015C                    NaN            1.019  downstream  chrXIII   300869   \n",
       "YML008C                    NaN            0.911  downstream  chrXIII   251839   \n",
       "YHR072W                    NaN              NaN  downstream  chrVIII   239098   \n",
       "YHR042W  reduction of function            1.006  downstream  chrVIII   190543   \n",
       "YKL150W                    NaN            1.000         NaN    chrXI   166400   \n",
       "\n",
       "             end strand                                        description  \\\n",
       "gene_id                                                                      \n",
       "YPL028W   499292      +  Acetyl-CoA C-acetyltransferase (acetoacetyl-Co...   \n",
       "YMR208W   685798      +  Mevalonate kinase, acts in the biosynthesis of...   \n",
       "YML126C    20535      -  3-hydroxy-3-methylglutaryl-CoA (HMG-CoA) synth...   \n",
       "YJL167W   106072      +  Farnesyl pyrophosphate synthetase, has both di...   \n",
       "YMR220W   713671      +  Phosphomevalonate kinase, an essential cytosol...   \n",
       "YHR190W   486179      +  Farnesyl-diphosphate farnesyl transferase (squ...   \n",
       "YML075C   118898      -  One of two isozymes of HMG-CoA reductase that ...   \n",
       "YLR450W  1035764      +  One of two isozymes of HMG-CoA reductase that ...   \n",
       "YPL117C   328730      -  Isopentenyl diphosphate%3Adimethylallyl diphos...   \n",
       "YNR043W   703085      +  Mevalonate pyrophosphate decarboxylase, essent...   \n",
       "YGR175C   848423      -  Squalene epoxidase, catalyzes the epoxidation ...   \n",
       "YHR007C   121683      -  Lanosterol 14-alpha-demethylase, catalyzes the...   \n",
       "YMR202W   668205      +  C-8 sterol isomerase, catalyzes the isomerizat...   \n",
       "YNL280C   110411      -  C-14 sterol reductase, acts in ergosterol bios...   \n",
       "YGR060W   611493      +  C-4 methyl sterol oxidase, catalyzes the first...   \n",
       "YGL001C   496502      -  C-3 sterol dehydrogenase, catalyzes the second...   \n",
       "YLR100W   342853      +  3-keto sterol reductase, catalyzes the last of...   \n",
       "YER044C   238016      -  Endoplasmic reticulum membrane protein, may fa...   \n",
       "YMR134W   538551      +  Protein of unknown function that may be involv...   \n",
       "YLR056W   254958      +  C-5 sterol desaturase, catalyzes the introduct...   \n",
       "YGL012W   474276      +  C-24(28) sterol reductase, catalyzes the final...   \n",
       "YMR015C   302485      -  C-22 sterol desaturase, a cytochrome P450 enzy...   \n",
       "YML008C   252990      -  Delta(24)-sterol C-methyltransferase, converts...   \n",
       "YHR072W   241293      +  Lanosterol synthase, an essential enzyme that ...   \n",
       "YHR042W   192618      +  NADP-cytochrome P450 reductase; involved in er...   \n",
       "YKL150W   167452      +  Mitochondrial NADH-cytochrome b5 reductase, in...   \n",
       "\n",
       "         positive_interactions  negative_interactions  \n",
       "gene_id                                                \n",
       "YPL028W                  152.0                  263.0  \n",
       "YMR208W                   27.0                  107.0  \n",
       "YML126C                  206.0                  316.0  \n",
       "YJL167W                    NaN                    NaN  \n",
       "YMR220W                  102.0                  150.0  \n",
       "YHR190W                  284.0                  423.0  \n",
       "YML075C                  108.0                  143.0  \n",
       "YLR450W                   44.0                  113.0  \n",
       "YPL117C                   78.0                  168.0  \n",
       "YNR043W                  176.0                  364.0  \n",
       "YGR175C                    NaN                    NaN  \n",
       "YHR007C                  226.0                  388.0  \n",
       "YMR202W                   57.0                  155.0  \n",
       "YNL280C                  224.0                  352.0  \n",
       "YGR060W                   60.0                  157.0  \n",
       "YGL001C                  106.0                  224.0  \n",
       "YLR100W                  307.0                  418.0  \n",
       "YER044C                    NaN                    NaN  \n",
       "YMR134W                    NaN                    NaN  \n",
       "YLR056W                  335.0                  654.0  \n",
       "YGL012W                  158.0                  228.0  \n",
       "YMR015C                  164.0                  184.0  \n",
       "YML008C                  303.0                  492.0  \n",
       "YHR072W                  214.0                  386.0  \n",
       "YHR042W                    NaN                    NaN  \n",
       "YKL150W                  113.0                   88.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read in ergosterol genes table\n",
    "ergosterol_genes = pd.read_csv(ergosterol_genes_file, sep='\\t')\n",
    "ergosterol_genes = ergosterol_genes.rename(columns={'Gene Systematic Name':'gene_id', 'Gene':'gene_name'}).set_index('gene_id')\n",
    "\n",
    "# annotate ergosterol genes\n",
    "ergosterol_genes = ergosterol_genes.merge(gene_info_df.drop('gene_name', axis=1), how='left', left_index=True, right_index=True)\n",
    "\n",
    "# add genetic interaction data\n",
    "genetic_interactions = pd.read_csv(genetic_interactions_summary_file, sep='\\t', index_col=0)\n",
    "ergosterol_genes = ergosterol_genes.merge(genetic_interactions, how='left', left_index=True, right_index=True)\n",
    "\n",
    "display(ergosterol_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search ergosterol genes for variants in each gvcf\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16',}\n",
    "\n",
    "\n",
    "\n",
    "genes_to_process = []\n",
    "for chrom in chromosome_dict.keys():\n",
    "    genes_df = ergosterol_genes.query('chrom==@chrom').sort_values('start')\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict[chrom])\n",
    "    genes_to_process.append((genes_df, gvcf_file, chromosome_dict))\n",
    "    \n",
    "# consolidate candidate variants\n",
    "num_of_cores = min(len(os.sched_getaffinity(0)), len(genes_to_process))\n",
    "with mp.Pool(num_of_cores) as pool:\n",
    "    results = pool.starmap(select_variants, genes_to_process)\n",
    "results = pd.concat(results).reset_index(drop=True)\n",
    "\n",
    "# write to new vcf file\n",
    "ergosterol_variants_file = \"ergosterol_variants.vcf\"\n",
    "template = vcf.Reader(filename=genes_to_process[0][1])\n",
    "# change chromosome naming convention to a single Roman numeral (follows ref genome, VEP formatting)\n",
    "chromosome_dict_rev = {v:k for k, v in chromosome_dict.items()} \n",
    "template.contigs = OrderedDict((chromosome_dict_rev[k][3:], v._replace(id=chromosome_dict_rev[k][3:])) for k, v in template.contigs.items())\n",
    "# remove 1011 genomes samples data\n",
    "template.samples=[]\n",
    "\n",
    "output = vcf.Writer(open(ergosterol_variants_file, 'w'), template)\n",
    "for record in results:\n",
    "    # update chromosome naming\n",
    "    record.CHROM = chromosome_dict_rev[record.CHROM][3:]\n",
    "    # write to file\n",
    "    output.write_record(record)\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to genome fasta files\n",
    "ref_genome_fasta = \"/home/users/rang/yeast/genomes/Saccharomyces_cerevisiae.R64-1-1.dna.chromosome.I.fa\"\n",
    "rm_genome_fasta = \"/home/users/rang/yeast/genomes/RM11-1A_SGD_2015_JRIP00000000.fsa\"\n",
    "yjm_genome_fasta = \"/home/users/rang/yeast/genomes/YJM789_Stanford_2007_AAFW02000000_highQuality31.fsa\"\n",
    "yps_genome_fasta = \"/home/users/rang/yeast/genomes/YPS128.genome.fa\"\n",
    "\n",
    "# path to input and output VCF files\n",
    "ergosterol_variants_file = \"ergosterol_variants.vcf\"\n",
    "ergosterol_variants_select_file = ergosterol_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# load genome files\n",
    "ref_genome = SeqIO.to_dict(SeqIO.parse(ref_genome_fasta, \"fasta\"))\n",
    "rm_genome = SeqIO.to_dict(SeqIO.parse(rm_genome_fasta, \"fasta\"))\n",
    "yjm_genome = SeqIO.to_dict(SeqIO.parse(yjm_genome_fasta, \"fasta\"))\n",
    "yps_genome = SeqIO.to_dict(SeqIO.parse(yps_genome_fasta, \"fasta\"))\n",
    "genomes_list = [ref_genome, rm_genome, yjm_genome, yps_genome] # IMPORTANT: SET REFERENCE GENOME AS FIRST IN LIST\n",
    "\n",
    "\n",
    "threshold = 10 # threshold value for searching NGGs\n",
    "window = 61 # size of flanking sequences to search for unique hits\n",
    "\n",
    "# read in all candidate variants to filter\n",
    "ergosterol_variants = vcf.Reader(filename=ergosterol_variants_file)\n",
    "ergosterol_variants = [rec for rec in ergosterol_variants]\n",
    "# split variants into chunks for multiprocessing\n",
    "n = len(ergosterol_variants)//len(os.sched_getaffinity(0)) + 1\n",
    "ergosterol_variants = [ergosterol_variants[i:i+n] for i in range(0, len(ergosterol_variants), n)]\n",
    "\n",
    "# find targetable variants (TGT info field indicates which genomes variant is targetable in)\n",
    "to_process = [(rec_list, genomes_list, threshold, window) for rec_list in ergosterol_variants]\n",
    "with mp.Pool(len(os.sched_getaffinity(0))) as pool:\n",
    "    result = pool.starmap(find_targetable_variants, to_process) \n",
    "\n",
    "# write targetable variants to file\n",
    "header = vcf.Reader(filename=ergosterol_variants_file)\n",
    "# add \"TGT\" info field into VCF header\n",
    "header_info = namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "header.infos['TGT'] = header_info_tgt\n",
    "# write to file\n",
    "ergosterol_variants_select = vcf.Writer(open(ergosterol_variants_select_file, 'w'), header)\n",
    "to_write = [rec for rec_list in result for rec in rec_list]\n",
    "for rec in to_write:\n",
    "    ergosterol_variants_select.write_record(rec)\n",
    "ergosterol_variants_select.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select variants by allele frequency (currently disabled singleton/doubleton filter)\n",
    "### if multiple alternate alleles, SPLIT VARIANT ENTRY INTO SEPARATE ENTRIES in VCF (useful for oligo design pipeline)\n",
    "### assign unitl variant ID prefix and number (use format VAR_XXXXX)\n",
    "\n",
    "# candidate variants editable in all strains\n",
    "ergosterol_variants_select = vcf.Reader(filename=ergosterol_variants_select_file)\n",
    "\n",
    "# open output file\n",
    "ergosterol_variants_select_AFfilter_file = ergosterol_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "ergosterol_variants_select_AFfilter = vcf.Writer(open(ergosterol_variants_select_AFfilter_file, 'w'), ergosterol_variants_select)\n",
    "\n",
    "var_counter = {}\n",
    "for record in ergosterol_variants_select:        \n",
    "    # (optional step to skip singletons and doubletons -- DISABLED by setting ac<1)\n",
    "    if all([ac<1 for ac in record.INFO['AC']]):\n",
    "        continue\n",
    "    # if at least one alternate allele meets minimum count, write to VCF\n",
    "    else:\n",
    "        alt = record.ALT[:]\n",
    "        ac = record.INFO['AC'][:]\n",
    "        af = record.INFO['AF'][:]\n",
    "        mleac = record.INFO['MLEAC'][:]\n",
    "        mleaf = record.INFO['MLEAF'][:]\n",
    "        \n",
    "        # split each alt allele into its own variant entry (useful if a locus has multiple possible alleles)\n",
    "        for i in range(len(alt)):\n",
    "            record.ALT = [alt[i]]\n",
    "            record.INFO['AC'] = [ac[i]]\n",
    "            record.INFO['AF'] = [af[i]]\n",
    "            record.INFO['MLEAC'] = [mleac[i]]\n",
    "            record.INFO['MLEAF'] = [mleaf[i]]\n",
    "\n",
    "            # inspect TGT field and assign variant ID prefix\n",
    "            ### if tgt field contains all 1 (i.e. targetable by all queried genomes), assign \"ERG\"\n",
    "            ### otherwise, assign unique prefix \"EG-\" with the last digit determined by converting binary to hexadecimal (works well for up to 4 genomes)\n",
    "            if all([i==1 for i in record.INFO['TGT']]):\n",
    "                var_id_prefix = \"ERG\"\n",
    "            else:\n",
    "                binary_str = ''.join([str(x) for x in record.INFO['TGT']])\n",
    "                hex_digit = hex(int(binary_str, 2))[2:].upper()\n",
    "                var_id_prefix = \"EG{}\".format(hex_digit)\n",
    "\n",
    "            # add to counter\n",
    "            if var_id_prefix not in var_counter:\n",
    "                var_counter[var_id_prefix] = 1\n",
    "            else:\n",
    "                var_counter[var_id_prefix] += 1\n",
    "\n",
    "            # add unique variant ID (follow format VAR_XXXXX)\n",
    "            record.ID='{}_{:05d}'.format(var_id_prefix, var_counter[var_id_prefix])\n",
    "\n",
    "            # write to output\n",
    "            ergosterol_variants_select_AFfilter.write_record(record)\n",
    "            \n",
    "ergosterol_variants_select_AFfilter.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant ID prefix and which genomes they can be edited in:\n",
    "- ERG: BY | RM | YJM | YPS\n",
    "- EGE: BY | RM | YJM | ---\n",
    "- EGD: BY | RM | --- | YPS\n",
    "- EGC: BY | RM | --- | ---\n",
    "- EGB: BY | --- | YJM | YPS\n",
    "- EGA: BY | --- | YJM | ---\n",
    "- EG9: BY | --- | --- | YPS\n",
    "- EG8: BY | --- | --- | ---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit ergosterol_variants_select_AFfilter VCF to VEP to get annotations, and download VEP output in TXT format\n",
    "http://uswest.ensembl.org/Saccharomyces_cerevisiae/Tools/VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 875,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00001</th>\n",
       "      <td>V</td>\n",
       "      <td>237113</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2373|1622|1347|457|2919</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gct/Act</td>\n",
       "      <td>HAMAP:MF_00563,Gene3D:3.40.50.1480,PIRSF:PIRSF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00002</th>\n",
       "      <td>V</td>\n",
       "      <td>237121</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>2</td>\n",
       "      <td>2381|1630|1339|449|2911</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>449.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>3</td>\n",
       "      <td>2382|1631|1338|448|2910</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>448.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>17</td>\n",
       "      <td>2396|1645|1324|434|2896</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00001</th>\n",
       "      <td>V</td>\n",
       "      <td>237149</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>43</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>30</td>\n",
       "      <td>2409|1658|1311|421|2883</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>30.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>421.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGB_00245</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499756</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>157</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.078000</td>\n",
       "      <td>3761</td>\n",
       "      <td>4583|919|464|2037|4510</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>aGt/aAt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGB_00246</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499757</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.000990</td>\n",
       "      <td>3760</td>\n",
       "      <td>4582|918|465|2038|4511</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>agT/agC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGB_00247</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499761</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>3756</td>\n",
       "      <td>4578|914|469|2042|4515</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Ggt/Cgt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_02770</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499778</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>151</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.075000</td>\n",
       "      <td>3739</td>\n",
       "      <td>4561|897|486|2059|4532</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>aaA/aaG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_02771</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499786</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>55</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.027000</td>\n",
       "      <td>3731</td>\n",
       "      <td>4553|889|494|2067|4540</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gGt/gAt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4702 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                       \n",
       "ERG_00001     V  237113   C   T    1  2022  0.000495                   NaN   \n",
       "ERG_00002     V  237121   G   A    9  2022  0.004451                     2   \n",
       "ERG_00003     V  237122   G   A    9  2022  0.004451                     3   \n",
       "ERG_00004     V  237136   T   C   26  2022  0.013000                    17   \n",
       "EGE_00001     V  237149   A   C   43  2022  0.021000                    30   \n",
       "...         ...     ...  ..  ..  ...   ...       ...                   ...   \n",
       "EGB_00245   XVI  499756   G   A  157  2018  0.078000                  3761   \n",
       "EGB_00246   XVI  499757   T   C    2  2020  0.000990                  3760   \n",
       "EGB_00247   XVI  499761   G   C    1  2020  0.000495                  3756   \n",
       "ERG_02770   XVI  499778   A   G  151  2020  0.075000                  3739   \n",
       "ERG_02771   XVI  499786   G   A   55  2020  0.027000                  3731   \n",
       "\n",
       "           downstream_distance_str                   region  ...   Codons  \\\n",
       "var_id                                                       ...            \n",
       "ERG_00001  2373|1622|1347|457|2919         missense_variant  ...  Gct/Act   \n",
       "ERG_00002  2381|1630|1339|449|2911  unidirectional_promoter  ...      NaN   \n",
       "ERG_00003  2382|1631|1338|448|2910  unidirectional_promoter  ...      NaN   \n",
       "ERG_00004  2396|1645|1324|434|2896  unidirectional_promoter  ...      NaN   \n",
       "EGE_00001  2409|1658|1311|421|2883  unidirectional_promoter  ...      NaN   \n",
       "...                            ...                      ...  ...      ...   \n",
       "EGB_00245   4583|919|464|2037|4510         missense_variant  ...  aGt/aAt   \n",
       "EGB_00246   4582|918|465|2038|4511       synonymous_variant  ...  agT/agC   \n",
       "EGB_00247   4578|914|469|2042|4515         missense_variant  ...  Ggt/Cgt   \n",
       "ERG_02770   4561|897|486|2059|4532       synonymous_variant  ...  aaA/aaG   \n",
       "ERG_02771   4553|889|494|2067|4540         missense_variant  ...  gGt/gAt   \n",
       "\n",
       "                                                     DOMAINS  \\\n",
       "var_id                                                         \n",
       "ERG_00001  HAMAP:MF_00563,Gene3D:3.40.50.1480,PIRSF:PIRSF...   \n",
       "ERG_00002                                                NaN   \n",
       "ERG_00003                                                NaN   \n",
       "ERG_00004                                                NaN   \n",
       "EGE_00001                                                NaN   \n",
       "...                                                      ...   \n",
       "EGB_00245                                                NaN   \n",
       "EGB_00246                                                NaN   \n",
       "EGB_00247                                                NaN   \n",
       "ERG_02770                                                NaN   \n",
       "ERG_02771                                                NaN   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00001                     NaN                   NaN   \n",
       "ERG_00002                    SAH1               YER043C   \n",
       "ERG_00003                    SAH1               YER043C   \n",
       "ERG_00004                    SAH1               YER043C   \n",
       "EGE_00001                    SAH1               YER043C   \n",
       "...                           ...                   ...   \n",
       "EGB_00245                     NaN                   NaN   \n",
       "EGB_00246                     NaN                   NaN   \n",
       "EGB_00247                     NaN                   NaN   \n",
       "ERG_02770                     NaN                   NaN   \n",
       "ERG_02771                     NaN                   NaN   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "ERG_00001                      NaN                    NaN   \n",
       "ERG_00002    upstream_gene_variant                    2.0   \n",
       "ERG_00003    upstream_gene_variant                    3.0   \n",
       "ERG_00004    upstream_gene_variant                   17.0   \n",
       "EGE_00001    upstream_gene_variant                   30.0   \n",
       "...                            ...                    ...   \n",
       "EGB_00245                      NaN                    NaN   \n",
       "EGB_00246                      NaN                    NaN   \n",
       "EGB_00247                      NaN                    NaN   \n",
       "ERG_02770                      NaN                    NaN   \n",
       "ERG_02771                      NaN                    NaN   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00001                     NaN                   NaN   \n",
       "ERG_00002                   ERG28               YER044C   \n",
       "ERG_00003                   ERG28               YER044C   \n",
       "ERG_00004                   ERG28               YER044C   \n",
       "EGE_00001                   ERG28               YER044C   \n",
       "...                           ...                   ...   \n",
       "EGB_00245                     NaN                   NaN   \n",
       "EGB_00246                     NaN                   NaN   \n",
       "EGB_00247                     NaN                   NaN   \n",
       "ERG_02770                     NaN                   NaN   \n",
       "ERG_02771                     NaN                   NaN   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "ERG_00001                      NaN                    NaN  \n",
       "ERG_00002  downstream_gene_variant                  449.0  \n",
       "ERG_00003  downstream_gene_variant                  448.0  \n",
       "ERG_00004  downstream_gene_variant                  434.0  \n",
       "EGE_00001  downstream_gene_variant                  421.0  \n",
       "...                            ...                    ...  \n",
       "EGB_00245                      NaN                    NaN  \n",
       "EGB_00246                      NaN                    NaN  \n",
       "EGB_00247                      NaN                    NaN  \n",
       "ERG_02770                      NaN                    NaN  \n",
       "ERG_02771                      NaN                    NaN  \n",
       "\n",
       "[4702 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00002</th>\n",
       "      <td>V</td>\n",
       "      <td>237121</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>2</td>\n",
       "      <td>2381|1630|1339|449|2911</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>449.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>3</td>\n",
       "      <td>2382|1631|1338|448|2910</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>448.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>17</td>\n",
       "      <td>2396|1645|1324|434|2896</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00001</th>\n",
       "      <td>V</td>\n",
       "      <td>237149</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>43</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>30</td>\n",
       "      <td>2409|1658|1311|421|2883</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>30.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>421.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00002</th>\n",
       "      <td>V</td>\n",
       "      <td>237151</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>32</td>\n",
       "      <td>2411|1660|1309|419|2881</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>32.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>419.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00041</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00042</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>35</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.017000</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00043</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499656</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>3861|9</td>\n",
       "      <td>4683|1019|364|1937|4410</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>364.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00044</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499659</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>3858|6</td>\n",
       "      <td>4680|1016|367|1940|4413</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>367.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00045</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499664</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1670</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.827000</td>\n",
       "      <td>3853|1</td>\n",
       "      <td>4675|1011|372|1945|4418</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>372.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4082 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT    AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                        \n",
       "ERG_00002     V  237121   G   A     9  2022  0.004451                     2   \n",
       "ERG_00003     V  237122   G   A     9  2022  0.004451                     3   \n",
       "ERG_00004     V  237136   T   C    26  2022  0.013000                    17   \n",
       "EGE_00001     V  237149   A   C    43  2022  0.021000                    30   \n",
       "EGE_00002     V  237151   T   C     2  2022  0.000989                    32   \n",
       "...         ...     ...  ..  ..   ...   ...       ...                   ...   \n",
       "EGA_00041   XVI  499655   G   A     4  2022  0.001978               3862|10   \n",
       "EGA_00042   XVI  499655   G   T    35  2022  0.017000               3862|10   \n",
       "EGA_00043   XVI  499656   A   G    26  2022  0.013000                3861|9   \n",
       "EGA_00044   XVI  499659   G   A     1  2022  0.000495                3858|6   \n",
       "EGA_00045   XVI  499664   A   G  1670  2020  0.827000                3853|1   \n",
       "\n",
       "           downstream_distance_str                   region  ... Codons  \\\n",
       "var_id                                                       ...          \n",
       "ERG_00002  2381|1630|1339|449|2911  unidirectional_promoter  ...    NaN   \n",
       "ERG_00003  2382|1631|1338|448|2910  unidirectional_promoter  ...    NaN   \n",
       "ERG_00004  2396|1645|1324|434|2896  unidirectional_promoter  ...    NaN   \n",
       "EGE_00001  2409|1658|1311|421|2883  unidirectional_promoter  ...    NaN   \n",
       "EGE_00002  2411|1660|1309|419|2881  unidirectional_promoter  ...    NaN   \n",
       "...                            ...                      ...  ...    ...   \n",
       "EGA_00041  4684|1020|363|1936|4409  unidirectional_promoter  ...    NaN   \n",
       "EGA_00042  4684|1020|363|1936|4409  unidirectional_promoter  ...    NaN   \n",
       "EGA_00043  4683|1019|364|1937|4410  unidirectional_promoter  ...    NaN   \n",
       "EGA_00044  4680|1016|367|1940|4413  unidirectional_promoter  ...    NaN   \n",
       "EGA_00045  4675|1011|372|1945|4418  unidirectional_promoter  ...    NaN   \n",
       "\n",
       "          DOMAINS closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                            \n",
       "ERG_00002     NaN                    SAH1               YER043C   \n",
       "ERG_00003     NaN                    SAH1               YER043C   \n",
       "ERG_00004     NaN                    SAH1               YER043C   \n",
       "EGE_00001     NaN                    SAH1               YER043C   \n",
       "EGE_00002     NaN                    SAH1               YER043C   \n",
       "...           ...                     ...                   ...   \n",
       "EGA_00041     NaN                   ERG10               YPL028W   \n",
       "EGA_00042     NaN                   ERG10               YPL028W   \n",
       "EGA_00043     NaN                   ERG10               YPL028W   \n",
       "EGA_00044     NaN                   ERG10               YPL028W   \n",
       "EGA_00045     NaN                   ERG10               YPL028W   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "ERG_00002    upstream_gene_variant                    2.0   \n",
       "ERG_00003    upstream_gene_variant                    3.0   \n",
       "ERG_00004    upstream_gene_variant                   17.0   \n",
       "EGE_00001    upstream_gene_variant                   30.0   \n",
       "EGE_00002    upstream_gene_variant                   32.0   \n",
       "...                            ...                    ...   \n",
       "EGA_00041  downstream_gene_variant                  363.0   \n",
       "EGA_00042  downstream_gene_variant                  363.0   \n",
       "EGA_00043  downstream_gene_variant                  364.0   \n",
       "EGA_00044  downstream_gene_variant                  367.0   \n",
       "EGA_00045  downstream_gene_variant                  372.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00002                   ERG28               YER044C   \n",
       "ERG_00003                   ERG28               YER044C   \n",
       "ERG_00004                   ERG28               YER044C   \n",
       "EGE_00001                   ERG28               YER044C   \n",
       "EGE_00002                   ERG28               YER044C   \n",
       "...                           ...                   ...   \n",
       "EGA_00041                    SMA1               YPL027W   \n",
       "EGA_00042                    SMA1               YPL027W   \n",
       "EGA_00043                    SMA1               YPL027W   \n",
       "EGA_00044                    SMA1               YPL027W   \n",
       "EGA_00045                    SMA1               YPL027W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "ERG_00002  downstream_gene_variant                  449.0  \n",
       "ERG_00003  downstream_gene_variant                  448.0  \n",
       "ERG_00004  downstream_gene_variant                  434.0  \n",
       "EGE_00001  downstream_gene_variant                  421.0  \n",
       "EGE_00002  downstream_gene_variant                  419.0  \n",
       "...                            ...                    ...  \n",
       "EGA_00041    upstream_gene_variant                   10.0  \n",
       "EGA_00042    upstream_gene_variant                   10.0  \n",
       "EGA_00043    upstream_gene_variant                    9.0  \n",
       "EGA_00044    upstream_gene_variant                    6.0  \n",
       "EGA_00045    upstream_gene_variant                    1.0  \n",
       "\n",
       "[4082 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add annotations to variants and remove variants that fall in non-ergosterol genes.\n",
    "\n",
    "ergosterol_variants_file = \"ergosterol_variants.vcf\"\n",
    "ergosterol_variants_select_file = ergosterol_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# get VEP annotations and remove variants that fall in non-ergosterol pathway genes\n",
    "ergosterol_variants_select_AFfilter_file = ergosterol_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "vep_output_filename = ergosterol_variants_select_AFfilter_file.replace('.vcf', '_VEPoutput.txt')\n",
    "variants_annotated_filename = vep_output_filename.replace('_VEPoutput.txt', '_annotated.txt')\n",
    "variants_annotated_df = annotate_variants_by_VEPoutput(ergosterol_variants_select_AFfilter_file, \n",
    "                                                       vep_output_filename, \n",
    "                                                       gff_file, \n",
    "                                                       variants_annotated_filename, \n",
    "                                                       id_colname='var_id')\n",
    "display(variants_annotated_df)\n",
    "\n",
    "# annotated file can be read in directly if annotation function has been run once before\n",
    "variants_annotated_df = pd.read_csv(variants_annotated_filename, sep='\\t', index_col=0)\n",
    "display(variants_annotated_df)\n",
    "\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "ergosterol_design_oligos = variants_annotated_df.query('(Gene_ID.isin(@ergosterol_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(ergosterol_design_oligos)\n",
    "\n",
    "# write variants to ergosterol_design_oligos VCF as a preliminary list to design oligos for\n",
    "ergosterol_design_oligos_vcf_file = ergosterol_variants_file.replace('.vcf', '_design_oligos_initial.vcf')\n",
    "\n",
    "ergosterol_variants_select_AFfilter = vcf.Reader(filename=ergosterol_variants_select_AFfilter_file)\n",
    "ergosterol_design_oligos_vcf = vcf.Writer(open(ergosterol_design_oligos_vcf_file, 'w'), ergosterol_variants_select_AFfilter)\n",
    "for record in ergosterol_variants_select_AFfilter:\n",
    "    if record.ID in ergosterol_design_oligos.index:\n",
    "        ergosterol_design_oligos_vcf.write_record(record)\n",
    "ergosterol_design_oligos_vcf.close()\n",
    "\n",
    "# # count variants according to which strains they can be edited in\n",
    "# display(pd.Series([x[:3] for x in ergosterol_design_oligos.index]).value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design oligos for initial list, then filter down to variants with at least 2 oligos designed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00002</th>\n",
       "      <td>V</td>\n",
       "      <td>237121</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>2</td>\n",
       "      <td>2381|1630|1339|449|2911</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>449.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>3</td>\n",
       "      <td>2382|1631|1338|448|2910</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>448.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>17</td>\n",
       "      <td>2396|1645|1324|434|2896</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00001</th>\n",
       "      <td>V</td>\n",
       "      <td>237149</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>43</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>30</td>\n",
       "      <td>2409|1658|1311|421|2883</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>30.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>421.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00002</th>\n",
       "      <td>V</td>\n",
       "      <td>237151</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>32</td>\n",
       "      <td>2411|1660|1309|419|2881</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>32.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>419.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00041</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00042</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>35</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.017000</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00043</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499656</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>3861|9</td>\n",
       "      <td>4683|1019|364|1937|4410</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>364.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00044</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499659</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>3858|6</td>\n",
       "      <td>4680|1016|367|1940|4413</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>367.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00045</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499664</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1670</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.827000</td>\n",
       "      <td>3853|1</td>\n",
       "      <td>4675|1011|372|1945|4418</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>372.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4082 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT    AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                        \n",
       "ERG_00002     V  237121   G   A     9  2022  0.004451                     2   \n",
       "ERG_00003     V  237122   G   A     9  2022  0.004451                     3   \n",
       "ERG_00004     V  237136   T   C    26  2022  0.013000                    17   \n",
       "EGE_00001     V  237149   A   C    43  2022  0.021000                    30   \n",
       "EGE_00002     V  237151   T   C     2  2022  0.000989                    32   \n",
       "...         ...     ...  ..  ..   ...   ...       ...                   ...   \n",
       "EGA_00041   XVI  499655   G   A     4  2022  0.001978               3862|10   \n",
       "EGA_00042   XVI  499655   G   T    35  2022  0.017000               3862|10   \n",
       "EGA_00043   XVI  499656   A   G    26  2022  0.013000                3861|9   \n",
       "EGA_00044   XVI  499659   G   A     1  2022  0.000495                3858|6   \n",
       "EGA_00045   XVI  499664   A   G  1670  2020  0.827000                3853|1   \n",
       "\n",
       "           downstream_distance_str                   region  ... Codons  \\\n",
       "var_id                                                       ...          \n",
       "ERG_00002  2381|1630|1339|449|2911  unidirectional_promoter  ...    NaN   \n",
       "ERG_00003  2382|1631|1338|448|2910  unidirectional_promoter  ...    NaN   \n",
       "ERG_00004  2396|1645|1324|434|2896  unidirectional_promoter  ...    NaN   \n",
       "EGE_00001  2409|1658|1311|421|2883  unidirectional_promoter  ...    NaN   \n",
       "EGE_00002  2411|1660|1309|419|2881  unidirectional_promoter  ...    NaN   \n",
       "...                            ...                      ...  ...    ...   \n",
       "EGA_00041  4684|1020|363|1936|4409  unidirectional_promoter  ...    NaN   \n",
       "EGA_00042  4684|1020|363|1936|4409  unidirectional_promoter  ...    NaN   \n",
       "EGA_00043  4683|1019|364|1937|4410  unidirectional_promoter  ...    NaN   \n",
       "EGA_00044  4680|1016|367|1940|4413  unidirectional_promoter  ...    NaN   \n",
       "EGA_00045  4675|1011|372|1945|4418  unidirectional_promoter  ...    NaN   \n",
       "\n",
       "          DOMAINS closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                            \n",
       "ERG_00002     NaN                    SAH1               YER043C   \n",
       "ERG_00003     NaN                    SAH1               YER043C   \n",
       "ERG_00004     NaN                    SAH1               YER043C   \n",
       "EGE_00001     NaN                    SAH1               YER043C   \n",
       "EGE_00002     NaN                    SAH1               YER043C   \n",
       "...           ...                     ...                   ...   \n",
       "EGA_00041     NaN                   ERG10               YPL028W   \n",
       "EGA_00042     NaN                   ERG10               YPL028W   \n",
       "EGA_00043     NaN                   ERG10               YPL028W   \n",
       "EGA_00044     NaN                   ERG10               YPL028W   \n",
       "EGA_00045     NaN                   ERG10               YPL028W   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "ERG_00002    upstream_gene_variant                    2.0   \n",
       "ERG_00003    upstream_gene_variant                    3.0   \n",
       "ERG_00004    upstream_gene_variant                   17.0   \n",
       "EGE_00001    upstream_gene_variant                   30.0   \n",
       "EGE_00002    upstream_gene_variant                   32.0   \n",
       "...                            ...                    ...   \n",
       "EGA_00041  downstream_gene_variant                  363.0   \n",
       "EGA_00042  downstream_gene_variant                  363.0   \n",
       "EGA_00043  downstream_gene_variant                  364.0   \n",
       "EGA_00044  downstream_gene_variant                  367.0   \n",
       "EGA_00045  downstream_gene_variant                  372.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00002                   ERG28               YER044C   \n",
       "ERG_00003                   ERG28               YER044C   \n",
       "ERG_00004                   ERG28               YER044C   \n",
       "EGE_00001                   ERG28               YER044C   \n",
       "EGE_00002                   ERG28               YER044C   \n",
       "...                           ...                   ...   \n",
       "EGA_00041                    SMA1               YPL027W   \n",
       "EGA_00042                    SMA1               YPL027W   \n",
       "EGA_00043                    SMA1               YPL027W   \n",
       "EGA_00044                    SMA1               YPL027W   \n",
       "EGA_00045                    SMA1               YPL027W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "ERG_00002  downstream_gene_variant                  449.0  \n",
       "ERG_00003  downstream_gene_variant                  448.0  \n",
       "ERG_00004  downstream_gene_variant                  434.0  \n",
       "EGE_00001  downstream_gene_variant                  421.0  \n",
       "EGE_00002  downstream_gene_variant                  419.0  \n",
       "...                            ...                    ...  \n",
       "EGA_00041    upstream_gene_variant                   10.0  \n",
       "EGA_00042    upstream_gene_variant                   10.0  \n",
       "EGA_00043    upstream_gene_variant                    9.0  \n",
       "EGA_00044    upstream_gene_variant                    6.0  \n",
       "EGA_00045    upstream_gene_variant                    1.0  \n",
       "\n",
       "[4082 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERG_00003</th>\n",
       "      <td>V</td>\n",
       "      <td>237122</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>3</td>\n",
       "      <td>2382|1631|1338|448|2910</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>448.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERG_00004</th>\n",
       "      <td>V</td>\n",
       "      <td>237136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>17</td>\n",
       "      <td>2396|1645|1324|434|2896</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>434.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00008</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TG</td>\n",
       "      <td>661</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.327000</td>\n",
       "      <td>72</td>\n",
       "      <td>2451|1700|1268|378|2840</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>72.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00009</th>\n",
       "      <td>V</td>\n",
       "      <td>237191</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>72</td>\n",
       "      <td>2451|1700|1268|378|2840</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>72.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGE_00010</th>\n",
       "      <td>V</td>\n",
       "      <td>237196</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>69</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.034000</td>\n",
       "      <td>77</td>\n",
       "      <td>2456|1705|1264|374|2836</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAH1</td>\n",
       "      <td>YER043C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>77.0</td>\n",
       "      <td>ERG28</td>\n",
       "      <td>YER044C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00039</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499649</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>3868|16</td>\n",
       "      <td>4690|1026|357|1930|4403</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>357.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00040</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499650</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>3867|15</td>\n",
       "      <td>4689|1025|358|1931|4404</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>358.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00041</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00042</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499655</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>35</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.017000</td>\n",
       "      <td>3862|10</td>\n",
       "      <td>4684|1020|363|1936|4409</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>363.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGA_00044</th>\n",
       "      <td>XVI</td>\n",
       "      <td>499659</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>3858|6</td>\n",
       "      <td>4680|1016|367|1940|4413</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG10</td>\n",
       "      <td>YPL028W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>367.0</td>\n",
       "      <td>SMA1</td>\n",
       "      <td>YPL027W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1580 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                       \n",
       "ERG_00003     V  237122   G   A    9  2022  0.004451                     3   \n",
       "ERG_00004     V  237136   T   C   26  2022  0.013000                    17   \n",
       "EGE_00008     V  237191   T  TG  661  2022  0.327000                    72   \n",
       "EGE_00009     V  237191   T  TA    6  2022  0.002967                    72   \n",
       "EGE_00010     V  237196   A   T   69  2022  0.034000                    77   \n",
       "...         ...     ...  ..  ..  ...   ...       ...                   ...   \n",
       "EGA_00039   XVI  499649   G   A    4  2022  0.001978               3868|16   \n",
       "EGA_00040   XVI  499650   G   C    5  2022  0.002473               3867|15   \n",
       "EGA_00041   XVI  499655   G   A    4  2022  0.001978               3862|10   \n",
       "EGA_00042   XVI  499655   G   T   35  2022  0.017000               3862|10   \n",
       "EGA_00044   XVI  499659   G   A    1  2022  0.000495                3858|6   \n",
       "\n",
       "           downstream_distance_str                   region  ... DOMAINS  \\\n",
       "var_id                                                       ...           \n",
       "ERG_00003  2382|1631|1338|448|2910  unidirectional_promoter  ...     NaN   \n",
       "ERG_00004  2396|1645|1324|434|2896  unidirectional_promoter  ...     NaN   \n",
       "EGE_00008  2451|1700|1268|378|2840  unidirectional_promoter  ...     NaN   \n",
       "EGE_00009  2451|1700|1268|378|2840  unidirectional_promoter  ...     NaN   \n",
       "EGE_00010  2456|1705|1264|374|2836  unidirectional_promoter  ...     NaN   \n",
       "...                            ...                      ...  ...     ...   \n",
       "EGA_00039  4690|1026|357|1930|4403  unidirectional_promoter  ...     NaN   \n",
       "EGA_00040  4689|1025|358|1931|4404  unidirectional_promoter  ...     NaN   \n",
       "EGA_00041  4684|1020|363|1936|4409  unidirectional_promoter  ...     NaN   \n",
       "EGA_00042  4684|1020|363|1936|4409  unidirectional_promoter  ...     NaN   \n",
       "EGA_00044  4680|1016|367|1940|4413  unidirectional_promoter  ...     NaN   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00003                    SAH1               YER043C   \n",
       "ERG_00004                    SAH1               YER043C   \n",
       "EGE_00008                    SAH1               YER043C   \n",
       "EGE_00009                    SAH1               YER043C   \n",
       "EGE_00010                    SAH1               YER043C   \n",
       "...                           ...                   ...   \n",
       "EGA_00039                   ERG10               YPL028W   \n",
       "EGA_00040                   ERG10               YPL028W   \n",
       "EGA_00041                   ERG10               YPL028W   \n",
       "EGA_00042                   ERG10               YPL028W   \n",
       "EGA_00044                   ERG10               YPL028W   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "ERG_00003    upstream_gene_variant                    3.0   \n",
       "ERG_00004    upstream_gene_variant                   17.0   \n",
       "EGE_00008    upstream_gene_variant                   72.0   \n",
       "EGE_00009    upstream_gene_variant                   72.0   \n",
       "EGE_00010    upstream_gene_variant                   77.0   \n",
       "...                            ...                    ...   \n",
       "EGA_00039  downstream_gene_variant                  357.0   \n",
       "EGA_00040  downstream_gene_variant                  358.0   \n",
       "EGA_00041  downstream_gene_variant                  363.0   \n",
       "EGA_00042  downstream_gene_variant                  363.0   \n",
       "EGA_00044  downstream_gene_variant                  367.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "ERG_00003                   ERG28               YER044C   \n",
       "ERG_00004                   ERG28               YER044C   \n",
       "EGE_00008                   ERG28               YER044C   \n",
       "EGE_00009                   ERG28               YER044C   \n",
       "EGE_00010                   ERG28               YER044C   \n",
       "...                           ...                   ...   \n",
       "EGA_00039                    SMA1               YPL027W   \n",
       "EGA_00040                    SMA1               YPL027W   \n",
       "EGA_00041                    SMA1               YPL027W   \n",
       "EGA_00042                    SMA1               YPL027W   \n",
       "EGA_00044                    SMA1               YPL027W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \n",
       "var_id                                                                   \n",
       "ERG_00003  downstream_gene_variant                  448.0           2.0  \n",
       "ERG_00004  downstream_gene_variant                  434.0           2.0  \n",
       "EGE_00008  downstream_gene_variant                  378.0           2.0  \n",
       "EGE_00009  downstream_gene_variant                  378.0           2.0  \n",
       "EGE_00010  downstream_gene_variant                  374.0           2.0  \n",
       "...                            ...                    ...           ...  \n",
       "EGA_00039    upstream_gene_variant                   16.0           2.0  \n",
       "EGA_00040    upstream_gene_variant                   15.0           2.0  \n",
       "EGA_00041    upstream_gene_variant                   10.0           2.0  \n",
       "EGA_00042    upstream_gene_variant                   10.0           2.0  \n",
       "EGA_00044    upstream_gene_variant                    6.0           2.0  \n",
       "\n",
       "[1580 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "oligo_file = os.path.expanduser(\"~/crispey3/initial_design/Output/all_SNPs_ergosterol_initial_GG_9bp_OLIGO.tab\")\n",
    "ergosterol_oligo_df = pd.read_csv(oligo_file, sep='\\t')\n",
    "\n",
    "ergosterol_design_oligos = pd.read_csv(\"/home/users/rang/scratch/yeast/ergosterol/ergosterol_variants_select_AFfilter_annotated.txt\", sep='\\t', index_col=0)\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "ergosterol_design_oligos = ergosterol_design_oligos.query('(Gene_ID.isin(@ergosterol_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(ergosterol_design_oligos)\n",
    "\n",
    "ergosterol_design_oligos['num_of_oligos'] = ergosterol_oligo_df.groupby('var_id').apply(len)\n",
    "ergosterol_oligos_final = ergosterol_design_oligos.query('num_of_oligos>1').copy()\n",
    "display(ergosterol_oligos_final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write ergosterol_oligos_final table to file\n",
    "ergosterol_oligos_final.to_csv(\"/home/users/rang/scratch/yeast/ergosterol/ergosterol_variants_final.txt\", sep='\\t')\n",
    "# use this file in separate script to organize into pools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of ergosterol variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants: 1580\n",
      "Number of oligos: 3968\n",
      "\n",
      "Distribution of variants among ergosterol genes\n",
      "count     26.000000\n",
      "mean      49.384615\n",
      "std       26.023185\n",
      "min       19.000000\n",
      "25%       32.250000\n",
      "50%       42.000000\n",
      "75%       57.000000\n",
      "max      114.000000\n",
      "Name: Gene_ID, dtype: float64\n",
      "\n",
      "Distribution of oligos among ergosterol variants\n",
      "count    1580.000000\n",
      "mean        2.511392\n",
      "std         0.858882\n",
      "min         2.000000\n",
      "25%         2.000000\n",
      "50%         2.000000\n",
      "75%         3.000000\n",
      "max         7.000000\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Proportion of variant annotation classes\n",
      "synonymous_variant         0.372152\n",
      "missense_variant           0.236076\n",
      "unidirectional_promoter    0.183544\n",
      "bidirectional_promoter     0.149367\n",
      "intergenic                 0.053165\n",
      "frameshift_variant         0.002532\n",
      "stop_gained                0.001899\n",
      "inframe_deletion           0.001266\n",
      "Name: region, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "### SUMMARY OF ERGOSTEROL SET ###\n",
    "print(\"Number of variants:\", len(ergosterol_oligos_final))\n",
    "print(\"Number of oligos:\", int(sum(ergosterol_oligos_final.num_of_oligos)))\n",
    "print()\n",
    "print(\"Distribution of variants among ergosterol genes\")\n",
    "print(ergosterol_oligos_final.Gene_ID.value_counts().filter(ergosterol_genes.index).describe())\n",
    "print()\n",
    "print(\"Distribution of oligos among ergosterol variants\")\n",
    "print(ergosterol_oligos_final.num_of_oligos.describe())\n",
    "print()\n",
    "print(\"Proportion of variant annotation classes\")\n",
    "print(ergosterol_oligos_final.region.value_counts()/len(ergosterol_oligos_final))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine number of pools to use for ergosterol oligos\n",
    "If a variant set has too few oligos to form its own pool, we can group it with another variant set with similar editable genomes\n",
    "- ERG --> assign excess to EGE/EGD/EGB\n",
    "- EGE, EGD --> assign excess to EGC \n",
    "- EGC, EGB, EGA, EG9 --> assign excess to EG8\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ERG    2479.0\n",
       "EGE     469.0\n",
       "EG8     253.0\n",
       "EGB     209.0\n",
       "EG9     196.0\n",
       "EGC     193.0\n",
       "EGD     134.0\n",
       "EGA      35.0\n",
       "Name: num_of_oligos, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos to fit: 3968\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(33, 36)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use 33 pools:\n",
      "Number of oligos per pool: 118\n",
      "Number of oligos leftover: 74\n",
      "Number of technical oligos: 99\n",
      "\n",
      "Use 34 pools:\n",
      "Number of oligos per pool: 116\n",
      "Number of oligos leftover: 24\n",
      "Number of technical oligos: 170\n",
      "\n",
      "Use 35 pools:\n",
      "Number of oligos per pool: 113\n",
      "Number of oligos leftover: 13\n",
      "Number of technical oligos: 280\n",
      "\n",
      "Use 36 pools:\n",
      "Number of oligos per pool: 110\n",
      "Number of oligos leftover: 8\n",
      "Number of technical oligos: 396\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([33, 34, 35, 36])"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count OLIGOS in each variant prefix set\n",
    "display(ergosterol_oligos_final.groupby(lambda x: x[:3]).num_of_oligos.sum().sort_values(ascending=False))\n",
    "\n",
    "# determine number of pools to use\n",
    "num_of_oligos = int(ergosterol_oligos_final.num_of_oligos.sum())\n",
    "min_pool_size = 109 # allows 12 technical oligos\n",
    "max_pool_size = 118 # allows 3 technical oligos\n",
    "complete_pool_size = 121 # maximum oligos synthesized per pool\n",
    "\n",
    "# find number of pools to use\n",
    "find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant ID prefix and which genomes they can be edited in:\n",
    "- ERG: BY | RM | YJM | YPS\n",
    "- EGE: BY | RM | YJM | ---\n",
    "- EGD: BY | RM | --- | YPS\n",
    "- EGC: BY | RM | --- | ---\n",
    "- EGB: BY | --- | YJM | YPS\n",
    "- EGA: BY | --- | YJM | ---\n",
    "- EG9: BY | --- | --- | YPS\n",
    "- EG8: BY | --- | --- | ---\n",
    "\n",
    "\n",
    "### We should be able to fit all ergosterol variants in 34 pools\n",
    "- ERG: 21 pools (assign spillover to EGE)\n",
    "- EGE: 4 pools (assign spillover to EGC)\n",
    "- EGD: 1 pool, (assign spillover to EGC)\n",
    "- EGC: 2 pools, (assign spillover to EG8)\n",
    "- EGB: 1 pool (assign spillover to EGA)\n",
    "- EGA: 1 pool, (assign spillover to EG8)\n",
    "- EG9: 1 pool, (assign spillover to EG8)\n",
    "- EG8: 3 pools\n",
    "\n",
    "### Based on the aforementioned oligo arrangement into pools, each yeast strain will be assayed by the following pools:\n",
    "- BY: 34 pools (all pools)\n",
    "- RM: 28 pools (21 ERG, 4 EGE, 1 EGD, 2 EGC pools)\n",
    "- YJM: 27 pools (21 ERG, 4 EGE, 1 EGB, 1 EGA pools)\n",
    "- YPS: 24 pools (21 ERG, 1 EGD, 1 EGB, 1 EG9 pools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selecting variants in genes with high genetic interactions\n",
    "Survey natural variation in \"hub\" genes to identify GxG effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/\"\n",
    "# summary of genetic interactions output file\n",
    "genetic_interactions_summary_file = working_dir+'genetic_interactions_summary_by_gene.txt'\n",
    "os.chdir(working_dir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>essential</th>\n",
       "      <th>paralog</th>\n",
       "      <th>KA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFL039C</th>\n",
       "      <td>1324</td>\n",
       "      <td>2220</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>53260</td>\n",
       "      <td>54696</td>\n",
       "      <td>-</td>\n",
       "      <td>ACT1</td>\n",
       "      <td>Actin, structural protein involved in cell pol...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL034C-B</th>\n",
       "      <td>1150</td>\n",
       "      <td>1897</td>\n",
       "      <td>chrVI</td>\n",
       "      <td>63016</td>\n",
       "      <td>63993</td>\n",
       "      <td>-</td>\n",
       "      <td>MOB2</td>\n",
       "      <td>Component of the RAM signaling network, that a...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJR076C</th>\n",
       "      <td>749</td>\n",
       "      <td>1342</td>\n",
       "      <td>chrX</td>\n",
       "      <td>575354</td>\n",
       "      <td>576601</td>\n",
       "      <td>-</td>\n",
       "      <td>CDC11</td>\n",
       "      <td>Component of the septin ring of the mother-bud...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBL105C</th>\n",
       "      <td>841</td>\n",
       "      <td>1139</td>\n",
       "      <td>chrII</td>\n",
       "      <td>14241</td>\n",
       "      <td>17696</td>\n",
       "      <td>-</td>\n",
       "      <td>PKC1</td>\n",
       "      <td>Protein serine/threonine kinase essential for ...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YER157W</th>\n",
       "      <td>925</td>\n",
       "      <td>967</td>\n",
       "      <td>chrV</td>\n",
       "      <td>484788</td>\n",
       "      <td>487193</td>\n",
       "      <td>+</td>\n",
       "      <td>COG3</td>\n",
       "      <td>Essential component of the conserved oligomeri...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLR038C</th>\n",
       "      <td>178</td>\n",
       "      <td>352</td>\n",
       "      <td>chrXII</td>\n",
       "      <td>224921</td>\n",
       "      <td>225172</td>\n",
       "      <td>-</td>\n",
       "      <td>COX12</td>\n",
       "      <td>Subunit VIb of cytochrome c oxidase, which is ...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YOR038C</th>\n",
       "      <td>200</td>\n",
       "      <td>330</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>402761</td>\n",
       "      <td>405388</td>\n",
       "      <td>-</td>\n",
       "      <td>HIR2</td>\n",
       "      <td>Subunit of the HIR complex, a nucleosome assem...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YHR200W</th>\n",
       "      <td>200</td>\n",
       "      <td>330</td>\n",
       "      <td>chrVIII</td>\n",
       "      <td>499079</td>\n",
       "      <td>499885</td>\n",
       "      <td>+</td>\n",
       "      <td>RPN10</td>\n",
       "      <td>Non-ATPase base subunit of the 19S regulatory ...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YOR083W</th>\n",
       "      <td>223</td>\n",
       "      <td>306</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>479533</td>\n",
       "      <td>480420</td>\n",
       "      <td>+</td>\n",
       "      <td>WHI5</td>\n",
       "      <td>Repressor of G1 transcription that binds to SC...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1.484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGR097W</th>\n",
       "      <td>159</td>\n",
       "      <td>370</td>\n",
       "      <td>chrVII</td>\n",
       "      <td>678695</td>\n",
       "      <td>682135</td>\n",
       "      <td>+</td>\n",
       "      <td>ASK10</td>\n",
       "      <td>Component of RNA polymerase II holoenzyme, pho...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.561</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>800 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           positive_interactions  negative_interactions    chrom   start  \\\n",
       "gene_id                                                                    \n",
       "YFL039C                     1324                   2220    chrVI   53260   \n",
       "YFL034C-B                   1150                   1897    chrVI   63016   \n",
       "YJR076C                      749                   1342     chrX  575354   \n",
       "YBL105C                      841                   1139    chrII   14241   \n",
       "YER157W                      925                    967     chrV  484788   \n",
       "...                          ...                    ...      ...     ...   \n",
       "YLR038C                      178                    352   chrXII  224921   \n",
       "YOR038C                      200                    330    chrXV  402761   \n",
       "YHR200W                      200                    330  chrVIII  499079   \n",
       "YOR083W                      223                    306    chrXV  479533   \n",
       "YGR097W                      159                    370   chrVII  678695   \n",
       "\n",
       "              end strand gene_name  \\\n",
       "gene_id                              \n",
       "YFL039C     54696      -      ACT1   \n",
       "YFL034C-B   63993      -      MOB2   \n",
       "YJR076C    576601      -     CDC11   \n",
       "YBL105C     17696      -      PKC1   \n",
       "YER157W    487193      +      COG3   \n",
       "...           ...    ...       ...   \n",
       "YLR038C    225172      -     COX12   \n",
       "YOR038C    405388      -      HIR2   \n",
       "YHR200W    499885      +     RPN10   \n",
       "YOR083W    480420      +      WHI5   \n",
       "YGR097W    682135      +     ASK10   \n",
       "\n",
       "                                                 description  essential  \\\n",
       "gene_id                                                                   \n",
       "YFL039C    Actin, structural protein involved in cell pol...       True   \n",
       "YFL034C-B  Component of the RAM signaling network, that a...       True   \n",
       "YJR076C    Component of the septin ring of the mother-bud...       True   \n",
       "YBL105C    Protein serine/threonine kinase essential for ...       True   \n",
       "YER157W    Essential component of the conserved oligomeri...       True   \n",
       "...                                                      ...        ...   \n",
       "YLR038C    Subunit VIb of cytochrome c oxidase, which is ...      False   \n",
       "YOR038C    Subunit of the HIR complex, a nucleosome assem...      False   \n",
       "YHR200W    Non-ATPase base subunit of the 19S regulatory ...      False   \n",
       "YOR083W    Repressor of G1 transcription that binds to SC...      False   \n",
       "YGR097W    Component of RNA polymerase II holoenzyme, pho...      False   \n",
       "\n",
       "           paralog     KA  \n",
       "gene_id                    \n",
       "YFL039C      False    NaN  \n",
       "YFL034C-B    False    NaN  \n",
       "YJR076C      False    NaN  \n",
       "YBL105C      False    NaN  \n",
       "YER157W      False    NaN  \n",
       "...            ...    ...  \n",
       "YLR038C      False    NaN  \n",
       "YOR038C      False    NaN  \n",
       "YHR200W      False    NaN  \n",
       "YOR083W       True  1.484  \n",
       "YGR097W       True  0.561  \n",
       "\n",
       "[800 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# examine top genes with most genetic interactions\n",
    "genes_cutoff = 800\n",
    "\n",
    "num_of_interactions_df = pd.read_csv(genetic_interactions_summary_file.replace('.txt', '_annotated.txt'), sep='\\t', header = 0, index_col=0)\n",
    "num_of_interactions_df = num_of_interactions_df.dropna().astype({'start':'int', 'end':'int'})\n",
    "\n",
    "# remove dubious orfs\n",
    "dubious_orfs_file = working_dir+'dubious_orfs.tsv'\n",
    "dubious_orfs = pd.read_csv(dubious_orfs_file, sep='\\t', header=None, names=['gene_id', 'description'])\n",
    "num_of_interactions_df = num_of_interactions_df.query('~gene_id.isin(@dubious_orfs.gene_id)')\n",
    "\n",
    "\n",
    "# remove genes in excluded list\n",
    "genes_to_exclude = ['YDL227C', # HO\n",
    "                    'YOR202W', # HIS3\n",
    "                    'YEL021W', # URA3\n",
    "                    'YCL018W', # LEU2\n",
    "                    'YBR115C', # LYS2\n",
    "                    'YBR020W', # GAL1\n",
    "                    'YDR009W', # GAL3\n",
    "                    'YPL248C', # GAL4\n",
    "                    'YBR018C', # GAL7\n",
    "                    'YBR019C', # GAL10\n",
    "                    'YML051W', # GAL80\n",
    "                    'YLR256W'] # HAP1\n",
    "genes_to_exclude = genes_to_exclude + ergosterol_genes.index.tolist() # exclude ergosterol pathway genes that already have oligos synthesized in separate set\n",
    "num_of_interactions_df = num_of_interactions_df.query('~gene_id.isin(@genes_to_exclude)')\n",
    "\n",
    "\n",
    "# find top genes with most genetic interactions\n",
    "interaction_rich_genes = num_of_interactions_df.loc[num_of_interactions_df[['positive_interactions', 'negative_interactions']].sum(axis=1).sort_values(ascending=False).index].head(genes_cutoff)\n",
    "\n",
    "# annotate genes by essentiality\n",
    "essential_genes_file = working_dir+\"S288C_essential_genes.txt\"\n",
    "essential_genes_df = pd.read_csv(essential_genes_file, sep='\\t')\n",
    "interaction_rich_genes.loc[:,'essential'] = interaction_rich_genes.index.isin(essential_genes_df['Gene Systematic Name'])\n",
    "\n",
    "# annotate genes by paralogy\n",
    "paralogs_file = working_dir+\"S288C_paralogs.txt\"\n",
    "paralogs_df = pd.read_csv(paralogs_file, sep='\\t', index_col='Byrne')\n",
    "paralogs_ka_list = pd.concat([paralogs_df[['Gene 1', 'KA']].rename(columns={'Gene 1':'Gene'}), paralogs_df[['Gene 2', 'KA']].rename(columns={'Gene 2':'Gene'})], axis=0, ignore_index=True).set_index('Gene')\n",
    "interaction_rich_genes.loc[:,'paralog'] = interaction_rich_genes.index.isin(paralogs_ka_list.index)\n",
    "interaction_rich_genes.loc[interaction_rich_genes['paralog'],'KA'] = paralogs_ka_list.loc[interaction_rich_genes.index[interaction_rich_genes['paralog']], 'KA']\n",
    "\n",
    "display(interaction_rich_genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search hub genes for variants in each gvcf\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16',}\n",
    "\n",
    "genes_to_process = []\n",
    "for chrom in chromosome_dict.keys():\n",
    "    genes_df = interaction_rich_genes.query('chrom==@chrom').sort_values('start')\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict[chrom])\n",
    "    genes_to_process.append((genes_df, gvcf_file, chromosome_dict))\n",
    "\n",
    "# consolidate candidate variants\n",
    "num_of_cores = min(len(os.sched_getaffinity(0)), len(genes_to_process))\n",
    "with mp.Pool(num_of_cores) as pool:\n",
    "    results = pool.starmap(select_variants, genes_to_process)\n",
    "results = pd.concat(results).reset_index(drop=True)\n",
    "\n",
    "# write to new vcf file\n",
    "gxg_variants_file = \"gxg_variants.vcf\"\n",
    "template = vcf.Reader(filename=genes_to_process[0][1])\n",
    "# change chromosome naming convention to a single Roman numeral (follows ref genome, VEP formatting)\n",
    "chromosome_dict_rev = {v:k for k, v in chromosome_dict.items()} \n",
    "template.contigs = OrderedDict((chromosome_dict_rev[k][3:], v._replace(id=chromosome_dict_rev[k][3:])) for k, v in template.contigs.items())\n",
    "# remove 1011 genomes samples data\n",
    "template.samples=[]\n",
    "\n",
    "output = vcf.Writer(open(gxg_variants_file, 'w'), template)\n",
    "for record in results:\n",
    "    # update chromosome naming\n",
    "    record.CHROM = chromosome_dict_rev[record.CHROM][3:]\n",
    "    # write to file\n",
    "    output.write_record(record)\n",
    "output.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant at chrom IV, pos 434292 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 434294 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 434307 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIV, pos 574011 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIV, pos 574012 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIV, pos 574018 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 86643 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 86648 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XI, pos 162514 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117522 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117533 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117538 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117552 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117675 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117677 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117678 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117679 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117701 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1336 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1337 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1339 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1341 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1344 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1359 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1362 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 117998 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1415 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1416 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1417 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1419 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1420 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1422 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1449 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1450 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1453 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1454 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1456 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1458 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1462 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1463 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1465 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1467 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1468 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1470 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1471 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1480 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1487 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1494 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1497 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1501 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1511 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1516 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1517 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1523 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1525 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1526 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1527 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1528 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1529 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1530 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1531 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1534 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1535 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1536 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1537 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1539 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1540 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1668 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1669 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1672 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1675 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1677 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1694 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1797 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1798 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1939 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1944 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1950 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1957 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1963 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1964 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1975 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1982 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1984 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1987 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1989 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1999 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2000 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2001 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2002 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2006 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2007 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2016 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2018 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2022 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2024 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2028 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2029 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2031 skipped: No unique matches found in tested genome(s)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant at chrom IV, pos 2033 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2035 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2036 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2040 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2041 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2043 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2044 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2045 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2046 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2047 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2049 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2052 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2054 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2055 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2056 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2058 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2060 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2062 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2063 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2064 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2066 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2067 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2068 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2072 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2073 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2074 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2076 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2095 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2615 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2616 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2617 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2618 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2619 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2620 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2628 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIII, pos 321186 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2629 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2632 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2635 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2638 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2656 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2709 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2710 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2713 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2841 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2842 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2843 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2845 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2846 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 2848 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 3030 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 3032 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 3043 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401557 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401738 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401740 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401746 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401748 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401750 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401754 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401762 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401764 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401765 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401768 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401769 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401782 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 401785 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIII, pos 363111 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 162027 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 162263 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VI, pos 162269 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 307028 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 307052 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 307100 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 875405 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 875426 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 875431 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 875442 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 391073 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 591805 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 591844 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 591877 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 591901 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 591929 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 591934 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 591936 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 591988 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 592112 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 592129 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 592354 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XV, pos 575205 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XV, pos 575211 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XV, pos 575238 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XV, pos 575241 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XV, pos 575250 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 620000 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 620017 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 941339 skipped: No unique matches found in tested genome(s)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant at chrom XII, pos 941425 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 941431 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 941477 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 941478 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 941479 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 941481 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 543689 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 543691 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 543697 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 543698 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 543704 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150870 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150879 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150887 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150888 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150891 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150892 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150899 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150902 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 1150910 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 469506 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 727664 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VII, pos 727665 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 487359 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 487361 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 492383 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XVI, pos 211798 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757436 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757439 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757443 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757444 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757445 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757451 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757452 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757454 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757459 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757478 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757479 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757481 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757484 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757487 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757505 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757506 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757508 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757521 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757532 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757538 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757544 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757558 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757560 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757562 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757564 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757670 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757698 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 757700 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758156 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758162 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758176 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758180 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758182 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758189 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758219 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758220 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758222 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758229 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758234 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758240 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758249 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758253 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758372 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758411 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758432 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758445 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758480 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758508 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758534 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758540 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758546 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758556 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758564 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758700 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 758756 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759030 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759032 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759086 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759302 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759306 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759308 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759371 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759392 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759405 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759414 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759416 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759440 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759441 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759468 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759470 skipped: No unique matches found in tested genome(s)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant at chrom IV, pos 759494 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759495 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759500 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759512 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759524 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759526 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759632 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759633 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 759638 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 153588 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 153741 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 153885 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154001 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154035 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154038 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154068 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154071 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154152 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154230 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIII, pos 747922 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154531 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 154551 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XI, pos 618041 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 48411 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 48416 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 197347 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 197348 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 197352 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 197353 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom X, pos 197584 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64305 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64307 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64308 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64317 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64319 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64320 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64322 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64325 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64332 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64335 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64338 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64341 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64763 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64764 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64775 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64788 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64794 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XII, pos 64797 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 475747 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 475748 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 347643 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 347652 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom II, pos 347653 skipped: No unique matches found in tested genome(s)\n"
     ]
    }
   ],
   "source": [
    "# path to genome fasta files\n",
    "ref_genome_fasta = \"/home/users/rang/yeast/genomes/Saccharomyces_cerevisiae.R64-1-1.dna.chromosome.I.fa\"\n",
    "rm_genome_fasta = \"/home/users/rang/yeast/genomes/RM11-1A_SGD_2015_JRIP00000000.fsa\"\n",
    "yjm_genome_fasta = \"/home/users/rang/yeast/genomes/YJM789_Stanford_2007_AAFW02000000_highQuality31.fsa\"\n",
    "yps_genome_fasta = \"/home/users/rang/yeast/genomes/YPS128.genome.fa\"\n",
    "\n",
    "# path to input and output VCF files\n",
    "gxg_variants_file = \"gxg_variants.vcf\"\n",
    "gxg_variants_select_file = gxg_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# load genome files\n",
    "ref_genome = SeqIO.to_dict(SeqIO.parse(ref_genome_fasta, \"fasta\"))\n",
    "rm_genome = SeqIO.to_dict(SeqIO.parse(rm_genome_fasta, \"fasta\"))\n",
    "yjm_genome = SeqIO.to_dict(SeqIO.parse(yjm_genome_fasta, \"fasta\"))\n",
    "yps_genome = SeqIO.to_dict(SeqIO.parse(yps_genome_fasta, \"fasta\"))\n",
    "genomes_list = [ref_genome, rm_genome, yjm_genome, yps_genome] # IMPORTANT: SET REFERENCE GENOME AS FIRST IN LIST\n",
    "\n",
    "\n",
    "threshold = 10 # threshold value for searching NGGs\n",
    "window = 61 # size of flanking sequences to search for unique hits\n",
    "\n",
    "# read in all candidate variants to filter\n",
    "gxg_variants = vcf.Reader(filename=gxg_variants_file)\n",
    "gxg_variants = [rec for rec in gxg_variants]\n",
    "# split variants into chunks for multiprocessing\n",
    "n = len(gxg_variants)//len(os.sched_getaffinity(0)) + 1\n",
    "gxg_variants = [gxg_variants[i:i+n] for i in range(0, len(gxg_variants), n)]\n",
    "\n",
    "# find targetable variants (TGT info field indicates which genomes variant is targetable in)\n",
    "to_process = [(rec_list, genomes_list, threshold, window) for rec_list in gxg_variants]\n",
    "with mp.Pool(len(os.sched_getaffinity(0))) as pool:\n",
    "    result = pool.starmap(find_targetable_variants, to_process) \n",
    "\n",
    "# write targetable variants to file\n",
    "header = vcf.Reader(filename=gxg_variants_file)\n",
    "# add \"TGT\" info field into VCF header\n",
    "header_info = namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "header.infos['TGT'] = header_info_tgt\n",
    "# write to file\n",
    "gxg_variants_select = vcf.Writer(open(gxg_variants_select_file, 'w'), header)\n",
    "to_write = [rec for rec_list in result for rec in rec_list]\n",
    "for rec in to_write:\n",
    "    gxg_variants_select.write_record(rec)\n",
    "gxg_variants_select.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select variants by allele frequency (note difference from ergosterol variant selection step)\n",
    "### remove singletons and doubletons\n",
    "### if multiple alternate alleles, select most common alternate allele\n",
    "### assign unitl variant ID prefix and number (use format VAR_XXXXX)\n",
    "\n",
    "# candidate variants editable in all strains\n",
    "gxg_variants_select = vcf.Reader(filename=gxg_variants_select_file)\n",
    "\n",
    "# open output file\n",
    "gxg_variants_select_AFfilter_file = gxg_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "gxg_variants_select_AFfilter = vcf.Writer(open(gxg_variants_select_AFfilter_file, 'w'), gxg_variants_select)\n",
    "\n",
    "var_counter = {}\n",
    "for record in gxg_variants_select:        \n",
    "    # skip variant if all possible alternate allele count is less than 3\n",
    "    if all([ac<3 for ac in record.INFO['AC']]):\n",
    "        continue\n",
    "    # if at least one alternate allele meets minimum count, select most common alternate allele to study\n",
    "    else:\n",
    "        i = np.argmax(record.INFO['AC']) # most common alt allele\n",
    "        # update record entry\n",
    "        record.ALT = [record.ALT[i]]\n",
    "        record.INFO['AC'] = [record.INFO['AC'][i]]\n",
    "        record.INFO['AF'] = [record.INFO['AF'][i]]\n",
    "        record.INFO['MLEAC'] = [record.INFO['MLEAC'][i]]\n",
    "        record.INFO['MLEAF'] = [record.INFO['MLEAF'][i]]\n",
    "        \n",
    "        # inspect TGT field and assign variant ID prefix\n",
    "        ### if tgt field contains all 1 (i.e. targetable by all queried genomes), assign \"GXG\"\n",
    "        ### otherwise, skip variant\n",
    "        if all([i==1 for i in record.INFO['TGT']]):\n",
    "            var_id_prefix = \"GXG\"\n",
    "        else:\n",
    "#             print('variant does not pass TGT check: skipping')\n",
    "            continue\n",
    "#             # assign unique prefix \"GG-\" with the last digit determined by converting binary to hexadecimal (works well for up to 4 genomes)\n",
    "#             binary_str = ''.join([str(x) for x in record.INFO['TGT']])\n",
    "#             hex_digit = hex(int(binary_str, 2))[2:].upper()\n",
    "#             var_id_prefix = \"GG{}\".format(hex_digit)\n",
    "            \n",
    "        # add to counter\n",
    "        if var_id_prefix not in var_counter:\n",
    "            var_counter[var_id_prefix] = 1\n",
    "        else:\n",
    "            var_counter[var_id_prefix] += 1\n",
    "        \n",
    "        # add unique variant ID (follow format VAR_XXXXX)\n",
    "        record.ID='{}_{:05d}'.format(var_id_prefix, var_counter[var_id_prefix])\n",
    "        \n",
    "        # write to output\n",
    "        gxg_variants_select_AFfilter.write_record(record)\n",
    "gxg_variants_select_AFfilter.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant ID prefix and which genomes they can be edited in:\n",
    "- GXG: BY | RM | YJM | YPS\n",
    "- GGE: BY | RM | YJM | ---\n",
    "- GGD: BY | RM | --- | YPS\n",
    "- GGC: BY | RM | --- | ---\n",
    "- GGB: BY | --- | YJM | YPS\n",
    "- GGA: BY | --- | YJM | ---\n",
    "- GG9: BY | --- | --- | YPS\n",
    "- GG8: BY | --- | --- | ---\n",
    "\n",
    "Only variants targetable in all 4 yeast strains (GXG prefix-ed) will be included in this library"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit gxg_variants_select_AFfilter VCF to VEP to get annotations, and download VEP output in TXT format\n",
    "http://uswest.ensembl.org/Saccharomyces_cerevisiae/Tools/VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parsing entry: 5000\n",
      "parsing entry: 10000\n",
      "parsing entry: 15000\n",
      "parsing entry: 20000\n",
      "parsing entry: 25000\n",
      "parsing entry: 30000\n",
      "parsing entry: 35000\n",
      "parsing entry: 40000\n",
      "parsing entry: 45000\n",
      "parsing entry: 50000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00001</th>\n",
       "      <td>I</td>\n",
       "      <td>41691</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.005935</td>\n",
       "      <td>4208|486</td>\n",
       "      <td>1190|2719|4544</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>tcC/tcT</td>\n",
       "      <td>PANTHER:PTHR23244</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00002</th>\n",
       "      <td>I</td>\n",
       "      <td>41705</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>4194|472</td>\n",
       "      <td>1176|2733|4558</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gAt/gCt</td>\n",
       "      <td>PANTHER:PTHR23244</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00003</th>\n",
       "      <td>I</td>\n",
       "      <td>41847</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>4052|330</td>\n",
       "      <td>1034|2875|4700</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>taT/taC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00004</th>\n",
       "      <td>I</td>\n",
       "      <td>41850</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>45</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.022000</td>\n",
       "      <td>4049|327</td>\n",
       "      <td>1031|2878|4703</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>ttA/ttG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00005</th>\n",
       "      <td>I</td>\n",
       "      <td>42019</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>61</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3880|158</td>\n",
       "      <td>862|118|3047|4872</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>118.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50559</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920406</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>1365|1454</td>\n",
       "      <td>3444|2503|3898</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gcC/gcA</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50560</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920425</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>18</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.008920</td>\n",
       "      <td>1384|1435</td>\n",
       "      <td>3463|2484|3879</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gga/Aga</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50561</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920442</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>90</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.045000</td>\n",
       "      <td>1401|1418</td>\n",
       "      <td>3480|2467|3862</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gtA/gtG</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50562</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920446</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.007929</td>\n",
       "      <td>1405|1414</td>\n",
       "      <td>3484|2463|3858</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gtc/Atc</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50563</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920703</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>1662|1157</td>\n",
       "      <td>3741|216|2206|3601</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QCR2</td>\n",
       "      <td>YPR191W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>216.0</td>\n",
       "      <td>AQY1</td>\n",
       "      <td>YPR192W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>1157.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50563 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT  AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                      \n",
       "GXG_00001     I   41691   C   T  12  2022  0.005935              4208|486   \n",
       "GXG_00002     I   41705   A   C   4  2022  0.001978              4194|472   \n",
       "GXG_00003     I   41847   T   C   4  2022  0.001978              4052|330   \n",
       "GXG_00004     I   41850   A   G  45  2022  0.022000              4049|327   \n",
       "GXG_00005     I   42019   A   C  61  2022  0.030000              3880|158   \n",
       "...         ...     ...  ..  ..  ..   ...       ...                   ...   \n",
       "GXG_50559   XVI  920406   C   A   4  2018  0.001982             1365|1454   \n",
       "GXG_50560   XVI  920425   G   A  18  2018  0.008920             1384|1435   \n",
       "GXG_50561   XVI  920442   A   G  90  2016  0.045000             1401|1418   \n",
       "GXG_50562   XVI  920446   G   A  16  2018  0.007929             1405|1414   \n",
       "GXG_50563   XVI  920703   T   C   4  2018  0.001982             1662|1157   \n",
       "\n",
       "          downstream_distance_str                   region  ...   Codons  \\\n",
       "var_id                                                      ...            \n",
       "GXG_00001          1190|2719|4544       synonymous_variant  ...  tcC/tcT   \n",
       "GXG_00002          1176|2733|4558         missense_variant  ...  gAt/gCt   \n",
       "GXG_00003          1034|2875|4700       synonymous_variant  ...  taT/taC   \n",
       "GXG_00004          1031|2878|4703       synonymous_variant  ...  ttA/ttG   \n",
       "GXG_00005       862|118|3047|4872  unidirectional_promoter  ...      NaN   \n",
       "...                           ...                      ...  ...      ...   \n",
       "GXG_50559          3444|2503|3898       synonymous_variant  ...  gcC/gcA   \n",
       "GXG_50560          3463|2484|3879         missense_variant  ...  Gga/Aga   \n",
       "GXG_50561          3480|2467|3862       synonymous_variant  ...  gtA/gtG   \n",
       "GXG_50562          3484|2463|3858         missense_variant  ...  Gtc/Atc   \n",
       "GXG_50563      3741|216|2206|3601  unidirectional_promoter  ...      NaN   \n",
       "\n",
       "                                           DOMAINS closest_gene1_Gene_Name  \\\n",
       "var_id                                                                       \n",
       "GXG_00001                        PANTHER:PTHR23244                     NaN   \n",
       "GXG_00002                        PANTHER:PTHR23244                     NaN   \n",
       "GXG_00003                                      NaN                     NaN   \n",
       "GXG_00004                                      NaN                     NaN   \n",
       "GXG_00005                                      NaN                    GPB2   \n",
       "...                                            ...                     ...   \n",
       "GXG_50559  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50560  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50561  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50562  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50563                                      NaN                    QCR2   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00001                   NaN                      NaN   \n",
       "GXG_00002                   NaN                      NaN   \n",
       "GXG_00003                   NaN                      NaN   \n",
       "GXG_00004                   NaN                      NaN   \n",
       "GXG_00005               YAL056W  downstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "GXG_50559                   NaN                      NaN   \n",
       "GXG_50560                   NaN                      NaN   \n",
       "GXG_50561                   NaN                      NaN   \n",
       "GXG_50562                   NaN                      NaN   \n",
       "GXG_50563               YPR191W  downstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXG_00001                    NaN                     NaN   \n",
       "GXG_00002                    NaN                     NaN   \n",
       "GXG_00003                    NaN                     NaN   \n",
       "GXG_00004                    NaN                     NaN   \n",
       "GXG_00005                  118.0                   PEX22   \n",
       "...                          ...                     ...   \n",
       "GXG_50559                    NaN                     NaN   \n",
       "GXG_50560                    NaN                     NaN   \n",
       "GXG_50561                    NaN                     NaN   \n",
       "GXG_50562                    NaN                     NaN   \n",
       "GXG_50563                  216.0                    AQY1   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00001                   NaN                      NaN   \n",
       "GXG_00002                   NaN                      NaN   \n",
       "GXG_00003                   NaN                      NaN   \n",
       "GXG_00004                   NaN                      NaN   \n",
       "GXG_00005               YAL055W    upstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "GXG_50559                   NaN                      NaN   \n",
       "GXG_50560                   NaN                      NaN   \n",
       "GXG_50561                   NaN                      NaN   \n",
       "GXG_50562                   NaN                      NaN   \n",
       "GXG_50563               YPR192W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance  \n",
       "var_id                            \n",
       "GXG_00001                    NaN  \n",
       "GXG_00002                    NaN  \n",
       "GXG_00003                    NaN  \n",
       "GXG_00004                    NaN  \n",
       "GXG_00005                  158.0  \n",
       "...                          ...  \n",
       "GXG_50559                    NaN  \n",
       "GXG_50560                    NaN  \n",
       "GXG_50561                    NaN  \n",
       "GXG_50562                    NaN  \n",
       "GXG_50563                 1157.0  \n",
       "\n",
       "[50563 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00005</th>\n",
       "      <td>I</td>\n",
       "      <td>42019</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>61</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3880|158</td>\n",
       "      <td>862|118|3047|4872</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>118.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00006</th>\n",
       "      <td>I</td>\n",
       "      <td>42023</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.005941</td>\n",
       "      <td>3876|154</td>\n",
       "      <td>858|122|3051|4876</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>122.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00007</th>\n",
       "      <td>I</td>\n",
       "      <td>42033</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>60</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3865|143</td>\n",
       "      <td>847|132|3061|4886</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>132.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00008</th>\n",
       "      <td>I</td>\n",
       "      <td>42036</td>\n",
       "      <td>TTGG</td>\n",
       "      <td>CTGG</td>\n",
       "      <td>6</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.002976</td>\n",
       "      <td>3860|138</td>\n",
       "      <td>842|135|3064|4889</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>135.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>138.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00009</th>\n",
       "      <td>I</td>\n",
       "      <td>42037</td>\n",
       "      <td>TGG</td>\n",
       "      <td>T</td>\n",
       "      <td>30</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>3860|138</td>\n",
       "      <td>842|137|3066|4891</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>137.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>138.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50559</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920406</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>1365|1454</td>\n",
       "      <td>3444|2503|3898</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gcC/gcA</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50560</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920425</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>18</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.008920</td>\n",
       "      <td>1384|1435</td>\n",
       "      <td>3463|2484|3879</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gga/Aga</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50561</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920442</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>90</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.045000</td>\n",
       "      <td>1401|1418</td>\n",
       "      <td>3480|2467|3862</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gtA/gtG</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50562</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920446</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.007929</td>\n",
       "      <td>1405|1414</td>\n",
       "      <td>3484|2463|3858</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gtc/Atc</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50563</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920703</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>1662|1157</td>\n",
       "      <td>3741|216|2206|3601</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QCR2</td>\n",
       "      <td>YPR191W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>216.0</td>\n",
       "      <td>AQY1</td>\n",
       "      <td>YPR192W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>1157.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44835 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS   REF   ALT  AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                          \n",
       "GXG_00005     I   42019     A     C  61  2022  0.030000              3880|158   \n",
       "GXG_00006     I   42023     C     T  12  2020  0.005941              3876|154   \n",
       "GXG_00007     I   42033     A    AT  60  2012  0.030000              3865|143   \n",
       "GXG_00008     I   42036  TTGG  CTGG   6  2016  0.002976              3860|138   \n",
       "GXG_00009     I   42037   TGG     T  30  2022  0.015000              3860|138   \n",
       "...         ...     ...   ...   ...  ..   ...       ...                   ...   \n",
       "GXG_50559   XVI  920406     C     A   4  2018  0.001982             1365|1454   \n",
       "GXG_50560   XVI  920425     G     A  18  2018  0.008920             1384|1435   \n",
       "GXG_50561   XVI  920442     A     G  90  2016  0.045000             1401|1418   \n",
       "GXG_50562   XVI  920446     G     A  16  2018  0.007929             1405|1414   \n",
       "GXG_50563   XVI  920703     T     C   4  2018  0.001982             1662|1157   \n",
       "\n",
       "          downstream_distance_str                   region  ...   Codons  \\\n",
       "var_id                                                      ...            \n",
       "GXG_00005       862|118|3047|4872  unidirectional_promoter  ...      NaN   \n",
       "GXG_00006       858|122|3051|4876  unidirectional_promoter  ...      NaN   \n",
       "GXG_00007       847|132|3061|4886  unidirectional_promoter  ...      NaN   \n",
       "GXG_00008       842|135|3064|4889  unidirectional_promoter  ...      NaN   \n",
       "GXG_00009       842|137|3066|4891  unidirectional_promoter  ...      NaN   \n",
       "...                           ...                      ...  ...      ...   \n",
       "GXG_50559          3444|2503|3898       synonymous_variant  ...  gcC/gcA   \n",
       "GXG_50560          3463|2484|3879         missense_variant  ...  Gga/Aga   \n",
       "GXG_50561          3480|2467|3862       synonymous_variant  ...  gtA/gtG   \n",
       "GXG_50562          3484|2463|3858         missense_variant  ...  Gtc/Atc   \n",
       "GXG_50563      3741|216|2206|3601  unidirectional_promoter  ...      NaN   \n",
       "\n",
       "                                           DOMAINS closest_gene1_Gene_Name  \\\n",
       "var_id                                                                       \n",
       "GXG_00005                                      NaN                    GPB2   \n",
       "GXG_00006                                      NaN                    GPB2   \n",
       "GXG_00007                                      NaN                    GPB2   \n",
       "GXG_00008                                      NaN                    GPB2   \n",
       "GXG_00009                                      NaN                    GPB2   \n",
       "...                                            ...                     ...   \n",
       "GXG_50559  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50560  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50561  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50562  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50563                                      NaN                    QCR2   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00005               YAL056W  downstream_gene_variant   \n",
       "GXG_00006               YAL056W  downstream_gene_variant   \n",
       "GXG_00007               YAL056W  downstream_gene_variant   \n",
       "GXG_00008               YAL056W  downstream_gene_variant   \n",
       "GXG_00009               YAL056W  downstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "GXG_50559                   NaN                      NaN   \n",
       "GXG_50560                   NaN                      NaN   \n",
       "GXG_50561                   NaN                      NaN   \n",
       "GXG_50562                   NaN                      NaN   \n",
       "GXG_50563               YPR191W  downstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXG_00005                  118.0                   PEX22   \n",
       "GXG_00006                  122.0                   PEX22   \n",
       "GXG_00007                  132.0                   PEX22   \n",
       "GXG_00008                  135.0                   PEX22   \n",
       "GXG_00009                  137.0                   PEX22   \n",
       "...                          ...                     ...   \n",
       "GXG_50559                    NaN                     NaN   \n",
       "GXG_50560                    NaN                     NaN   \n",
       "GXG_50561                    NaN                     NaN   \n",
       "GXG_50562                    NaN                     NaN   \n",
       "GXG_50563                  216.0                    AQY1   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00005               YAL055W    upstream_gene_variant   \n",
       "GXG_00006               YAL055W    upstream_gene_variant   \n",
       "GXG_00007               YAL055W    upstream_gene_variant   \n",
       "GXG_00008               YAL055W    upstream_gene_variant   \n",
       "GXG_00009               YAL055W    upstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "GXG_50559                   NaN                      NaN   \n",
       "GXG_50560                   NaN                      NaN   \n",
       "GXG_50561                   NaN                      NaN   \n",
       "GXG_50562                   NaN                      NaN   \n",
       "GXG_50563               YPR192W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance  \n",
       "var_id                            \n",
       "GXG_00005                  158.0  \n",
       "GXG_00006                  154.0  \n",
       "GXG_00007                  143.0  \n",
       "GXG_00008                  138.0  \n",
       "GXG_00009                  138.0  \n",
       "...                          ...  \n",
       "GXG_50559                    NaN  \n",
       "GXG_50560                    NaN  \n",
       "GXG_50561                    NaN  \n",
       "GXG_50562                    NaN  \n",
       "GXG_50563                 1157.0  \n",
       "\n",
       "[44835 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GXG    44835\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add annotations to variants and remove variants that fall in genes not listed in interaction_rich_genes\n",
    "gxg_variants_file = \"gxg_variants.vcf\"\n",
    "gxg_variants_select_file = gxg_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# get VEP annotations and remove variants that fall within genes not in approved list\n",
    "gxg_variants_select_AFfilter_file = gxg_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "vep_output_filename = gxg_variants_select_AFfilter_file.replace('.vcf', '_VEPoutput.txt')\n",
    "variants_annotated_filename = vep_output_filename.replace('_VEPoutput.txt', '_annotated.txt')\n",
    "variants_annotated_df = annotate_variants_by_VEPoutput(gxg_variants_select_AFfilter_file, \n",
    "                                                       vep_output_filename, \n",
    "                                                       gff_file, \n",
    "                                                       variants_annotated_filename, \n",
    "                                                       id_colname='var_id')\n",
    "\n",
    "# annotated file can be read in directly if annotation function has been run once before\n",
    "variants_annotated_df = pd.read_csv(variants_annotated_filename, sep='\\t', index_col=0)\n",
    "display(variants_annotated_df)\n",
    "\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "gxg_design_oligos = variants_annotated_df.query('(Gene_ID.isin(@interaction_rich_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "\n",
    "##### (step not required if variants pre-filtered) ADDITIONALLY, since this set is focused on GxG, only include variants editable in at least 3 out of 4 genomes #####\n",
    "# gxg_design_oligos = gxg_design_oligos.filter(regex='GXG|GGE|GGD|GGB', axis=0)\n",
    "\n",
    "display(gxg_design_oligos)\n",
    "# count variants according to which strains they can be edited in\n",
    "display(pd.Series([x[:3] for x in gxg_design_oligos.index]).value_counts())\n",
    "\n",
    "\n",
    "# write variants to gxg_design_oligos VCF as a preliminary list to design oligos for\n",
    "gxg_design_oligos_vcf_file = gxg_variants_file.replace('.vcf', '_design_oligos_initial.vcf')\n",
    "gxg_variants_select_AFfilter = vcf.Reader(filename=gxg_variants_select_AFfilter_file)\n",
    "gxg_design_oligos_vcf = vcf.Writer(open(gxg_design_oligos_vcf_file, 'w'), gxg_variants_select_AFfilter)\n",
    "for record in gxg_variants_select_AFfilter:\n",
    "    if record.ID in gxg_design_oligos.index:\n",
    "        gxg_design_oligos_vcf.write_record(record)\n",
    "gxg_design_oligos_vcf.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design oligos for initial list, then filter down to variants with at least 2 oligos designed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00005</th>\n",
       "      <td>I</td>\n",
       "      <td>42019</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>61</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3880|158</td>\n",
       "      <td>862|118|3047|4872</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>118.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00006</th>\n",
       "      <td>I</td>\n",
       "      <td>42023</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.005941</td>\n",
       "      <td>3876|154</td>\n",
       "      <td>858|122|3051|4876</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>122.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00007</th>\n",
       "      <td>I</td>\n",
       "      <td>42033</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>60</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3865|143</td>\n",
       "      <td>847|132|3061|4886</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>132.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00008</th>\n",
       "      <td>I</td>\n",
       "      <td>42036</td>\n",
       "      <td>TTGG</td>\n",
       "      <td>CTGG</td>\n",
       "      <td>6</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.002976</td>\n",
       "      <td>3860|138</td>\n",
       "      <td>842|135|3064|4889</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>135.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>138.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00009</th>\n",
       "      <td>I</td>\n",
       "      <td>42037</td>\n",
       "      <td>TGG</td>\n",
       "      <td>T</td>\n",
       "      <td>30</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>3860|138</td>\n",
       "      <td>842|137|3066|4891</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>137.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>138.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50559</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920406</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>1365|1454</td>\n",
       "      <td>3444|2503|3898</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gcC/gcA</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50560</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920425</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>18</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.008920</td>\n",
       "      <td>1384|1435</td>\n",
       "      <td>3463|2484|3879</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gga/Aga</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50561</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920442</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>90</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.045000</td>\n",
       "      <td>1401|1418</td>\n",
       "      <td>3480|2467|3862</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gtA/gtG</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50562</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920446</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.007929</td>\n",
       "      <td>1405|1414</td>\n",
       "      <td>3484|2463|3858</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gtc/Atc</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50563</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920703</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>1662|1157</td>\n",
       "      <td>3741|216|2206|3601</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QCR2</td>\n",
       "      <td>YPR191W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>216.0</td>\n",
       "      <td>AQY1</td>\n",
       "      <td>YPR192W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>1157.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44835 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS   REF   ALT  AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                          \n",
       "GXG_00005     I   42019     A     C  61  2022  0.030000              3880|158   \n",
       "GXG_00006     I   42023     C     T  12  2020  0.005941              3876|154   \n",
       "GXG_00007     I   42033     A    AT  60  2012  0.030000              3865|143   \n",
       "GXG_00008     I   42036  TTGG  CTGG   6  2016  0.002976              3860|138   \n",
       "GXG_00009     I   42037   TGG     T  30  2022  0.015000              3860|138   \n",
       "...         ...     ...   ...   ...  ..   ...       ...                   ...   \n",
       "GXG_50559   XVI  920406     C     A   4  2018  0.001982             1365|1454   \n",
       "GXG_50560   XVI  920425     G     A  18  2018  0.008920             1384|1435   \n",
       "GXG_50561   XVI  920442     A     G  90  2016  0.045000             1401|1418   \n",
       "GXG_50562   XVI  920446     G     A  16  2018  0.007929             1405|1414   \n",
       "GXG_50563   XVI  920703     T     C   4  2018  0.001982             1662|1157   \n",
       "\n",
       "          downstream_distance_str                   region  ...   Codons  \\\n",
       "var_id                                                      ...            \n",
       "GXG_00005       862|118|3047|4872  unidirectional_promoter  ...      NaN   \n",
       "GXG_00006       858|122|3051|4876  unidirectional_promoter  ...      NaN   \n",
       "GXG_00007       847|132|3061|4886  unidirectional_promoter  ...      NaN   \n",
       "GXG_00008       842|135|3064|4889  unidirectional_promoter  ...      NaN   \n",
       "GXG_00009       842|137|3066|4891  unidirectional_promoter  ...      NaN   \n",
       "...                           ...                      ...  ...      ...   \n",
       "GXG_50559          3444|2503|3898       synonymous_variant  ...  gcC/gcA   \n",
       "GXG_50560          3463|2484|3879         missense_variant  ...  Gga/Aga   \n",
       "GXG_50561          3480|2467|3862       synonymous_variant  ...  gtA/gtG   \n",
       "GXG_50562          3484|2463|3858         missense_variant  ...  Gtc/Atc   \n",
       "GXG_50563      3741|216|2206|3601  unidirectional_promoter  ...      NaN   \n",
       "\n",
       "                                           DOMAINS closest_gene1_Gene_Name  \\\n",
       "var_id                                                                       \n",
       "GXG_00005                                      NaN                    GPB2   \n",
       "GXG_00006                                      NaN                    GPB2   \n",
       "GXG_00007                                      NaN                    GPB2   \n",
       "GXG_00008                                      NaN                    GPB2   \n",
       "GXG_00009                                      NaN                    GPB2   \n",
       "...                                            ...                     ...   \n",
       "GXG_50559  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50560  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50561  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50562  Gene3D:3.30.830.10,Superfamily:SSF63411                     NaN   \n",
       "GXG_50563                                      NaN                    QCR2   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00005               YAL056W  downstream_gene_variant   \n",
       "GXG_00006               YAL056W  downstream_gene_variant   \n",
       "GXG_00007               YAL056W  downstream_gene_variant   \n",
       "GXG_00008               YAL056W  downstream_gene_variant   \n",
       "GXG_00009               YAL056W  downstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "GXG_50559                   NaN                      NaN   \n",
       "GXG_50560                   NaN                      NaN   \n",
       "GXG_50561                   NaN                      NaN   \n",
       "GXG_50562                   NaN                      NaN   \n",
       "GXG_50563               YPR191W  downstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXG_00005                  118.0                   PEX22   \n",
       "GXG_00006                  122.0                   PEX22   \n",
       "GXG_00007                  132.0                   PEX22   \n",
       "GXG_00008                  135.0                   PEX22   \n",
       "GXG_00009                  137.0                   PEX22   \n",
       "...                          ...                     ...   \n",
       "GXG_50559                    NaN                     NaN   \n",
       "GXG_50560                    NaN                     NaN   \n",
       "GXG_50561                    NaN                     NaN   \n",
       "GXG_50562                    NaN                     NaN   \n",
       "GXG_50563                  216.0                    AQY1   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXG_00005               YAL055W    upstream_gene_variant   \n",
       "GXG_00006               YAL055W    upstream_gene_variant   \n",
       "GXG_00007               YAL055W    upstream_gene_variant   \n",
       "GXG_00008               YAL055W    upstream_gene_variant   \n",
       "GXG_00009               YAL055W    upstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "GXG_50559                   NaN                      NaN   \n",
       "GXG_50560                   NaN                      NaN   \n",
       "GXG_50561                   NaN                      NaN   \n",
       "GXG_50562                   NaN                      NaN   \n",
       "GXG_50563               YPR192W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance  \n",
       "var_id                            \n",
       "GXG_00005                  158.0  \n",
       "GXG_00006                  154.0  \n",
       "GXG_00007                  143.0  \n",
       "GXG_00008                  138.0  \n",
       "GXG_00009                  138.0  \n",
       "...                          ...  \n",
       "GXG_50559                    NaN  \n",
       "GXG_50560                    NaN  \n",
       "GXG_50561                    NaN  \n",
       "GXG_50562                    NaN  \n",
       "GXG_50563                 1157.0  \n",
       "\n",
       "[44835 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00006</th>\n",
       "      <td>I</td>\n",
       "      <td>42023</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.005941</td>\n",
       "      <td>3876|154</td>\n",
       "      <td>858|122|3051|4876</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>122.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>154.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00007</th>\n",
       "      <td>I</td>\n",
       "      <td>42033</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>60</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>3865|143</td>\n",
       "      <td>847|132|3061|4886</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPB2</td>\n",
       "      <td>YAL056W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>132.0</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>143.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00012</th>\n",
       "      <td>I</td>\n",
       "      <td>42184</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>3715</td>\n",
       "      <td>697|283|3212</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00013</th>\n",
       "      <td>I</td>\n",
       "      <td>42186</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>59</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.029000</td>\n",
       "      <td>3713</td>\n",
       "      <td>695|285|3214</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00014</th>\n",
       "      <td>I</td>\n",
       "      <td>42452</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>3447</td>\n",
       "      <td>429|551|3480</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Pfam:PF12827,Gene3D:3.40.50.11730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50554</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920332</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>29</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1291|1528</td>\n",
       "      <td>3370|2577|3972</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>PANTHER:PTHR11851:SF191,PANTHER:PTHR11851,Gene...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50556</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920361</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>6</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>1320|1499</td>\n",
       "      <td>3399|2548|3943</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>PANTHER:PTHR11851:SF191,PANTHER:PTHR11851,Gene...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50558</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920384</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>1343|1476</td>\n",
       "      <td>3422|2525|3920</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50559</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920406</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>1365|1454</td>\n",
       "      <td>3444|2503|3898</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50560</th>\n",
       "      <td>XVI</td>\n",
       "      <td>920425</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>18</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.008920</td>\n",
       "      <td>1384|1435</td>\n",
       "      <td>3463|2484|3879</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene3D:3.30.830.10,Superfamily:SSF63411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18255 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT  AC    AN        AF upstream_distance_str  \\\n",
       "var_id                                                                      \n",
       "GXG_00006     I   42023   C   T  12  2020  0.005941              3876|154   \n",
       "GXG_00007     I   42033   A  AT  60  2012  0.030000              3865|143   \n",
       "GXG_00012     I   42184   C   T   5  2022  0.002473                  3715   \n",
       "GXG_00013     I   42186   C   T  59  2022  0.029000                  3713   \n",
       "GXG_00014     I   42452   C   T   8  2022  0.003956                  3447   \n",
       "...         ...     ...  ..  ..  ..   ...       ...                   ...   \n",
       "GXG_50554   XVI  920332   A   G  29  2014  0.014000             1291|1528   \n",
       "GXG_50556   XVI  920361   C   T   6  2018  0.002973             1320|1499   \n",
       "GXG_50558   XVI  920384   C   A   4  2018  0.001982             1343|1476   \n",
       "GXG_50559   XVI  920406   C   A   4  2018  0.001982             1365|1454   \n",
       "GXG_50560   XVI  920425   G   A  18  2018  0.008920             1384|1435   \n",
       "\n",
       "          downstream_distance_str                   region  ...  \\\n",
       "var_id                                                      ...   \n",
       "GXG_00006       858|122|3051|4876  unidirectional_promoter  ...   \n",
       "GXG_00007       847|132|3061|4886  unidirectional_promoter  ...   \n",
       "GXG_00012            697|283|3212         missense_variant  ...   \n",
       "GXG_00013            695|285|3214         missense_variant  ...   \n",
       "GXG_00014            429|551|3480       synonymous_variant  ...   \n",
       "...                           ...                      ...  ...   \n",
       "GXG_50554          3370|2577|3972         missense_variant  ...   \n",
       "GXG_50556          3399|2548|3943       synonymous_variant  ...   \n",
       "GXG_50558          3422|2525|3920         missense_variant  ...   \n",
       "GXG_50559          3444|2503|3898       synonymous_variant  ...   \n",
       "GXG_50560          3463|2484|3879         missense_variant  ...   \n",
       "\n",
       "                                                     DOMAINS  \\\n",
       "var_id                                                         \n",
       "GXG_00006                                                NaN   \n",
       "GXG_00007                                                NaN   \n",
       "GXG_00012                                                NaN   \n",
       "GXG_00013                                                NaN   \n",
       "GXG_00014                  Pfam:PF12827,Gene3D:3.40.50.11730   \n",
       "...                                                      ...   \n",
       "GXG_50554  PANTHER:PTHR11851:SF191,PANTHER:PTHR11851,Gene...   \n",
       "GXG_50556  PANTHER:PTHR11851:SF191,PANTHER:PTHR11851,Gene...   \n",
       "GXG_50558            Gene3D:3.30.830.10,Superfamily:SSF63411   \n",
       "GXG_50559            Gene3D:3.30.830.10,Superfamily:SSF63411   \n",
       "GXG_50560            Gene3D:3.30.830.10,Superfamily:SSF63411   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXG_00006                    GPB2               YAL056W   \n",
       "GXG_00007                    GPB2               YAL056W   \n",
       "GXG_00012                     NaN                   NaN   \n",
       "GXG_00013                     NaN                   NaN   \n",
       "GXG_00014                     NaN                   NaN   \n",
       "...                           ...                   ...   \n",
       "GXG_50554                     NaN                   NaN   \n",
       "GXG_50556                     NaN                   NaN   \n",
       "GXG_50558                     NaN                   NaN   \n",
       "GXG_50559                     NaN                   NaN   \n",
       "GXG_50560                     NaN                   NaN   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "GXG_00006  downstream_gene_variant                  122.0   \n",
       "GXG_00007  downstream_gene_variant                  132.0   \n",
       "GXG_00012                      NaN                    NaN   \n",
       "GXG_00013                      NaN                    NaN   \n",
       "GXG_00014                      NaN                    NaN   \n",
       "...                            ...                    ...   \n",
       "GXG_50554                      NaN                    NaN   \n",
       "GXG_50556                      NaN                    NaN   \n",
       "GXG_50558                      NaN                    NaN   \n",
       "GXG_50559                      NaN                    NaN   \n",
       "GXG_50560                      NaN                    NaN   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXG_00006                   PEX22               YAL055W   \n",
       "GXG_00007                   PEX22               YAL055W   \n",
       "GXG_00012                     NaN                   NaN   \n",
       "GXG_00013                     NaN                   NaN   \n",
       "GXG_00014                     NaN                   NaN   \n",
       "...                           ...                   ...   \n",
       "GXG_50554                     NaN                   NaN   \n",
       "GXG_50556                     NaN                   NaN   \n",
       "GXG_50558                     NaN                   NaN   \n",
       "GXG_50559                     NaN                   NaN   \n",
       "GXG_50560                     NaN                   NaN   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \n",
       "var_id                                                                   \n",
       "GXG_00006    upstream_gene_variant                  154.0           2.0  \n",
       "GXG_00007    upstream_gene_variant                  143.0           2.0  \n",
       "GXG_00012                      NaN                    NaN           2.0  \n",
       "GXG_00013                      NaN                    NaN           3.0  \n",
       "GXG_00014                      NaN                    NaN           3.0  \n",
       "...                            ...                    ...           ...  \n",
       "GXG_50554                      NaN                    NaN           3.0  \n",
       "GXG_50556                      NaN                    NaN           2.0  \n",
       "GXG_50558                      NaN                    NaN           2.0  \n",
       "GXG_50559                      NaN                    NaN           2.0  \n",
       "GXG_50560                      NaN                    NaN           2.0  \n",
       "\n",
       "[18255 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gxg_oligo_file = os.path.expanduser(\"~/crispey3/initial_design/Output/all_SNPs_gxg_initial_GG_9bp_OLIGO.tab\")\n",
    "gxg_oligo_df = pd.read_csv(gxg_oligo_file, sep='\\t')\n",
    "\n",
    "gxg_design_oligos = pd.read_csv(\"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/gxg_variants_select_AFfilter_annotated.txt\", sep='\\t', index_col=0)\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "gxg_design_oligos = gxg_design_oligos.query('(Gene_ID.isin(@interaction_rich_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(gxg_design_oligos)\n",
    "\n",
    "gxg_design_oligos['num_of_oligos'] = gxg_oligo_df.groupby('var_id').apply(len)\n",
    "# select oligos that have at least 2 oligos designed\n",
    "gxg_oligos_final = gxg_design_oligos.query('num_of_oligos>1').copy()\n",
    "# select only variants targetable in all 4 yeast strains\n",
    "gxg_oligos_final = gxg_oligos_final.filter(regex='GXG', axis=0)\n",
    "\n",
    "display(gxg_oligos_final)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## From this list, I will select the final set of 1000 common and 1000 rare variants (~5k oligos) based on the following criteria:\n",
    "1. Determine max number of oligos allowable in library (e.g. 5000 oligos)\n",
    "2. Estimate number of variants allowed (5000/2.5 = 2000 variants)\n",
    "3. Set AF cutoff for what variants classify as \"common\" (e.g. 0.02, which is ~40 allele counts in pool of 2000).\n",
    "4. Rank genes by number of common variants, followed by number of rare variants (so among genes with an equal number of common variants, we select genes which have \"ultra-rare\" variants to sample).\n",
    "5. For each gene, sort the variants by allele frequency then allele count, in descending order. Select and count common variants that meet allele frequency cutoff. Then select an equal number of top rare variants to match common variants.\n",
    "6. Consolidate selected variants and associated genes, check final number of genes, variants and oligos to ensure they fit library specs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>essential</th>\n",
       "      <th>paralog</th>\n",
       "      <th>KA</th>\n",
       "      <th>left_limit</th>\n",
       "      <th>right_limit</th>\n",
       "      <th>region_length</th>\n",
       "      <th>num_of_vars</th>\n",
       "      <th>num_of_vars_common</th>\n",
       "      <th>num_of_vars_rare</th>\n",
       "      <th>var_density</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YNL262W</th>\n",
       "      <td>252</td>\n",
       "      <td>319</td>\n",
       "      <td>XIV</td>\n",
       "      <td>148212</td>\n",
       "      <td>154880</td>\n",
       "      <td>+</td>\n",
       "      <td>POL2</td>\n",
       "      <td>Catalytic subunit of DNA polymerase (II) epsil...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147712</td>\n",
       "      <td>155380</td>\n",
       "      <td>7669</td>\n",
       "      <td>99</td>\n",
       "      <td>24</td>\n",
       "      <td>15</td>\n",
       "      <td>1.290911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPL231W</th>\n",
       "      <td>348</td>\n",
       "      <td>548</td>\n",
       "      <td>XVI</td>\n",
       "      <td>108652</td>\n",
       "      <td>114315</td>\n",
       "      <td>+</td>\n",
       "      <td>FAS2</td>\n",
       "      <td>Alpha subunit of fatty acid synthetase, which ...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>108152</td>\n",
       "      <td>114815</td>\n",
       "      <td>6664</td>\n",
       "      <td>99</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.485594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR207C</th>\n",
       "      <td>242</td>\n",
       "      <td>298</td>\n",
       "      <td>XIII</td>\n",
       "      <td>677193</td>\n",
       "      <td>683564</td>\n",
       "      <td>-</td>\n",
       "      <td>HFA1</td>\n",
       "      <td>Mitochondrial acetyl-coenzyme A carboxylase, c...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.366</td>\n",
       "      <td>676693</td>\n",
       "      <td>684064</td>\n",
       "      <td>7372</td>\n",
       "      <td>81</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>1.098752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGR097W</th>\n",
       "      <td>159</td>\n",
       "      <td>370</td>\n",
       "      <td>VII</td>\n",
       "      <td>678695</td>\n",
       "      <td>682135</td>\n",
       "      <td>+</td>\n",
       "      <td>ASK10</td>\n",
       "      <td>Component of RNA polymerase II holoenzyme, pho...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.561</td>\n",
       "      <td>678195</td>\n",
       "      <td>682635</td>\n",
       "      <td>4441</td>\n",
       "      <td>52</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>1.170907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFL033C</th>\n",
       "      <td>276</td>\n",
       "      <td>396</td>\n",
       "      <td>VI</td>\n",
       "      <td>69115</td>\n",
       "      <td>74427</td>\n",
       "      <td>-</td>\n",
       "      <td>RIM15</td>\n",
       "      <td>Glucose-repressible protein kinase involved in...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68615</td>\n",
       "      <td>74927</td>\n",
       "      <td>6313</td>\n",
       "      <td>61</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>0.966260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YDL248W</th>\n",
       "      <td>300</td>\n",
       "      <td>344</td>\n",
       "      <td>IV</td>\n",
       "      <td>1802</td>\n",
       "      <td>2953</td>\n",
       "      <td>+</td>\n",
       "      <td>COS7</td>\n",
       "      <td>Protein of unknown function, member of the DUP...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1302</td>\n",
       "      <td>3453</td>\n",
       "      <td>2152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBR181C</th>\n",
       "      <td>305</td>\n",
       "      <td>338</td>\n",
       "      <td>II</td>\n",
       "      <td>591712</td>\n",
       "      <td>592774</td>\n",
       "      <td>-</td>\n",
       "      <td>RPS6B</td>\n",
       "      <td>Protein component of the small (40S) ribosomal...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000</td>\n",
       "      <td>591212</td>\n",
       "      <td>593274</td>\n",
       "      <td>2063</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YIR005W</th>\n",
       "      <td>275</td>\n",
       "      <td>346</td>\n",
       "      <td>IX</td>\n",
       "      <td>364889</td>\n",
       "      <td>365335</td>\n",
       "      <td>+</td>\n",
       "      <td>IST3</td>\n",
       "      <td>Component of the U2 snRNP, required for the fi...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>364389</td>\n",
       "      <td>365835</td>\n",
       "      <td>1447</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.138217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR256C</th>\n",
       "      <td>176</td>\n",
       "      <td>355</td>\n",
       "      <td>XIII</td>\n",
       "      <td>778945</td>\n",
       "      <td>779127</td>\n",
       "      <td>-</td>\n",
       "      <td>COX7</td>\n",
       "      <td>Subunit VII of cytochrome c oxidase, which is ...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>778445</td>\n",
       "      <td>779627</td>\n",
       "      <td>1183</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.084531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLR038C</th>\n",
       "      <td>178</td>\n",
       "      <td>352</td>\n",
       "      <td>XII</td>\n",
       "      <td>224921</td>\n",
       "      <td>225172</td>\n",
       "      <td>-</td>\n",
       "      <td>COX12</td>\n",
       "      <td>Subunit VIb of cytochrome c oxidase, which is ...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>224421</td>\n",
       "      <td>225672</td>\n",
       "      <td>1252</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.319489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>800 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         positive_interactions  negative_interactions chrom   start     end  \\\n",
       "gene_id                                                                       \n",
       "YNL262W                    252                    319   XIV  148212  154880   \n",
       "YPL231W                    348                    548   XVI  108652  114315   \n",
       "YMR207C                    242                    298  XIII  677193  683564   \n",
       "YGR097W                    159                    370   VII  678695  682135   \n",
       "YFL033C                    276                    396    VI   69115   74427   \n",
       "...                        ...                    ...   ...     ...     ...   \n",
       "YDL248W                    300                    344    IV    1802    2953   \n",
       "YBR181C                    305                    338    II  591712  592774   \n",
       "YIR005W                    275                    346    IX  364889  365335   \n",
       "YMR256C                    176                    355  XIII  778945  779127   \n",
       "YLR038C                    178                    352   XII  224921  225172   \n",
       "\n",
       "        strand gene_name                                        description  \\\n",
       "gene_id                                                                       \n",
       "YNL262W      +      POL2  Catalytic subunit of DNA polymerase (II) epsil...   \n",
       "YPL231W      +      FAS2  Alpha subunit of fatty acid synthetase, which ...   \n",
       "YMR207C      -      HFA1  Mitochondrial acetyl-coenzyme A carboxylase, c...   \n",
       "YGR097W      +     ASK10  Component of RNA polymerase II holoenzyme, pho...   \n",
       "YFL033C      -     RIM15  Glucose-repressible protein kinase involved in...   \n",
       "...        ...       ...                                                ...   \n",
       "YDL248W      +      COS7  Protein of unknown function, member of the DUP...   \n",
       "YBR181C      -     RPS6B  Protein component of the small (40S) ribosomal...   \n",
       "YIR005W      +      IST3  Component of the U2 snRNP, required for the fi...   \n",
       "YMR256C      -      COX7  Subunit VII of cytochrome c oxidase, which is ...   \n",
       "YLR038C      -     COX12  Subunit VIb of cytochrome c oxidase, which is ...   \n",
       "\n",
       "         essential  paralog     KA  left_limit  right_limit  region_length  \\\n",
       "gene_id                                                                      \n",
       "YNL262W       True    False    NaN      147712       155380           7669   \n",
       "YPL231W       True    False    NaN      108152       114815           6664   \n",
       "YMR207C      False     True  0.366      676693       684064           7372   \n",
       "YGR097W      False     True  0.561      678195       682635           4441   \n",
       "YFL033C      False    False    NaN       68615        74927           6313   \n",
       "...            ...      ...    ...         ...          ...            ...   \n",
       "YDL248W      False    False    NaN        1302         3453           2152   \n",
       "YBR181C      False     True  0.000      591212       593274           2063   \n",
       "YIR005W      False    False    NaN      364389       365835           1447   \n",
       "YMR256C      False    False    NaN      778445       779627           1183   \n",
       "YLR038C      False    False    NaN      224421       225672           1252   \n",
       "\n",
       "         num_of_vars  num_of_vars_common  num_of_vars_rare  var_density  \n",
       "gene_id                                                                  \n",
       "YNL262W           99                  24                15     1.290911  \n",
       "YPL231W           99                  21                21     1.485594  \n",
       "YMR207C           81                  19                22     1.098752  \n",
       "YGR097W           52                  19                12     1.170907  \n",
       "YFL033C           61                  18                12     0.966260  \n",
       "...              ...                 ...               ...          ...  \n",
       "YDL248W            0                   0                 0     0.000000  \n",
       "YBR181C            0                   0                 0     0.000000  \n",
       "YIR005W            2                   0                 0     0.138217  \n",
       "YMR256C            1                   0                 0     0.084531  \n",
       "YLR038C            4                   0                 0     0.319489  \n",
       "\n",
       "[800 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create a dataframe of hub genes with annotations and number of variants (common, rare)\n",
    "\n",
    "left_limit=500 # distance from left of gene to search\n",
    "right_limit=500 # distance from right of gene to search\n",
    "\n",
    "genes_df = interaction_rich_genes.copy()\n",
    "genes_df.loc[:,'chrom'] = genes_df['chrom'].str.replace('chr', '')\n",
    "genes_df.loc[:,'left_limit'] = genes_df['start'] - left_limit\n",
    "genes_df.loc[:,'right_limit'] = genes_df['end'] + right_limit\n",
    "genes_df.loc[:, 'region_length'] = genes_df['right_limit'] - genes_df['left_limit'] + 1\n",
    "\n",
    "# associate each variant with genes of interest if it falls within gene limits\n",
    "gxg_oligos_final.loc[:,'assoc_gene'] = gxg_oligos_final.apply(lambda x: genes_df.query('chrom==@x.CHROM & left_limit<=@x.POS & right_limit>=@x.POS').index.tolist(), axis=1)\n",
    "# remove variants without gene associations\n",
    "gxg_oligos_final = gxg_oligos_final[gxg_oligos_final.assoc_gene.apply(len)>0]\n",
    "# for coding variants with multiple associations (i.e. genes of interest adjacent to each other),\n",
    "# adjust to associate only gene whose coding region contains variant\n",
    "selection = (gxg_oligos_final.assoc_gene.apply(len)>1) & (~gxg_oligos_final.region.isin(intergenic_regions_list))\n",
    "gxg_oligos_final.loc[selection, 'assoc_gene'] = gxg_oligos_final.loc[selection, 'Gene_ID'].apply(lambda x: [x])\n",
    "\n",
    "# # finally, exclude remaining variants that are associated with more than one gene (should be only noncoding variants)\n",
    "# gxg_oligos_final = gxg_oligos_final[gxg_oligos_final['assoc_gene'].apply(lambda x: len(x)==1)]\n",
    "\n",
    "# count the number of variants associated with each gene\n",
    "af_cutoff_common=0.02\n",
    "af_cutoff_rare=0.002\n",
    "genes_df.loc[:, 'num_of_vars'] = [gxg_oligos_final['assoc_gene'].apply(lambda x: gene in x).sum() for gene in genes_df.index]\n",
    "genes_df.loc[:, 'num_of_vars_common'] = [gxg_oligos_final.query('AF>=@af_cutoff_common')['assoc_gene'].apply(lambda x: gene in x).sum() for gene in genes_df.index]\n",
    "genes_df.loc[:, 'num_of_vars_rare'] = [gxg_oligos_final.query('AF<=@af_cutoff_rare')['assoc_gene'].apply(lambda x: gene in x).sum() for gene in genes_df.index]\n",
    "\n",
    "# estimate density of variants (num of variants per 100bp) surveyed in each gene_region\n",
    "genes_df.loc[:, 'var_density'] = genes_df['num_of_vars'] / genes_df['region_length'] * 100\n",
    "\n",
    "# sort by number of common variants\n",
    "genes_df = genes_df.sort_values(['num_of_vars_common', 'num_of_vars_rare'], ascending=False)\n",
    "\n",
    "display(genes_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'YLR086W'} added to exclude list\n",
      "{'YOL005C'} added to exclude list\n",
      "{'YDR126W'} added to exclude list\n",
      "{'YCR094W'} added to exclude list\n",
      "{'YJR060W'} added to exclude list\n",
      "YJR060W is excluded\n",
      "{'YPR162C'} added to exclude list\n",
      "{'YLR086W'} added to exclude list\n",
      "{'YCR034W'} added to exclude list\n",
      "{'YBL008W'} added to exclude list\n",
      "{'YDR335W'} added to exclude list\n",
      "{'YFL037W', 'YFL039C'} added to exclude list\n",
      "{'YML127W'} added to exclude list\n",
      "{'YML071C'} added to exclude list\n",
      "YDR335W is excluded\n",
      "YML127W is excluded\n",
      "{'YLL036C'} added to exclude list\n",
      "{'YDL035C'} added to exclude list\n",
      "YDL035C is excluded\n",
      "{'YDL002C'} added to exclude list\n",
      "{'YAL025C'} added to exclude list\n",
      "YLL036C is excluded\n",
      "{'YPR182W'} added to exclude list\n",
      "YCR094W is excluded\n",
      "Number of hub genes selected: 103\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "      <th>assoc_gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXG_00321</th>\n",
       "      <td>I</td>\n",
       "      <td>101530</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>385|1833</td>\n",
       "      <td>1662|4742|35</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>MAK16</td>\n",
       "      <td>YAL025C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>385.0</td>\n",
       "      <td>LTE1</td>\n",
       "      <td>YAL024C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>35.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[YAL024C, YAL025C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00329</th>\n",
       "      <td>I</td>\n",
       "      <td>101816</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001980</td>\n",
       "      <td>671|2119</td>\n",
       "      <td>1948|4456</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[YAL024C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00332</th>\n",
       "      <td>I</td>\n",
       "      <td>101901</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>90</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.045000</td>\n",
       "      <td>756|2204</td>\n",
       "      <td>2033|4371</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[YAL024C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00339</th>\n",
       "      <td>I</td>\n",
       "      <td>102261</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>78</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.039000</td>\n",
       "      <td>1116|2564</td>\n",
       "      <td>2393|4011</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[YAL024C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_00346</th>\n",
       "      <td>I</td>\n",
       "      <td>102699</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>310</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.154000</td>\n",
       "      <td>1554|3002</td>\n",
       "      <td>2831|3573</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[YAL024C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50437</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899389</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>3625|805|1366|2655</td>\n",
       "      <td>2385</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[YPR181C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50440</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899419</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>196</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.097000</td>\n",
       "      <td>3655|775|1336|2625</td>\n",
       "      <td>2415</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[YPR181C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50446</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001487</td>\n",
       "      <td>3757|673|1234|2523</td>\n",
       "      <td>2517</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[YPR181C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50447</th>\n",
       "      <td>XVI</td>\n",
       "      <td>899679</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>68</td>\n",
       "      <td>2010</td>\n",
       "      <td>0.034000</td>\n",
       "      <td>3915|12|515|1076|2365</td>\n",
       "      <td>2675</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SEC23</td>\n",
       "      <td>YPR181C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>12.0</td>\n",
       "      <td>SMX3</td>\n",
       "      <td>YPR182W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>515.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[YPR181C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXG_50454</th>\n",
       "      <td>XVI</td>\n",
       "      <td>900065</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.001487</td>\n",
       "      <td>4301|398|129|690|1979</td>\n",
       "      <td>3061</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>SEC23</td>\n",
       "      <td>YPR181C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>398.0</td>\n",
       "      <td>SMX3</td>\n",
       "      <td>YPR182W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>129.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[YPR182W, YPR181C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT   AC    AN        AF  upstream_distance_str  \\\n",
       "var_id                                                                        \n",
       "GXG_00321     I  101530   G   C    4  2022  0.001978               385|1833   \n",
       "GXG_00329     I  101816   C   T    4  2020  0.001980               671|2119   \n",
       "GXG_00332     I  101901   C   A   90  2022  0.045000               756|2204   \n",
       "GXG_00339     I  102261   C   T   78  2020  0.039000              1116|2564   \n",
       "GXG_00346     I  102699   G   A  310  2016  0.154000              1554|3002   \n",
       "...         ...     ...  ..  ..  ...   ...       ...                    ...   \n",
       "GXG_50437   XVI  899389   G   A    6  2018  0.002973     3625|805|1366|2655   \n",
       "GXG_50440   XVI  899419   A   G  196  2014  0.097000     3655|775|1336|2625   \n",
       "GXG_50446   XVI  899521   G   C    3  2018  0.001487     3757|673|1234|2523   \n",
       "GXG_50447   XVI  899679   A   G   68  2010  0.034000  3915|12|515|1076|2365   \n",
       "GXG_50454   XVI  900065   C   G    3  2018  0.001487  4301|398|129|690|1979   \n",
       "\n",
       "          downstream_distance_str                   region  ...  \\\n",
       "var_id                                                      ...   \n",
       "GXG_00321            1662|4742|35  unidirectional_promoter  ...   \n",
       "GXG_00329               1948|4456         missense_variant  ...   \n",
       "GXG_00332               2033|4371       synonymous_variant  ...   \n",
       "GXG_00339               2393|4011       synonymous_variant  ...   \n",
       "GXG_00346               2831|3573       synonymous_variant  ...   \n",
       "...                           ...                      ...  ...   \n",
       "GXG_50437                    2385       synonymous_variant  ...   \n",
       "GXG_50440                    2415       synonymous_variant  ...   \n",
       "GXG_50446                    2517       synonymous_variant  ...   \n",
       "GXG_50447                    2675   bidirectional_promoter  ...   \n",
       "GXG_50454                    3061   bidirectional_promoter  ...   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXG_00321                   MAK16               YAL025C   \n",
       "GXG_00329                     NaN                   NaN   \n",
       "GXG_00332                     NaN                   NaN   \n",
       "GXG_00339                     NaN                   NaN   \n",
       "GXG_00346                     NaN                   NaN   \n",
       "...                           ...                   ...   \n",
       "GXG_50437                     NaN                   NaN   \n",
       "GXG_50440                     NaN                   NaN   \n",
       "GXG_50446                     NaN                   NaN   \n",
       "GXG_50447                   SEC23               YPR181C   \n",
       "GXG_50454                   SEC23               YPR181C   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "GXG_00321    upstream_gene_variant                  385.0   \n",
       "GXG_00329                      NaN                    NaN   \n",
       "GXG_00332                      NaN                    NaN   \n",
       "GXG_00339                      NaN                    NaN   \n",
       "GXG_00346                      NaN                    NaN   \n",
       "...                            ...                    ...   \n",
       "GXG_50437                      NaN                    NaN   \n",
       "GXG_50440                      NaN                    NaN   \n",
       "GXG_50446                      NaN                    NaN   \n",
       "GXG_50447    upstream_gene_variant                   12.0   \n",
       "GXG_50454    upstream_gene_variant                  398.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXG_00321                    LTE1               YAL024C   \n",
       "GXG_00329                     NaN                   NaN   \n",
       "GXG_00332                     NaN                   NaN   \n",
       "GXG_00339                     NaN                   NaN   \n",
       "GXG_00346                     NaN                   NaN   \n",
       "...                           ...                   ...   \n",
       "GXG_50437                     NaN                   NaN   \n",
       "GXG_50440                     NaN                   NaN   \n",
       "GXG_50446                     NaN                   NaN   \n",
       "GXG_50447                    SMX3               YPR182W   \n",
       "GXG_50454                    SMX3               YPR182W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \\\n",
       "var_id                                                                    \n",
       "GXG_00321  downstream_gene_variant                   35.0           3.0   \n",
       "GXG_00329                      NaN                    NaN           4.0   \n",
       "GXG_00332                      NaN                    NaN           2.0   \n",
       "GXG_00339                      NaN                    NaN           2.0   \n",
       "GXG_00346                      NaN                    NaN           3.0   \n",
       "...                            ...                    ...           ...   \n",
       "GXG_50437                      NaN                    NaN           2.0   \n",
       "GXG_50440                      NaN                    NaN           3.0   \n",
       "GXG_50446                      NaN                    NaN           2.0   \n",
       "GXG_50447    upstream_gene_variant                  515.0           4.0   \n",
       "GXG_50454    upstream_gene_variant                  129.0           3.0   \n",
       "\n",
       "                   assoc_gene  \n",
       "var_id                         \n",
       "GXG_00321  [YAL024C, YAL025C]  \n",
       "GXG_00329           [YAL024C]  \n",
       "GXG_00332           [YAL024C]  \n",
       "GXG_00339           [YAL024C]  \n",
       "GXG_00346           [YAL024C]  \n",
       "...                       ...  \n",
       "GXG_50437           [YPR181C]  \n",
       "GXG_50440           [YPR181C]  \n",
       "GXG_50446           [YPR181C]  \n",
       "GXG_50447           [YPR181C]  \n",
       "GXG_50454  [YPR182W, YPR181C]  \n",
       "\n",
       "[2000 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "synonymous_variant         0.4395\n",
       "missense_variant           0.3470\n",
       "unidirectional_promoter    0.1020\n",
       "bidirectional_promoter     0.0835\n",
       "intergenic                 0.0175\n",
       "inframe_insertion          0.0030\n",
       "intron_variant             0.0030\n",
       "inframe_deletion           0.0025\n",
       "stop_gained                0.0010\n",
       "frameshift_variant         0.0010\n",
       "Name: region, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of oligos from gxg gene set: 4957\n"
     ]
    }
   ],
   "source": [
    "# maximum number of oligos to design for: 5000\n",
    "# let's aim for 2000 variants (1000 common, 1000 rare), in as few genes as possible\n",
    "cutoff = 2000\n",
    "\n",
    "var_list = []\n",
    "gene_list = []\n",
    "exclude = []\n",
    "for gene in genes_df.index:\n",
    "    # skip blacklisted genes in exclusion list\n",
    "    if gene in exclude:\n",
    "        print(gene, 'is excluded')\n",
    "        continue\n",
    "        \n",
    "    df = gxg_oligos_final[gxg_oligos_final['assoc_gene'].apply(lambda x: gene in x)].sort_values(['AF', 'AC'], ascending=False)\n",
    "    com_var = df.query('AF>=@af_cutoff_common').index.tolist()\n",
    "    # check if number of common variants makes up >50% of all targetable variants in gene (unlikely, but possible)\n",
    "    # if so, reduce common variants selected to 50% of all targetable variants in this gene -- ensures balanced representation of common vs rare alleles\n",
    "    if len(com_var) > len(df)//2:\n",
    "        print(\"Common variants ({}) are >50% of all variants ({}). Capping common variants to {}\".format(len(com_var), len(df), len(df)//2))\n",
    "        com_var = com_var[:len(df)//2]        \n",
    "    rar_var = df.tail(len(com_var)).index.tolist()\n",
    "    com_rar_var = com_var + rar_var\n",
    "    \n",
    "    # check if selected variants are associated with genes other than the one being parsed\n",
    "    # if so, add those other genes to the exclude list to reduce overlapping variants\n",
    "    other_genes = set([g for l in df.loc[com_rar_var, 'assoc_gene'] for g in l if g != gene])\n",
    "    if len(other_genes)>0:\n",
    "        print(other_genes, \"added to exclude list\")\n",
    "        exclude += list(other_genes)\n",
    "    \n",
    "    # add selected variants to var_list\n",
    "    var_list += com_rar_var\n",
    "    # add gene to gene_list\n",
    "    gene_list.append(gene)\n",
    "    \n",
    "    if len(var_list)>= cutoff:\n",
    "        break\n",
    "\n",
    "# number of genes selected\n",
    "hub_genes = genes_df.loc[gene_list]\n",
    "print('Number of hub genes selected:', len(hub_genes))\n",
    "\n",
    "# select final list of variants that will be targeted\n",
    "gxg_oligos_final = gxg_oligos_final.query('var_id.isin(@var_list)')\n",
    "display(gxg_oligos_final)\n",
    "\n",
    "# distribution of variant annotation types\n",
    "display((gxg_oligos_final.region.value_counts()/len(gxg_oligos_final)))\n",
    "\n",
    "# how many oligos are we expecting from this list of variants?\n",
    "print(\"Number of oligos from gxg gene set:\", int(gxg_oligos_final.num_of_oligos.sum()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write gxg_oligos_final table to file\n",
    "gxg_oligos_final.to_csv(\"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/gxg_variants_final.txt\", sep='\\t')\n",
    "\n",
    "# use this file in separate script to organize into pools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive_interactions</th>\n",
       "      <th>negative_interactions</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>essential</th>\n",
       "      <th>paralog</th>\n",
       "      <th>KA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YAL024C</th>\n",
       "      <td>189</td>\n",
       "      <td>412</td>\n",
       "      <td>chrI</td>\n",
       "      <td>101565</td>\n",
       "      <td>105872</td>\n",
       "      <td>-</td>\n",
       "      <td>LTE1</td>\n",
       "      <td>Protein similar to GDP/GTP exchange factors bu...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         positive_interactions  negative_interactions chrom   start     end  \\\n",
       "gene_id                                                                       \n",
       "YAL024C                    189                    412  chrI  101565  105872   \n",
       "\n",
       "        strand gene_name                                        description  \\\n",
       "gene_id                                                                       \n",
       "YAL024C      -      LTE1  Protein similar to GDP/GTP exchange factors bu...   \n",
       "\n",
       "         essential  paralog  KA  \n",
       "gene_id                          \n",
       "YAL024C      False    False NaN  "
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interaction_rich_genes.query('gene_name==\"LTE1\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of GXG variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants: 2000\n",
      "Number of oligos: 4957\n",
      "\n",
      "Distribution of (common) variants among hub genes\n",
      "count    103.000000\n",
      "mean       9.708738\n",
      "std        3.285831\n",
      "min        7.000000\n",
      "25%        8.000000\n",
      "50%        9.000000\n",
      "75%       11.000000\n",
      "max       24.000000\n",
      "Name: num_of_vars_common, dtype: float64\n",
      "\n",
      "Distribution of oligos among GxG variants\n",
      "count    2000.000000\n",
      "mean        2.478500\n",
      "std         0.740143\n",
      "min         2.000000\n",
      "25%         2.000000\n",
      "50%         2.000000\n",
      "75%         3.000000\n",
      "max         7.000000\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Proportion of variant annotation classes\n",
      "synonymous_variant         0.4395\n",
      "missense_variant           0.3470\n",
      "unidirectional_promoter    0.1020\n",
      "bidirectional_promoter     0.0835\n",
      "intergenic                 0.0175\n",
      "inframe_insertion          0.0030\n",
      "intron_variant             0.0030\n",
      "inframe_deletion           0.0025\n",
      "stop_gained                0.0010\n",
      "frameshift_variant         0.0010\n",
      "Name: region, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "### SUMMARY OF GXG SET ###\n",
    "print(\"Number of variants:\", len(gxg_oligos_final))\n",
    "print(\"Number of oligos:\", int(sum(gxg_oligos_final.num_of_oligos)))\n",
    "print()\n",
    "print(\"Distribution of (common) variants among hub genes\")\n",
    "# print(gxg_oligos_final.Gene_ID.value_counts().filter(hub_genes.index).describe())\n",
    "print(hub_genes.num_of_vars_common.describe())\n",
    "print()\n",
    "print(\"Distribution of oligos among GxG variants\")\n",
    "print(gxg_oligos_final.num_of_oligos.describe())\n",
    "print()\n",
    "print(\"Proportion of variant annotation classes\")\n",
    "print(gxg_oligos_final.region.value_counts()/len(gxg_oligos_final))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine number of pools to use for GxG oligos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GXG    4957.0\n",
       "Name: num_of_oligos, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos to fit: 4957\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(42, 45)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use 42 pools:\n",
      "Number of oligos per pool: 118\n",
      "Number of oligos leftover: 1\n",
      "Number of technical oligos: 126\n",
      "\n",
      "Use 43 pools:\n",
      "Number of oligos per pool: 115\n",
      "Number of oligos leftover: 12\n",
      "Number of technical oligos: 258\n",
      "\n",
      "Use 44 pools:\n",
      "Number of oligos per pool: 112\n",
      "Number of oligos leftover: 29\n",
      "Number of technical oligos: 396\n",
      "\n",
      "Use 45 pools:\n",
      "Number of oligos per pool: 110\n",
      "Number of oligos leftover: 7\n",
      "Number of technical oligos: 495\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([42, 43, 44, 45])"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count OLIGOS in each variant prefix set\n",
    "display(gxg_oligos_final.groupby(lambda x: x[:3]).num_of_oligos.sum().sort_values(ascending=False))\n",
    "\n",
    "# determine number of pools to use\n",
    "num_of_oligos = int(gxg_oligos_final.num_of_oligos.sum())\n",
    "min_pool_size = 109 # allows 12 technical oligos\n",
    "max_pool_size = 118 # allows 3 technical oligos\n",
    "complete_pool_size = 121 # maximum oligos synthesized per pool\n",
    "\n",
    "# find number of pools to use\n",
    "find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant ID prefix and which genomes they can be edited in:\n",
    "- GXG: BY | RM | YJM | YPS\n",
    "\n",
    "### We should be able to fit GxG variants in 42 pools\n",
    "- GXG: 42 pools, 117-118 oligos each - 4,946 oligos\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selecting variants from previous CRISPEY screen with known effects\n",
    "These oligos showed effects in initial screen in 4 yeast strains, and are good candidates for demonstrating GxG fitness effects (e.g. GAT2 frameshift mutation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/epistasis/\"\n",
    "\n",
    "# path to oligo-variant info table (taken from CRISPEY library creation sheet)\n",
    "oligo_var_path = working_dir+\"library_oligo_to_variant_v2_withAnno.txt\"\n",
    "\n",
    "# path to epistais set variants VCF file\n",
    "epi_vcf_file = '/home/users/rang/tmp/crispr/forLibDesign_170918.vcf'\n",
    "# path to promoter set variants VCF file\n",
    "prom_vcf_file = '/home/users/rang/tmp/crispr/forLibDesign_promoter625.vcf'\n",
    "\n",
    "# path to GxG variants final set (check to avoid duplication)\n",
    "gxg_oligos_final_file = \"/home/users/rang/scratch/yeast/genetic_interactions/costanzo_2016/gxg_variants_final.txt\"\n",
    "\n",
    "os.chdir(working_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>SNP_chr_pos</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNPEff_Annotation</th>\n",
       "      <th>SNPEff_Annotation_Impact</th>\n",
       "      <th>SNPEff_Gene_Name</th>\n",
       "      <th>SNPEff_Gene_ID</th>\n",
       "      <th>SNPEff_region</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RAR_13-C1</th>\n",
       "      <td>I</td>\n",
       "      <td>42183</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAR_13</th>\n",
       "      <td>I</td>\n",
       "      <td>42184</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>PEX22</td>\n",
       "      <td>YAL055W</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAR_21-C1</th>\n",
       "      <td>I</td>\n",
       "      <td>50078</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>OAF1</td>\n",
       "      <td>YAL051W</td>\n",
       "      <td>synonymous_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAR_21</th>\n",
       "      <td>I</td>\n",
       "      <td>50081</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>OAF1</td>\n",
       "      <td>YAL051W</td>\n",
       "      <td>synonymous_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COM_8</th>\n",
       "      <td>I</td>\n",
       "      <td>57455</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAL045C</td>\n",
       "      <td>unidirectional promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROM_1871</th>\n",
       "      <td>XII</td>\n",
       "      <td>819307</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>RPL26A</td>\n",
       "      <td>YLR344W</td>\n",
       "      <td>unidirectional promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROM_1872</th>\n",
       "      <td>XII</td>\n",
       "      <td>819308</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>RPL26A</td>\n",
       "      <td>YLR344W</td>\n",
       "      <td>unidirectional promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROM_1873</th>\n",
       "      <td>XII</td>\n",
       "      <td>819309</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>RPL26A</td>\n",
       "      <td>YLR344W</td>\n",
       "      <td>unidirectional promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROM_1874</th>\n",
       "      <td>XII</td>\n",
       "      <td>819310</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>RPL26A</td>\n",
       "      <td>YLR344W</td>\n",
       "      <td>unidirectional promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROM_1875</th>\n",
       "      <td>XII</td>\n",
       "      <td>819311</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>RPL26A</td>\n",
       "      <td>YLR344W</td>\n",
       "      <td>unidirectional promoter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13699 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom  SNP_chr_pos REF  ALT        SNPEff_Annotation  \\\n",
       "var_id                                                           \n",
       "RAR_13-C1     I        42183   C  [T]         missense_variant   \n",
       "RAR_13        I        42184   C  [T]         missense_variant   \n",
       "RAR_21-C1     I        50078   C  [T]       synonymous_variant   \n",
       "RAR_21        I        50081   C  [T]       synonymous_variant   \n",
       "COM_8         I        57455   C  [T]  downstream_gene_variant   \n",
       "...         ...          ...  ..  ...                      ...   \n",
       "PROM_1871   XII       819307   T  [C]    upstream_gene_variant   \n",
       "PROM_1872   XII       819308   C  [T]    upstream_gene_variant   \n",
       "PROM_1873   XII       819309   A  [G]    upstream_gene_variant   \n",
       "PROM_1874   XII       819310   G  [A]    upstream_gene_variant   \n",
       "PROM_1875   XII       819311   A  [G]    upstream_gene_variant   \n",
       "\n",
       "          SNPEff_Annotation_Impact SNPEff_Gene_Name SNPEff_Gene_ID  \\\n",
       "var_id                                                               \n",
       "RAR_13-C1                 MODERATE            PEX22        YAL055W   \n",
       "RAR_13                    MODERATE            PEX22        YAL055W   \n",
       "RAR_21-C1                      LOW             OAF1        YAL051W   \n",
       "RAR_21                         LOW             OAF1        YAL051W   \n",
       "COM_8                     MODIFIER              NaN        YAL045C   \n",
       "...                            ...              ...            ...   \n",
       "PROM_1871                 MODIFIER           RPL26A        YLR344W   \n",
       "PROM_1872                 MODIFIER           RPL26A        YLR344W   \n",
       "PROM_1873                 MODIFIER           RPL26A        YLR344W   \n",
       "PROM_1874                 MODIFIER           RPL26A        YLR344W   \n",
       "PROM_1875                 MODIFIER           RPL26A        YLR344W   \n",
       "\n",
       "                     SNPEff_region  \n",
       "var_id                              \n",
       "RAR_13-C1         missense_variant  \n",
       "RAR_13            missense_variant  \n",
       "RAR_21-C1       synonymous_variant  \n",
       "RAR_21          synonymous_variant  \n",
       "COM_8      unidirectional promoter  \n",
       "...                            ...  \n",
       "PROM_1871  unidirectional promoter  \n",
       "PROM_1872  unidirectional promoter  \n",
       "PROM_1873  unidirectional promoter  \n",
       "PROM_1874  unidirectional promoter  \n",
       "PROM_1875  unidirectional promoter  \n",
       "\n",
       "[13699 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load oligo-variant info table\n",
    "oligo_info_df = pd.read_csv(oligo_var_path, sep='\\t', header = 0, index_col=0, low_memory=False)\n",
    "oligo_info_df = oligo_info_df.reset_index().rename(columns={'oligo_id':'oligo_num'})\n",
    "oligo_info_df.index = oligo_info_df.oligo_num.astype('str')+\"#\"+oligo_info_df.set_name\n",
    "oligo_info_df.index.name = 'oligo_id'\n",
    "var_info_df = oligo_info_df.loc[:,['SNP_id', 'chrom', 'SNP_chr_pos', 'REF', 'ALT', 'SNPEff_Annotation', 'SNPEff_Annotation_Impact', 'SNPEff_Gene_Name', 'SNPEff_Gene_ID', 'SNPEff_region']].copy().drop_duplicates().set_index('SNP_id') #.reset_index(drop=True)#\n",
    "var_info_df.index.name = 'var_id'\n",
    "display(var_info_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select TDH3 promoter oligos targeting variants with existing data\n",
    "These variants have data published by the Wittkopp lab and are useful for calibrating CRISPEY fitness readouts\n",
    "- fitness effects\n",
    "- 3 TFBS locations\n",
    "- expression data\n",
    "\n",
    "Ensure each variant is targeted by 2 or more oligos whose edit position in guide is between -9 and +2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>oligo_num</th>\n",
       "      <th>oligo_seq</th>\n",
       "      <th>set_name</th>\n",
       "      <th>set_priority</th>\n",
       "      <th>chrom</th>\n",
       "      <th>SNP_chr_pos</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP_id</th>\n",
       "      <th>guide_chr_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>SNPEff_CDS_pos</th>\n",
       "      <th>SNPEff_AA_pos</th>\n",
       "      <th>VEP_Amino_acids</th>\n",
       "      <th>VEP_Codons</th>\n",
       "      <th>VEP_HGVSc</th>\n",
       "      <th>VEP_HGVSp</th>\n",
       "      <th>VEP_SWISSPROT</th>\n",
       "      <th>VEP_UNIPARC</th>\n",
       "      <th>VEP_DOMAINS</th>\n",
       "      <th>VEP_BLOSUM62</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oligo_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28474#promoterMut</th>\n",
       "      <td>28474</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCAGAGAGAGCAGAAGGTAATAGAAGG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884068</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_883</td>\n",
       "      <td>884073</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28475#promoterMut</th>\n",
       "      <td>28475</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAGAGAGAGCAGAAGGTAATAGAAGGT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884068</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_883</td>\n",
       "      <td>884074</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28484#promoterMut</th>\n",
       "      <td>28484</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCAGAGAGAGCAGAAGGTAATAGAAGG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884073</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_888</td>\n",
       "      <td>884073</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28485#promoterMut</th>\n",
       "      <td>28485</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAGAGAGAGCAGAAGGTAATAGAAGGT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884073</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_888</td>\n",
       "      <td>884074</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28539#promoterMut</th>\n",
       "      <td>28539</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAAAATGAGATAGATACATGCGTGGGT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884105</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_920</td>\n",
       "      <td>884105</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28540#promoterMut</th>\n",
       "      <td>28540</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTGAGATAGATACATGCGTGGGTCAAT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884105</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_920</td>\n",
       "      <td>884109</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28541#promoterMut</th>\n",
       "      <td>28541</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGGCACAAACAGGCAAAAAACGGGCAC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884105</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_920</td>\n",
       "      <td>884105</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28544#promoterMut</th>\n",
       "      <td>28544</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAAAATGAGATAGATACATGCGTGGGT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884107</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_922</td>\n",
       "      <td>884105</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28545#promoterMut</th>\n",
       "      <td>28545</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTGAGATAGATACATGCGTGGGTCAAT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884107</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_922</td>\n",
       "      <td>884109</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28546#promoterMut</th>\n",
       "      <td>28546</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGGCACAAACAGGCAAAAAACGGGCAC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884107</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_922</td>\n",
       "      <td>884105</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28547#promoterMut</th>\n",
       "      <td>28547</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAAAATGAGATAGATACATGCGTGGGT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884108</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_923</td>\n",
       "      <td>884105</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28548#promoterMut</th>\n",
       "      <td>28548</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTGAGATAGATACATGCGTGGGTCAAT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884108</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_923</td>\n",
       "      <td>884109</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28549#promoterMut</th>\n",
       "      <td>28549</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGGCACAAACAGGCAAAAAACGGGCAC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884108</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_923</td>\n",
       "      <td>884105</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28578#promoterMut</th>\n",
       "      <td>28578</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGGTCAATTGCCTTGTGTCATCATTTA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884123</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_938</td>\n",
       "      <td>884128</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28579#promoterMut</th>\n",
       "      <td>28579</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCGCAACTACAGAGAACAGGGGCACAA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884123</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_938</td>\n",
       "      <td>884124</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28589#promoterMut</th>\n",
       "      <td>28589</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGGTCAATTGCCTTGTGTCATCATTTA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884128</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_943</td>\n",
       "      <td>884128</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28590#promoterMut</th>\n",
       "      <td>28590</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCGCAACTACAGAGAACAGGGGCACAA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884128</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_943</td>\n",
       "      <td>884124</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28656#promoterMut</th>\n",
       "      <td>28656</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTCACCAACCATCAGTTCATAGGTCCA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884158</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_973</td>\n",
       "      <td>884159</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28657#promoterMut</th>\n",
       "      <td>28657</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTCAGTTCATAGGTCCATTCTCTTAGC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884158</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_973</td>\n",
       "      <td>884149</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28662#promoterMut</th>\n",
       "      <td>28662</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCTTCACCAACCATCAGTTCATAGGTC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884160</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_975</td>\n",
       "      <td>884161</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28663#promoterMut</th>\n",
       "      <td>28663</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTTCACCAACCATCAGTTCATAGGTCC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884160</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_975</td>\n",
       "      <td>884160</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28776#promoterMut</th>\n",
       "      <td>28776</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGTTTGTGCCCCTGTTCTCTGTAGTTG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884200</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1015</td>\n",
       "      <td>884203</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28777#promoterMut</th>\n",
       "      <td>28777</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGTGCCCCTGTTCTCTGTAGTTGCGCT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884200</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1015</td>\n",
       "      <td>884207</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28778#promoterMut</th>\n",
       "      <td>28778</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAAAAAAGAATCCCAGCACCAAAATAT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884200</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1015</td>\n",
       "      <td>884193</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28974#promoterMut</th>\n",
       "      <td>28974</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGATTCTTTTTTTTTCTGGATGCCAGC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884289</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_1104</td>\n",
       "      <td>884289</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28975#promoterMut</th>\n",
       "      <td>28975</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884289</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_1104</td>\n",
       "      <td>884297</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28979#promoterMut</th>\n",
       "      <td>28979</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGATTCTTTTTTTTTCTGGATGCCAGC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884291</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_1106</td>\n",
       "      <td>884289</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28980#promoterMut</th>\n",
       "      <td>28980</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884291</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_1106</td>\n",
       "      <td>884297</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28982#promoterMut</th>\n",
       "      <td>28982</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGATTCTTTTTTTTTCTGGATGCCAGC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884292</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_1107</td>\n",
       "      <td>884289</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28983#promoterMut</th>\n",
       "      <td>28983</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884292</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_1107</td>\n",
       "      <td>884297</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28991#promoterMut</th>\n",
       "      <td>28991</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884296</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1111</td>\n",
       "      <td>884297</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28992#promoterMut</th>\n",
       "      <td>28992</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCGGGTTACACAGAATATATAACATCG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884296</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1111</td>\n",
       "      <td>884297</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28994#promoterMut</th>\n",
       "      <td>28994</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884297</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1112</td>\n",
       "      <td>884297</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28995#promoterMut</th>\n",
       "      <td>28995</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCGGGTTACACAGAATATATAACATCG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884297</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1112</td>\n",
       "      <td>884297</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29016#promoterMut</th>\n",
       "      <td>29016</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTACATGCCCAAAATAGGGGGCGGGTTA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884315</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1130</td>\n",
       "      <td>884316</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29017#promoterMut</th>\n",
       "      <td>29017</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCATGCCCAAAATAGGGGGCGGGTTAC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884315</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1130</td>\n",
       "      <td>884315</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29022#promoterMut</th>\n",
       "      <td>29022</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTACATGCCCAAAATAGGGGGCGGGTTA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884318</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_1133</td>\n",
       "      <td>884316</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29023#promoterMut</th>\n",
       "      <td>29023</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCATGCCCAAAATAGGGGGCGGGTTAC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884318</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_1133</td>\n",
       "      <td>884315</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29024#promoterMut</th>\n",
       "      <td>29024</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTACATGCCCAAAATAGGGGGCGGGTTA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884319</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>PROM_1134</td>\n",
       "      <td>884316</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29025#promoterMut</th>\n",
       "      <td>29025</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCATGCCCAAAATAGGGGGCGGGTTAC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884319</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>PROM_1134</td>\n",
       "      <td>884315</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29026#promoterMut</th>\n",
       "      <td>29026</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTACATGCCCAAAATAGGGGGCGGGTTA...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884320</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1135</td>\n",
       "      <td>884316</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29027#promoterMut</th>\n",
       "      <td>29027</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCATGCCCAAAATAGGGGGCGGGTTAC...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884320</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_1135</td>\n",
       "      <td>884315</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P16451</td>\n",
       "      <td>UPI0000052FB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   oligo_num  \\\n",
       "oligo_id                       \n",
       "28474#promoterMut      28474   \n",
       "28475#promoterMut      28475   \n",
       "28484#promoterMut      28484   \n",
       "28485#promoterMut      28485   \n",
       "28539#promoterMut      28539   \n",
       "28540#promoterMut      28540   \n",
       "28541#promoterMut      28541   \n",
       "28544#promoterMut      28544   \n",
       "28545#promoterMut      28545   \n",
       "28546#promoterMut      28546   \n",
       "28547#promoterMut      28547   \n",
       "28548#promoterMut      28548   \n",
       "28549#promoterMut      28549   \n",
       "28578#promoterMut      28578   \n",
       "28579#promoterMut      28579   \n",
       "28589#promoterMut      28589   \n",
       "28590#promoterMut      28590   \n",
       "28656#promoterMut      28656   \n",
       "28657#promoterMut      28657   \n",
       "28662#promoterMut      28662   \n",
       "28663#promoterMut      28663   \n",
       "28776#promoterMut      28776   \n",
       "28777#promoterMut      28777   \n",
       "28778#promoterMut      28778   \n",
       "28974#promoterMut      28974   \n",
       "28975#promoterMut      28975   \n",
       "28979#promoterMut      28979   \n",
       "28980#promoterMut      28980   \n",
       "28982#promoterMut      28982   \n",
       "28983#promoterMut      28983   \n",
       "28991#promoterMut      28991   \n",
       "28992#promoterMut      28992   \n",
       "28994#promoterMut      28994   \n",
       "28995#promoterMut      28995   \n",
       "29016#promoterMut      29016   \n",
       "29017#promoterMut      29017   \n",
       "29022#promoterMut      29022   \n",
       "29023#promoterMut      29023   \n",
       "29024#promoterMut      29024   \n",
       "29025#promoterMut      29025   \n",
       "29026#promoterMut      29026   \n",
       "29027#promoterMut      29027   \n",
       "\n",
       "                                                           oligo_seq  \\\n",
       "oligo_id                                                               \n",
       "28474#promoterMut  GAGTTACTGTCTGTTTTCCTCAGAGAGAGCAGAAGGTAATAGAAGG...   \n",
       "28475#promoterMut  GAGTTACTGTCTGTTTTCCTAGAGAGAGCAGAAGGTAATAGAAGGT...   \n",
       "28484#promoterMut  GAGTTACTGTCTGTTTTCCTCAGAGAGAGCAGAAGGTAATAGAAGG...   \n",
       "28485#promoterMut  GAGTTACTGTCTGTTTTCCTAGAGAGAGCAGAAGGTAATAGAAGGT...   \n",
       "28539#promoterMut  GAGTTACTGTCTGTTTTCCTAAAATGAGATAGATACATGCGTGGGT...   \n",
       "28540#promoterMut  GAGTTACTGTCTGTTTTCCTTGAGATAGATACATGCGTGGGTCAAT...   \n",
       "28541#promoterMut  GAGTTACTGTCTGTTTTCCTGGCACAAACAGGCAAAAAACGGGCAC...   \n",
       "28544#promoterMut  GAGTTACTGTCTGTTTTCCTAAAATGAGATAGATACATGCGTGGGT...   \n",
       "28545#promoterMut  GAGTTACTGTCTGTTTTCCTTGAGATAGATACATGCGTGGGTCAAT...   \n",
       "28546#promoterMut  GAGTTACTGTCTGTTTTCCTGGCACAAACAGGCAAAAAACGGGCAC...   \n",
       "28547#promoterMut  GAGTTACTGTCTGTTTTCCTAAAATGAGATAGATACATGCGTGGGT...   \n",
       "28548#promoterMut  GAGTTACTGTCTGTTTTCCTTGAGATAGATACATGCGTGGGTCAAT...   \n",
       "28549#promoterMut  GAGTTACTGTCTGTTTTCCTGGCACAAACAGGCAAAAAACGGGCAC...   \n",
       "28578#promoterMut  GAGTTACTGTCTGTTTTCCTGGTCAATTGCCTTGTGTCATCATTTA...   \n",
       "28579#promoterMut  GAGTTACTGTCTGTTTTCCTCGCAACTACAGAGAACAGGGGCACAA...   \n",
       "28589#promoterMut  GAGTTACTGTCTGTTTTCCTGGTCAATTGCCTTGTGTCATCATTTA...   \n",
       "28590#promoterMut  GAGTTACTGTCTGTTTTCCTCGCAACTACAGAGAACAGGGGCACAA...   \n",
       "28656#promoterMut  GAGTTACTGTCTGTTTTCCTTCACCAACCATCAGTTCATAGGTCCA...   \n",
       "28657#promoterMut  GAGTTACTGTCTGTTTTCCTTCAGTTCATAGGTCCATTCTCTTAGC...   \n",
       "28662#promoterMut  GAGTTACTGTCTGTTTTCCTCTTCACCAACCATCAGTTCATAGGTC...   \n",
       "28663#promoterMut  GAGTTACTGTCTGTTTTCCTTTCACCAACCATCAGTTCATAGGTCC...   \n",
       "28776#promoterMut  GAGTTACTGTCTGTTTTCCTGTTTGTGCCCCTGTTCTCTGTAGTTG...   \n",
       "28777#promoterMut  GAGTTACTGTCTGTTTTCCTGTGCCCCTGTTCTCTGTAGTTGCGCT...   \n",
       "28778#promoterMut  GAGTTACTGTCTGTTTTCCTAAAAAAGAATCCCAGCACCAAAATAT...   \n",
       "28974#promoterMut  GAGTTACTGTCTGTTTTCCTGATTCTTTTTTTTTCTGGATGCCAGC...   \n",
       "28975#promoterMut  GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...   \n",
       "28979#promoterMut  GAGTTACTGTCTGTTTTCCTGATTCTTTTTTTTTCTGGATGCCAGC...   \n",
       "28980#promoterMut  GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...   \n",
       "28982#promoterMut  GAGTTACTGTCTGTTTTCCTGATTCTTTTTTTTTCTGGATGCCAGC...   \n",
       "28983#promoterMut  GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...   \n",
       "28991#promoterMut  GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...   \n",
       "28992#promoterMut  GAGTTACTGTCTGTTTTCCTCGGGTTACACAGAATATATAACATCG...   \n",
       "28994#promoterMut  GAGTTACTGTCTGTTTTCCTTTTTTTCTGGATGCCAGCTTAAAAAG...   \n",
       "28995#promoterMut  GAGTTACTGTCTGTTTTCCTCGGGTTACACAGAATATATAACATCG...   \n",
       "29016#promoterMut  GAGTTACTGTCTGTTTTCCTACATGCCCAAAATAGGGGGCGGGTTA...   \n",
       "29017#promoterMut  GAGTTACTGTCTGTTTTCCTCATGCCCAAAATAGGGGGCGGGTTAC...   \n",
       "29022#promoterMut  GAGTTACTGTCTGTTTTCCTACATGCCCAAAATAGGGGGCGGGTTA...   \n",
       "29023#promoterMut  GAGTTACTGTCTGTTTTCCTCATGCCCAAAATAGGGGGCGGGTTAC...   \n",
       "29024#promoterMut  GAGTTACTGTCTGTTTTCCTACATGCCCAAAATAGGGGGCGGGTTA...   \n",
       "29025#promoterMut  GAGTTACTGTCTGTTTTCCTCATGCCCAAAATAGGGGGCGGGTTAC...   \n",
       "29026#promoterMut  GAGTTACTGTCTGTTTTCCTACATGCCCAAAATAGGGGGCGGGTTA...   \n",
       "29027#promoterMut  GAGTTACTGTCTGTTTTCCTCATGCCCAAAATAGGGGGCGGGTTAC...   \n",
       "\n",
       "                      set_name  set_priority chrom  SNP_chr_pos REF  ALT  \\\n",
       "oligo_id                                                                   \n",
       "28474#promoterMut  promoterMut             5   VII       884068   C  [T]   \n",
       "28475#promoterMut  promoterMut             5   VII       884068   C  [T]   \n",
       "28484#promoterMut  promoterMut             5   VII       884073   G  [A]   \n",
       "28485#promoterMut  promoterMut             5   VII       884073   G  [A]   \n",
       "28539#promoterMut  promoterMut             5   VII       884105   C  [T]   \n",
       "28540#promoterMut  promoterMut             5   VII       884105   C  [T]   \n",
       "28541#promoterMut  promoterMut             5   VII       884105   C  [T]   \n",
       "28544#promoterMut  promoterMut             5   VII       884107   G  [A]   \n",
       "28545#promoterMut  promoterMut             5   VII       884107   G  [A]   \n",
       "28546#promoterMut  promoterMut             5   VII       884107   G  [A]   \n",
       "28547#promoterMut  promoterMut             5   VII       884108   G  [A]   \n",
       "28548#promoterMut  promoterMut             5   VII       884108   G  [A]   \n",
       "28549#promoterMut  promoterMut             5   VII       884108   G  [A]   \n",
       "28578#promoterMut  promoterMut             5   VII       884123   C  [T]   \n",
       "28579#promoterMut  promoterMut             5   VII       884123   C  [T]   \n",
       "28589#promoterMut  promoterMut             5   VII       884128   G  [A]   \n",
       "28590#promoterMut  promoterMut             5   VII       884128   G  [A]   \n",
       "28656#promoterMut  promoterMut             5   VII       884158   C  [T]   \n",
       "28657#promoterMut  promoterMut             5   VII       884158   C  [T]   \n",
       "28662#promoterMut  promoterMut             5   VII       884160   C  [T]   \n",
       "28663#promoterMut  promoterMut             5   VII       884160   C  [T]   \n",
       "28776#promoterMut  promoterMut             5   VII       884200   C  [T]   \n",
       "28777#promoterMut  promoterMut             5   VII       884200   C  [T]   \n",
       "28778#promoterMut  promoterMut             5   VII       884200   C  [T]   \n",
       "28974#promoterMut  promoterMut             5   VII       884289   G  [A]   \n",
       "28975#promoterMut  promoterMut             5   VII       884289   G  [A]   \n",
       "28979#promoterMut  promoterMut             5   VII       884291   G  [A]   \n",
       "28980#promoterMut  promoterMut             5   VII       884291   G  [A]   \n",
       "28982#promoterMut  promoterMut             5   VII       884292   G  [A]   \n",
       "28983#promoterMut  promoterMut             5   VII       884292   G  [A]   \n",
       "28991#promoterMut  promoterMut             5   VII       884296   C  [T]   \n",
       "28992#promoterMut  promoterMut             5   VII       884296   C  [T]   \n",
       "28994#promoterMut  promoterMut             5   VII       884297   C  [T]   \n",
       "28995#promoterMut  promoterMut             5   VII       884297   C  [T]   \n",
       "29016#promoterMut  promoterMut             5   VII       884315   C  [T]   \n",
       "29017#promoterMut  promoterMut             5   VII       884315   C  [T]   \n",
       "29022#promoterMut  promoterMut             5   VII       884318   G  [A]   \n",
       "29023#promoterMut  promoterMut             5   VII       884318   G  [A]   \n",
       "29024#promoterMut  promoterMut             5   VII       884319   A  [G]   \n",
       "29025#promoterMut  promoterMut             5   VII       884319   A  [G]   \n",
       "29026#promoterMut  promoterMut             5   VII       884320   C  [T]   \n",
       "29027#promoterMut  promoterMut             5   VII       884320   C  [T]   \n",
       "\n",
       "                      SNP_id  guide_chr_pos  ... SNPEff_CDS_pos  \\\n",
       "oligo_id                                     ...                  \n",
       "28474#promoterMut   PROM_883         884073  ...            NaN   \n",
       "28475#promoterMut   PROM_883         884074  ...            NaN   \n",
       "28484#promoterMut   PROM_888         884073  ...            NaN   \n",
       "28485#promoterMut   PROM_888         884074  ...            NaN   \n",
       "28539#promoterMut   PROM_920         884105  ...            NaN   \n",
       "28540#promoterMut   PROM_920         884109  ...            NaN   \n",
       "28541#promoterMut   PROM_920         884105  ...            NaN   \n",
       "28544#promoterMut   PROM_922         884105  ...            NaN   \n",
       "28545#promoterMut   PROM_922         884109  ...            NaN   \n",
       "28546#promoterMut   PROM_922         884105  ...            NaN   \n",
       "28547#promoterMut   PROM_923         884105  ...            NaN   \n",
       "28548#promoterMut   PROM_923         884109  ...            NaN   \n",
       "28549#promoterMut   PROM_923         884105  ...            NaN   \n",
       "28578#promoterMut   PROM_938         884128  ...            NaN   \n",
       "28579#promoterMut   PROM_938         884124  ...            NaN   \n",
       "28589#promoterMut   PROM_943         884128  ...            NaN   \n",
       "28590#promoterMut   PROM_943         884124  ...            NaN   \n",
       "28656#promoterMut   PROM_973         884159  ...            NaN   \n",
       "28657#promoterMut   PROM_973         884149  ...            NaN   \n",
       "28662#promoterMut   PROM_975         884161  ...            NaN   \n",
       "28663#promoterMut   PROM_975         884160  ...            NaN   \n",
       "28776#promoterMut  PROM_1015         884203  ...            NaN   \n",
       "28777#promoterMut  PROM_1015         884207  ...            NaN   \n",
       "28778#promoterMut  PROM_1015         884193  ...            NaN   \n",
       "28974#promoterMut  PROM_1104         884289  ...            NaN   \n",
       "28975#promoterMut  PROM_1104         884297  ...            NaN   \n",
       "28979#promoterMut  PROM_1106         884289  ...            NaN   \n",
       "28980#promoterMut  PROM_1106         884297  ...            NaN   \n",
       "28982#promoterMut  PROM_1107         884289  ...            NaN   \n",
       "28983#promoterMut  PROM_1107         884297  ...            NaN   \n",
       "28991#promoterMut  PROM_1111         884297  ...            NaN   \n",
       "28992#promoterMut  PROM_1111         884297  ...            NaN   \n",
       "28994#promoterMut  PROM_1112         884297  ...            NaN   \n",
       "28995#promoterMut  PROM_1112         884297  ...            NaN   \n",
       "29016#promoterMut  PROM_1130         884316  ...            NaN   \n",
       "29017#promoterMut  PROM_1130         884315  ...            NaN   \n",
       "29022#promoterMut  PROM_1133         884316  ...            NaN   \n",
       "29023#promoterMut  PROM_1133         884315  ...            NaN   \n",
       "29024#promoterMut  PROM_1134         884316  ...            NaN   \n",
       "29025#promoterMut  PROM_1134         884315  ...            NaN   \n",
       "29026#promoterMut  PROM_1135         884316  ...            NaN   \n",
       "29027#promoterMut  PROM_1135         884315  ...            NaN   \n",
       "\n",
       "                   SNPEff_AA_pos VEP_Amino_acids VEP_Codons VEP_HGVSc  \\\n",
       "oligo_id                                                                \n",
       "28474#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28475#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28484#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28485#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28539#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28540#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28541#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28544#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28545#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28546#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28547#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28548#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28549#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28578#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28579#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28589#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28590#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28656#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28657#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28662#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28663#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28776#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28777#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28778#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28974#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28975#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28979#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28980#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28982#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28983#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28991#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28992#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28994#promoterMut            NaN             NaN        NaN       NaN   \n",
       "28995#promoterMut            NaN             NaN        NaN       NaN   \n",
       "29016#promoterMut            NaN             NaN        NaN       NaN   \n",
       "29017#promoterMut            NaN             NaN        NaN       NaN   \n",
       "29022#promoterMut            NaN             NaN        NaN       NaN   \n",
       "29023#promoterMut            NaN             NaN        NaN       NaN   \n",
       "29024#promoterMut            NaN             NaN        NaN       NaN   \n",
       "29025#promoterMut            NaN             NaN        NaN       NaN   \n",
       "29026#promoterMut            NaN             NaN        NaN       NaN   \n",
       "29027#promoterMut            NaN             NaN        NaN       NaN   \n",
       "\n",
       "                  VEP_HGVSp VEP_SWISSPROT    VEP_UNIPARC VEP_DOMAINS  \\\n",
       "oligo_id                                                               \n",
       "28474#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28475#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28484#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28485#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28539#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28540#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28541#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28544#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28545#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28546#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28547#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28548#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28549#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28578#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28579#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28589#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28590#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28656#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28657#promoterMut       NaN        P00359  UPI000004F91A         NaN   \n",
       "28662#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28663#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28776#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28777#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28778#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28974#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28975#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28979#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28980#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28982#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28983#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28991#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28992#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28994#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "28995#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "29016#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "29017#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "29022#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "29023#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "29024#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "29025#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "29026#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "29027#promoterMut       NaN        P16451  UPI0000052FB2         NaN   \n",
       "\n",
       "                  VEP_BLOSUM62  \n",
       "oligo_id                        \n",
       "28474#promoterMut          NaN  \n",
       "28475#promoterMut          NaN  \n",
       "28484#promoterMut          NaN  \n",
       "28485#promoterMut          NaN  \n",
       "28539#promoterMut          NaN  \n",
       "28540#promoterMut          NaN  \n",
       "28541#promoterMut          NaN  \n",
       "28544#promoterMut          NaN  \n",
       "28545#promoterMut          NaN  \n",
       "28546#promoterMut          NaN  \n",
       "28547#promoterMut          NaN  \n",
       "28548#promoterMut          NaN  \n",
       "28549#promoterMut          NaN  \n",
       "28578#promoterMut          NaN  \n",
       "28579#promoterMut          NaN  \n",
       "28589#promoterMut          NaN  \n",
       "28590#promoterMut          NaN  \n",
       "28656#promoterMut          NaN  \n",
       "28657#promoterMut          NaN  \n",
       "28662#promoterMut          NaN  \n",
       "28663#promoterMut          NaN  \n",
       "28776#promoterMut          NaN  \n",
       "28777#promoterMut          NaN  \n",
       "28778#promoterMut          NaN  \n",
       "28974#promoterMut          NaN  \n",
       "28975#promoterMut          NaN  \n",
       "28979#promoterMut          NaN  \n",
       "28980#promoterMut          NaN  \n",
       "28982#promoterMut          NaN  \n",
       "28983#promoterMut          NaN  \n",
       "28991#promoterMut          NaN  \n",
       "28992#promoterMut          NaN  \n",
       "28994#promoterMut          NaN  \n",
       "28995#promoterMut          NaN  \n",
       "29016#promoterMut          NaN  \n",
       "29017#promoterMut          NaN  \n",
       "29022#promoterMut          NaN  \n",
       "29023#promoterMut          NaN  \n",
       "29024#promoterMut          NaN  \n",
       "29025#promoterMut          NaN  \n",
       "29026#promoterMut          NaN  \n",
       "29027#promoterMut          NaN  \n",
       "\n",
       "[42 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(42, 42)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load Wittkopp lab expression and fitness data on TDH3 promoter variants\n",
    "# Wittkopp fitness data\n",
    "wittkopp_fitness_df = pd.read_csv('tdh3_fitness_wittkopp.txt', sep='\\t', index_col=0)\n",
    "\n",
    "# Wittkopp expression data\n",
    "wittkopp_expr_df = pd.read_csv('tdh3_expression_wittkopp.txt', sep='\\t')\n",
    "# keep variants that reduced expression by >2%\n",
    "wittkopp_expr_df = wittkopp_expr_df[wittkopp_expr_df['YFP.MEAN.RELATIVE']<0.98].iloc[:,[6,3]]\n",
    "\n",
    "\n",
    "# get TDH3 oligos from previous screen\n",
    "promoter_name = 'TDH3'\n",
    "promoter_chrom = 'VII'\n",
    "promoter_cds_start = 883810\n",
    "promoter_strand = 1 # positive strand\n",
    "tdh3_oligo_info = oligo_info_df.loc[(oligo_info_df.SNP_id.str.contains('PROM_')) & (oligo_info_df.chrom==promoter_chrom)].copy() #['SNP_id', 'chrom', 'SNP_chr_pos', 'REF', 'ALT', 'SNPEff_Gene_Name', 'SNPEff_Gene_ID', 'SNPEff_region', 'SNP_pos_in_guide', 'guide_strand']\n",
    "tdh3_oligo_info['relative_pos'] = (promoter_cds_start - tdh3_oligo_info.SNP_chr_pos)*promoter_strand\n",
    "# tdh3_oligo_info.rename(columns={'SNP_id':'var_id'}, inplace=True)\n",
    "\n",
    "\n",
    "# filter out oligos where SNP_pos_in_guide < -10\n",
    "tdh3_oligo_info = tdh3_oligo_info.query('SNP_pos_in_guide>=-9')\n",
    "\n",
    "# keep only variants targeted by 2 or more oligos\n",
    "tdh3_oligo_info = tdh3_oligo_info.groupby('SNP_id').filter(lambda x: len(x)>1)\n",
    "\n",
    "# select oligos targeting variants in known TFBS and/or have fitness data and/or have >2% reduction in TDH3 expression\n",
    "tdh3_oligo_info = tdh3_oligo_info.query('(relative_pos<=-441 & relative_pos>=-449) | \\\n",
    "                               (relative_pos<=-479 & relative_pos>=-487) | \\\n",
    "                               (relative_pos<=-502 & relative_pos>=-510) | \\\n",
    "                               (relative_pos.isin(@wittkopp_fitness_df.Position)) | \\\n",
    "                               (relative_pos.isin(@wittkopp_expr_df.POSITION))')\n",
    "tdh3_oligo_info = tdh3_oligo_info.drop('relative_pos', axis=1)\n",
    "display(tdh3_oligo_info)\n",
    "display(tdh3_oligo_info.shape)\n",
    "\n",
    "# retrieve variants from vcf file\n",
    "prom_vcf = vcf.Reader(filename=prom_vcf_file)\n",
    "tdh3_var_list = []\n",
    "tdh_num=0\n",
    "for record in prom_vcf:\n",
    "    if record.ID in tdh3_oligo_info.SNP_id.values:\n",
    "        tdh_num+=1\n",
    "        # append old record ID into INFO\n",
    "        record.INFO['OVID'] = record.ID\n",
    "        # add new variant ID to epi_oligo_df and rename record ID\n",
    "        new_var_id = 'TDH_{:05d}'.format(tdh_num)\n",
    "        tdh3_oligo_info.loc[tdh3_oligo_info.SNP_id==record.ID,'var_id'] = new_var_id\n",
    "        record.ID = new_var_id\n",
    "        # add VCF record to list\n",
    "        tdh3_var_list.append(record)\n",
    "tdh3_var_list = pd.Series(tdh3_var_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select variants from previous screen that show fitness effects (validation sets)\n",
    "These variants are potentially useful for detecting GxG effects. Fitness effect was calculated from stacked barcode counts fitted into limma-voom (limma ran on 4 strains together, gen logFC estimated for 18.5k oligos meeting 10k read cutoff across 4 strains)\n",
    "\n",
    "Selection criteria:\n",
    "- Significant hit in at least 3 strains (adjusted p-value<0.05) - test set for measuring GxG\n",
    "- Very large effect in one strain (abs(logFC)>0.01) - strain-specific effect validation\n",
    "\n",
    "Ensure each variant is targeted by 2 or more oligos whose edit position in guide is between -9 and +2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "g_list = ['BY', 'RM', 'YJM', 'YPS']\n",
    "hits_df = []\n",
    "for g in g_list:\n",
    "    limma_output_file= working_dir+\"limma/4-indiv-comp-oct2019/limma_results_{}_stacked_070820.txt\".format(g)\n",
    "    var_fitness_df = pd.read_csv(limma_output_file, sep='\\t', index_col=0, header=0)\n",
    "    var_fitness_df.index.name = 'oligo_id'\n",
    "    # annotate variant fitness file\n",
    "    var_fitness_df = var_fitness_df.merge(oligo_info_df[['set_name', 'SNP_id', 'chrom', 'SNP_chr_pos', 'SNP_pos_in_guide', 'guide_strand', 'SNPEff_region']], left_on='oligo_id', right_index=True, how='inner').sort_index()\n",
    "    var_fitness_df.rename(columns={'SNP_id':'var_id', 'P.Value':'pval', 'adj.P.Val':'padj'}, inplace=True)\n",
    "    # extract hits from epistasis main set\n",
    "    df = var_fitness_df.query('padj<0.05 & oligo_id.str.contains(\"SNP_avoidExcludedSeqs\")').sort_values('logFC', ascending=True)[['var_id', 'logFC', 'padj']].rename(columns={'logFC':'logFC_{}'.format(g), 'padj':'padj_{}'.format(g)}).reset_index()\n",
    "    hits_df.append(df)\n",
    "\n",
    "hits_df = pd.concat([x.set_index(['oligo_id', 'var_id']) for x in hits_df], axis=1, join='outer').reset_index(level=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping variant Record(CHROM=IV, POS=319482, REF=G, ALT=[A]): already in GxG set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/groups/hbfraser/modules/packages/conda/4.6.14/envs/crispey-bar/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item] = s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping variant Record(CHROM=V, POS=88751, REF=C, ALT=[T]): already in GxG set\n",
      "Skipping variant Record(CHROM=VI, POS=74406, REF=C, ALT=[T]): already in GxG set\n",
      "Skipping variant Record(CHROM=VII, POS=619256, REF=G, ALT=[A]): already in GxG set\n",
      "Skipping variant Record(CHROM=IX, POS=112910, REF=C, ALT=[T]): already in GxG set\n",
      "Skipping variant Record(CHROM=X, POS=519926, REF=G, ALT=[A]): already in GxG set\n",
      "Skipping variant Record(CHROM=XIII, POS=682727, REF=C, ALT=[T]): already in GxG set\n"
     ]
    }
   ],
   "source": [
    "# Validation set 1: significant hits in at least 3 strains\n",
    "set1 = hits_df[hits_df.isna().sum(axis=1)<=2]\n",
    "# filter out oligos where SNP_pos_in_guide < -10\n",
    "sig_hits_set = oligo_info_df.query('SNP_id.isin(@set1.var_id) & SNP_pos_in_guide>-10').copy()\n",
    "# keep only variants targeted by 2 or more oligos\n",
    "sig_hits_set = sig_hits_set.groupby('SNP_id').filter(lambda x: len(x)>1)\n",
    "\n",
    "# display(set1)\n",
    "# display(sig_hits_set)\n",
    "\n",
    "# Validation set 2: large effects observed in at least one strain NOT IN SET 1\n",
    "# effect size cut off is arbitrary - picked to keep final number of oligos manageable\n",
    "set2 = hits_df[(hits_df.filter(regex=(\"logFC_*\")).abs()>0.0405).any(axis=1)].query('~var_id.isin(@set1.var_id)')\n",
    "# filter out oligos where SNP_pos_in_guide < -10\n",
    "large_effect_set = oligo_info_df.query('SNP_id.isin(@set2.var_id) & SNP_pos_in_guide>-10').copy()\n",
    "# keep only variants targeted by 2 or more oligos\n",
    "large_effect_set = large_effect_set.groupby('SNP_id').filter(lambda x: len(x)>1)\n",
    "# display(set2)\n",
    "# display(large_effect_set)\n",
    "\n",
    "# set 1 and set 2 oligos\n",
    "epi_oligo_df = pd.concat([sig_hits_set, large_effect_set])\n",
    "\n",
    "\n",
    "# read in gxg_oligos_final to check for dups\n",
    "gxg_oligos_final = pd.read_csv(gxg_oligos_final_file, sep='\\t', index_col=0)\n",
    "\n",
    "# retrieve variants from vcf file\n",
    "epi_vcf = vcf.Reader(filename=epi_vcf_file)\n",
    "epi_var_list = []\n",
    "val_num=0\n",
    "for record in epi_vcf:\n",
    "    if (record.ID in epi_oligo_df.SNP_id.values):\n",
    "        # add variant if it is not in main GxG set\n",
    "        if gxg_oligos_final.query('CHROM==@record.CHROM & POS==@record.POS & ALT==@record.ALT[0]').empty:\n",
    "            val_num+=1\n",
    "            # append old record ID into INFO\n",
    "            record.INFO['OVID'] = record.ID\n",
    "            # add new variant ID to epi_oligo_df and rename record ID\n",
    "            new_var_id = 'VAL_{:05d}'.format(val_num)\n",
    "            epi_oligo_df.loc[epi_oligo_df.SNP_id==record.ID,'var_id'] = new_var_id\n",
    "            record.ID = new_var_id\n",
    "            # add VCF record to list\n",
    "            epi_var_list.append(record)\n",
    "        # skip variant if already in GxG set\n",
    "        else:\n",
    "            print('Skipping variant {}: already in GxG set'.format(record))\n",
    "            # remove oligos targeting this variant\n",
    "            epi_oligo_df = epi_oligo_df.loc[epi_oligo_df['SNP_id']!=record.ID,:]\n",
    "\n",
    "epi_var_list = pd.Series(epi_var_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>oligo_num</th>\n",
       "      <th>oligo_seq</th>\n",
       "      <th>set_name</th>\n",
       "      <th>set_priority</th>\n",
       "      <th>chrom</th>\n",
       "      <th>SNP_chr_pos</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP_id</th>\n",
       "      <th>guide_chr_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>SNPEff_AA_pos</th>\n",
       "      <th>VEP_Amino_acids</th>\n",
       "      <th>VEP_Codons</th>\n",
       "      <th>VEP_HGVSc</th>\n",
       "      <th>VEP_HGVSp</th>\n",
       "      <th>VEP_SWISSPROT</th>\n",
       "      <th>VEP_UNIPARC</th>\n",
       "      <th>VEP_DOMAINS</th>\n",
       "      <th>VEP_BLOSUM62</th>\n",
       "      <th>var_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oligo_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28474#promoterMut</th>\n",
       "      <td>28474</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCAGAGAGAGCAGAAGGTAATAGAAGG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884068</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_883</td>\n",
       "      <td>884073</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TDH_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28475#promoterMut</th>\n",
       "      <td>28475</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAGAGAGAGCAGAAGGTAATAGAAGGT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884068</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_883</td>\n",
       "      <td>884074</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TDH_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28484#promoterMut</th>\n",
       "      <td>28484</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTCAGAGAGAGCAGAAGGTAATAGAAGG...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884073</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_888</td>\n",
       "      <td>884073</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TDH_00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28485#promoterMut</th>\n",
       "      <td>28485</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAGAGAGAGCAGAAGGTAATAGAAGGT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884073</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>PROM_888</td>\n",
       "      <td>884074</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TDH_00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28539#promoterMut</th>\n",
       "      <td>28539</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTAAAATGAGATAGATACATGCGTGGGT...</td>\n",
       "      <td>promoterMut</td>\n",
       "      <td>5</td>\n",
       "      <td>VII</td>\n",
       "      <td>884105</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>PROM_920</td>\n",
       "      <td>884105</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00359</td>\n",
       "      <td>UPI000004F91A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TDH_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24211#SNP_avoidExcludedSeqs</th>\n",
       "      <td>24211</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTGATGTCATCCTATAATTGTTTCCGTA...</td>\n",
       "      <td>SNP_avoidExcludedSeqs</td>\n",
       "      <td>0</td>\n",
       "      <td>XVI</td>\n",
       "      <td>563805</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>RAR_14853-C1</td>\n",
       "      <td>563813</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P53394</td>\n",
       "      <td>UPI0000136210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAL_00240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24791#SNP_avoidExcludedSeqs</th>\n",
       "      <td>24791</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTGTCATGATTTACTAAAGCTAACCAT...</td>\n",
       "      <td>SNP_avoidExcludedSeqs</td>\n",
       "      <td>0</td>\n",
       "      <td>XVI</td>\n",
       "      <td>820797</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>COM_4811</td>\n",
       "      <td>820797</td>\n",
       "      <td>...</td>\n",
       "      <td>209</td>\n",
       "      <td>Q</td>\n",
       "      <td>caG/caA</td>\n",
       "      <td>YPR144C_mRNA.:c.627G&gt;A</td>\n",
       "      <td>YPR144C.:p.Gln209%3D</td>\n",
       "      <td>Q06512</td>\n",
       "      <td>UPI0000053033</td>\n",
       "      <td>PANTHER:PTHR12455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAL_00242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24792#SNP_avoidExcludedSeqs</th>\n",
       "      <td>24792</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTATGATTTACTAAAGCTAACCATTTAC...</td>\n",
       "      <td>SNP_avoidExcludedSeqs</td>\n",
       "      <td>0</td>\n",
       "      <td>XVI</td>\n",
       "      <td>820797</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>COM_4811</td>\n",
       "      <td>820801</td>\n",
       "      <td>...</td>\n",
       "      <td>209</td>\n",
       "      <td>Q</td>\n",
       "      <td>caG/caA</td>\n",
       "      <td>YPR144C_mRNA.:c.627G&gt;A</td>\n",
       "      <td>YPR144C.:p.Gln209%3D</td>\n",
       "      <td>Q06512</td>\n",
       "      <td>UPI0000053033</td>\n",
       "      <td>PANTHER:PTHR12455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAL_00242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24793#SNP_avoidExcludedSeqs</th>\n",
       "      <td>24793</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTTTACTAAAGCTAACCATTTACCAAC...</td>\n",
       "      <td>SNP_avoidExcludedSeqs</td>\n",
       "      <td>0</td>\n",
       "      <td>XVI</td>\n",
       "      <td>820797</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>COM_4811</td>\n",
       "      <td>820805</td>\n",
       "      <td>...</td>\n",
       "      <td>209</td>\n",
       "      <td>Q</td>\n",
       "      <td>caG/caA</td>\n",
       "      <td>YPR144C_mRNA.:c.627G&gt;A</td>\n",
       "      <td>YPR144C.:p.Gln209%3D</td>\n",
       "      <td>Q06512</td>\n",
       "      <td>UPI0000053033</td>\n",
       "      <td>PANTHER:PTHR12455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAL_00242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24794#SNP_avoidExcludedSeqs</th>\n",
       "      <td>24794</td>\n",
       "      <td>GAGTTACTGTCTGTTTTCCTTCAATACTACTTTCAATCAGAATTTA...</td>\n",
       "      <td>SNP_avoidExcludedSeqs</td>\n",
       "      <td>0</td>\n",
       "      <td>XVI</td>\n",
       "      <td>820797</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>COM_4811</td>\n",
       "      <td>820797</td>\n",
       "      <td>...</td>\n",
       "      <td>209</td>\n",
       "      <td>Q</td>\n",
       "      <td>caG/caA</td>\n",
       "      <td>YPR144C_mRNA.:c.627G&gt;A</td>\n",
       "      <td>YPR144C.:p.Gln209%3D</td>\n",
       "      <td>Q06512</td>\n",
       "      <td>UPI0000053033</td>\n",
       "      <td>PANTHER:PTHR12455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAL_00242</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>665 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             oligo_num  \\\n",
       "oligo_id                                 \n",
       "28474#promoterMut                28474   \n",
       "28475#promoterMut                28475   \n",
       "28484#promoterMut                28484   \n",
       "28485#promoterMut                28485   \n",
       "28539#promoterMut                28539   \n",
       "...                                ...   \n",
       "24211#SNP_avoidExcludedSeqs      24211   \n",
       "24791#SNP_avoidExcludedSeqs      24791   \n",
       "24792#SNP_avoidExcludedSeqs      24792   \n",
       "24793#SNP_avoidExcludedSeqs      24793   \n",
       "24794#SNP_avoidExcludedSeqs      24794   \n",
       "\n",
       "                                                                     oligo_seq  \\\n",
       "oligo_id                                                                         \n",
       "28474#promoterMut            GAGTTACTGTCTGTTTTCCTCAGAGAGAGCAGAAGGTAATAGAAGG...   \n",
       "28475#promoterMut            GAGTTACTGTCTGTTTTCCTAGAGAGAGCAGAAGGTAATAGAAGGT...   \n",
       "28484#promoterMut            GAGTTACTGTCTGTTTTCCTCAGAGAGAGCAGAAGGTAATAGAAGG...   \n",
       "28485#promoterMut            GAGTTACTGTCTGTTTTCCTAGAGAGAGCAGAAGGTAATAGAAGGT...   \n",
       "28539#promoterMut            GAGTTACTGTCTGTTTTCCTAAAATGAGATAGATACATGCGTGGGT...   \n",
       "...                                                                        ...   \n",
       "24211#SNP_avoidExcludedSeqs  GAGTTACTGTCTGTTTTCCTGATGTCATCCTATAATTGTTTCCGTA...   \n",
       "24791#SNP_avoidExcludedSeqs  GAGTTACTGTCTGTTTTCCTTGTCATGATTTACTAAAGCTAACCAT...   \n",
       "24792#SNP_avoidExcludedSeqs  GAGTTACTGTCTGTTTTCCTATGATTTACTAAAGCTAACCATTTAC...   \n",
       "24793#SNP_avoidExcludedSeqs  GAGTTACTGTCTGTTTTCCTTTTACTAAAGCTAACCATTTACCAAC...   \n",
       "24794#SNP_avoidExcludedSeqs  GAGTTACTGTCTGTTTTCCTTCAATACTACTTTCAATCAGAATTTA...   \n",
       "\n",
       "                                          set_name  set_priority chrom  \\\n",
       "oligo_id                                                                 \n",
       "28474#promoterMut                      promoterMut             5   VII   \n",
       "28475#promoterMut                      promoterMut             5   VII   \n",
       "28484#promoterMut                      promoterMut             5   VII   \n",
       "28485#promoterMut                      promoterMut             5   VII   \n",
       "28539#promoterMut                      promoterMut             5   VII   \n",
       "...                                            ...           ...   ...   \n",
       "24211#SNP_avoidExcludedSeqs  SNP_avoidExcludedSeqs             0   XVI   \n",
       "24791#SNP_avoidExcludedSeqs  SNP_avoidExcludedSeqs             0   XVI   \n",
       "24792#SNP_avoidExcludedSeqs  SNP_avoidExcludedSeqs             0   XVI   \n",
       "24793#SNP_avoidExcludedSeqs  SNP_avoidExcludedSeqs             0   XVI   \n",
       "24794#SNP_avoidExcludedSeqs  SNP_avoidExcludedSeqs             0   XVI   \n",
       "\n",
       "                             SNP_chr_pos REF  ALT        SNP_id  \\\n",
       "oligo_id                                                          \n",
       "28474#promoterMut                 884068   C  [T]      PROM_883   \n",
       "28475#promoterMut                 884068   C  [T]      PROM_883   \n",
       "28484#promoterMut                 884073   G  [A]      PROM_888   \n",
       "28485#promoterMut                 884073   G  [A]      PROM_888   \n",
       "28539#promoterMut                 884105   C  [T]      PROM_920   \n",
       "...                                  ...  ..  ...           ...   \n",
       "24211#SNP_avoidExcludedSeqs       563805   G  [A]  RAR_14853-C1   \n",
       "24791#SNP_avoidExcludedSeqs       820797   C  [T]      COM_4811   \n",
       "24792#SNP_avoidExcludedSeqs       820797   C  [T]      COM_4811   \n",
       "24793#SNP_avoidExcludedSeqs       820797   C  [T]      COM_4811   \n",
       "24794#SNP_avoidExcludedSeqs       820797   C  [T]      COM_4811   \n",
       "\n",
       "                             guide_chr_pos  ... SNPEff_AA_pos  \\\n",
       "oligo_id                                    ...                 \n",
       "28474#promoterMut                   884073  ...           NaN   \n",
       "28475#promoterMut                   884074  ...           NaN   \n",
       "28484#promoterMut                   884073  ...           NaN   \n",
       "28485#promoterMut                   884074  ...           NaN   \n",
       "28539#promoterMut                   884105  ...           NaN   \n",
       "...                                    ...  ...           ...   \n",
       "24211#SNP_avoidExcludedSeqs         563813  ...           NaN   \n",
       "24791#SNP_avoidExcludedSeqs         820797  ...           209   \n",
       "24792#SNP_avoidExcludedSeqs         820801  ...           209   \n",
       "24793#SNP_avoidExcludedSeqs         820805  ...           209   \n",
       "24794#SNP_avoidExcludedSeqs         820797  ...           209   \n",
       "\n",
       "                             VEP_Amino_acids VEP_Codons  \\\n",
       "oligo_id                                                  \n",
       "28474#promoterMut                        NaN        NaN   \n",
       "28475#promoterMut                        NaN        NaN   \n",
       "28484#promoterMut                        NaN        NaN   \n",
       "28485#promoterMut                        NaN        NaN   \n",
       "28539#promoterMut                        NaN        NaN   \n",
       "...                                      ...        ...   \n",
       "24211#SNP_avoidExcludedSeqs              NaN        NaN   \n",
       "24791#SNP_avoidExcludedSeqs                Q    caG/caA   \n",
       "24792#SNP_avoidExcludedSeqs                Q    caG/caA   \n",
       "24793#SNP_avoidExcludedSeqs                Q    caG/caA   \n",
       "24794#SNP_avoidExcludedSeqs                Q    caG/caA   \n",
       "\n",
       "                                          VEP_HGVSc             VEP_HGVSp  \\\n",
       "oligo_id                                                                    \n",
       "28474#promoterMut                               NaN                   NaN   \n",
       "28475#promoterMut                               NaN                   NaN   \n",
       "28484#promoterMut                               NaN                   NaN   \n",
       "28485#promoterMut                               NaN                   NaN   \n",
       "28539#promoterMut                               NaN                   NaN   \n",
       "...                                             ...                   ...   \n",
       "24211#SNP_avoidExcludedSeqs                     NaN                   NaN   \n",
       "24791#SNP_avoidExcludedSeqs  YPR144C_mRNA.:c.627G>A  YPR144C.:p.Gln209%3D   \n",
       "24792#SNP_avoidExcludedSeqs  YPR144C_mRNA.:c.627G>A  YPR144C.:p.Gln209%3D   \n",
       "24793#SNP_avoidExcludedSeqs  YPR144C_mRNA.:c.627G>A  YPR144C.:p.Gln209%3D   \n",
       "24794#SNP_avoidExcludedSeqs  YPR144C_mRNA.:c.627G>A  YPR144C.:p.Gln209%3D   \n",
       "\n",
       "                            VEP_SWISSPROT    VEP_UNIPARC        VEP_DOMAINS  \\\n",
       "oligo_id                                                                      \n",
       "28474#promoterMut                  P00359  UPI000004F91A                NaN   \n",
       "28475#promoterMut                  P00359  UPI000004F91A                NaN   \n",
       "28484#promoterMut                  P00359  UPI000004F91A                NaN   \n",
       "28485#promoterMut                  P00359  UPI000004F91A                NaN   \n",
       "28539#promoterMut                  P00359  UPI000004F91A                NaN   \n",
       "...                                   ...            ...                ...   \n",
       "24211#SNP_avoidExcludedSeqs        P53394  UPI0000136210                NaN   \n",
       "24791#SNP_avoidExcludedSeqs        Q06512  UPI0000053033  PANTHER:PTHR12455   \n",
       "24792#SNP_avoidExcludedSeqs        Q06512  UPI0000053033  PANTHER:PTHR12455   \n",
       "24793#SNP_avoidExcludedSeqs        Q06512  UPI0000053033  PANTHER:PTHR12455   \n",
       "24794#SNP_avoidExcludedSeqs        Q06512  UPI0000053033  PANTHER:PTHR12455   \n",
       "\n",
       "                            VEP_BLOSUM62     var_id  \n",
       "oligo_id                                             \n",
       "28474#promoterMut                    NaN  TDH_00001  \n",
       "28475#promoterMut                    NaN  TDH_00001  \n",
       "28484#promoterMut                    NaN  TDH_00002  \n",
       "28485#promoterMut                    NaN  TDH_00002  \n",
       "28539#promoterMut                    NaN  TDH_00003  \n",
       "...                                  ...        ...  \n",
       "24211#SNP_avoidExcludedSeqs          NaN  VAL_00240  \n",
       "24791#SNP_avoidExcludedSeqs          NaN  VAL_00242  \n",
       "24792#SNP_avoidExcludedSeqs          NaN  VAL_00242  \n",
       "24793#SNP_avoidExcludedSeqs          NaN  VAL_00242  \n",
       "24794#SNP_avoidExcludedSeqs          NaN  VAL_00242  \n",
       "\n",
       "[665 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# combine selected TDH3 promoter oligos and epi set oligos\n",
    "known_effects_oligo_df = pd.concat([tdh3_oligo_info, epi_oligo_df])\n",
    "display(known_effects_oligo_df)\n",
    "\n",
    "# write selected TDH3 promoter variants and epi set variants to VCF, run on oligo design pipeline\n",
    "known_effects_var_list = pd.concat([tdh3_var_list, epi_var_list]).reset_index(drop=True)\n",
    "known_effects_vcf_file = '/home/users/rang/scratch/yeast/epistasis/epival_variants_final.vcf'\n",
    "epi_vcf = vcf.Reader(filename=epi_vcf_file)\n",
    "known_effects_vcf = vcf.Writer(open(known_effects_vcf_file, 'w'), epi_vcf)\n",
    "for record in known_effects_var_list:\n",
    "    known_effects_vcf.write_record(record)\n",
    "known_effects_vcf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>SNP_chr_pos</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>guide_id</th>\n",
       "      <th>guide_strand</th>\n",
       "      <th>guide0_chr_pos</th>\n",
       "      <th>guide_cut_chr_pos</th>\n",
       "      <th>SNP_pos_in_guide</th>\n",
       "      <th>guide_index_in_snp</th>\n",
       "      <th>...</th>\n",
       "      <th>donor_seq_shift</th>\n",
       "      <th>donor_mut_pos_in_guide</th>\n",
       "      <th>donor_info_str</th>\n",
       "      <th>set_name</th>\n",
       "      <th>contains_excluded_seqs</th>\n",
       "      <th>pool</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>barcode_id</th>\n",
       "      <th>oligo_seq</th>\n",
       "      <th>oligo_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TDH_00001</th>\n",
       "      <td>VII</td>\n",
       "      <td>884068</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>guide_1</td>\n",
       "      <td>+</td>\n",
       "      <td>884073</td>\n",
       "      <td>884070</td>\n",
       "      <td>-6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-6</td>\n",
       "      <td>REF2ALT:C&gt;T</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_1#guide_1:REF2ALT:offset14:donorI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TDH_00001</th>\n",
       "      <td>VII</td>\n",
       "      <td>884068</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>guide_2</td>\n",
       "      <td>+</td>\n",
       "      <td>884074</td>\n",
       "      <td>884071</td>\n",
       "      <td>-7</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-7</td>\n",
       "      <td>REF2ALT:C&gt;T</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_2#guide_2:REF2ALT:offset14:donorI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TDH_00002</th>\n",
       "      <td>VII</td>\n",
       "      <td>884073</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>guide_3</td>\n",
       "      <td>+</td>\n",
       "      <td>884073</td>\n",
       "      <td>884070</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-1</td>\n",
       "      <td>REF2ALT:G&gt;A</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_3#guide_3:REF2ALT:offset14:donorI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TDH_00002</th>\n",
       "      <td>VII</td>\n",
       "      <td>884073</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>guide_4</td>\n",
       "      <td>+</td>\n",
       "      <td>884074</td>\n",
       "      <td>884071</td>\n",
       "      <td>-2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-2</td>\n",
       "      <td>REF2ALT:G&gt;A</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_4#guide_4:REF2ALT:offset14:donorI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TDH_00003</th>\n",
       "      <td>VII</td>\n",
       "      <td>884105</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>guide_5</td>\n",
       "      <td>+</td>\n",
       "      <td>884105</td>\n",
       "      <td>884102</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-1</td>\n",
       "      <td>REF2ALT:C&gt;T</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_5#guide_5:REF2ALT:offset14:donorI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00243</th>\n",
       "      <td>XVI</td>\n",
       "      <td>868939</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>guide_704</td>\n",
       "      <td>+</td>\n",
       "      <td>868941</td>\n",
       "      <td>868938</td>\n",
       "      <td>-3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-3</td>\n",
       "      <td>REF2ALT:A&gt;G</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_704#guide_704:REF2ALT:offset14:do...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00243</th>\n",
       "      <td>XVI</td>\n",
       "      <td>868939</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>guide_705</td>\n",
       "      <td>-</td>\n",
       "      <td>868936</td>\n",
       "      <td>868940</td>\n",
       "      <td>-2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>-14</td>\n",
       "      <td>-2</td>\n",
       "      <td>REF2ALT:A&gt;G</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_705#guide_705:REF2ALT:offset-14:d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00243</th>\n",
       "      <td>XVI</td>\n",
       "      <td>868939</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>guide_706</td>\n",
       "      <td>-</td>\n",
       "      <td>868931</td>\n",
       "      <td>868935</td>\n",
       "      <td>-7</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>-14</td>\n",
       "      <td>-7</td>\n",
       "      <td>REF2ALT:A&gt;G</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_706#guide_706:REF2ALT:offset-14:d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00243</th>\n",
       "      <td>XVI</td>\n",
       "      <td>868939</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>guide_707</td>\n",
       "      <td>-</td>\n",
       "      <td>868930</td>\n",
       "      <td>868934</td>\n",
       "      <td>-8</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>-14</td>\n",
       "      <td>-8</td>\n",
       "      <td>REF2ALT:A&gt;G</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_707#guide_707:REF2ALT:offset-14:d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAL_00243</th>\n",
       "      <td>XVI</td>\n",
       "      <td>868939</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>guide_708</td>\n",
       "      <td>-</td>\n",
       "      <td>868929</td>\n",
       "      <td>868933</td>\n",
       "      <td>-9</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>-14</td>\n",
       "      <td>-9</td>\n",
       "      <td>REF2ALT:A&gt;G</td>\n",
       "      <td>epival</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>epival#guide_708#guide_708:REF2ALT:offset-14:d...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>692 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom  SNP_chr_pos REF  ALT   guide_id guide_strand  guide0_chr_pos  \\\n",
       "var_id                                                                          \n",
       "TDH_00001   VII       884068   C  [T]    guide_1            +          884073   \n",
       "TDH_00001   VII       884068   C  [T]    guide_2            +          884074   \n",
       "TDH_00002   VII       884073   G  [A]    guide_3            +          884073   \n",
       "TDH_00002   VII       884073   G  [A]    guide_4            +          884074   \n",
       "TDH_00003   VII       884105   C  [T]    guide_5            +          884105   \n",
       "...         ...          ...  ..  ...        ...          ...             ...   \n",
       "VAL_00243   XVI       868939   A  [G]  guide_704            +          868941   \n",
       "VAL_00243   XVI       868939   A  [G]  guide_705            -          868936   \n",
       "VAL_00243   XVI       868939   A  [G]  guide_706            -          868931   \n",
       "VAL_00243   XVI       868939   A  [G]  guide_707            -          868930   \n",
       "VAL_00243   XVI       868939   A  [G]  guide_708            -          868929   \n",
       "\n",
       "           guide_cut_chr_pos  SNP_pos_in_guide  guide_index_in_snp  ...  \\\n",
       "var_id                                                              ...   \n",
       "TDH_00001             884070                -6                   1  ...   \n",
       "TDH_00001             884071                -7                   2  ...   \n",
       "TDH_00002             884070                -1                   1  ...   \n",
       "TDH_00002             884071                -2                   2  ...   \n",
       "TDH_00003             884102                -1                   1  ...   \n",
       "...                      ...               ...                 ...  ...   \n",
       "VAL_00243             868938                -3                   1  ...   \n",
       "VAL_00243             868940                -2                   2  ...   \n",
       "VAL_00243             868935                -7                   3  ...   \n",
       "VAL_00243             868934                -8                   4  ...   \n",
       "VAL_00243             868933                -9                   5  ...   \n",
       "\n",
       "          donor_seq_shift donor_mut_pos_in_guide donor_info_str set_name  \\\n",
       "var_id                                                                     \n",
       "TDH_00001              14                     -6    REF2ALT:C>T   epival   \n",
       "TDH_00001              14                     -7    REF2ALT:C>T   epival   \n",
       "TDH_00002              14                     -1    REF2ALT:G>A   epival   \n",
       "TDH_00002              14                     -2    REF2ALT:G>A   epival   \n",
       "TDH_00003              14                     -1    REF2ALT:C>T   epival   \n",
       "...                   ...                    ...            ...      ...   \n",
       "VAL_00243              14                     -3    REF2ALT:A>G   epival   \n",
       "VAL_00243             -14                     -2    REF2ALT:A>G   epival   \n",
       "VAL_00243             -14                     -7    REF2ALT:A>G   epival   \n",
       "VAL_00243             -14                     -8    REF2ALT:A>G   epival   \n",
       "VAL_00243             -14                     -9    REF2ALT:A>G   epival   \n",
       "\n",
       "          contains_excluded_seqs  pool  barcode_seq  barcode_id  \\\n",
       "var_id                                                            \n",
       "TDH_00001                  False    -1          NaN        None   \n",
       "TDH_00001                  False    -1          NaN        None   \n",
       "TDH_00002                  False    -1          NaN        None   \n",
       "TDH_00002                  False    -1          NaN        None   \n",
       "TDH_00003                  False    -1          NaN        None   \n",
       "...                          ...   ...          ...         ...   \n",
       "VAL_00243                  False    -1          NaN        None   \n",
       "VAL_00243                  False    -1          NaN        None   \n",
       "VAL_00243                  False    -1          NaN        None   \n",
       "VAL_00243                  False    -1          NaN        None   \n",
       "VAL_00243                  False    -1          NaN        None   \n",
       "\n",
       "                                                   oligo_seq  \\\n",
       "var_id                                                         \n",
       "TDH_00001  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "TDH_00001  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "TDH_00002  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "TDH_00002  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "TDH_00003  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "...                                                      ...   \n",
       "VAL_00243  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "VAL_00243  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "VAL_00243  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "VAL_00243  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "VAL_00243  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "\n",
       "                                                    oligo_id  \n",
       "var_id                                                        \n",
       "TDH_00001  epival#guide_1#guide_1:REF2ALT:offset14:donorI...  \n",
       "TDH_00001  epival#guide_2#guide_2:REF2ALT:offset14:donorI...  \n",
       "TDH_00002  epival#guide_3#guide_3:REF2ALT:offset14:donorI...  \n",
       "TDH_00002  epival#guide_4#guide_4:REF2ALT:offset14:donorI...  \n",
       "TDH_00003  epival#guide_5#guide_5:REF2ALT:offset14:donorI...  \n",
       "...                                                      ...  \n",
       "VAL_00243  epival#guide_704#guide_704:REF2ALT:offset14:do...  \n",
       "VAL_00243  epival#guide_705#guide_705:REF2ALT:offset-14:d...  \n",
       "VAL_00243  epival#guide_706#guide_706:REF2ALT:offset-14:d...  \n",
       "VAL_00243  epival#guide_707#guide_707:REF2ALT:offset-14:d...  \n",
       "VAL_00243  epival#guide_708#guide_708:REF2ALT:offset-14:d...  \n",
       "\n",
       "[692 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TDH3 promoter oligos: 42\n",
      "Number of validation set oligos: 650\n",
      "Total number of oligos to fit: 692\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(5, 6)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use 5 pools:\n",
      "Number of oligos per pool: 118\n",
      "Number of oligos leftover: 102\n",
      "Number of technical oligos: 15\n",
      "\n",
      "Use 6 pools:\n",
      "Number of oligos per pool: 115\n",
      "Number of oligos leftover: 2\n",
      "Number of technical oligos: 36\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# after redesigning oligos for the selected validation variants, inspect them\n",
    "known_effects_oligo_updated_file = '/home/users/rang/crispey3/initial_design/Output/all_SNPs_epival_GG_9bp_OLIGO.tab'\n",
    "known_effects_oligo_df = pd.read_csv(known_effects_oligo_updated_file, sep='\\t', index_col=0)\n",
    "display(known_effects_oligo_df)\n",
    "\n",
    "# count OLIGOS in each set\n",
    "print(\"Number of TDH3 promoter oligos:\", len(known_effects_oligo_df.filter(like='TDH',axis=0)))\n",
    "print(\"Number of validation set oligos:\", len(known_effects_oligo_df.filter(like='VAL',axis=0)))\n",
    "\n",
    "# determine number of pools to use\n",
    "num_of_oligos = len(known_effects_oligo_df)\n",
    "min_pool_size = 109 # allows 12 technical oligos\n",
    "max_pool_size = 118 # allows 3 technical oligos\n",
    "complete_pool_size = 121 # maximum oligos synthesized per pool\n",
    "\n",
    "# find number of pools to use\n",
    "find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size)\n",
    "\n",
    "# # write known_effects_oligo_df to file for pooling assignment\n",
    "known_effects_oligo_df.to_csv('/home/users/rang/scratch/yeast/epistasis/epival_variants_final.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant ID prefix and which genomes they can be edited in:\n",
    "- TDH: BY | --- | --- | ---\n",
    "- VAL: BY | RM | YJM | YPS\n",
    "\n",
    "### We should be able to fit known effects variants in 6 pools\n",
    "- TDH, VAL: 6 pools\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selecting variants for GxE screen\n",
    "Shi-An provided a list of QTLs identified with condition-specific fitness effects. Search QTL (and gene(s) within it) for variants to screen for GxE project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_variants_from_qtls(qtls_df, vcf_file, chrom_dict, upstream_length=500, downstream_length=500):\n",
    "    '''\n",
    "    modified version of select_variants function that accepts a dataframe of QTLs\n",
    "    (requires qtl_left and qtl_right columns defined, and assumes each QTL is associated with one gene)\n",
    "    defines search boundaries for each QTL and then searches vcf for variants within it\n",
    "    can be parallelized by splitting genes and vcf by chromosome\n",
    "    '''\n",
    "    # find gene upstream and downstream boundaries\n",
    "    qtls_df.loc[qtls_df['strand']=='+', 'gene_left'] = qtls_df['start']-upstream_length\n",
    "    qtls_df.loc[qtls_df['strand']=='+', 'gene_right'] = qtls_df['end']+downstream_length\n",
    "    qtls_df.loc[qtls_df['strand']=='-', 'gene_left'] = qtls_df['start']-downstream_length\n",
    "    qtls_df.loc[qtls_df['strand']=='-', 'gene_right'] = qtls_df['end']+upstream_length\n",
    "    \n",
    "    # define search boundaries by either gene or QTL boundary, whichever is larger\n",
    "    qtls_df.loc[:,'search_left'] = qtls_df[['gene_left', 'qtl_left']].min(axis=1)\n",
    "    qtls_df.loc[:,'search_right'] = qtls_df[['gene_right', 'qtl_right']].max(axis=1)\n",
    "    qtls_df.loc[:,'search_left'] = qtls_df['search_left'].astype(int)\n",
    "    qtls_df.loc[:,'search_right'] = qtls_df['search_right'].astype(int)\n",
    "    \n",
    "    # from these search boundaries, remove duplicates\n",
    "    search_boundaries_df = qtls_df[['chrom', 'search_left', 'search_right']].drop_duplicates()\n",
    "\n",
    "    # read in vcf file to parse for variants\n",
    "    allVars = vcf.Reader(filename=vcf_file)\n",
    "    candVars = []\n",
    "    for i, row in search_boundaries_df.iterrows():\n",
    "        retrieve = allVars.fetch(chrom=chrom_dict[row['chrom']], \n",
    "                                 start=row['search_left']-1,\n",
    "                                 end=row['search_right'])\n",
    "        for rec in retrieve:\n",
    "            rec.samples = []\n",
    "            rec.FORMAT=None\n",
    "            candVars.append(rec)\n",
    "    candVars = pd.Series(candVars).drop_duplicates().reset_index(drop=True)\n",
    "    return candVars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/GxE/\"\n",
    "# list of QTLs to parse for GxE variants\n",
    "gxe_qtls_file = working_dir+\"GxE_gene_selection_20201130.tsv\"\n",
    "\n",
    "os.chdir(working_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>cross</th>\n",
       "      <th>forward_scan_marker</th>\n",
       "      <th>index</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>LOD</th>\n",
       "      <th>relocalized_peak</th>\n",
       "      <th>...</th>\n",
       "      <th>Beta_abs</th>\n",
       "      <th>qtl_genes</th>\n",
       "      <th>num_qtl_genes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Caffeine;15mM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>624258</td>\n",
       "      <td>B</td>\n",
       "      <td>chrXV_620334_A_G_61508</td>\n",
       "      <td>18833</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>74.281303</td>\n",
       "      <td>chrXV_624258_G_A_61750</td>\n",
       "      <td>...</td>\n",
       "      <td>0.909699</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fluconazole;100uM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>621444</td>\n",
       "      <td>B</td>\n",
       "      <td>chrXV_621444_A_G_61540</td>\n",
       "      <td>18839</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>117.931452</td>\n",
       "      <td>chrXV_621444_A_G_61540</td>\n",
       "      <td>...</td>\n",
       "      <td>0.920556</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fluconazole;100uM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619941</td>\n",
       "      <td>377</td>\n",
       "      <td>chrXV_619941_G_T_47212</td>\n",
       "      <td>23734</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>168.252715</td>\n",
       "      <td>chrXV_619941_G_T_47212</td>\n",
       "      <td>...</td>\n",
       "      <td>0.711397</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fluconazole;100uM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>622413</td>\n",
       "      <td>3049</td>\n",
       "      <td>chrXV_622413_A_G_72894</td>\n",
       "      <td>25133</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>120.009983</td>\n",
       "      <td>chrXV_622413_A_G_72894</td>\n",
       "      <td>...</td>\n",
       "      <td>0.607320</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fluconazole;100uM;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>624320</td>\n",
       "      <td>377</td>\n",
       "      <td>chrXV_474477_C_A_46484</td>\n",
       "      <td>23322</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>7.932991</td>\n",
       "      <td>chrXV_624320_TTTTC_T_47455</td>\n",
       "      <td>...</td>\n",
       "      <td>0.107987</td>\n",
       "      <td>['YOR153W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOR153W</td>\n",
       "      <td>PDR5</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>619840</td>\n",
       "      <td>624375</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane ATP-binding cassette (ABC) tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>74126</td>\n",
       "      <td>381</td>\n",
       "      <td>chrXV_74126_C_T_49092</td>\n",
       "      <td>19288</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>79.558526</td>\n",
       "      <td>chrXV_74126_C_T_49092</td>\n",
       "      <td>...</td>\n",
       "      <td>0.345004</td>\n",
       "      <td>['YOL130W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YOL130W</td>\n",
       "      <td>ALR1</td>\n",
       "      <td>chrXV</td>\n",
       "      <td>74400</td>\n",
       "      <td>76979</td>\n",
       "      <td>+</td>\n",
       "      <td>Plasma membrane Mg(2%2B) transporter, expressi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>545729</td>\n",
       "      <td>3000</td>\n",
       "      <td>chrIV_545729_C_A_11919</td>\n",
       "      <td>4657</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>41.519803</td>\n",
       "      <td>chrIV_545729_C_A_11919</td>\n",
       "      <td>...</td>\n",
       "      <td>0.329880</td>\n",
       "      <td>['YDR044W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YDR044W</td>\n",
       "      <td>HEM13</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>546642</td>\n",
       "      <td>547628</td>\n",
       "      <td>+</td>\n",
       "      <td>Coproporphyrinogen III oxidase, an oxygen requ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>590540</td>\n",
       "      <td>B</td>\n",
       "      <td>chrIV_590540_T_C_12080</td>\n",
       "      <td>3384</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>44.270147</td>\n",
       "      <td>chrIV_590540_T_C_12080</td>\n",
       "      <td>...</td>\n",
       "      <td>0.301252</td>\n",
       "      <td>['YDR072C']</td>\n",
       "      <td>1</td>\n",
       "      <td>YDR072C</td>\n",
       "      <td>IPT1</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>589761</td>\n",
       "      <td>591344</td>\n",
       "      <td>-</td>\n",
       "      <td>Inositolphosphotransferase, involved in synthe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>chrXII</td>\n",
       "      <td>256508</td>\n",
       "      <td>393</td>\n",
       "      <td>chrXII_256370_C_G_49854</td>\n",
       "      <td>18962</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>47.917048</td>\n",
       "      <td>chrXII_256508_A_G_49856</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282061</td>\n",
       "      <td>['YLR057W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YLR057W</td>\n",
       "      <td>MNL2</td>\n",
       "      <td>chrXII</td>\n",
       "      <td>255306</td>\n",
       "      <td>257855</td>\n",
       "      <td>+</td>\n",
       "      <td>Putative protein of unknown function; YLR050W ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>chrXVI</td>\n",
       "      <td>181552</td>\n",
       "      <td>3028</td>\n",
       "      <td>chrXVI_181552_T_C_85082</td>\n",
       "      <td>28588</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>13.472733</td>\n",
       "      <td>chrXVI_181552_T_C_85082</td>\n",
       "      <td>...</td>\n",
       "      <td>0.135279</td>\n",
       "      <td>['YPL193W']</td>\n",
       "      <td>1</td>\n",
       "      <td>YPL193W</td>\n",
       "      <td>RSA1</td>\n",
       "      <td>chrXVI</td>\n",
       "      <td>181403</td>\n",
       "      <td>182548</td>\n",
       "      <td>+</td>\n",
       "      <td>Protein involved in the assembly of 60S riboso...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  trait     chr     pos cross      forward_scan_marker  index  \\\n",
       "0       Caffeine;15mM;2   chrXV  624258     B   chrXV_620334_A_G_61508  18833   \n",
       "1   Fluconazole;100uM;2   chrXV  621444     B   chrXV_621444_A_G_61540  18839   \n",
       "2   Fluconazole;100uM;2   chrXV  619941   377   chrXV_619941_G_T_47212  23734   \n",
       "3   Fluconazole;100uM;2   chrXV  622413  3049   chrXV_622413_A_G_72894  25133   \n",
       "4   Fluconazole;100uM;2   chrXV  624320   377   chrXV_474477_C_A_46484  23322   \n",
       "..                  ...     ...     ...   ...                      ...    ...   \n",
       "94    Neomycin;5mg/mL;2   chrXV   74126   381    chrXV_74126_C_T_49092  19288   \n",
       "95    Neomycin;5mg/mL;2   chrIV  545729  3000   chrIV_545729_C_A_11919   4657   \n",
       "96    Neomycin;5mg/mL;2   chrIV  590540     B   chrIV_590540_T_C_12080   3384   \n",
       "97    Neomycin;5mg/mL;2  chrXII  256508   393  chrXII_256370_C_G_49854  18962   \n",
       "98    Neomycin;5mg/mL;2  chrXVI  181552  3028  chrXVI_181552_T_C_85082  28588   \n",
       "\n",
       "    p-value  q-value         LOD            relocalized_peak  ...  Beta_abs  \\\n",
       "0    0.0001   0.0001   74.281303      chrXV_624258_G_A_61750  ...  0.909699   \n",
       "1    0.0001   0.0001  117.931452      chrXV_621444_A_G_61540  ...  0.920556   \n",
       "2    0.0001   0.0001  168.252715      chrXV_619941_G_T_47212  ...  0.711397   \n",
       "3    0.0001   0.0001  120.009983      chrXV_622413_A_G_72894  ...  0.607320   \n",
       "4    0.0009   0.0003    7.932991  chrXV_624320_TTTTC_T_47455  ...  0.107987   \n",
       "..      ...      ...         ...                         ...  ...       ...   \n",
       "94   0.0001   0.0001   79.558526       chrXV_74126_C_T_49092  ...  0.345004   \n",
       "95   0.0001   0.0001   41.519803      chrIV_545729_C_A_11919  ...  0.329880   \n",
       "96   0.0001   0.0001   44.270147      chrIV_590540_T_C_12080  ...  0.301252   \n",
       "97   0.0001   0.0001   47.917048     chrXII_256508_A_G_49856  ...  0.282061   \n",
       "98   0.0001   0.0001   13.472733     chrXVI_181552_T_C_85082  ...  0.135279   \n",
       "\n",
       "      qtl_genes  num_qtl_genes  gene_id  gene_name   chrom   start     end  \\\n",
       "0   ['YOR153W']              1  YOR153W       PDR5   chrXV  619840  624375   \n",
       "1   ['YOR153W']              1  YOR153W       PDR5   chrXV  619840  624375   \n",
       "2   ['YOR153W']              1  YOR153W       PDR5   chrXV  619840  624375   \n",
       "3   ['YOR153W']              1  YOR153W       PDR5   chrXV  619840  624375   \n",
       "4   ['YOR153W']              1  YOR153W       PDR5   chrXV  619840  624375   \n",
       "..          ...            ...      ...        ...     ...     ...     ...   \n",
       "94  ['YOL130W']              1  YOL130W       ALR1   chrXV   74400   76979   \n",
       "95  ['YDR044W']              1  YDR044W      HEM13   chrIV  546642  547628   \n",
       "96  ['YDR072C']              1  YDR072C       IPT1   chrIV  589761  591344   \n",
       "97  ['YLR057W']              1  YLR057W       MNL2  chrXII  255306  257855   \n",
       "98  ['YPL193W']              1  YPL193W       RSA1  chrXVI  181403  182548   \n",
       "\n",
       "   strand                                        description  \n",
       "0       +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "1       +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "2       +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "3       +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "4       +  Plasma membrane ATP-binding cassette (ABC) tra...  \n",
       "..    ...                                                ...  \n",
       "94      +  Plasma membrane Mg(2%2B) transporter, expressi...  \n",
       "95      +  Coproporphyrinogen III oxidase, an oxygen requ...  \n",
       "96      -  Inositolphosphotransferase, involved in synthe...  \n",
       "97      +  Putative protein of unknown function; YLR050W ...  \n",
       "98      +  Protein involved in the assembly of 60S riboso...  \n",
       "\n",
       "[92 rows x 33 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read in GxE QTL table\n",
    "gxe_qtls = pd.read_csv(gxe_qtls_file, sep='\\t', index_col=0)\n",
    "gxe_qtls = gxe_qtls.rename(columns={'Systematic Name':'gene_id', 'Gene ID':'gene_name', 'Chr':'chrom'})#.set_index('gene_id')\n",
    "# add gene strand and description annotations to gxe QTLs\n",
    "gxe_qtls = gxe_qtls.merge(gene_info_df[['strand', 'description']], how='left', left_on='gene_id', right_index=True)\n",
    "\n",
    "# remove QTLs that involve blacklisted genes\n",
    "genes_to_exclude = ['YDL227C', # HO\n",
    "                    'YOR202W', # HIS3\n",
    "                    'YEL021W', # URA3\n",
    "                    'YCL018W', # LEU2\n",
    "                    'YBR115C', # LYS2\n",
    "                    'YBR020W', # GAL1\n",
    "                    'YDR009W', # GAL3\n",
    "                    'YPL248C', # GAL4\n",
    "                    'YBR018C', # GAL7\n",
    "                    'YBR019C', # GAL10\n",
    "                    'YML051W', # GAL80\n",
    "                    'YLR256W'] # HAP1\n",
    "                   \n",
    "genes_to_exclude = genes_to_exclude # note: ERG1 and ERG11 oligos will be duplicated in GxE set, has ~180 oligos\n",
    "gxe_qtls = gxe_qtls.query('~gene_id.isin(@genes_to_exclude)')\n",
    "\n",
    "\n",
    "# get list of 1,011 yeast strains used in QTL mapping\n",
    "yeast_strains_id = gxe_qtls['Strain ID of Parent 1 in Peter et al. 2018'].dropna().unique()\n",
    "yeast_strains_name_to_id = dict(zip(gxe_qtls['Parent 1'].apply(lambda x: x.split(' ')[0]),\n",
    "                                    gxe_qtls['Strain ID of Parent 1 in Peter et al. 2018']))\n",
    "display(gxe_qtls)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search GxE QTLs for variants in each gvcf\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16'}\n",
    "\n",
    "\n",
    "\n",
    "qtls_to_process = []\n",
    "for chrom in chromosome_dict.keys():\n",
    "    qtls_df = gxe_qtls.query('chrom==@chrom').sort_values('pos')\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict[chrom])\n",
    "    qtls_to_process.append((qtls_df, gvcf_file, chromosome_dict))\n",
    "    \n",
    "# consolidate candidate variants\n",
    "num_of_cores = min(len(os.sched_getaffinity(0)), len(qtls_to_process))\n",
    "with mp.Pool(num_of_cores) as pool:\n",
    "    results = pool.starmap(select_variants_from_qtls, qtls_to_process)\n",
    "results = pd.concat(results).reset_index(drop=True)\n",
    "\n",
    "# write to new vcf file\n",
    "gxe_variants_file = \"gxe_variants.vcf\"\n",
    "template = vcf.Reader(filename=qtls_to_process[0][1])\n",
    "# change chromosome naming convention to a single Roman numeral (follows ref genome, VEP formatting)\n",
    "chromosome_dict_rev = {v:k for k, v in chromosome_dict.items()} \n",
    "template.contigs = OrderedDict((chromosome_dict_rev[k][3:], v._replace(id=chromosome_dict_rev[k][3:])) for k, v in template.contigs.items())\n",
    "# remove 1011 genomes strain allele info\n",
    "template.samples = []\n",
    "\n",
    "output = vcf.Writer(open(gxe_variants_file, 'w'), template)\n",
    "for record in results:\n",
    "    # update chromosome naming\n",
    "    record.CHROM = chromosome_dict_rev[record.CHROM][3:]\n",
    "    # write to file\n",
    "    output.write_record(record)\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant at chrom V, pos 86643 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom V, pos 86648 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146273 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146275 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146276 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146413 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146415 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146416 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146429 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146430 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146483 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146492 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146524 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146534 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146538 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146549 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146555 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146556 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146560 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom VIII, pos 146562 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527563 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527564 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527584 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527586 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527628 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527672 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527710 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527719 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527743 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527746 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527761 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527763 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527765 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527905 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527910 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527918 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527931 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527936 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527947 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527952 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527962 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527965 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527974 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527977 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527979 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 527989 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528087 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528139 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528141 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528142 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528241 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528427 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528433 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528451 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528645 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 528774 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 534842 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 535356 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 535615 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 535633 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 535840 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 536011 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537442 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537448 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537451 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537460 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537463 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537466 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537499 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537508 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537511 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537533 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537565 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537568 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537578 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537581 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537586 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537612 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537613 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537633 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537670 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537672 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537694 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537706 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537709 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537713 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537715 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537722 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537724 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537738 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537750 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537808 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537811 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537817 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537828 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537829 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537865 skipped: No unique matches found in tested genome(s)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant at chrom IV, pos 537889 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537892 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537897 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537913 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537915 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 537934 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538008 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538009 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538024 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538030 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538063 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538066 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538069 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538081 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538082 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538084 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538119 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538120 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538141 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538144 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538149 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538151 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538152 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538153 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538162 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538163 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538165 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538166 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom IV, pos 538167 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIV, pos 374914 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIV, pos 374916 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIV, pos 374922 skipped: No unique matches found in tested genome(s)\n"
     ]
    }
   ],
   "source": [
    "# path to genome fasta files\n",
    "ref_genome_fasta = \"/home/users/rang/yeast/genomes/Saccharomyces_cerevisiae.R64-1-1.dna.chromosome.I.fa\"\n",
    "\n",
    "# path to input and output VCF files\n",
    "gxe_variants_file = \"gxe_variants.vcf\"\n",
    "gxe_variants_select_file = gxe_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# load genome files\n",
    "ref_genome = SeqIO.to_dict(SeqIO.parse(ref_genome_fasta, \"fasta\"))\n",
    "genomes_list = [ref_genome] # IMPORTANT: SET REFERENCE GENOME AS FIRST IN LIST\n",
    "\n",
    "\n",
    "threshold = 10 # threshold value for searching NGGs\n",
    "window = 61 # size of flanking sequences to search for unique hits\n",
    "\n",
    "# read in all candidate variants to filter\n",
    "gxe_variants = vcf.Reader(filename=gxe_variants_file)\n",
    "gxe_variants = [rec for rec in gxe_variants]\n",
    "# split variants into chunks for multiprocessing\n",
    "n = len(gxe_variants)//len(os.sched_getaffinity(0)) + 1\n",
    "gxe_variants = [gxe_variants[i:i+n] for i in range(0, len(gxe_variants), n)]\n",
    "\n",
    "# find targetable variants (TGT info field indicates which genomes variant is targetable in)\n",
    "to_process = [(rec_list, genomes_list, threshold, window) for rec_list in gxe_variants]\n",
    "with mp.Pool(len(os.sched_getaffinity(0))) as pool:\n",
    "    result = pool.starmap(find_targetable_variants, to_process) \n",
    "\n",
    "# write targetable variants to file\n",
    "header = vcf.Reader(filename=gxe_variants_file)\n",
    "# add \"TGT\" info field into VCF header\n",
    "header_info = namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "header.infos['TGT'] = header_info_tgt\n",
    "# write to file\n",
    "gxe_variants_select = vcf.Writer(open(gxe_variants_select_file, 'w'), header)\n",
    "to_write = [rec for rec_list in result for rec in rec_list]\n",
    "for rec in to_write:\n",
    "    gxe_variants_select.write_record(rec)\n",
    "gxe_variants_select.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select variants by allele frequency (currently disabled singleton/doubleton filter)\n",
    "### if multiple alternate alleles, SPLIT VARIANT ENTRY INTO SEPARATE ENTRIES in VCF (useful for oligo design pipeline)\n",
    "### assign unitl variant ID prefix and number (use format VAR_XXXXX)\n",
    "\n",
    "# candidate variants editable in all strains\n",
    "gxe_variants_select = vcf.Reader(filename=gxe_variants_select_file)\n",
    "\n",
    "# open output file\n",
    "gxe_variants_select_AFfilter_file = gxe_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "gxe_variants_select_AFfilter = vcf.Writer(open(gxe_variants_select_AFfilter_file, 'w'), gxe_variants_select)\n",
    "\n",
    "var_counter = {}\n",
    "for record in gxe_variants_select:        \n",
    "    # (optional step to skip singletons and doubletons -- DISABLED by setting ac<1)\n",
    "    if all([ac<1 for ac in record.INFO['AC']]):\n",
    "        continue\n",
    "    # if at least one alternate allele meets minimum count, write to VCF\n",
    "    else:\n",
    "        alt = record.ALT[:]\n",
    "        ac = record.INFO['AC'][:]\n",
    "        af = record.INFO['AF'][:]\n",
    "        mleac = record.INFO['MLEAC'][:]\n",
    "        mleaf = record.INFO['MLEAF'][:]\n",
    "        \n",
    "        # split each alt allele into its own variant entry (useful if a locus has multiple possible alleles)\n",
    "        for i in range(len(alt)):\n",
    "            record.ALT = [alt[i]]\n",
    "            record.INFO['AC'] = [ac[i]]\n",
    "            record.INFO['AF'] = [af[i]]\n",
    "            record.INFO['MLEAC'] = [mleac[i]]\n",
    "            record.INFO['MLEAF'] = [mleaf[i]]\n",
    "\n",
    "            # inspect TGT field and assign variant ID prefix\n",
    "            ### if tgt field contains all 1 (i.e. targetable by all queried genomes), assign \"GXE\"\n",
    "            ### otherwise, skip variant\n",
    "            if all([i==1 for i in record.INFO['TGT']]):\n",
    "                var_id_prefix = \"GXE\"\n",
    "            else:\n",
    "                print('variant does not pass TGT check: skipping')\n",
    "                continue\n",
    "#                 binary_str = ''.join([str(x) for x in record.INFO['TGT']])\n",
    "#                 hex_digit = hex(int(binary_str, 2))[2:].upper()\n",
    "#                 var_id_prefix = \"EG{}\".format(hex_digit)\n",
    "\n",
    "            # add to counter\n",
    "            if var_id_prefix not in var_counter:\n",
    "                var_counter[var_id_prefix] = 1\n",
    "            else:\n",
    "                var_counter[var_id_prefix] += 1\n",
    "\n",
    "            # add unique variant ID (follow format VAR_XXXXX)\n",
    "            record.ID='{}_{:05d}'.format(var_id_prefix, var_counter[var_id_prefix])\n",
    "\n",
    "            # write to output\n",
    "            gxe_variants_select_AFfilter.write_record(record)\n",
    "            \n",
    "gxe_variants_select_AFfilter.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit gxe_variants_select_AFfilter VCF to VEP to get annotations, and download VEP output in TXT format\n",
    "http://uswest.ensembl.org/Saccharomyces_cerevisiae/Tools/VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parsing entry: 5000\n",
      "parsing entry: 10000\n",
      "parsing entry: 15000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00001</th>\n",
       "      <td>I</td>\n",
       "      <td>192136</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.000991</td>\n",
       "      <td>483</td>\n",
       "      <td>4404|3325|201</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>cgG/cgA</td>\n",
       "      <td>Low_complexity_(Seg):seg,PANTHER:PTHR22589,PAN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00002</th>\n",
       "      <td>I</td>\n",
       "      <td>192158</td>\n",
       "      <td>CCCGCTGCCAGGA</td>\n",
       "      <td>C</td>\n",
       "      <td>114</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.056000</td>\n",
       "      <td>449</td>\n",
       "      <td>4427|3348|167</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00003</th>\n",
       "      <td>I</td>\n",
       "      <td>192162</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.000990</td>\n",
       "      <td>457</td>\n",
       "      <td>4430|3351|175</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gCt/gTt</td>\n",
       "      <td>Low_complexity_(Seg):seg,PANTHER:PTHR22589,PAN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00004</th>\n",
       "      <td>I</td>\n",
       "      <td>192164</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002475</td>\n",
       "      <td>455</td>\n",
       "      <td>4432|3353|173</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gcc/Ccc</td>\n",
       "      <td>Low_complexity_(Seg):seg,PANTHER:PTHR22589,PAN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00005</th>\n",
       "      <td>I</td>\n",
       "      <td>192167</td>\n",
       "      <td>AGG</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>450</td>\n",
       "      <td>4436|3357|168</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17204</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375568</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>16</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.007913</td>\n",
       "      <td>4905</td>\n",
       "      <td>3552|1723|399|2647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Cat/Gat</td>\n",
       "      <td>PIRSF:PIRSF000350,PANTHER:PTHR42737,Superfamil...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17205</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375603</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>4940</td>\n",
       "      <td>3517|1688|434|2682</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>ggG/ggA</td>\n",
       "      <td>Gene3D:3.50.50.60,Pfam:PF07992,PIRSF:PIRSF0003...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17206</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375623</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>4960</td>\n",
       "      <td>3497|1668|454|2702</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>aGa/aAa</td>\n",
       "      <td>Gene3D:3.50.50.60,Pfam:PF07992,PIRSF:PIRSF0003...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17207</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375637</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>4974</td>\n",
       "      <td>3483|1654|468|2716</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Ggt/Agt</td>\n",
       "      <td>Gene3D:3.50.50.60,Pfam:PF07992,PIRSF:PIRSF0003...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17208</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375669</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001978</td>\n",
       "      <td></td>\n",
       "      <td>3451|1622|500|2748</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>gcT/gcC</td>\n",
       "      <td>Gene3D:3.50.50.60,Pfam:PF07992,PIRSF:PIRSF0003...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17208 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS            REF ALT   AC    AN        AF  \\\n",
       "var_id                                                            \n",
       "GXE_00001     I  192136              G   A    2  2018  0.000991   \n",
       "GXE_00002     I  192158  CCCGCTGCCAGGA   C  114  2020  0.056000   \n",
       "GXE_00003     I  192162              C   T    2  2020  0.000990   \n",
       "GXE_00004     I  192164              G   C    5  2020  0.002475   \n",
       "GXE_00005     I  192167            AGG   A    1  2020  0.000495   \n",
       "...         ...     ...            ...  ..  ...   ...       ...   \n",
       "GXE_17204   XVI  375568              C   G   16  2022  0.007913   \n",
       "GXE_17205   XVI  375603              G   A    4  2022  0.001978   \n",
       "GXE_17206   XVI  375623              G   A    2  2022  0.000989   \n",
       "GXE_17207   XVI  375637              G   A    6  2022  0.002967   \n",
       "GXE_17208   XVI  375669              T   C    4  2022  0.001978   \n",
       "\n",
       "          upstream_distance_str downstream_distance_str              region  \\\n",
       "var_id                                                                        \n",
       "GXE_00001                   483           4404|3325|201  synonymous_variant   \n",
       "GXE_00002                   449           4427|3348|167    inframe_deletion   \n",
       "GXE_00003                   457           4430|3351|175    missense_variant   \n",
       "GXE_00004                   455           4432|3353|173    missense_variant   \n",
       "GXE_00005                   450           4436|3357|168  frameshift_variant   \n",
       "...                         ...                     ...                 ...   \n",
       "GXE_17204                  4905      3552|1723|399|2647    missense_variant   \n",
       "GXE_17205                  4940      3517|1688|434|2682  synonymous_variant   \n",
       "GXE_17206                  4960      3497|1668|454|2702    missense_variant   \n",
       "GXE_17207                  4974      3483|1654|468|2716    missense_variant   \n",
       "GXE_17208                            3451|1622|500|2748  synonymous_variant   \n",
       "\n",
       "           ...   Codons                                            DOMAINS  \\\n",
       "var_id     ...                                                               \n",
       "GXE_00001  ...  cgG/cgA  Low_complexity_(Seg):seg,PANTHER:PTHR22589,PAN...   \n",
       "GXE_00002  ...      NaN                                                NaN   \n",
       "GXE_00003  ...  gCt/gTt  Low_complexity_(Seg):seg,PANTHER:PTHR22589,PAN...   \n",
       "GXE_00004  ...  Gcc/Ccc  Low_complexity_(Seg):seg,PANTHER:PTHR22589,PAN...   \n",
       "GXE_00005  ...      NaN                                                NaN   \n",
       "...        ...      ...                                                ...   \n",
       "GXE_17204  ...  Cat/Gat  PIRSF:PIRSF000350,PANTHER:PTHR42737,Superfamil...   \n",
       "GXE_17205  ...  ggG/ggA  Gene3D:3.50.50.60,Pfam:PF07992,PIRSF:PIRSF0003...   \n",
       "GXE_17206  ...  aGa/aAa  Gene3D:3.50.50.60,Pfam:PF07992,PIRSF:PIRSF0003...   \n",
       "GXE_17207  ...  Ggt/Agt  Gene3D:3.50.50.60,Pfam:PF07992,PIRSF:PIRSF0003...   \n",
       "GXE_17208  ...  gcT/gcC  Gene3D:3.50.50.60,Pfam:PF07992,PIRSF:PIRSF0003...   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXE_00001                     NaN                   NaN   \n",
       "GXE_00002                     NaN                   NaN   \n",
       "GXE_00003                     NaN                   NaN   \n",
       "GXE_00004                     NaN                   NaN   \n",
       "GXE_00005                     NaN                   NaN   \n",
       "...                           ...                   ...   \n",
       "GXE_17204                     NaN                   NaN   \n",
       "GXE_17205                     NaN                   NaN   \n",
       "GXE_17206                     NaN                   NaN   \n",
       "GXE_17207                     NaN                   NaN   \n",
       "GXE_17208                     NaN                   NaN   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "GXE_00001                      NaN                    NaN   \n",
       "GXE_00002                      NaN                    NaN   \n",
       "GXE_00003                      NaN                    NaN   \n",
       "GXE_00004                      NaN                    NaN   \n",
       "GXE_00005                      NaN                    NaN   \n",
       "...                            ...                    ...   \n",
       "GXE_17204                      NaN                    NaN   \n",
       "GXE_17205                      NaN                    NaN   \n",
       "GXE_17206                      NaN                    NaN   \n",
       "GXE_17207                      NaN                    NaN   \n",
       "GXE_17208                      NaN                    NaN   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXE_00001                     NaN                   NaN   \n",
       "GXE_00002                     NaN                   NaN   \n",
       "GXE_00003                     NaN                   NaN   \n",
       "GXE_00004                     NaN                   NaN   \n",
       "GXE_00005                     NaN                   NaN   \n",
       "...                           ...                   ...   \n",
       "GXE_17204                     NaN                   NaN   \n",
       "GXE_17205                     NaN                   NaN   \n",
       "GXE_17206                     NaN                   NaN   \n",
       "GXE_17207                     NaN                   NaN   \n",
       "GXE_17208                     NaN                   NaN   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "GXE_00001                      NaN                    NaN  \n",
       "GXE_00002                      NaN                    NaN  \n",
       "GXE_00003                      NaN                    NaN  \n",
       "GXE_00004                      NaN                    NaN  \n",
       "GXE_00005                      NaN                    NaN  \n",
       "...                            ...                    ...  \n",
       "GXE_17204                      NaN                    NaN  \n",
       "GXE_17205                      NaN                    NaN  \n",
       "GXE_17206                      NaN                    NaN  \n",
       "GXE_17207                      NaN                    NaN  \n",
       "GXE_17208                      NaN                    NaN  \n",
       "\n",
       "[17208 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00017</th>\n",
       "      <td>I</td>\n",
       "      <td>192257</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2006</td>\n",
       "      <td>0.003988</td>\n",
       "      <td>362</td>\n",
       "      <td>4525|3446|80|1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAT1</td>\n",
       "      <td>YAR035W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>200.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00025</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.000993</td>\n",
       "      <td>8|181</td>\n",
       "      <td>4693|3614|169</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>8.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17195</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375403</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>99|4740</td>\n",
       "      <td>3717|1888|234|2482</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>234.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17196</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375403</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>99|4740</td>\n",
       "      <td>3717|1888|234|2482</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>234.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17197</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375405</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.005935</td>\n",
       "      <td>97|4742</td>\n",
       "      <td>3715|1886|236|2484</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>236.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17198</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375411</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>33</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.016000</td>\n",
       "      <td>91|4748</td>\n",
       "      <td>3709|1880|242|2490</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>242.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17199</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375464</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>38|4801</td>\n",
       "      <td>3656|1827|295|2543</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>295.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15390 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT   AC    AN  \\\n",
       "var_id                                                                 \n",
       "GXE_00017     I  192257                G                A    8  2006   \n",
       "GXE_00023     I  192419                C                A    6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT   39  2014   \n",
       "GXE_00025     I  192424  GTTTGGATTACCTCT                G    2  2014   \n",
       "GXE_00026     I  192427                T                C  902  2000   \n",
       "...         ...     ...              ...              ...  ...   ...   \n",
       "GXE_17195   XVI  375403                C                T   26  2022   \n",
       "GXE_17196   XVI  375403                C                G    2  2022   \n",
       "GXE_17197   XVI  375405                C                T   12  2022   \n",
       "GXE_17198   XVI  375411                G                T   33  2022   \n",
       "GXE_17199   XVI  375464                G                A    2  2022   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00017  0.003988                   362          4525|3446|80|1   \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00025  0.000993                 8|181           4693|3614|169   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "...             ...                   ...                     ...   \n",
       "GXE_17195  0.013000               99|4740      3717|1888|234|2482   \n",
       "GXE_17196  0.000989               99|4740      3717|1888|234|2482   \n",
       "GXE_17197  0.005935               97|4742      3715|1886|236|2484   \n",
       "GXE_17198  0.016000               91|4748      3709|1880|242|2490   \n",
       "GXE_17199  0.000989               38|4801      3656|1827|295|2543   \n",
       "\n",
       "                            region  ... Codons DOMAINS  \\\n",
       "var_id                              ...                  \n",
       "GXE_00017               intergenic  ...    NaN     NaN   \n",
       "GXE_00023   bidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_00024   bidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_00025   bidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_00026   bidirectional_promoter  ...    NaN     NaN   \n",
       "...                            ...  ...    ...     ...   \n",
       "GXE_17195  unidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_17196  unidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_17197  unidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_17198  unidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_17199  unidirectional_promoter  ...    NaN     NaN   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXE_00017                    YAT1               YAR035W   \n",
       "GXE_00023                     NaN             YAR035C-A   \n",
       "GXE_00024                     NaN             YAR035C-A   \n",
       "GXE_00025                     NaN             YAR035C-A   \n",
       "GXE_00026                     NaN             YAR035C-A   \n",
       "...                           ...                   ...   \n",
       "GXE_17195                    SSU1               YPL092W   \n",
       "GXE_17196                    SSU1               YPL092W   \n",
       "GXE_17197                    SSU1               YPL092W   \n",
       "GXE_17198                    SSU1               YPL092W   \n",
       "GXE_17199                    SSU1               YPL092W   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "GXE_00017  downstream_gene_variant                    1.0   \n",
       "GXE_00023    upstream_gene_variant                    2.0   \n",
       "GXE_00024    upstream_gene_variant                    7.0   \n",
       "GXE_00025    upstream_gene_variant                    8.0   \n",
       "GXE_00026    upstream_gene_variant                   10.0   \n",
       "...                            ...                    ...   \n",
       "GXE_17195  downstream_gene_variant                  234.0   \n",
       "GXE_17196  downstream_gene_variant                  234.0   \n",
       "GXE_17197  downstream_gene_variant                  236.0   \n",
       "GXE_17198  downstream_gene_variant                  242.0   \n",
       "GXE_17199  downstream_gene_variant                  295.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXE_00017                     NaN             YAR035C-A   \n",
       "GXE_00023                    SWH1               YAR042W   \n",
       "GXE_00024                    SWH1               YAR042W   \n",
       "GXE_00025                    SWH1               YAR042W   \n",
       "GXE_00026                    SWH1               YAR042W   \n",
       "...                           ...                   ...   \n",
       "GXE_17195                    GLR1               YPL091W   \n",
       "GXE_17196                    GLR1               YPL091W   \n",
       "GXE_17197                    GLR1               YPL091W   \n",
       "GXE_17198                    GLR1               YPL091W   \n",
       "GXE_17199                    GLR1               YPL091W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "GXE_00017  downstream_gene_variant                   80.0  \n",
       "GXE_00023    upstream_gene_variant                  200.0  \n",
       "GXE_00024    upstream_gene_variant                  181.0  \n",
       "GXE_00025    upstream_gene_variant                  181.0  \n",
       "GXE_00026    upstream_gene_variant                  192.0  \n",
       "...                            ...                    ...  \n",
       "GXE_17195    upstream_gene_variant                   99.0  \n",
       "GXE_17196    upstream_gene_variant                   99.0  \n",
       "GXE_17197    upstream_gene_variant                   97.0  \n",
       "GXE_17198    upstream_gene_variant                   91.0  \n",
       "GXE_17199    upstream_gene_variant                   38.0  \n",
       "\n",
       "[15390 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add annotations to variants and remove variants that fall in non-gxe genes.\n",
    "\n",
    "gxe_variants_file = \"gxe_variants.vcf\"\n",
    "gxe_variants_select_file = gxe_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# get VEP annotations and remove variants that fall in non-gxe pathway genes\n",
    "gxe_variants_select_AFfilter_file = gxe_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "vep_output_filename = gxe_variants_select_AFfilter_file.replace('.vcf', '_VEPoutput.txt')\n",
    "variants_annotated_filename = vep_output_filename.replace('_VEPoutput.txt', '_annotated.txt')\n",
    "variants_annotated_df = annotate_variants_by_VEPoutput(gxe_variants_select_AFfilter_file, \n",
    "                                                       vep_output_filename, \n",
    "                                                       gff_file, \n",
    "                                                       variants_annotated_filename, \n",
    "                                                       id_colname='var_id')\n",
    "display(variants_annotated_df)\n",
    "\n",
    "# annotated file can be read in directly if annotation function has been run once before\n",
    "variants_annotated_df = pd.read_csv(variants_annotated_filename, sep='\\t', index_col=0)\n",
    "display(variants_annotated_df)\n",
    "\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "gxe_design_oligos = variants_annotated_df.query('(Gene_ID.isin(@gxe_qtls.gene_id.unique()) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(gxe_design_oligos)\n",
    "\n",
    "# write variants to gxe_design_oligos VCF as a preliminary list to design oligos for\n",
    "gxe_design_oligos_vcf_file = gxe_variants_file.replace('.vcf', '_design_oligos_initial.vcf')\n",
    "\n",
    "gxe_variants_select_AFfilter = vcf.Reader(filename=gxe_variants_select_AFfilter_file)\n",
    "gxe_design_oligos_vcf = vcf.Writer(open(gxe_design_oligos_vcf_file, 'w'), gxe_variants_select_AFfilter)\n",
    "for record in gxe_variants_select_AFfilter:\n",
    "    if record.ID in gxe_design_oligos.index:\n",
    "        gxe_design_oligos_vcf.write_record(record)\n",
    "gxe_design_oligos_vcf.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design oligos for initial list, then filter down to variants with at least 2 oligos designed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00017</th>\n",
       "      <td>I</td>\n",
       "      <td>192257</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2006</td>\n",
       "      <td>0.003988</td>\n",
       "      <td>362</td>\n",
       "      <td>4525|3446|80|1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAT1</td>\n",
       "      <td>YAR035W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>200.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00025</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.000993</td>\n",
       "      <td>8|181</td>\n",
       "      <td>4693|3614|169</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>8.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17195</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375403</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>26</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>99|4740</td>\n",
       "      <td>3717|1888|234|2482</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>234.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17196</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375403</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>99|4740</td>\n",
       "      <td>3717|1888|234|2482</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>234.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17197</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375405</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.005935</td>\n",
       "      <td>97|4742</td>\n",
       "      <td>3715|1886|236|2484</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>236.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17198</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375411</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>33</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.016000</td>\n",
       "      <td>91|4748</td>\n",
       "      <td>3709|1880|242|2490</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>242.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17199</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375464</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>38|4801</td>\n",
       "      <td>3656|1827|295|2543</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>295.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15390 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT   AC    AN  \\\n",
       "var_id                                                                 \n",
       "GXE_00017     I  192257                G                A    8  2006   \n",
       "GXE_00023     I  192419                C                A    6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT   39  2014   \n",
       "GXE_00025     I  192424  GTTTGGATTACCTCT                G    2  2014   \n",
       "GXE_00026     I  192427                T                C  902  2000   \n",
       "...         ...     ...              ...              ...  ...   ...   \n",
       "GXE_17195   XVI  375403                C                T   26  2022   \n",
       "GXE_17196   XVI  375403                C                G    2  2022   \n",
       "GXE_17197   XVI  375405                C                T   12  2022   \n",
       "GXE_17198   XVI  375411                G                T   33  2022   \n",
       "GXE_17199   XVI  375464                G                A    2  2022   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00017  0.003988                   362          4525|3446|80|1   \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00025  0.000993                 8|181           4693|3614|169   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "...             ...                   ...                     ...   \n",
       "GXE_17195  0.013000               99|4740      3717|1888|234|2482   \n",
       "GXE_17196  0.000989               99|4740      3717|1888|234|2482   \n",
       "GXE_17197  0.005935               97|4742      3715|1886|236|2484   \n",
       "GXE_17198  0.016000               91|4748      3709|1880|242|2490   \n",
       "GXE_17199  0.000989               38|4801      3656|1827|295|2543   \n",
       "\n",
       "                            region  ... Codons DOMAINS  \\\n",
       "var_id                              ...                  \n",
       "GXE_00017               intergenic  ...    NaN     NaN   \n",
       "GXE_00023   bidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_00024   bidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_00025   bidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_00026   bidirectional_promoter  ...    NaN     NaN   \n",
       "...                            ...  ...    ...     ...   \n",
       "GXE_17195  unidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_17196  unidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_17197  unidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_17198  unidirectional_promoter  ...    NaN     NaN   \n",
       "GXE_17199  unidirectional_promoter  ...    NaN     NaN   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXE_00017                    YAT1               YAR035W   \n",
       "GXE_00023                     NaN             YAR035C-A   \n",
       "GXE_00024                     NaN             YAR035C-A   \n",
       "GXE_00025                     NaN             YAR035C-A   \n",
       "GXE_00026                     NaN             YAR035C-A   \n",
       "...                           ...                   ...   \n",
       "GXE_17195                    SSU1               YPL092W   \n",
       "GXE_17196                    SSU1               YPL092W   \n",
       "GXE_17197                    SSU1               YPL092W   \n",
       "GXE_17198                    SSU1               YPL092W   \n",
       "GXE_17199                    SSU1               YPL092W   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "GXE_00017  downstream_gene_variant                    1.0   \n",
       "GXE_00023    upstream_gene_variant                    2.0   \n",
       "GXE_00024    upstream_gene_variant                    7.0   \n",
       "GXE_00025    upstream_gene_variant                    8.0   \n",
       "GXE_00026    upstream_gene_variant                   10.0   \n",
       "...                            ...                    ...   \n",
       "GXE_17195  downstream_gene_variant                  234.0   \n",
       "GXE_17196  downstream_gene_variant                  234.0   \n",
       "GXE_17197  downstream_gene_variant                  236.0   \n",
       "GXE_17198  downstream_gene_variant                  242.0   \n",
       "GXE_17199  downstream_gene_variant                  295.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "GXE_00017                     NaN             YAR035C-A   \n",
       "GXE_00023                    SWH1               YAR042W   \n",
       "GXE_00024                    SWH1               YAR042W   \n",
       "GXE_00025                    SWH1               YAR042W   \n",
       "GXE_00026                    SWH1               YAR042W   \n",
       "...                           ...                   ...   \n",
       "GXE_17195                    GLR1               YPL091W   \n",
       "GXE_17196                    GLR1               YPL091W   \n",
       "GXE_17197                    GLR1               YPL091W   \n",
       "GXE_17198                    GLR1               YPL091W   \n",
       "GXE_17199                    GLR1               YPL091W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "GXE_00017  downstream_gene_variant                   80.0  \n",
       "GXE_00023    upstream_gene_variant                  200.0  \n",
       "GXE_00024    upstream_gene_variant                  181.0  \n",
       "GXE_00025    upstream_gene_variant                  181.0  \n",
       "GXE_00026    upstream_gene_variant                  192.0  \n",
       "...                            ...                    ...  \n",
       "GXE_17195    upstream_gene_variant                   99.0  \n",
       "GXE_17196    upstream_gene_variant                   99.0  \n",
       "GXE_17197    upstream_gene_variant                   97.0  \n",
       "GXE_17198    upstream_gene_variant                   91.0  \n",
       "GXE_17199    upstream_gene_variant                   38.0  \n",
       "\n",
       "[15390 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>200.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00025</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.000993</td>\n",
       "      <td>8|181</td>\n",
       "      <td>4693|3614|169</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>8.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>181.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>192.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00027</th>\n",
       "      <td>I</td>\n",
       "      <td>192429</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1506</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.754000</td>\n",
       "      <td>12|190</td>\n",
       "      <td>4697|3618|173</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAR035C-A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>12.0</td>\n",
       "      <td>SWH1</td>\n",
       "      <td>YAR042W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>190.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17168</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375080</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>422|4417</td>\n",
       "      <td>4040|2211|2159</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>PANTHER:PTHR31686,PANTHER:PTHR31686:SF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17188</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375331</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>171|4668</td>\n",
       "      <td>3789|1960|162|2410</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>162.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>171.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17189</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375333</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>168|4670</td>\n",
       "      <td>3786|1957|164|2412</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>164.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>168.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17192</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375357</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>145|4694</td>\n",
       "      <td>3763|1934|188|2436</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>188.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>145.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17193</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375360</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>142|4697</td>\n",
       "      <td>3760|1931|191|2439</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSU1</td>\n",
       "      <td>YPL092W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>191.0</td>\n",
       "      <td>GLR1</td>\n",
       "      <td>YPL091W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>142.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6480 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT    AC    AN  \\\n",
       "var_id                                                                  \n",
       "GXE_00023     I  192419                C                A     6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT    39  2014   \n",
       "GXE_00025     I  192424  GTTTGGATTACCTCT                G     2  2014   \n",
       "GXE_00026     I  192427                T                C   902  2000   \n",
       "GXE_00027     I  192429                G                C  1506  1998   \n",
       "...         ...     ...              ...              ...   ...   ...   \n",
       "GXE_17168   XVI  375080                G                A    31  2022   \n",
       "GXE_17188   XVI  375331                C                A     8  2020   \n",
       "GXE_17189   XVI  375333                A               AT     8  2022   \n",
       "GXE_17192   XVI  375357                C                T     2  2022   \n",
       "GXE_17193   XVI  375360                T                G     1  2022   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00025  0.000993                 8|181           4693|3614|169   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "GXE_00027  0.754000                12|190           4697|3618|173   \n",
       "...             ...                   ...                     ...   \n",
       "GXE_17168  0.015000              422|4417          4040|2211|2159   \n",
       "GXE_17188  0.003960              171|4668      3789|1960|162|2410   \n",
       "GXE_17189  0.003956              168|4670      3786|1957|164|2412   \n",
       "GXE_17192  0.000989              145|4694      3763|1934|188|2436   \n",
       "GXE_17193  0.000495              142|4697      3760|1931|191|2439   \n",
       "\n",
       "                            region  ...  \\\n",
       "var_id                              ...   \n",
       "GXE_00023   bidirectional_promoter  ...   \n",
       "GXE_00024   bidirectional_promoter  ...   \n",
       "GXE_00025   bidirectional_promoter  ...   \n",
       "GXE_00026   bidirectional_promoter  ...   \n",
       "GXE_00027   bidirectional_promoter  ...   \n",
       "...                            ...  ...   \n",
       "GXE_17168         missense_variant  ...   \n",
       "GXE_17188  unidirectional_promoter  ...   \n",
       "GXE_17189  unidirectional_promoter  ...   \n",
       "GXE_17192  unidirectional_promoter  ...   \n",
       "GXE_17193  unidirectional_promoter  ...   \n",
       "\n",
       "                                           DOMAINS closest_gene1_Gene_Name  \\\n",
       "var_id                                                                       \n",
       "GXE_00023                                      NaN                     NaN   \n",
       "GXE_00024                                      NaN                     NaN   \n",
       "GXE_00025                                      NaN                     NaN   \n",
       "GXE_00026                                      NaN                     NaN   \n",
       "GXE_00027                                      NaN                     NaN   \n",
       "...                                            ...                     ...   \n",
       "GXE_17168  PANTHER:PTHR31686,PANTHER:PTHR31686:SF1                     NaN   \n",
       "GXE_17188                                      NaN                    SSU1   \n",
       "GXE_17189                                      NaN                    SSU1   \n",
       "GXE_17192                                      NaN                    SSU1   \n",
       "GXE_17193                                      NaN                    SSU1   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "GXE_00023             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00024             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00025             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00026             YAR035C-A    upstream_gene_variant   \n",
       "GXE_00027             YAR035C-A    upstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "GXE_17168                   NaN                      NaN   \n",
       "GXE_17188               YPL092W  downstream_gene_variant   \n",
       "GXE_17189               YPL092W  downstream_gene_variant   \n",
       "GXE_17192               YPL092W  downstream_gene_variant   \n",
       "GXE_17193               YPL092W  downstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "GXE_00023                    2.0                    SWH1   \n",
       "GXE_00024                    7.0                    SWH1   \n",
       "GXE_00025                    8.0                    SWH1   \n",
       "GXE_00026                   10.0                    SWH1   \n",
       "GXE_00027                   12.0                    SWH1   \n",
       "...                          ...                     ...   \n",
       "GXE_17168                    NaN                     NaN   \n",
       "GXE_17188                  162.0                    GLR1   \n",
       "GXE_17189                  164.0                    GLR1   \n",
       "GXE_17192                  188.0                    GLR1   \n",
       "GXE_17193                  191.0                    GLR1   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "GXE_00023               YAR042W    upstream_gene_variant   \n",
       "GXE_00024               YAR042W    upstream_gene_variant   \n",
       "GXE_00025               YAR042W    upstream_gene_variant   \n",
       "GXE_00026               YAR042W    upstream_gene_variant   \n",
       "GXE_00027               YAR042W    upstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "GXE_17168                   NaN                      NaN   \n",
       "GXE_17188               YPL091W    upstream_gene_variant   \n",
       "GXE_17189               YPL091W    upstream_gene_variant   \n",
       "GXE_17192               YPL091W    upstream_gene_variant   \n",
       "GXE_17193               YPL091W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance num_of_oligos  \n",
       "var_id                                          \n",
       "GXE_00023                  200.0           3.0  \n",
       "GXE_00024                  181.0           3.0  \n",
       "GXE_00025                  181.0           3.0  \n",
       "GXE_00026                  192.0           2.0  \n",
       "GXE_00027                  190.0           3.0  \n",
       "...                          ...           ...  \n",
       "GXE_17168                    NaN           2.0  \n",
       "GXE_17188                  171.0           2.0  \n",
       "GXE_17189                  168.0           3.0  \n",
       "GXE_17192                  145.0           2.0  \n",
       "GXE_17193                  142.0           2.0  \n",
       "\n",
       "[6480 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gxe_oligo_file = os.path.expanduser(\"~/crispey3/initial_design/Output/all_SNPs_gxe_initial_GG_9bp_OLIGO.tab\")\n",
    "gxe_oligo_df = pd.read_csv(gxe_oligo_file, sep='\\t')\n",
    "\n",
    "gxe_design_oligos = pd.read_csv(\"/home/users/rang/scratch/yeast/GxE/gxe_variants_select_AFfilter_annotated.txt\", sep='\\t', index_col=0)\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "gxe_design_oligos = gxe_design_oligos.query('(Gene_ID.isin(@gxe_qtls.gene_id.unique()) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(gxe_design_oligos)\n",
    "\n",
    "\n",
    "gxe_design_oligos['num_of_oligos'] = gxe_oligo_df.groupby('var_id').apply(len)\n",
    "gxe_oligos_final = gxe_design_oligos.query('num_of_oligos>1').copy()\n",
    "display(gxe_oligos_final)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove singletons & doubletons, but preserve those that appear in at least one of the 15 QTL parental strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500 variants processed\n",
      "1000 variants processed\n",
      "1500 variants processed\n",
      "2000 variants processed\n",
      "2500 variants processed\n",
      "3000 variants processed\n",
      "3500 variants processed\n",
      "4000 variants processed\n",
      "4500 variants processed\n",
      "5000 variants processed\n",
      "5500 variants processed\n",
      "6000 variants processed\n"
     ]
    }
   ],
   "source": [
    "# search GVCF and add allele information for parent strains in QTL crosses\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16'}\n",
    "\n",
    "gxe_variants_file = \"gxe_variants.vcf\"\n",
    "i=0\n",
    "for chromosome, df in gxe_oligos_final.groupby('CHROM'):\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict['chr'+chromosome])\n",
    "    allVars = vcf.Reader(filename=gvcf_file)\n",
    "    \n",
    "    for v, row in df.iterrows():\n",
    "        # find variant in 1011 genomes GVCF\n",
    "        retrieve = allVars.fetch(chrom=chromosome_dict['chr'+row.CHROM], \n",
    "                                 start=row.POS-1,\n",
    "                                 end=row.POS)\n",
    "        # select record that matches \n",
    "        for r in retrieve:\n",
    "            if (r.POS==row.POS) & (r.REF==row.REF):\n",
    "                record = r\n",
    "                break\n",
    "                \n",
    "        # get all alternate alleles at this position\n",
    "        gxe_oligos_final.loc[v, 'all_alt_alleles'] = ','.join([str(a) for a in record.ALT])\n",
    "        # get allele calls for strains in yeast_strains_id\n",
    "        for s in record.samples:\n",
    "            if (s.sample in yeast_strains_id) and (None not in s.gt_alleles):\n",
    "                gxe_oligos_final.loc[v, s.sample] = '/'.join(s.gt_alleles)\n",
    "        \n",
    "        i+=1\n",
    "        if i % 500 == 0:\n",
    "            print(i, \"variants processed\")\n",
    "        \n",
    "# write updated gxe_oligos_final table\n",
    "gxe_oligos_final.to_csv(gxe_variants_file.replace('.vcf', '_with_parents.txt'), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACI</th>\n",
       "      <th>ACQ</th>\n",
       "      <th>ACS</th>\n",
       "      <th>ADE</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADG</th>\n",
       "      <th>ADI</th>\n",
       "      <th>ADR</th>\n",
       "      <th>SACE_MAA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00027</th>\n",
       "      <td>I</td>\n",
       "      <td>192429</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1506</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.754000</td>\n",
       "      <td>12|190</td>\n",
       "      <td>4697|3618|173</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00033</th>\n",
       "      <td>I</td>\n",
       "      <td>192466</td>\n",
       "      <td>AAAGGG</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001485</td>\n",
       "      <td>50|148</td>\n",
       "      <td>4735|3656|211</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17166</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375077</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001484</td>\n",
       "      <td>425|4414</td>\n",
       "      <td>4043|2214|2156</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17167</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375078</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>424|4415</td>\n",
       "      <td>4042|2213|2157</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17168</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375080</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>422|4417</td>\n",
       "      <td>4040|2211|2159</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17188</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375331</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>171|4668</td>\n",
       "      <td>3789|1960|162|2410</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17189</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375333</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>168|4670</td>\n",
       "      <td>3786|1957|164|2412</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4384 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT    AC    AN  \\\n",
       "var_id                                                                  \n",
       "GXE_00023     I  192419                C                A     6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT    39  2014   \n",
       "GXE_00026     I  192427                T                C   902  2000   \n",
       "GXE_00027     I  192429                G                C  1506  1998   \n",
       "GXE_00033     I  192466           AAAGGG                A     3  2020   \n",
       "...         ...     ...              ...              ...   ...   ...   \n",
       "GXE_17166   XVI  375077                T                C     3  2022   \n",
       "GXE_17167   XVI  375078                C                A     8  2020   \n",
       "GXE_17168   XVI  375080                G                A    31  2022   \n",
       "GXE_17188   XVI  375331                C                A     8  2020   \n",
       "GXE_17189   XVI  375333                A               AT     8  2022   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "GXE_00027  0.754000                12|190           4697|3618|173   \n",
       "GXE_00033  0.001485                50|148           4735|3656|211   \n",
       "...             ...                   ...                     ...   \n",
       "GXE_17166  0.001484              425|4414          4043|2214|2156   \n",
       "GXE_17167  0.003960              424|4415          4042|2213|2157   \n",
       "GXE_17168  0.015000              422|4417          4040|2211|2159   \n",
       "GXE_17188  0.003960              171|4668      3789|1960|162|2410   \n",
       "GXE_17189  0.003956              168|4670      3786|1957|164|2412   \n",
       "\n",
       "                            region  ...  ACG  ACI  ACQ  ACS  ADE  ADF  ADG  \\\n",
       "var_id                              ...                                      \n",
       "GXE_00023   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_00024   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  NaN  0/0  0/0   \n",
       "GXE_00026   bidirectional_promoter  ...  1/1  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_00027   bidirectional_promoter  ...  1/1  1/1  1/1  1/1  NaN  1/1  1/1   \n",
       "GXE_00033   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "...                            ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "GXE_17166         missense_variant  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17167              stop_gained  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17168         missense_variant  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17188  unidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17189  unidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "\n",
       "           ADI  ADR SACE_MAA  \n",
       "var_id                        \n",
       "GXE_00023  0/0  0/0      0/0  \n",
       "GXE_00024  0/0  0/0      0/0  \n",
       "GXE_00026  1/1  1/1      0/0  \n",
       "GXE_00027  1/1  1/1      0/0  \n",
       "GXE_00033  0/0  0/0      0/0  \n",
       "...        ...  ...      ...  \n",
       "GXE_17166  0/0  0/0      0/0  \n",
       "GXE_17167  0/0  0/0      0/0  \n",
       "GXE_17168  0/0  0/0      0/0  \n",
       "GXE_17188  0/0  0/0      0/0  \n",
       "GXE_17189  0/0  0/0      0/0  \n",
       "\n",
       "[4384 rows x 58 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# find singletons and doubletons where at least 1 alt allele found in 15 parental strains used to map QTLs\n",
    "df = gxe_oligos_final[~( (gxe_oligos_final.loc[:,yeast_strains_id] == '0/0') | (gxe_oligos_final.loc[:,yeast_strains_id].isna()) ).all(axis=1)].query('AC<3')\n",
    "\n",
    "# check if parental strains carry alt allele that matches variant entry\n",
    "singletons_doubletons_include_list = []\n",
    "for v, row in df.iterrows():\n",
    "    alt_allele_num = row.all_alt_alleles.split(',').index(row.ALT)+1\n",
    "    if any([str(alt_allele_num) in gt for gt in row[yeast_strains_id].fillna('unknown')]):\n",
    "        singletons_doubletons_include_list.append(v)\n",
    "        \n",
    "# finally, subset gxe_oligos_final to remove singletons and doubletons, unless variant is in singletons_doubletons_include_list\n",
    "gxe_oligos_final_no_singletons = gxe_oligos_final.query('AC>=3 | var_id.isin(@singletons_doubletons_include_list)')\n",
    "display(gxe_oligos_final_no_singletons)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>qtl_left</th>\n",
       "      <th>qtl_right</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Caffeine;15mM;2</td>\n",
       "      <td>207664</td>\n",
       "      <td>208615</td>\n",
       "      <td>208157</td>\n",
       "      <td>209806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Cobalt_Chloride;2mM;2</td>\n",
       "      <td>208615</td>\n",
       "      <td>209367</td>\n",
       "      <td>208157</td>\n",
       "      <td>209806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Cobalt_Chloride;2mM;2</td>\n",
       "      <td>373834</td>\n",
       "      <td>375203</td>\n",
       "      <td>373793</td>\n",
       "      <td>375169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Cobalt_Chloride;2mM;2</td>\n",
       "      <td>373834</td>\n",
       "      <td>375203</td>\n",
       "      <td>373793</td>\n",
       "      <td>375169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Fructose;;1</td>\n",
       "      <td>373674</td>\n",
       "      <td>374422</td>\n",
       "      <td>373793</td>\n",
       "      <td>375169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>373699</td>\n",
       "      <td>374377</td>\n",
       "      <td>373793</td>\n",
       "      <td>375169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>181440</td>\n",
       "      <td>181552</td>\n",
       "      <td>181403</td>\n",
       "      <td>182548</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    trait  qtl_left  qtl_right   start     end\n",
       "20        Caffeine;15mM;2    207664     208615  208157  209806\n",
       "21  Cobalt_Chloride;2mM;2    208615     209367  208157  209806\n",
       "28  Cobalt_Chloride;2mM;2    373834     375203  373793  375169\n",
       "29  Cobalt_Chloride;2mM;2    373834     375203  373793  375169\n",
       "30            Fructose;;1    373674     374422  373793  375169\n",
       "31      Neomycin;5mg/mL;2    373699     374377  373793  375169\n",
       "98      Neomycin;5mg/mL;2    181440     181552  181403  182548"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>chrom</th>\n",
       "      <th>left_limit</th>\n",
       "      <th>right_limit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Caffeine;15mM;2</td>\n",
       "      <td>XVI</td>\n",
       "      <td>207657</td>\n",
       "      <td>210306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Cobalt_Chloride;2mM;2</td>\n",
       "      <td>XVI</td>\n",
       "      <td>207657</td>\n",
       "      <td>210306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Cobalt_Chloride;2mM;2</td>\n",
       "      <td>XVI</td>\n",
       "      <td>373293</td>\n",
       "      <td>375669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Fructose;;1</td>\n",
       "      <td>XVI</td>\n",
       "      <td>373293</td>\n",
       "      <td>375669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>XVI</td>\n",
       "      <td>373293</td>\n",
       "      <td>375669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Neomycin;5mg/mL;2</td>\n",
       "      <td>XVI</td>\n",
       "      <td>180903</td>\n",
       "      <td>183048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    trait chrom  left_limit  right_limit\n",
       "20        Caffeine;15mM;2   XVI      207657       210306\n",
       "21  Cobalt_Chloride;2mM;2   XVI      207657       210306\n",
       "28  Cobalt_Chloride;2mM;2   XVI      373293       375669\n",
       "30            Fructose;;1   XVI      373293       375669\n",
       "31      Neomycin;5mg/mL;2   XVI      373293       375669\n",
       "98      Neomycin;5mg/mL;2   XVI      180903       183048"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(gxe_qtls.query('chr==\"chrXVI\"')[['trait', 'qtl_left','qtl_right', 'start', 'end']])\n",
    "display(qtls_df.query('chrom==\"XVI\"'))#[['trait', 'qtl_left','qtl_right', 'start', 'end']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/groups/hbfraser/modules/packages/conda/4.6.14/envs/crispey-bar/lib/python3.7/site-packages/pandas/core/indexing.py:376: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/home/groups/hbfraser/modules/packages/conda/4.6.14/envs/crispey-bar/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item] = s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>ACQ</th>\n",
       "      <th>ACS</th>\n",
       "      <th>ADE</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADG</th>\n",
       "      <th>ADI</th>\n",
       "      <th>ADR</th>\n",
       "      <th>SACE_MAA</th>\n",
       "      <th>assoc_traits</th>\n",
       "      <th>assoc_traits_str</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GXE_00023</th>\n",
       "      <td>I</td>\n",
       "      <td>192419</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.002970</td>\n",
       "      <td>2|200</td>\n",
       "      <td>4687|3608|163</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00024</th>\n",
       "      <td>I</td>\n",
       "      <td>192424</td>\n",
       "      <td>GTTTGGATTACCTCT</td>\n",
       "      <td>ATTTGGATTACCTCT</td>\n",
       "      <td>39</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>7|181</td>\n",
       "      <td>4692|3613|168</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00026</th>\n",
       "      <td>I</td>\n",
       "      <td>192427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>902</td>\n",
       "      <td>2000</td>\n",
       "      <td>0.451000</td>\n",
       "      <td>10|192</td>\n",
       "      <td>4695|3616|171</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00027</th>\n",
       "      <td>I</td>\n",
       "      <td>192429</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1506</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.754000</td>\n",
       "      <td>12|190</td>\n",
       "      <td>4697|3618|173</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_00033</th>\n",
       "      <td>I</td>\n",
       "      <td>192466</td>\n",
       "      <td>AAAGGG</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.001485</td>\n",
       "      <td>50|148</td>\n",
       "      <td>4735|3656|211</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17166</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375077</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.001484</td>\n",
       "      <td>425|4414</td>\n",
       "      <td>4043|2214|2156</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17167</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375078</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>424|4415</td>\n",
       "      <td>4042|2213|2157</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17168</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375080</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>422|4417</td>\n",
       "      <td>4040|2211|2159</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17188</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375331</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>171|4668</td>\n",
       "      <td>3789|1960|162|2410</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GXE_17189</th>\n",
       "      <td>XVI</td>\n",
       "      <td>375333</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>8</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>168|4670</td>\n",
       "      <td>3786|1957|164|2412</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>[Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...</td>\n",
       "      <td>Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4384 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS              REF              ALT    AC    AN  \\\n",
       "var_id                                                                  \n",
       "GXE_00023     I  192419                C                A     6  2020   \n",
       "GXE_00024     I  192424  GTTTGGATTACCTCT  ATTTGGATTACCTCT    39  2014   \n",
       "GXE_00026     I  192427                T                C   902  2000   \n",
       "GXE_00027     I  192429                G                C  1506  1998   \n",
       "GXE_00033     I  192466           AAAGGG                A     3  2020   \n",
       "...         ...     ...              ...              ...   ...   ...   \n",
       "GXE_17166   XVI  375077                T                C     3  2022   \n",
       "GXE_17167   XVI  375078                C                A     8  2020   \n",
       "GXE_17168   XVI  375080                G                A    31  2022   \n",
       "GXE_17188   XVI  375331                C                A     8  2020   \n",
       "GXE_17189   XVI  375333                A               AT     8  2022   \n",
       "\n",
       "                 AF upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                              \n",
       "GXE_00023  0.002970                 2|200           4687|3608|163   \n",
       "GXE_00024  0.019000                 7|181           4692|3613|168   \n",
       "GXE_00026  0.451000                10|192           4695|3616|171   \n",
       "GXE_00027  0.754000                12|190           4697|3618|173   \n",
       "GXE_00033  0.001485                50|148           4735|3656|211   \n",
       "...             ...                   ...                     ...   \n",
       "GXE_17166  0.001484              425|4414          4043|2214|2156   \n",
       "GXE_17167  0.003960              424|4415          4042|2213|2157   \n",
       "GXE_17168  0.015000              422|4417          4040|2211|2159   \n",
       "GXE_17188  0.003960              171|4668      3789|1960|162|2410   \n",
       "GXE_17189  0.003956              168|4670      3786|1957|164|2412   \n",
       "\n",
       "                            region  ...  ACQ  ACS  ADE  ADF  ADG  ADI  ADR  \\\n",
       "var_id                              ...                                      \n",
       "GXE_00023   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_00024   bidirectional_promoter  ...  0/0  0/0  NaN  0/0  0/0  0/0  0/0   \n",
       "GXE_00026   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  1/1  1/1   \n",
       "GXE_00027   bidirectional_promoter  ...  1/1  1/1  NaN  1/1  1/1  1/1  1/1   \n",
       "GXE_00033   bidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "...                            ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "GXE_17166         missense_variant  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17167              stop_gained  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17168         missense_variant  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17188  unidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "GXE_17189  unidirectional_promoter  ...  0/0  0/0  0/0  0/0  0/0  0/0  0/0   \n",
       "\n",
       "          SACE_MAA                                       assoc_traits  \\\n",
       "var_id                                                                  \n",
       "GXE_00023      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "GXE_00024      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "GXE_00026      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "GXE_00027      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "GXE_00033      0/0  [Cobalt_Chloride;2mM;2, Fluconazole;100uM;2, N...   \n",
       "...            ...                                                ...   \n",
       "GXE_17166      0/0  [Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...   \n",
       "GXE_17167      0/0  [Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...   \n",
       "GXE_17168      0/0  [Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...   \n",
       "GXE_17188      0/0  [Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...   \n",
       "GXE_17189      0/0  [Cobalt_Chloride;2mM;2, Fructose;;1, Neomycin;...   \n",
       "\n",
       "                                            assoc_traits_str  \n",
       "var_id                                                        \n",
       "GXE_00023  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "GXE_00024  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "GXE_00026  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "GXE_00027  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "GXE_00033  Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neom...  \n",
       "...                                                      ...  \n",
       "GXE_17166  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...  \n",
       "GXE_17167  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...  \n",
       "GXE_17168  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...  \n",
       "GXE_17188  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...  \n",
       "GXE_17189  Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg...  \n",
       "\n",
       "[4384 rows x 60 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# use QTL/gene coordinates to associate each variant with traits in gxe_qtls\n",
    "qtls_df = gxe_qtls[['trait', 'chrom', 'start', 'end', 'qtl_left', 'qtl_right']].copy()\n",
    "qtls_df.loc[:,'chrom'] = qtls_df['chrom'].str.replace('chr', '')\n",
    "# find gene upstream and downstream boundaries\n",
    "qtls_df.loc[:, 'gene_left'] = qtls_df['start']-500\n",
    "qtls_df.loc[:, 'gene_right'] = qtls_df['end']+500\n",
    "# define search boundaries by either gene or QTL boundary, whichever is larger\n",
    "qtls_df.loc[:,'left_limit'] = qtls_df[['gene_left', 'qtl_left']].min(axis=1)\n",
    "qtls_df.loc[:,'right_limit'] = qtls_df[['gene_right', 'qtl_right']].max(axis=1)\n",
    "qtls_df.loc[:,'left_limit'] = qtls_df['left_limit'].astype(int)\n",
    "qtls_df.loc[:,'right_limit'] = qtls_df['right_limit'].astype(int)\n",
    "qtls_df.loc[:, 'region_length'] = qtls_df['right_limit'] - qtls_df['left_limit'] + 1\n",
    "# remove search boundaries duplicates\n",
    "qtls_df = qtls_df[['trait', 'chrom', 'left_limit', 'right_limit']].drop_duplicates()\n",
    "\n",
    "# associate each variant with traits of interest if it falls within QTL/gene search boundaries\n",
    "gxe_oligos_final_no_singletons.loc[:,'assoc_traits'] = gxe_oligos_final_no_singletons.apply(lambda x: qtls_df.query('chrom==@x.CHROM & left_limit<=@x.POS & right_limit>=@x.POS').trait.unique().tolist(), axis=1)\n",
    "gxe_oligos_final_no_singletons.loc[:,'assoc_traits_str'] = gxe_oligos_final_no_singletons['assoc_traits'].apply(lambda x: '_'.join(x))\n",
    "\n",
    "display(gxe_oligos_final_no_singletons)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write gxe_oligos_final_no_singletons table to file\n",
    "gxe_oligos_final_no_singletons.to_csv(gxe_variants_file.replace('.vcf', '_final.txt'), sep='\\t')\n",
    "\n",
    "# use this file in separate script to organize into pools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of GxE variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants: 4384\n",
      "Number of oligos: 10983\n",
      "\n",
      "Distribution of variants among gxe genes\n",
      "count     53.000000\n",
      "mean      73.245283\n",
      "std       44.006951\n",
      "min       10.000000\n",
      "25%       39.000000\n",
      "50%       68.000000\n",
      "75%       98.000000\n",
      "max      189.000000\n",
      "Name: Gene_ID, dtype: float64\n",
      "\n",
      "Distribution of oligos among gxe variants\n",
      "count    4384.000000\n",
      "mean        2.505246\n",
      "std         0.808900\n",
      "min         2.000000\n",
      "25%         2.000000\n",
      "50%         2.000000\n",
      "75%         3.000000\n",
      "max         7.000000\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Proportion of variant annotation classes\n",
      "synonymous_variant         0.377965\n",
      "missense_variant           0.288549\n",
      "bidirectional_promoter     0.140511\n",
      "unidirectional_promoter    0.123859\n",
      "intergenic                 0.055201\n",
      "frameshift_variant         0.008212\n",
      "stop_gained                0.003650\n",
      "inframe_insertion          0.001369\n",
      "inframe_deletion           0.000456\n",
      "stop_lost                  0.000228\n",
      "Name: region, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "### SUMMARY OF GxE SET ###\n",
    "print(\"Number of variants:\", len(gxe_oligos_final_no_singletons))\n",
    "print(\"Number of oligos:\", int(sum(gxe_oligos_final_no_singletons.num_of_oligos)))\n",
    "print()\n",
    "print(\"Distribution of variants among gxe genes\")\n",
    "print(gxe_oligos_final_no_singletons.Gene_ID.value_counts().filter(gxe_qtls.gene_id.unique()).describe())\n",
    "print()\n",
    "print(\"Distribution of oligos among gxe variants\")\n",
    "print(gxe_oligos_final_no_singletons.num_of_oligos.describe())\n",
    "print()\n",
    "print(\"Proportion of variant annotation classes\")\n",
    "print(gxe_oligos_final_no_singletons.region.value_counts()/len(gxe_oligos_final_no_singletons))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine number of pools to use for GxE oligos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants per condition(s)\n",
      "assoc_traits_str\n",
      "Caffeine;15mM;2                                                879\n",
      "Caffeine;15mM;2_Cobalt_Chloride;2mM;2                           41\n",
      "Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2          163\n",
      "Caffeine;15mM;2_Fructose;;1                                      1\n",
      "Caffeine;15mM;2_Neomycin;5mg/mL;2                                5\n",
      "Cobalt_Chloride;2mM;2                                          611\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2                       74\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1            4\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2    172\n",
      "Cobalt_Chloride;2mM;2_Fructose;;1                               70\n",
      "Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2             48\n",
      "Fluconazole;100uM;2                                            531\n",
      "Fluconazole;100uM;2_Fructose;;1                                 32\n",
      "Fructose;;1                                                    632\n",
      "Lithium_Chloride;100mM;2                                       659\n",
      "Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2                      53\n",
      "Neomycin;5mg/mL;2                                              409\n",
      "dtype: int64\n",
      "\n",
      "Number of oligos per condition(s)\n",
      "assoc_traits_str\n",
      "Caffeine;15mM;2                                                2171.0\n",
      "Caffeine;15mM;2_Cobalt_Chloride;2mM;2                           117.0\n",
      "Caffeine;15mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2           373.0\n",
      "Caffeine;15mM;2_Fructose;;1                                       2.0\n",
      "Caffeine;15mM;2_Neomycin;5mg/mL;2                                12.0\n",
      "Cobalt_Chloride;2mM;2                                          1539.0\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2                       175.0\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Fructose;;1            10.0\n",
      "Cobalt_Chloride;2mM;2_Fluconazole;100uM;2_Neomycin;5mg/mL;2     457.0\n",
      "Cobalt_Chloride;2mM;2_Fructose;;1                               170.0\n",
      "Cobalt_Chloride;2mM;2_Fructose;;1_Neomycin;5mg/mL;2             140.0\n",
      "Fluconazole;100uM;2                                            1376.0\n",
      "Fluconazole;100uM;2_Fructose;;1                                  73.0\n",
      "Fructose;;1                                                    1534.0\n",
      "Lithium_Chloride;100mM;2                                       1660.0\n",
      "Lithium_Chloride;100mM;2_Neomycin;5mg/mL;2                      146.0\n",
      "Neomycin;5mg/mL;2                                              1028.0\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Trait: Caffeine;15mM;2\n",
      "Number of variants: 1089\n",
      "Number of oligos: 2675\n",
      "\n",
      "Trait: Fluconazole;100uM;2\n",
      "Number of variants: 976\n",
      "Number of oligos: 2464\n",
      "\n",
      "Trait: Neomycin;5mg/mL;2\n",
      "Number of variants: 850\n",
      "Number of oligos: 2156\n",
      "\n",
      "Trait: Cobalt_Chloride;2mM;2\n",
      "Number of variants: 1020\n",
      "Number of oligos: 2608\n",
      "\n",
      "Trait: Fructose;;1\n",
      "Number of variants: 787\n",
      "Number of oligos: 1929\n",
      "\n",
      "Trait: Lithium_Chloride;100mM;2\n",
      "Number of variants: 712\n",
      "Number of oligos: 1806\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('Number of variants per condition(s)')\n",
    "print(gxe_oligos_final_no_singletons.groupby('assoc_traits_str').size())\n",
    "print()\n",
    "print('Number of oligos per condition(s)')\n",
    "print(gxe_oligos_final_no_singletons.groupby('assoc_traits_str').num_of_oligos.sum())\n",
    "print()\n",
    "\n",
    "# count variants and oligos per trait\n",
    "for trait in gxe_qtls.trait.unique():\n",
    "    print(\"Trait:\", trait)\n",
    "    print('Number of variants:', len(gxe_oligos_final_no_singletons[gxe_oligos_final_no_singletons.assoc_traits.apply(lambda x: trait in x)]) )\n",
    "    print('Number of oligos:', int(gxe_oligos_final_no_singletons[gxe_oligos_final_no_singletons.assoc_traits.apply(lambda x: trait in x)]['num_of_oligos'].sum()) )\n",
    "    print()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GXE    10983.0\n",
       "Name: num_of_oligos, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos to fit: 10983\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(93, 100)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use 93 pools:\n",
      "Number of oligos per pool: 118\n",
      "Number of oligos leftover: 9\n",
      "Number of technical oligos: 279\n",
      "\n",
      "Use 94 pools:\n",
      "Number of oligos per pool: 116\n",
      "Number of oligos leftover: 79\n",
      "Number of technical oligos: 470\n",
      "\n",
      "Use 95 pools:\n",
      "Number of oligos per pool: 115\n",
      "Number of oligos leftover: 58\n",
      "Number of technical oligos: 570\n",
      "\n",
      "Use 96 pools:\n",
      "Number of oligos per pool: 114\n",
      "Number of oligos leftover: 39\n",
      "Number of technical oligos: 672\n",
      "\n",
      "Use 97 pools:\n",
      "Number of oligos per pool: 113\n",
      "Number of oligos leftover: 22\n",
      "Number of technical oligos: 776\n",
      "\n",
      "Use 98 pools:\n",
      "Number of oligos per pool: 112\n",
      "Number of oligos leftover: 7\n",
      "Number of technical oligos: 882\n",
      "\n",
      "Use 99 pools:\n",
      "Number of oligos per pool: 110\n",
      "Number of oligos leftover: 93\n",
      "Number of technical oligos: 1089\n",
      "\n",
      "Use 100 pools:\n",
      "Number of oligos per pool: 109\n",
      "Number of oligos leftover: 83\n",
      "Number of technical oligos: 1200\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([ 93,  94,  95,  96,  97,  98,  99, 100])"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count OLIGOS in each variant prefix set\n",
    "display(gxe_oligos_final_no_singletons.groupby(lambda x: x[:3]).num_of_oligos.sum().sort_values(ascending=False))\n",
    "\n",
    "# determine number of pools to use\n",
    "num_of_oligos = int(gxe_oligos_final_no_singletons.num_of_oligos.sum())\n",
    "min_pool_size = 109 # allows 12 technical oligos\n",
    "max_pool_size = 118 # allows 3 technical oligos\n",
    "complete_pool_size = 121 # maximum oligos synthesized per pool\n",
    "\n",
    "# find number of pools to use\n",
    "find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We should be able to fit all GxE variants in 94 pools\n",
    "- GXE: 94 pools\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create a small list of HSP90 variants as a potential rotation project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "working_dir = \"/home/users/rang/scratch/yeast/hsp90/\"\n",
    "os.chdir(working_dir)\n",
    "\n",
    "# load gene_info_df from GFF file (see code above)\n",
    "\n",
    "# find HSP90 genes (HSP82, HSC82)\n",
    "hsp90_genes = gene_info_df.loc[['YPL240C', 'YMR186W'],:].sort_values(['chrom', 'start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select variants\n",
    "chromosome_dict = {'chrI' : 'chromosome1',\n",
    "                   'chrII': 'chromosome2',\n",
    "                   'chrIII': 'chromosome3',\n",
    "                   'chrIV': 'chromosome4',\n",
    "                   'chrV': 'chromosome5',\n",
    "                   'chrVI': 'chromosome6',\n",
    "                   'chrVII': 'chromosome7',\n",
    "                   'chrVIII': 'chromosome8',\n",
    "                   'chrIX': 'chromosome9',\n",
    "                   'chrX': 'chromosome10',\n",
    "                   'chrXI': 'chromosome11',\n",
    "                   'chrXII': 'chromosome12',\n",
    "                   'chrXIII': 'chromosome13',\n",
    "                   'chrXIV': 'chromosome14',\n",
    "                   'chrXV': 'chromosome15',\n",
    "                   'chrXVI': 'chromosome16',}\n",
    "\n",
    "results = []\n",
    "for gene, row in hsp90_genes.iterrows():\n",
    "    gvcf_file = \"/home/users/rang/share/yeast/1011genomes/by_chrom/{}.gvcf.gz\".format(chromosome_dict[row.chrom])\n",
    "    \n",
    "    # read in gvcf file to parse for variants\n",
    "    allVars = vcf.Reader(filename=gvcf_file)\n",
    "    retrieve = allVars.fetch(chrom=chromosome_dict[row['chrom']], \n",
    "                             start=int(row['start']-500-1),\n",
    "                             end=int(row['end']+500))\n",
    "    for rec in retrieve:\n",
    "        rec.samples=[]\n",
    "        rec.FORMAT=None\n",
    "        results.append(rec)\n",
    "\n",
    "results = pd.Series(results).drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# write to new vcf file\n",
    "hsp90_variants_file = \"hsp90_variants.vcf\"\n",
    "template = vcf.Reader(filename=gvcf_file)\n",
    "# change chromosome naming convention to a single Roman numeral (follows ref genome, VEP formatting)\n",
    "chromosome_dict_rev = {v:k for k, v in chromosome_dict.items()} \n",
    "template.contigs = OrderedDict((chromosome_dict_rev[k][3:], v._replace(id=chromosome_dict_rev[k][3:])) for k, v in template.contigs.items())\n",
    "# remove 1011 genomes samples data\n",
    "template.samples=[]\n",
    "\n",
    "output = vcf.Writer(open(hsp90_variants_file, 'w'), template)\n",
    "for record in results:\n",
    "    # update chromosome naming\n",
    "    record.CHROM = chromosome_dict_rev[record.CHROM][3:]\n",
    "    # write to file\n",
    "    output.write_record(record)\n",
    "output.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant at chrom XIII, pos 634136 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIII, pos 634148 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIII, pos 634157 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIII, pos 634169 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XVI, pos 96797 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XVI, pos 96803 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XVI, pos 96820 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XVI, pos 97542 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XVI, pos 98281 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XIII, pos 632705 skipped: No unique matches found in tested genome(s)\n",
      "Variant at chrom XVI, pos 98284 skipped: No unique matches found in tested genome(s)\n"
     ]
    }
   ],
   "source": [
    "# path to genome fasta files\n",
    "ref_genome_fasta = \"/home/users/rang/yeast/genomes/Saccharomyces_cerevisiae.R64-1-1.dna.chromosome.I.fa\"\n",
    "rm_genome_fasta = \"/home/users/rang/yeast/genomes/RM11-1A_SGD_2015_JRIP00000000.fsa\"\n",
    "yjm_genome_fasta = \"/home/users/rang/yeast/genomes/YJM789_Stanford_2007_AAFW02000000_highQuality31.fsa\"\n",
    "yps_genome_fasta = \"/home/users/rang/yeast/genomes/YPS128.genome.fa\"\n",
    "\n",
    "# path to input and output VCF files\n",
    "hsp90_variants_file = \"hsp90_variants.vcf\"\n",
    "hsp90_variants_select_file = hsp90_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# load genome files\n",
    "ref_genome = SeqIO.to_dict(SeqIO.parse(ref_genome_fasta, \"fasta\"))\n",
    "rm_genome = SeqIO.to_dict(SeqIO.parse(rm_genome_fasta, \"fasta\"))\n",
    "yjm_genome = SeqIO.to_dict(SeqIO.parse(yjm_genome_fasta, \"fasta\"))\n",
    "yps_genome = SeqIO.to_dict(SeqIO.parse(yps_genome_fasta, \"fasta\"))\n",
    "genomes_list = [ref_genome, rm_genome, yjm_genome, yps_genome] # IMPORTANT: SET REFERENCE GENOME AS FIRST IN LIST\n",
    "\n",
    "\n",
    "threshold = 10 # threshold value for searching NGGs\n",
    "window = 61 # size of flanking sequences to search for unique hits\n",
    "\n",
    "# read in all candidate variants to filter\n",
    "hsp90_variants = vcf.Reader(filename=hsp90_variants_file)\n",
    "hsp90_variants = [rec for rec in hsp90_variants]\n",
    "# split variants into chunks for multiprocessing\n",
    "n = len(hsp90_variants)//len(os.sched_getaffinity(0)) + 1\n",
    "hsp90_variants = [hsp90_variants[i:i+n] for i in range(0, len(hsp90_variants), n)]\n",
    "\n",
    "# find targetable variants (TGT info field indicates which genomes variant is targetable in)\n",
    "to_process = [(rec_list, genomes_list, threshold, window) for rec_list in hsp90_variants]\n",
    "with mp.Pool(len(os.sched_getaffinity(0))) as pool:\n",
    "    result = pool.starmap(find_targetable_variants, to_process) \n",
    "\n",
    "# write targetable variants to file\n",
    "header = vcf.Reader(filename=hsp90_variants_file)\n",
    "# add \"TGT\" info field into VCF header\n",
    "header_info = namedtuple('Info', ['id', 'num', 'type', 'desc', 'source', 'version'])\n",
    "header_info_tgt = header_info('TGT', '.', 'Integer', 'Targetability of variant in tested genomes', None, None)\n",
    "header.infos['TGT'] = header_info_tgt\n",
    "# write to file\n",
    "hsp90_variants_select = vcf.Writer(open(hsp90_variants_select_file, 'w'), header)\n",
    "to_write = [rec for rec_list in result for rec in rec_list]\n",
    "for rec in to_write:\n",
    "    hsp90_variants_select.write_record(rec)\n",
    "hsp90_variants_select.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select variants by allele frequency (note difference from ergosterol variant selection step)\n",
    "### remove singletons and doubletons\n",
    "### if multiple alternate alleles, select most common alternate allele\n",
    "### assign unitl variant ID prefix and number (use format VAR_XXXXX)\n",
    "\n",
    "# candidate variants editable in all strains\n",
    "hsp90_variants_select = vcf.Reader(filename=hsp90_variants_select_file)\n",
    "\n",
    "# open output file\n",
    "hsp90_variants_select_AFfilter_file = hsp90_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "hsp90_variants_select_AFfilter = vcf.Writer(open(hsp90_variants_select_AFfilter_file, 'w'), hsp90_variants_select)\n",
    "\n",
    "var_counter = {}\n",
    "for record in hsp90_variants_select:        \n",
    "    # (optional) filter by allele count - currently disabled by setting to ac<1\n",
    "    if all([ac<1 for ac in record.INFO['AC']]):\n",
    "        continue\n",
    "    # if at least one alternate allele meets minimum count, select most common alternate allele to study\n",
    "    else:\n",
    "        i = np.argmax(record.INFO['AC']) # most common alt allele\n",
    "        # update record entry\n",
    "        record.ALT = [record.ALT[i]]\n",
    "        record.INFO['AC'] = [record.INFO['AC'][i]]\n",
    "        record.INFO['AF'] = [record.INFO['AF'][i]]\n",
    "        record.INFO['MLEAC'] = [record.INFO['MLEAC'][i]]\n",
    "        record.INFO['MLEAF'] = [record.INFO['MLEAF'][i]]\n",
    "        \n",
    "        # inspect TGT field and assign variant ID prefix\n",
    "        ### if tgt field contains 1 at first position (i.e. targetable by reference genome), assign \"HSP\"\n",
    "        if record.INFO['TGT'][0] == 1:\n",
    "            var_id_prefix = \"HSP\"\n",
    "        else:\n",
    "#             print('variant does not pass TGT check: skipping')\n",
    "            continue\n",
    "#             # assign unique prefix \"HS-\" with the last digit determined by converting binary to hexadecimal (works well for up to 4 genomes)\n",
    "#             binary_str = ''.join([str(x) for x in record.INFO['TGT']])\n",
    "#             hex_digit = hex(int(binary_str, 2))[2:].upper()\n",
    "#             var_id_prefix = \"HS{}\".format(hex_digit)\n",
    "            \n",
    "        # add to counter\n",
    "        if var_id_prefix not in var_counter:\n",
    "            var_counter[var_id_prefix] = 1\n",
    "        else:\n",
    "            var_counter[var_id_prefix] += 1\n",
    "        \n",
    "        # add unique variant ID (follow format VAR_XXXXX)\n",
    "        record.ID='{}_{:05d}'.format(var_id_prefix, var_counter[var_id_prefix])\n",
    "        \n",
    "        # write to output\n",
    "        hsp90_variants_select_AFfilter.write_record(record)\n",
    "hsp90_variants_select_AFfilter.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit hsp90_variants_select_AFfilter VCF to VEP to get annotations, and download VEP output in TXT format\n",
    "http://uswest.ensembl.org/Saccharomyces_cerevisiae/Tools/VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HSP_00001</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631917</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>4109|438</td>\n",
       "      <td>3132|2772|4374</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Tac/Cac</td>\n",
       "      <td>Pfam:PF10304,PANTHER:PTHR20959,Superfamily:SSF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00002</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631956</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>342</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.169000</td>\n",
       "      <td>4148|399</td>\n",
       "      <td>3171|2733|4335</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Cta/Tta</td>\n",
       "      <td>Pfam:PF10304,PANTHER:PTHR20959</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00003</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631959</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>4151|396</td>\n",
       "      <td>3174|2730|4332</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gat/Cat</td>\n",
       "      <td>Pfam:PF10304,PANTHER:PTHR20959</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00004</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631977</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>56</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>4169|378</td>\n",
       "      <td>3192|7|2712|4314</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00005</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631978</td>\n",
       "      <td>CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>4171|259</td>\n",
       "      <td>3194|9|2593|4195</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>259.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00425</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99064</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>47</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.023000</td>\n",
       "      <td>4168|1432|420|439|2806|3955</td>\n",
       "      <td>2544</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>439.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>420.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00426</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99066</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>7</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003465</td>\n",
       "      <td>4166|1430|418|441|2808|3957</td>\n",
       "      <td>2542</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>441.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>418.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00427</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99082</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>4150|1414|402|457|2824|3973</td>\n",
       "      <td>2526</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>457.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>402.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00428</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99086</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>4146|1410|398|461|2828|3977</td>\n",
       "      <td>2522</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>461.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00429</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99087</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>4145|1409|397|462|2829|3978</td>\n",
       "      <td>2521</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>462.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>397.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>429 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS                                                REF  \\\n",
       "var_id                                                                       \n",
       "HSP_00001  XIII  631917                                                  T   \n",
       "HSP_00002  XIII  631956                                                  C   \n",
       "HSP_00003  XIII  631959                                                  G   \n",
       "HSP_00004  XIII  631977                                                  A   \n",
       "HSP_00005  XIII  631978  CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...   \n",
       "...         ...     ...                                                ...   \n",
       "HSP_00425   XVI   99064                                                  T   \n",
       "HSP_00426   XVI   99066                                                  G   \n",
       "HSP_00427   XVI   99082                                                  A   \n",
       "HSP_00428   XVI   99086                                                  A   \n",
       "HSP_00429   XVI   99087                                                  G   \n",
       "\n",
       "          ALT   AC    AN        AF        upstream_distance_str  \\\n",
       "var_id                                                            \n",
       "HSP_00001   C    1  2022  0.000495                     4109|438   \n",
       "HSP_00002   T  342  2020  0.169000                     4148|399   \n",
       "HSP_00003   C    2  2022  0.000989                     4151|396   \n",
       "HSP_00004   G   56  2022  0.028000                     4169|378   \n",
       "HSP_00005   C    6  2022  0.002967                     4171|259   \n",
       "...        ..  ...   ...       ...                          ...   \n",
       "HSP_00425   C   47  2014  0.023000  4168|1432|420|439|2806|3955   \n",
       "HSP_00426   A    7  2020  0.003465  4166|1430|418|441|2808|3957   \n",
       "HSP_00427   G    2  2022  0.000989  4150|1414|402|457|2824|3973   \n",
       "HSP_00428   G    2  2022  0.000989  4146|1410|398|461|2828|3977   \n",
       "HSP_00429   A    1  2022  0.000495  4145|1409|397|462|2829|3978   \n",
       "\n",
       "          downstream_distance_str                   region  ...   Codons  \\\n",
       "var_id                                                      ...            \n",
       "HSP_00001          3132|2772|4374         missense_variant  ...  Tac/Cac   \n",
       "HSP_00002          3171|2733|4335       synonymous_variant  ...  Cta/Tta   \n",
       "HSP_00003          3174|2730|4332         missense_variant  ...  Gat/Cat   \n",
       "HSP_00004        3192|7|2712|4314  unidirectional_promoter  ...      NaN   \n",
       "HSP_00005        3194|9|2593|4195  unidirectional_promoter  ...      NaN   \n",
       "...                           ...                      ...  ...      ...   \n",
       "HSP_00425                    2544   bidirectional_promoter  ...      NaN   \n",
       "HSP_00426                    2542   bidirectional_promoter  ...      NaN   \n",
       "HSP_00427                    2526   bidirectional_promoter  ...      NaN   \n",
       "HSP_00428                    2522   bidirectional_promoter  ...      NaN   \n",
       "HSP_00429                    2521   bidirectional_promoter  ...      NaN   \n",
       "\n",
       "                                                     DOMAINS  \\\n",
       "var_id                                                         \n",
       "HSP_00001  Pfam:PF10304,PANTHER:PTHR20959,Superfamily:SSF...   \n",
       "HSP_00002                     Pfam:PF10304,PANTHER:PTHR20959   \n",
       "HSP_00003                     Pfam:PF10304,PANTHER:PTHR20959   \n",
       "HSP_00004                                                NaN   \n",
       "HSP_00005                                                NaN   \n",
       "...                                                      ...   \n",
       "HSP_00425                                                NaN   \n",
       "HSP_00426                                                NaN   \n",
       "HSP_00427                                                NaN   \n",
       "HSP_00428                                                NaN   \n",
       "HSP_00429                                                NaN   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "HSP_00001                     NaN                   NaN   \n",
       "HSP_00002                     NaN                   NaN   \n",
       "HSP_00003                     NaN                   NaN   \n",
       "HSP_00004                    RTP1               YMR185W   \n",
       "HSP_00005                    RTP1               YMR185W   \n",
       "...                           ...                   ...   \n",
       "HSP_00425                   HSP82               YPL240C   \n",
       "HSP_00426                   HSP82               YPL240C   \n",
       "HSP_00427                   HSP82               YPL240C   \n",
       "HSP_00428                   HSP82               YPL240C   \n",
       "HSP_00429                   HSP82               YPL240C   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "HSP_00001                      NaN                    NaN   \n",
       "HSP_00002                      NaN                    NaN   \n",
       "HSP_00003                      NaN                    NaN   \n",
       "HSP_00004  downstream_gene_variant                    7.0   \n",
       "HSP_00005  downstream_gene_variant                    9.0   \n",
       "...                            ...                    ...   \n",
       "HSP_00425    upstream_gene_variant                  439.0   \n",
       "HSP_00426    upstream_gene_variant                  441.0   \n",
       "HSP_00427    upstream_gene_variant                  457.0   \n",
       "HSP_00428    upstream_gene_variant                  461.0   \n",
       "HSP_00429    upstream_gene_variant                  462.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "HSP_00001                     NaN                   NaN   \n",
       "HSP_00002                     NaN                   NaN   \n",
       "HSP_00003                     NaN                   NaN   \n",
       "HSP_00004                   HSC82               YMR186W   \n",
       "HSP_00005                   HSC82               YMR186W   \n",
       "...                           ...                   ...   \n",
       "HSP_00425                    YAR1               YPL239W   \n",
       "HSP_00426                    YAR1               YPL239W   \n",
       "HSP_00427                    YAR1               YPL239W   \n",
       "HSP_00428                    YAR1               YPL239W   \n",
       "HSP_00429                    YAR1               YPL239W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "HSP_00001                      NaN                    NaN  \n",
       "HSP_00002                      NaN                    NaN  \n",
       "HSP_00003                      NaN                    NaN  \n",
       "HSP_00004    upstream_gene_variant                  378.0  \n",
       "HSP_00005    upstream_gene_variant                  259.0  \n",
       "...                            ...                    ...  \n",
       "HSP_00425    upstream_gene_variant                  420.0  \n",
       "HSP_00426    upstream_gene_variant                  418.0  \n",
       "HSP_00427    upstream_gene_variant                  402.0  \n",
       "HSP_00428    upstream_gene_variant                  398.0  \n",
       "HSP_00429    upstream_gene_variant                  397.0  \n",
       "\n",
       "[429 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HSP_00004</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631977</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>56</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>4169|378</td>\n",
       "      <td>3192|7|2712|4314</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00005</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631978</td>\n",
       "      <td>CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>4171|259</td>\n",
       "      <td>3194|9|2593|4195</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>259.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00006</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631979</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.005946</td>\n",
       "      <td>4171|376</td>\n",
       "      <td>3194|9|2710|4312</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>376.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00007</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>53</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.026000</td>\n",
       "      <td>4172|375</td>\n",
       "      <td>3195|10|2709|4311</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>375.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00008</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631981</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>4173|374</td>\n",
       "      <td>3196|11|2708|4310</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>11.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00425</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99064</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>47</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.023000</td>\n",
       "      <td>4168|1432|420|439|2806|3955</td>\n",
       "      <td>2544</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>439.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>420.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00426</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99066</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>7</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003465</td>\n",
       "      <td>4166|1430|418|441|2808|3957</td>\n",
       "      <td>2542</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>441.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>418.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00427</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99082</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>4150|1414|402|457|2824|3973</td>\n",
       "      <td>2526</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>457.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>402.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00428</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99086</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>4146|1410|398|461|2828|3977</td>\n",
       "      <td>2522</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>461.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00429</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99087</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>4145|1409|397|462|2829|3978</td>\n",
       "      <td>2521</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>462.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>397.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>403 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS                                                REF  \\\n",
       "var_id                                                                       \n",
       "HSP_00004  XIII  631977                                                  A   \n",
       "HSP_00005  XIII  631978  CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...   \n",
       "HSP_00006  XIII  631979                                                  G   \n",
       "HSP_00007  XIII  631980                                                  C   \n",
       "HSP_00008  XIII  631981                                                  A   \n",
       "...         ...     ...                                                ...   \n",
       "HSP_00425   XVI   99064                                                  T   \n",
       "HSP_00426   XVI   99066                                                  G   \n",
       "HSP_00427   XVI   99082                                                  A   \n",
       "HSP_00428   XVI   99086                                                  A   \n",
       "HSP_00429   XVI   99087                                                  G   \n",
       "\n",
       "          ALT  AC    AN        AF        upstream_distance_str  \\\n",
       "var_id                                                           \n",
       "HSP_00004   G  56  2022  0.028000                     4169|378   \n",
       "HSP_00005   C   6  2022  0.002967                     4171|259   \n",
       "HSP_00006   A  12  2018  0.005946                     4171|376   \n",
       "HSP_00007   T  53  2018  0.026000                     4172|375   \n",
       "HSP_00008   G  25  2020  0.012000                     4173|374   \n",
       "...        ..  ..   ...       ...                          ...   \n",
       "HSP_00425   C  47  2014  0.023000  4168|1432|420|439|2806|3955   \n",
       "HSP_00426   A   7  2020  0.003465  4166|1430|418|441|2808|3957   \n",
       "HSP_00427   G   2  2022  0.000989  4150|1414|402|457|2824|3973   \n",
       "HSP_00428   G   2  2022  0.000989  4146|1410|398|461|2828|3977   \n",
       "HSP_00429   A   1  2022  0.000495  4145|1409|397|462|2829|3978   \n",
       "\n",
       "          downstream_distance_str                   region  ... Codons  \\\n",
       "var_id                                                      ...          \n",
       "HSP_00004        3192|7|2712|4314  unidirectional_promoter  ...    NaN   \n",
       "HSP_00005        3194|9|2593|4195  unidirectional_promoter  ...    NaN   \n",
       "HSP_00006        3194|9|2710|4312  unidirectional_promoter  ...    NaN   \n",
       "HSP_00007       3195|10|2709|4311  unidirectional_promoter  ...    NaN   \n",
       "HSP_00008       3196|11|2708|4310  unidirectional_promoter  ...    NaN   \n",
       "...                           ...                      ...  ...    ...   \n",
       "HSP_00425                    2544   bidirectional_promoter  ...    NaN   \n",
       "HSP_00426                    2542   bidirectional_promoter  ...    NaN   \n",
       "HSP_00427                    2526   bidirectional_promoter  ...    NaN   \n",
       "HSP_00428                    2522   bidirectional_promoter  ...    NaN   \n",
       "HSP_00429                    2521   bidirectional_promoter  ...    NaN   \n",
       "\n",
       "          DOMAINS closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                            \n",
       "HSP_00004     NaN                    RTP1               YMR185W   \n",
       "HSP_00005     NaN                    RTP1               YMR185W   \n",
       "HSP_00006     NaN                    RTP1               YMR185W   \n",
       "HSP_00007     NaN                    RTP1               YMR185W   \n",
       "HSP_00008     NaN                    RTP1               YMR185W   \n",
       "...           ...                     ...                   ...   \n",
       "HSP_00425     NaN                   HSP82               YPL240C   \n",
       "HSP_00426     NaN                   HSP82               YPL240C   \n",
       "HSP_00427     NaN                   HSP82               YPL240C   \n",
       "HSP_00428     NaN                   HSP82               YPL240C   \n",
       "HSP_00429     NaN                   HSP82               YPL240C   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "HSP_00004  downstream_gene_variant                    7.0   \n",
       "HSP_00005  downstream_gene_variant                    9.0   \n",
       "HSP_00006  downstream_gene_variant                    9.0   \n",
       "HSP_00007  downstream_gene_variant                   10.0   \n",
       "HSP_00008  downstream_gene_variant                   11.0   \n",
       "...                            ...                    ...   \n",
       "HSP_00425    upstream_gene_variant                  439.0   \n",
       "HSP_00426    upstream_gene_variant                  441.0   \n",
       "HSP_00427    upstream_gene_variant                  457.0   \n",
       "HSP_00428    upstream_gene_variant                  461.0   \n",
       "HSP_00429    upstream_gene_variant                  462.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "HSP_00004                   HSC82               YMR186W   \n",
       "HSP_00005                   HSC82               YMR186W   \n",
       "HSP_00006                   HSC82               YMR186W   \n",
       "HSP_00007                   HSC82               YMR186W   \n",
       "HSP_00008                   HSC82               YMR186W   \n",
       "...                           ...                   ...   \n",
       "HSP_00425                    YAR1               YPL239W   \n",
       "HSP_00426                    YAR1               YPL239W   \n",
       "HSP_00427                    YAR1               YPL239W   \n",
       "HSP_00428                    YAR1               YPL239W   \n",
       "HSP_00429                    YAR1               YPL239W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "HSP_00004    upstream_gene_variant                  378.0  \n",
       "HSP_00005    upstream_gene_variant                  259.0  \n",
       "HSP_00006    upstream_gene_variant                  376.0  \n",
       "HSP_00007    upstream_gene_variant                  375.0  \n",
       "HSP_00008    upstream_gene_variant                  374.0  \n",
       "...                            ...                    ...  \n",
       "HSP_00425    upstream_gene_variant                  420.0  \n",
       "HSP_00426    upstream_gene_variant                  418.0  \n",
       "HSP_00427    upstream_gene_variant                  402.0  \n",
       "HSP_00428    upstream_gene_variant                  398.0  \n",
       "HSP_00429    upstream_gene_variant                  397.0  \n",
       "\n",
       "[403 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "HSP    403\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add annotations to variants and remove variants that fall in genes not listed in interaction_rich_genes\n",
    "hsp90_variants_file = \"hsp90_variants.vcf\"\n",
    "hsp90_variants_select_file = hsp90_variants_file.replace('.vcf', '_select.vcf')\n",
    "\n",
    "# get VEP annotations and remove variants that fall within genes not in approved list\n",
    "hsp90_variants_select_AFfilter_file = hsp90_variants_select_file.replace('.vcf', '_AFfilter.vcf')\n",
    "vep_output_filename = hsp90_variants_select_AFfilter_file.replace('.vcf', '_VEPoutput.txt')\n",
    "variants_annotated_filename = vep_output_filename.replace('_VEPoutput.txt', '_annotated.txt')\n",
    "variants_annotated_df = annotate_variants_by_VEPoutput(hsp90_variants_select_AFfilter_file, \n",
    "                                                       vep_output_filename, \n",
    "                                                       gff_file, \n",
    "                                                       variants_annotated_filename, \n",
    "                                                       id_colname='var_id')\n",
    "\n",
    "# annotated file can be read in directly if annotation function has been run once before\n",
    "variants_annotated_df = pd.read_csv(variants_annotated_filename, sep='\\t', index_col=0)\n",
    "display(variants_annotated_df)\n",
    "\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "hsp90_design_oligos = variants_annotated_df.query('(Gene_ID.isin(@hsp90_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "\n",
    "# ##### ADDITIONALLY, since this set is focused on hsp90, only include variants editable in at least 3 out of 4 genomes #####\n",
    "# hsp90_design_oligos = hsp90_design_oligos.filter(regex='HSP|HSE|HSD|HSB', axis=0)\n",
    "\n",
    "display(hsp90_design_oligos)\n",
    "# count variants according to which strains they can be edited in\n",
    "display(pd.Series([x[:3] for x in hsp90_design_oligos.index]).value_counts())\n",
    "\n",
    "\n",
    "# write variants to hsp90_design_oligos VCF as a preliminary list to design oligos for\n",
    "hsp90_design_oligos_vcf_file = hsp90_variants_file.replace('.vcf', '_design_oligos_initial.vcf')\n",
    "hsp90_variants_select_AFfilter = vcf.Reader(filename=hsp90_variants_select_AFfilter_file)\n",
    "hsp90_design_oligos_vcf = vcf.Writer(open(hsp90_design_oligos_vcf_file, 'w'), hsp90_variants_select_AFfilter)\n",
    "for record in hsp90_variants_select_AFfilter:\n",
    "    if record.ID in hsp90_design_oligos.index:\n",
    "        hsp90_design_oligos_vcf.write_record(record)\n",
    "hsp90_design_oligos_vcf.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design oligos for initial list, then filter down to variants with at least 2 oligos designed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HSP_00004</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631977</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>56</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>4169|378</td>\n",
       "      <td>3192|7|2712|4314</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00005</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631978</td>\n",
       "      <td>CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>4171|259</td>\n",
       "      <td>3194|9|2593|4195</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>259.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00006</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631979</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.005946</td>\n",
       "      <td>4171|376</td>\n",
       "      <td>3194|9|2710|4312</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>376.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00007</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>53</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.026000</td>\n",
       "      <td>4172|375</td>\n",
       "      <td>3195|10|2709|4311</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>375.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00008</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631981</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>4173|374</td>\n",
       "      <td>3196|11|2708|4310</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>11.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00425</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99064</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>47</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.023000</td>\n",
       "      <td>4168|1432|420|439|2806|3955</td>\n",
       "      <td>2544</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>439.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>420.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00426</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99066</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>7</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.003465</td>\n",
       "      <td>4166|1430|418|441|2808|3957</td>\n",
       "      <td>2542</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>441.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>418.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00427</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99082</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>4150|1414|402|457|2824|3973</td>\n",
       "      <td>2526</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>457.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>402.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00428</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99086</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>4146|1410|398|461|2828|3977</td>\n",
       "      <td>2522</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>461.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00429</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99087</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>4145|1409|397|462|2829|3978</td>\n",
       "      <td>2521</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>462.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>397.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>403 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS                                                REF  \\\n",
       "var_id                                                                       \n",
       "HSP_00004  XIII  631977                                                  A   \n",
       "HSP_00005  XIII  631978  CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...   \n",
       "HSP_00006  XIII  631979                                                  G   \n",
       "HSP_00007  XIII  631980                                                  C   \n",
       "HSP_00008  XIII  631981                                                  A   \n",
       "...         ...     ...                                                ...   \n",
       "HSP_00425   XVI   99064                                                  T   \n",
       "HSP_00426   XVI   99066                                                  G   \n",
       "HSP_00427   XVI   99082                                                  A   \n",
       "HSP_00428   XVI   99086                                                  A   \n",
       "HSP_00429   XVI   99087                                                  G   \n",
       "\n",
       "          ALT  AC    AN        AF        upstream_distance_str  \\\n",
       "var_id                                                           \n",
       "HSP_00004   G  56  2022  0.028000                     4169|378   \n",
       "HSP_00005   C   6  2022  0.002967                     4171|259   \n",
       "HSP_00006   A  12  2018  0.005946                     4171|376   \n",
       "HSP_00007   T  53  2018  0.026000                     4172|375   \n",
       "HSP_00008   G  25  2020  0.012000                     4173|374   \n",
       "...        ..  ..   ...       ...                          ...   \n",
       "HSP_00425   C  47  2014  0.023000  4168|1432|420|439|2806|3955   \n",
       "HSP_00426   A   7  2020  0.003465  4166|1430|418|441|2808|3957   \n",
       "HSP_00427   G   2  2022  0.000989  4150|1414|402|457|2824|3973   \n",
       "HSP_00428   G   2  2022  0.000989  4146|1410|398|461|2828|3977   \n",
       "HSP_00429   A   1  2022  0.000495  4145|1409|397|462|2829|3978   \n",
       "\n",
       "          downstream_distance_str                   region  ... Codons  \\\n",
       "var_id                                                      ...          \n",
       "HSP_00004        3192|7|2712|4314  unidirectional_promoter  ...    NaN   \n",
       "HSP_00005        3194|9|2593|4195  unidirectional_promoter  ...    NaN   \n",
       "HSP_00006        3194|9|2710|4312  unidirectional_promoter  ...    NaN   \n",
       "HSP_00007       3195|10|2709|4311  unidirectional_promoter  ...    NaN   \n",
       "HSP_00008       3196|11|2708|4310  unidirectional_promoter  ...    NaN   \n",
       "...                           ...                      ...  ...    ...   \n",
       "HSP_00425                    2544   bidirectional_promoter  ...    NaN   \n",
       "HSP_00426                    2542   bidirectional_promoter  ...    NaN   \n",
       "HSP_00427                    2526   bidirectional_promoter  ...    NaN   \n",
       "HSP_00428                    2522   bidirectional_promoter  ...    NaN   \n",
       "HSP_00429                    2521   bidirectional_promoter  ...    NaN   \n",
       "\n",
       "          DOMAINS closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                            \n",
       "HSP_00004     NaN                    RTP1               YMR185W   \n",
       "HSP_00005     NaN                    RTP1               YMR185W   \n",
       "HSP_00006     NaN                    RTP1               YMR185W   \n",
       "HSP_00007     NaN                    RTP1               YMR185W   \n",
       "HSP_00008     NaN                    RTP1               YMR185W   \n",
       "...           ...                     ...                   ...   \n",
       "HSP_00425     NaN                   HSP82               YPL240C   \n",
       "HSP_00426     NaN                   HSP82               YPL240C   \n",
       "HSP_00427     NaN                   HSP82               YPL240C   \n",
       "HSP_00428     NaN                   HSP82               YPL240C   \n",
       "HSP_00429     NaN                   HSP82               YPL240C   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "HSP_00004  downstream_gene_variant                    7.0   \n",
       "HSP_00005  downstream_gene_variant                    9.0   \n",
       "HSP_00006  downstream_gene_variant                    9.0   \n",
       "HSP_00007  downstream_gene_variant                   10.0   \n",
       "HSP_00008  downstream_gene_variant                   11.0   \n",
       "...                            ...                    ...   \n",
       "HSP_00425    upstream_gene_variant                  439.0   \n",
       "HSP_00426    upstream_gene_variant                  441.0   \n",
       "HSP_00427    upstream_gene_variant                  457.0   \n",
       "HSP_00428    upstream_gene_variant                  461.0   \n",
       "HSP_00429    upstream_gene_variant                  462.0   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "HSP_00004                   HSC82               YMR186W   \n",
       "HSP_00005                   HSC82               YMR186W   \n",
       "HSP_00006                   HSC82               YMR186W   \n",
       "HSP_00007                   HSC82               YMR186W   \n",
       "HSP_00008                   HSC82               YMR186W   \n",
       "...                           ...                   ...   \n",
       "HSP_00425                    YAR1               YPL239W   \n",
       "HSP_00426                    YAR1               YPL239W   \n",
       "HSP_00427                    YAR1               YPL239W   \n",
       "HSP_00428                    YAR1               YPL239W   \n",
       "HSP_00429                    YAR1               YPL239W   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance  \n",
       "var_id                                                     \n",
       "HSP_00004    upstream_gene_variant                  378.0  \n",
       "HSP_00005    upstream_gene_variant                  259.0  \n",
       "HSP_00006    upstream_gene_variant                  376.0  \n",
       "HSP_00007    upstream_gene_variant                  375.0  \n",
       "HSP_00008    upstream_gene_variant                  374.0  \n",
       "...                            ...                    ...  \n",
       "HSP_00425    upstream_gene_variant                  420.0  \n",
       "HSP_00426    upstream_gene_variant                  418.0  \n",
       "HSP_00427    upstream_gene_variant                  402.0  \n",
       "HSP_00428    upstream_gene_variant                  398.0  \n",
       "HSP_00429    upstream_gene_variant                  397.0  \n",
       "\n",
       "[403 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HSP_00004</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631977</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>56</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>4169|378</td>\n",
       "      <td>3192|7|2712|4314</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00005</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631978</td>\n",
       "      <td>CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>4171|259</td>\n",
       "      <td>3194|9|2593|4195</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>259.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00006</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631979</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.005946</td>\n",
       "      <td>4171|376</td>\n",
       "      <td>3194|9|2710|4312</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>376.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00007</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>53</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.026000</td>\n",
       "      <td>4172|375</td>\n",
       "      <td>3195|10|2709|4311</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>375.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00008</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631981</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>4173|374</td>\n",
       "      <td>3196|11|2708|4310</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>11.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00415</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99017</td>\n",
       "      <td>AGGG</td>\n",
       "      <td>GGGG</td>\n",
       "      <td>209</td>\n",
       "      <td>1914</td>\n",
       "      <td>0.109000</td>\n",
       "      <td>4212|1476|464|392|2759|3908</td>\n",
       "      <td>2588</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>392.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>464.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00416</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99018</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>243</td>\n",
       "      <td>1896</td>\n",
       "      <td>0.128000</td>\n",
       "      <td>4214|1478|466|393|2760|3909</td>\n",
       "      <td>2590</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>393.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>466.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00417</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99019</td>\n",
       "      <td>G</td>\n",
       "      <td>GGGGGGGA</td>\n",
       "      <td>2</td>\n",
       "      <td>1998</td>\n",
       "      <td>0.001001</td>\n",
       "      <td>4212|1476|464|394|2761|3910</td>\n",
       "      <td>2588</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>394.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>464.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00418</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99021</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>4211|1475|463|396|2763|3912</td>\n",
       "      <td>2587</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>396.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>463.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00422</th>\n",
       "      <td>XVI</td>\n",
       "      <td>99032</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCTAAAACATATAAATA</td>\n",
       "      <td>4</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.001984</td>\n",
       "      <td>4199|1463|451|407|2774|3923</td>\n",
       "      <td>2575</td>\n",
       "      <td>bidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSP82</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>407.0</td>\n",
       "      <td>YAR1</td>\n",
       "      <td>YPL239W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>451.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS                                                REF  \\\n",
       "var_id                                                                       \n",
       "HSP_00004  XIII  631977                                                  A   \n",
       "HSP_00005  XIII  631978  CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...   \n",
       "HSP_00006  XIII  631979                                                  G   \n",
       "HSP_00007  XIII  631980                                                  C   \n",
       "HSP_00008  XIII  631981                                                  A   \n",
       "...         ...     ...                                                ...   \n",
       "HSP_00415   XVI   99017                                               AGGG   \n",
       "HSP_00416   XVI   99018                                                  G   \n",
       "HSP_00417   XVI   99019                                                  G   \n",
       "HSP_00418   XVI   99021                                                  G   \n",
       "HSP_00422   XVI   99032                                                  C   \n",
       "\n",
       "                          ALT   AC    AN        AF  \\\n",
       "var_id                                               \n",
       "HSP_00004                   G   56  2022  0.028000   \n",
       "HSP_00005                   C    6  2022  0.002967   \n",
       "HSP_00006                   A   12  2018  0.005946   \n",
       "HSP_00007                   T   53  2018  0.026000   \n",
       "HSP_00008                   G   25  2020  0.012000   \n",
       "...                       ...  ...   ...       ...   \n",
       "HSP_00415                GGGG  209  1914  0.109000   \n",
       "HSP_00416                   A  243  1896  0.128000   \n",
       "HSP_00417            GGGGGGGA    2  1998  0.001001   \n",
       "HSP_00418                   A    1  2014  0.000496   \n",
       "HSP_00422  CCCTAAAACATATAAATA    4  2016  0.001984   \n",
       "\n",
       "                 upstream_distance_str downstream_distance_str  \\\n",
       "var_id                                                           \n",
       "HSP_00004                     4169|378        3192|7|2712|4314   \n",
       "HSP_00005                     4171|259        3194|9|2593|4195   \n",
       "HSP_00006                     4171|376        3194|9|2710|4312   \n",
       "HSP_00007                     4172|375       3195|10|2709|4311   \n",
       "HSP_00008                     4173|374       3196|11|2708|4310   \n",
       "...                                ...                     ...   \n",
       "HSP_00415  4212|1476|464|392|2759|3908                    2588   \n",
       "HSP_00416  4214|1478|466|393|2760|3909                    2590   \n",
       "HSP_00417  4212|1476|464|394|2761|3910                    2588   \n",
       "HSP_00418  4211|1475|463|396|2763|3912                    2587   \n",
       "HSP_00422  4199|1463|451|407|2774|3923                    2575   \n",
       "\n",
       "                            region  ... DOMAINS closest_gene1_Gene_Name  \\\n",
       "var_id                              ...                                   \n",
       "HSP_00004  unidirectional_promoter  ...     NaN                    RTP1   \n",
       "HSP_00005  unidirectional_promoter  ...     NaN                    RTP1   \n",
       "HSP_00006  unidirectional_promoter  ...     NaN                    RTP1   \n",
       "HSP_00007  unidirectional_promoter  ...     NaN                    RTP1   \n",
       "HSP_00008  unidirectional_promoter  ...     NaN                    RTP1   \n",
       "...                            ...  ...     ...                     ...   \n",
       "HSP_00415   bidirectional_promoter  ...     NaN                   HSP82   \n",
       "HSP_00416   bidirectional_promoter  ...     NaN                   HSP82   \n",
       "HSP_00417   bidirectional_promoter  ...     NaN                   HSP82   \n",
       "HSP_00418   bidirectional_promoter  ...     NaN                   HSP82   \n",
       "HSP_00422   bidirectional_promoter  ...     NaN                   HSP82   \n",
       "\n",
       "          closest_gene1_Gene_ID closest_gene1_Annotation  \\\n",
       "var_id                                                     \n",
       "HSP_00004               YMR185W  downstream_gene_variant   \n",
       "HSP_00005               YMR185W  downstream_gene_variant   \n",
       "HSP_00006               YMR185W  downstream_gene_variant   \n",
       "HSP_00007               YMR185W  downstream_gene_variant   \n",
       "HSP_00008               YMR185W  downstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "HSP_00415               YPL240C    upstream_gene_variant   \n",
       "HSP_00416               YPL240C    upstream_gene_variant   \n",
       "HSP_00417               YPL240C    upstream_gene_variant   \n",
       "HSP_00418               YPL240C    upstream_gene_variant   \n",
       "HSP_00422               YPL240C    upstream_gene_variant   \n",
       "\n",
       "          closest_gene1_Distance closest_gene2_Gene_Name  \\\n",
       "var_id                                                     \n",
       "HSP_00004                    7.0                   HSC82   \n",
       "HSP_00005                    9.0                   HSC82   \n",
       "HSP_00006                    9.0                   HSC82   \n",
       "HSP_00007                   10.0                   HSC82   \n",
       "HSP_00008                   11.0                   HSC82   \n",
       "...                          ...                     ...   \n",
       "HSP_00415                  392.0                    YAR1   \n",
       "HSP_00416                  393.0                    YAR1   \n",
       "HSP_00417                  394.0                    YAR1   \n",
       "HSP_00418                  396.0                    YAR1   \n",
       "HSP_00422                  407.0                    YAR1   \n",
       "\n",
       "          closest_gene2_Gene_ID closest_gene2_Annotation  \\\n",
       "var_id                                                     \n",
       "HSP_00004               YMR186W    upstream_gene_variant   \n",
       "HSP_00005               YMR186W    upstream_gene_variant   \n",
       "HSP_00006               YMR186W    upstream_gene_variant   \n",
       "HSP_00007               YMR186W    upstream_gene_variant   \n",
       "HSP_00008               YMR186W    upstream_gene_variant   \n",
       "...                         ...                      ...   \n",
       "HSP_00415               YPL239W    upstream_gene_variant   \n",
       "HSP_00416               YPL239W    upstream_gene_variant   \n",
       "HSP_00417               YPL239W    upstream_gene_variant   \n",
       "HSP_00418               YPL239W    upstream_gene_variant   \n",
       "HSP_00422               YPL239W    upstream_gene_variant   \n",
       "\n",
       "          closest_gene2_Distance num_of_oligos  \n",
       "var_id                                          \n",
       "HSP_00004                  378.0           2.0  \n",
       "HSP_00005                  259.0           2.0  \n",
       "HSP_00006                  376.0           2.0  \n",
       "HSP_00007                  375.0           2.0  \n",
       "HSP_00008                  374.0           2.0  \n",
       "...                          ...           ...  \n",
       "HSP_00415                  464.0           5.0  \n",
       "HSP_00416                  466.0           5.0  \n",
       "HSP_00417                  464.0           5.0  \n",
       "HSP_00418                  463.0           3.0  \n",
       "HSP_00422                  451.0           2.0  \n",
       "\n",
       "[173 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hsp90_oligo_file = os.path.expanduser(\"~/crispey3/initial_design/Output/all_SNPs_hsp90_initial_GG_9bp_OLIGO.tab\")\n",
    "hsp90_oligo_df = pd.read_csv(hsp90_oligo_file, sep='\\t')\n",
    "\n",
    "hsp90_design_oligos = pd.read_csv(\"/home/users/rang/scratch/yeast/hsp90/hsp90_variants_select_AFfilter_annotated.txt\", sep='\\t', index_col=0)\n",
    "# remove variants that fall within genes, but are not in the approved genes list\n",
    "intergenic_regions_list = ['bidirectional_promoter', 'unidirectional_promoter', 'intergenic']\n",
    "hsp90_design_oligos = hsp90_design_oligos.query('(Gene_ID.isin(@hsp90_genes.index) | region.isin(@intergenic_regions_list))').copy()\n",
    "display(hsp90_design_oligos)\n",
    "\n",
    "hsp90_design_oligos['num_of_oligos'] = hsp90_oligo_df.groupby('var_id').apply(len)\n",
    "# select oligos that have at least 2 oligos designed\n",
    "hsp90_oligos_final = hsp90_design_oligos.query('num_of_oligos>1').copy()\n",
    "\n",
    "display(hsp90_oligos_final)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We can also design synthetic variants for natural HSP90 variants. Take every coding variant and design alternative ALT alleles in its place"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "hsp90_design_oligos_vcf_file = \"/home/users/rang/scratch/yeast/hsp90/hsp90_variants_design_oligos_initial.vcf\"\n",
    "hsp90_synthetic_vars_vcf_file = hsp90_design_oligos_vcf_file.replace('.vcf', '_synthetic_vars.vcf')\n",
    "\n",
    "hsp90_design_oligos_vcf = vcf.Reader(filename=hsp90_design_oligos_vcf_file)\n",
    "hsp90_synthetic_vars_vcf = vcf.Writer(open(hsp90_synthetic_vars_vcf_file, 'w'), hsp90_design_oligos_vcf)\n",
    "\n",
    "syn_var_counter = 0\n",
    "syn_var_prefix = 'HSX'\n",
    "syn_alt_alleles = ['A', 'C', 'T', 'G']\n",
    "syn_insert_dict = {'A':'C',\n",
    "                   'C':'T',\n",
    "                   'T':'G',\n",
    "                   'G':'A',}\n",
    "\n",
    "for record in hsp90_design_oligos_vcf:\n",
    "    if record.ID in hsp90_oligos_final.query('~region.isin(@intergenic_regions_list)').index:\n",
    "        # synthesize single-base alt alleles different from natural alt allele\n",
    "        ref_allele = record.REF[0]\n",
    "        for alt_allele in syn_alt_alleles:\n",
    "            if (alt_allele != ref_allele) and (alt_allele != str(record.ALT[0])):\n",
    "                syn_var_counter += 1\n",
    "                syn = vcf.model._Record(CHROM=record.CHROM, POS=record.POS,\n",
    "                                        ID=\"{}_{:05d}\".format(syn_var_prefix, syn_var_counter), \n",
    "                                        REF=ref_allele, ALT=[vcf.model._Substitution(alt_allele)],\n",
    "                                        QUAL=1, FILTER='.',\n",
    "                                        INFO={'AC':[0], 'AF':[0], 'AN':record.INFO['AN'], 'TGT':record.INFO['TGT']},\n",
    "                                        FORMAT=None, sample_indexes=None)\n",
    "                # write record to synthetic variants VCF\n",
    "                hsp90_synthetic_vars_vcf.write_record(syn)\n",
    "        \n",
    "        # also synthesize 1bp insertion if record is SNP\n",
    "        if len(record.REF) == len(record.ALT[0]):\n",
    "            insertion = ref_allele + syn_insert_dict[ref_allele]\n",
    "            syn_var_counter += 1\n",
    "            syn = vcf.model._Record(CHROM=record.CHROM, POS=record.POS,\n",
    "                                    ID=\"{}_{:05d}\".format(syn_var_prefix, syn_var_counter), \n",
    "                                    REF=ref_allele, ALT=[vcf.model._Substitution(insertion)],\n",
    "                                    QUAL=1, FILTER='.',\n",
    "                                    INFO={'AC':[0], 'AF':[0], 'AN':record.INFO['AN'], 'TGT':record.INFO['TGT']},\n",
    "                                    FORMAT=None, sample_indexes=None)\n",
    "            # write record to synthetic variants VCF\n",
    "            hsp90_synthetic_vars_vcf.write_record(syn)\n",
    "            \n",
    "hsp90_synthetic_vars_vcf.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate synthetic variants with VEP, select subset to complement natural variants in hsp90_oligos_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>CDS_frac</th>\n",
       "      <th>AA_len</th>\n",
       "      <th>AA_frac</th>\n",
       "      <th>cDNA_pos</th>\n",
       "      <th>CDS_pos</th>\n",
       "      <th>AA_pos</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HSX_00001</th>\n",
       "      <td>XIII</td>\n",
       "      <td>632471</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4663</td>\n",
       "      <td>3686|501|2218|3820</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0.055241</td>\n",
       "      <td>706.0</td>\n",
       "      <td>0.055241</td>\n",
       "      <td>117.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>S</td>\n",
       "      <td>tcC/tcA</td>\n",
       "      <td>CDD:cd16927,PIRSF:PIRSF002583,Pfam:PF02518,Gen...</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00002</th>\n",
       "      <td>XIII</td>\n",
       "      <td>632471</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4663</td>\n",
       "      <td>3686|501|2218|3820</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0.055241</td>\n",
       "      <td>706.0</td>\n",
       "      <td>0.055241</td>\n",
       "      <td>117.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>S</td>\n",
       "      <td>tcC/tcG</td>\n",
       "      <td>CDD:cd16927,PIRSF:PIRSF002583,Pfam:PF02518,Gen...</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00003</th>\n",
       "      <td>XIII</td>\n",
       "      <td>632471</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>0</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4663</td>\n",
       "      <td>3686|501|2217|3819</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00004</th>\n",
       "      <td>XIII</td>\n",
       "      <td>632477</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4669</td>\n",
       "      <td>3692|507|2212|3814</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0.058074</td>\n",
       "      <td>706.0</td>\n",
       "      <td>0.058074</td>\n",
       "      <td>123.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>A</td>\n",
       "      <td>gcT/gcA</td>\n",
       "      <td>CDD:cd16927,PIRSF:PIRSF002583,Pfam:PF02518,Gen...</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00005</th>\n",
       "      <td>XIII</td>\n",
       "      <td>632477</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4669</td>\n",
       "      <td>3692|507|2212|3814</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0.058074</td>\n",
       "      <td>706.0</td>\n",
       "      <td>0.058074</td>\n",
       "      <td>123.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>A</td>\n",
       "      <td>gcT/gcC</td>\n",
       "      <td>CDD:cd16927,PIRSF:PIRSF002583,Pfam:PF02518,Gen...</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00339</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98464</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>2002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00340</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98464</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>0</td>\n",
       "      <td>2002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00341</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98503</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>2002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00342</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98503</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>2002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00343</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98503</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>0</td>\n",
       "      <td>2002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>343 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS REF ALT  AC    AN   AF upstream_distance_str  \\\n",
       "var_id                                                                 \n",
       "HSX_00001  XIII  632471   C   A   0  2018  0.0                  4663   \n",
       "HSX_00002  XIII  632471   C   G   0  2018  0.0                  4663   \n",
       "HSX_00003  XIII  632471   C  CT   0  2018  0.0                  4663   \n",
       "HSX_00004  XIII  632477   T   A   0  2022  0.0                  4669   \n",
       "HSX_00005  XIII  632477   T   C   0  2022  0.0                  4669   \n",
       "...         ...     ...  ..  ..  ..   ...  ...                   ...   \n",
       "HSX_00339   XVI   98464   C   G   0  2002  0.0                   NaN   \n",
       "HSX_00340   XVI   98464   C  CT   0  2002  0.0                   NaN   \n",
       "HSX_00341   XVI   98503   C   T   0  2002  0.0                   NaN   \n",
       "HSX_00342   XVI   98503   C   G   0  2002  0.0                   NaN   \n",
       "HSX_00343   XVI   98503   C  CT   0  2002  0.0                   NaN   \n",
       "\n",
       "          downstream_distance_str              region  ...  CDS_frac AA_len  \\\n",
       "var_id                                                 ...                    \n",
       "HSX_00001      3686|501|2218|3820  synonymous_variant  ...  0.055241  706.0   \n",
       "HSX_00002      3686|501|2218|3820  synonymous_variant  ...  0.055241  706.0   \n",
       "HSX_00003      3686|501|2217|3819  frameshift_variant  ...       NaN    NaN   \n",
       "HSX_00004      3692|507|2212|3814  synonymous_variant  ...  0.058074  706.0   \n",
       "HSX_00005      3692|507|2212|3814  synonymous_variant  ...  0.058074  706.0   \n",
       "...                           ...                 ...  ...       ...    ...   \n",
       "HSX_00339                     NaN                 NaN  ...       NaN    NaN   \n",
       "HSX_00340                     NaN                 NaN  ...       NaN    NaN   \n",
       "HSX_00341                     NaN                 NaN  ...       NaN    NaN   \n",
       "HSX_00342                     NaN                 NaN  ...       NaN    NaN   \n",
       "HSX_00343                     NaN                 NaN  ...       NaN    NaN   \n",
       "\n",
       "            AA_frac cDNA_pos CDS_pos  AA_pos Amino_acids   Codons  \\\n",
       "var_id                                                              \n",
       "HSX_00001  0.055241    117.0   117.0    39.0           S  tcC/tcA   \n",
       "HSX_00002  0.055241    117.0   117.0    39.0           S  tcC/tcG   \n",
       "HSX_00003       NaN      NaN     NaN     NaN         NaN      NaN   \n",
       "HSX_00004  0.058074    123.0   123.0    41.0           A  gcT/gcA   \n",
       "HSX_00005  0.058074    123.0   123.0    41.0           A  gcT/gcC   \n",
       "...             ...      ...     ...     ...         ...      ...   \n",
       "HSX_00339       NaN      NaN     NaN     NaN         NaN      NaN   \n",
       "HSX_00340       NaN      NaN     NaN     NaN         NaN      NaN   \n",
       "HSX_00341       NaN      NaN     NaN     NaN         NaN      NaN   \n",
       "HSX_00342       NaN      NaN     NaN     NaN         NaN      NaN   \n",
       "HSX_00343       NaN      NaN     NaN     NaN         NaN      NaN   \n",
       "\n",
       "                                                     DOMAINS num_of_oligos  \n",
       "var_id                                                                      \n",
       "HSX_00001  CDD:cd16927,PIRSF:PIRSF002583,Pfam:PF02518,Gen...           2.0  \n",
       "HSX_00002  CDD:cd16927,PIRSF:PIRSF002583,Pfam:PF02518,Gen...           2.0  \n",
       "HSX_00003                                                NaN           2.0  \n",
       "HSX_00004  CDD:cd16927,PIRSF:PIRSF002583,Pfam:PF02518,Gen...           2.0  \n",
       "HSX_00005  CDD:cd16927,PIRSF:PIRSF002583,Pfam:PF02518,Gen...           2.0  \n",
       "...                                                      ...           ...  \n",
       "HSX_00339                                                NaN           2.0  \n",
       "HSX_00340                                                NaN           2.0  \n",
       "HSX_00341                                                NaN           2.0  \n",
       "HSX_00342                                                NaN           2.0  \n",
       "HSX_00343                                                NaN           2.0  \n",
       "\n",
       "[343 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/groups/hbfraser/modules/packages/conda/4.6.14/envs/crispey-bar/lib/python3.7/site-packages/ipykernel_launcher.py:36: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>upstream_distance_str</th>\n",
       "      <th>downstream_distance_str</th>\n",
       "      <th>region</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>closest_gene1_Gene_Name</th>\n",
       "      <th>closest_gene1_Gene_ID</th>\n",
       "      <th>closest_gene1_Annotation</th>\n",
       "      <th>closest_gene1_Distance</th>\n",
       "      <th>closest_gene2_Gene_Name</th>\n",
       "      <th>closest_gene2_Gene_ID</th>\n",
       "      <th>closest_gene2_Annotation</th>\n",
       "      <th>closest_gene2_Distance</th>\n",
       "      <th>num_of_oligos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HSP_00004</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631977</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>56</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>4169|378</td>\n",
       "      <td>3192|7|2712|4314</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>7.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>378.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00005</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631978</td>\n",
       "      <td>CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>4171|259</td>\n",
       "      <td>3194|9|2593|4195</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>259.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00006</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631979</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.005946</td>\n",
       "      <td>4171|376</td>\n",
       "      <td>3194|9|2710|4312</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>9.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>376.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00007</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>53</td>\n",
       "      <td>2018</td>\n",
       "      <td>0.026000</td>\n",
       "      <td>4172|375</td>\n",
       "      <td>3195|10|2709|4311</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>10.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>375.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP_00008</th>\n",
       "      <td>XIII</td>\n",
       "      <td>631981</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25</td>\n",
       "      <td>2020</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>4173|374</td>\n",
       "      <td>3196|11|2708|4310</td>\n",
       "      <td>unidirectional_promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RTP1</td>\n",
       "      <td>YMR185W</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>11.0</td>\n",
       "      <td>HSC82</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>374.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00327</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98320</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>1996</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4864|2128|1116|2109|3258</td>\n",
       "      <td>3240</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00329</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98335</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4819|2083|1071|2155|3304</td>\n",
       "      <td>3195</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene3D:3.30.565.10,HAMAP:MF_00505,Pfam:PF02518...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00330</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98335</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>2012</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4818|2082|1070|2155|3304</td>\n",
       "      <td>3194</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00334</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98367</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>0</td>\n",
       "      <td>2016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4729|1993|981|2245|3394</td>\n",
       "      <td>3105</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene3D:3.30.565.10,HAMAP:MF_00505,Pfam:PF02518...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSX_00336</th>\n",
       "      <td>XVI</td>\n",
       "      <td>98413</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>2004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4728|1992|980|2245|3394</td>\n",
       "      <td>3104</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>332 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM     POS                                                REF  \\\n",
       "var_id                                                                       \n",
       "HSP_00004  XIII  631977                                                  A   \n",
       "HSP_00005  XIII  631978  CGCATTTGATTATAATTTGCTTCTTAGGCAAAATTAATATTTACGT...   \n",
       "HSP_00006  XIII  631979                                                  G   \n",
       "HSP_00007  XIII  631980                                                  C   \n",
       "HSP_00008  XIII  631981                                                  A   \n",
       "...         ...     ...                                                ...   \n",
       "HSX_00327   XVI   98320                                                  T   \n",
       "HSX_00329   XVI   98335                                                  G   \n",
       "HSX_00330   XVI   98335                                                  G   \n",
       "HSX_00334   XVI   98367                                                  C   \n",
       "HSX_00336   XVI   98413                                                  C   \n",
       "\n",
       "          ALT  AC    AN        AF     upstream_distance_str  \\\n",
       "var_id                                                        \n",
       "HSP_00004   G  56  2022  0.028000                  4169|378   \n",
       "HSP_00005   C   6  2022  0.002967                  4171|259   \n",
       "HSP_00006   A  12  2018  0.005946                  4171|376   \n",
       "HSP_00007   T  53  2018  0.026000                  4172|375   \n",
       "HSP_00008   G  25  2020  0.012000                  4173|374   \n",
       "...        ..  ..   ...       ...                       ...   \n",
       "HSX_00327   G   0  1996  0.000000  4864|2128|1116|2109|3258   \n",
       "HSX_00329   A   0  2012  0.000000  4819|2083|1071|2155|3304   \n",
       "HSX_00330   C   0  2012  0.000000  4818|2082|1070|2155|3304   \n",
       "HSX_00334  CT   0  2016  0.000000   4729|1993|981|2245|3394   \n",
       "HSX_00336   G   0  2004  0.000000   4728|1992|980|2245|3394   \n",
       "\n",
       "          downstream_distance_str                   region  ...  \\\n",
       "var_id                                                      ...   \n",
       "HSP_00004        3192|7|2712|4314  unidirectional_promoter  ...   \n",
       "HSP_00005        3194|9|2593|4195  unidirectional_promoter  ...   \n",
       "HSP_00006        3194|9|2710|4312  unidirectional_promoter  ...   \n",
       "HSP_00007       3195|10|2709|4311  unidirectional_promoter  ...   \n",
       "HSP_00008       3196|11|2708|4310  unidirectional_promoter  ...   \n",
       "...                           ...                      ...  ...   \n",
       "HSX_00327                    3240       frameshift_variant  ...   \n",
       "HSX_00329                    3195         missense_variant  ...   \n",
       "HSX_00330                    3194       frameshift_variant  ...   \n",
       "HSX_00334                    3105       synonymous_variant  ...   \n",
       "HSX_00336                    3104       frameshift_variant  ...   \n",
       "\n",
       "                                                     DOMAINS  \\\n",
       "var_id                                                         \n",
       "HSP_00004                                                NaN   \n",
       "HSP_00005                                                NaN   \n",
       "HSP_00006                                                NaN   \n",
       "HSP_00007                                                NaN   \n",
       "HSP_00008                                                NaN   \n",
       "...                                                      ...   \n",
       "HSX_00327                                                NaN   \n",
       "HSX_00329  Gene3D:3.30.565.10,HAMAP:MF_00505,Pfam:PF02518...   \n",
       "HSX_00330                                                NaN   \n",
       "HSX_00334  Gene3D:3.30.565.10,HAMAP:MF_00505,Pfam:PF02518...   \n",
       "HSX_00336                                                NaN   \n",
       "\n",
       "          closest_gene1_Gene_Name closest_gene1_Gene_ID  \\\n",
       "var_id                                                    \n",
       "HSP_00004                    RTP1               YMR185W   \n",
       "HSP_00005                    RTP1               YMR185W   \n",
       "HSP_00006                    RTP1               YMR185W   \n",
       "HSP_00007                    RTP1               YMR185W   \n",
       "HSP_00008                    RTP1               YMR185W   \n",
       "...                           ...                   ...   \n",
       "HSX_00327                     NaN                   NaN   \n",
       "HSX_00329                     NaN                   NaN   \n",
       "HSX_00330                     NaN                   NaN   \n",
       "HSX_00334                     NaN                   NaN   \n",
       "HSX_00336                     NaN                   NaN   \n",
       "\n",
       "          closest_gene1_Annotation closest_gene1_Distance  \\\n",
       "var_id                                                      \n",
       "HSP_00004  downstream_gene_variant                    7.0   \n",
       "HSP_00005  downstream_gene_variant                    9.0   \n",
       "HSP_00006  downstream_gene_variant                    9.0   \n",
       "HSP_00007  downstream_gene_variant                   10.0   \n",
       "HSP_00008  downstream_gene_variant                   11.0   \n",
       "...                            ...                    ...   \n",
       "HSX_00327                      NaN                    NaN   \n",
       "HSX_00329                      NaN                    NaN   \n",
       "HSX_00330                      NaN                    NaN   \n",
       "HSX_00334                      NaN                    NaN   \n",
       "HSX_00336                      NaN                    NaN   \n",
       "\n",
       "          closest_gene2_Gene_Name closest_gene2_Gene_ID  \\\n",
       "var_id                                                    \n",
       "HSP_00004                   HSC82               YMR186W   \n",
       "HSP_00005                   HSC82               YMR186W   \n",
       "HSP_00006                   HSC82               YMR186W   \n",
       "HSP_00007                   HSC82               YMR186W   \n",
       "HSP_00008                   HSC82               YMR186W   \n",
       "...                           ...                   ...   \n",
       "HSX_00327                     NaN                   NaN   \n",
       "HSX_00329                     NaN                   NaN   \n",
       "HSX_00330                     NaN                   NaN   \n",
       "HSX_00334                     NaN                   NaN   \n",
       "HSX_00336                     NaN                   NaN   \n",
       "\n",
       "          closest_gene2_Annotation closest_gene2_Distance num_of_oligos  \n",
       "var_id                                                                   \n",
       "HSP_00004    upstream_gene_variant                  378.0           2.0  \n",
       "HSP_00005    upstream_gene_variant                  259.0           2.0  \n",
       "HSP_00006    upstream_gene_variant                  376.0           2.0  \n",
       "HSP_00007    upstream_gene_variant                  375.0           2.0  \n",
       "HSP_00008    upstream_gene_variant                  374.0           2.0  \n",
       "...                            ...                    ...           ...  \n",
       "HSX_00327                      NaN                    NaN           2.0  \n",
       "HSX_00329                      NaN                    NaN           3.0  \n",
       "HSX_00330                      NaN                    NaN           3.0  \n",
       "HSX_00334                      NaN                    NaN           2.0  \n",
       "HSX_00336                      NaN                    NaN           2.0  \n",
       "\n",
       "[332 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get VEP annotations and remove variants that fall within genes not in approved list\n",
    "hsp90_synthetic_vars_vcf_file = hsp90_design_oligos_vcf_file.replace('.vcf', '_synthetic_vars.vcf')\n",
    "vep_output_filename = hsp90_synthetic_vars_vcf_file.replace('.vcf', '_VEPoutput.txt')\n",
    "variants_annotated_filename = vep_output_filename.replace('_VEPoutput.txt', '_annotated.txt')\n",
    "variants_annotated_df = annotate_variants_by_VEPoutput(hsp90_synthetic_vars_vcf_file, \n",
    "                                                       vep_output_filename, \n",
    "                                                       gff_file, \n",
    "                                                       variants_annotated_filename, \n",
    "                                                       id_colname='var_id')\n",
    "\n",
    "# annotated file can be read in directly if annotation function has been run once before\n",
    "variants_annotated_df = pd.read_csv(variants_annotated_filename, sep='\\t', index_col=0)\n",
    "# add number of oligos based on that generated for natural variants\n",
    "for v, row in hsp90_oligos_final.iterrows():\n",
    "    variants_annotated_df.loc[(variants_annotated_df.CHROM==row['CHROM']) & (variants_annotated_df.POS==row['POS']), 'num_of_oligos'] = row['num_of_oligos']\n",
    "\n",
    "display(variants_annotated_df)\n",
    "\n",
    "\n",
    "# select one synthetic variant to match each natural coding mutation\n",
    "syn_vars_to_keep = []\n",
    "for v, row in hsp90_oligos_final.iterrows():\n",
    "    if row['region'] not in intergenic_regions_list:\n",
    "        df = variants_annotated_df.query('CHROM==@row.CHROM & POS==@row.POS')\n",
    "        # for every synonymous variant, select a synthetic variants that generates a MODERATE or HIGH annotation impact\n",
    "        if row['region'] == 'synonymous_variant':\n",
    "            syn_vars_to_keep += df.query('Annotation_Impact==\"MODERATE\"').head(1).index.tolist()\n",
    "            syn_vars_to_keep += df.query('Annotation_Impact==\"HIGH\"').head(1).index.tolist()\n",
    "        # for all other regions, select a synonymous variant with LOW annotation impact\n",
    "        else:\n",
    "            syn_vars_to_keep += df.query('Annotation_Impact==\"LOW\"').head(1).index.tolist()\n",
    "\n",
    "# display(variants_annotated_df.loc[syn_vars_to_keep])\n",
    "            \n",
    "# add selected synthetic variants to final list\n",
    "hsp90_oligos_final = pd.concat([hsp90_oligos_final, variants_annotated_df.loc[syn_vars_to_keep]])[hsp90_oligos_final.columns]\n",
    "display(hsp90_oligos_final)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write hsp90_oligos_final table to file\n",
    "hsp90_oligos_final.to_csv(\"/home/users/rang/scratch/yeast/hsp90/hsp90_variants_final.txt\", sep='\\t')\n",
    "# use this file in separate script to organize into pools\n",
    "\n",
    "# write natural and synthetic variants into single VCF file\n",
    "template = vcf.Reader(filename=\"/home/users/rang/scratch/yeast/hsp90/hsp90_variants_design_oligos_initial.vcf\")\n",
    "hsp90_variants_final_vcf = vcf.Writer(open(\"/home/users/rang/scratch/yeast/hsp90/hsp90_variants_final.vcf\", 'w'), template)\n",
    " \n",
    "for record in template:\n",
    "    if record.ID in hsp90_oligos_final.index:\n",
    "        hsp90_variants_final_vcf.write_record(record)\n",
    "        \n",
    "for record in vcf.Reader(filename=\"/home/users/rang/scratch/yeast/hsp90/hsp90_variants_design_oligos_initial_synthetic_vars.vcf\"):\n",
    "    if record.ID in hsp90_oligos_final.index:\n",
    "        hsp90_variants_final_vcf.write_record(record)\n",
    "\n",
    "hsp90_variants_final_vcf.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of HSP90 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants: 332\n",
      "Number of oligos: 781\n",
      "Number of positions: 173\n",
      "\n",
      "Distribution of variants among HSP90 genes\n",
      "count      2.000000\n",
      "mean     156.500000\n",
      "std       36.062446\n",
      "min      131.000000\n",
      "25%      143.750000\n",
      "50%      156.500000\n",
      "75%      169.250000\n",
      "max      182.000000\n",
      "Name: Gene_ID, dtype: float64\n",
      "\n",
      "Distribution of oligos among HSP90 variants\n",
      "count    332.000000\n",
      "mean       2.352410\n",
      "std        0.772973\n",
      "min        2.000000\n",
      "25%        2.000000\n",
      "50%        2.000000\n",
      "75%        2.000000\n",
      "max        6.000000\n",
      "Name: num_of_oligos, dtype: float64\n",
      "\n",
      "Proportion of variant annotation classes\n",
      "synonymous_variant         0.307229\n",
      "frameshift_variant         0.271084\n",
      "missense_variant           0.234940\n",
      "unidirectional_promoter    0.093373\n",
      "bidirectional_promoter     0.066265\n",
      "intergenic                 0.018072\n",
      "stop_gained                0.009036\n",
      "Name: region, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "### SUMMARY OF HSP90 SET ###\n",
    "print(\"Number of variants:\", len(hsp90_oligos_final))\n",
    "print(\"Number of oligos:\", int(sum(hsp90_oligos_final.num_of_oligos)))\n",
    "print(\"Number of positions:\", len(hsp90_oligos_final[['CHROM','POS']].drop_duplicates()))\n",
    "print()\n",
    "print(\"Distribution of variants among HSP90 genes\")\n",
    "print(hsp90_oligos_final.Gene_ID.value_counts().filter(hsp90_genes.index).describe())\n",
    "print()\n",
    "print(\"Distribution of oligos among HSP90 variants\")\n",
    "print(hsp90_oligos_final.num_of_oligos.describe())\n",
    "print()\n",
    "print(\"Proportion of variant annotation classes\")\n",
    "print(hsp90_oligos_final.region.value_counts()/len(hsp90_oligos_final))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine number of pools to use for HSP90 oligos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HSP    419.0\n",
       "HSX    362.0\n",
       "Name: num_of_oligos, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of oligos to fit: 781\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(6, 7)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use 6 pools:\n",
      "Number of oligos per pool: 118\n",
      "Number of oligos leftover: 73\n",
      "Number of technical oligos: 18\n",
      "\n",
      "Use 7 pools:\n",
      "Number of oligos per pool: 111\n",
      "Number of oligos leftover: 4\n",
      "Number of technical oligos: 70\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([6, 7])"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count OLIGOS in each variant prefix set\n",
    "display(hsp90_oligos_final.groupby(lambda x: x[:3]).num_of_oligos.sum().sort_values(ascending=False))\n",
    "\n",
    "# determine number of pools to use\n",
    "num_of_oligos = int(hsp90_oligos_final.num_of_oligos.sum())\n",
    "min_pool_size = 105 # allows 12 technical oligos\n",
    "max_pool_size = 118 # allows 3 technical oligos\n",
    "complete_pool_size = 121 # maximum oligos synthesized per pool\n",
    "\n",
    "# find number of pools to use\n",
    "find_pool_size_options(num_of_oligos, min_pool_size, max_pool_size, complete_pool_size)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Humanized yeast pilot\n",
    "An assortment of oligos generated from two sets of variant files were combined to get the final list being synthesized\n",
    "- Fritz Roth curated variants list, taken direct from hYeast ORF fasta (Priority)\n",
    "- Shi-An's compiled variants from GnomAD (Secondary)\n",
    "\n",
    "Assemble oligos such that for each gene, we can synthesize 118 oligos.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>SNP_chr_pos</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>guide_id</th>\n",
       "      <th>guide_strand</th>\n",
       "      <th>guide0_chr_pos</th>\n",
       "      <th>guide_cut_chr_pos</th>\n",
       "      <th>SNP_pos_in_guide</th>\n",
       "      <th>...</th>\n",
       "      <th>donor_seq_shift</th>\n",
       "      <th>donor_mut_pos_in_guide</th>\n",
       "      <th>donor_info_str</th>\n",
       "      <th>set_name</th>\n",
       "      <th>contains_excluded_seqs</th>\n",
       "      <th>pool</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>barcode_id</th>\n",
       "      <th>oligo_seq</th>\n",
       "      <th>oligo_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DHFR_10007</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>457</td>\n",
       "      <td>A</td>\n",
       "      <td>[T]</td>\n",
       "      <td>guide_20499</td>\n",
       "      <td>+</td>\n",
       "      <td>460</td>\n",
       "      <td>457</td>\n",
       "      <td>-4</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-4</td>\n",
       "      <td>REF2ALT:A&gt;T</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>DHFR#guide_20499#guide_20499:REF2ALT:offset14:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DHFR_10007</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>457</td>\n",
       "      <td>A</td>\n",
       "      <td>[T]</td>\n",
       "      <td>guide_20500</td>\n",
       "      <td>-</td>\n",
       "      <td>449</td>\n",
       "      <td>453</td>\n",
       "      <td>-7</td>\n",
       "      <td>...</td>\n",
       "      <td>-14</td>\n",
       "      <td>-7</td>\n",
       "      <td>REF2ALT:A&gt;T</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>DHFR#guide_20500#guide_20500:REF2ALT:offset-14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DHFR_10008</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>237</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>guide_20501</td>\n",
       "      <td>+</td>\n",
       "      <td>241</td>\n",
       "      <td>238</td>\n",
       "      <td>-5</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-5</td>\n",
       "      <td>REF2ALT:C&gt;T</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>DHFR#guide_20501#guide_20501:REF2ALT:offset14:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DHFR_10104</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>186</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>guide_20502</td>\n",
       "      <td>-</td>\n",
       "      <td>180</td>\n",
       "      <td>184</td>\n",
       "      <td>-5</td>\n",
       "      <td>...</td>\n",
       "      <td>-14</td>\n",
       "      <td>-5</td>\n",
       "      <td>REF2ALT:G&gt;C</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>DHFR#guide_20502#guide_20502:REF2ALT:offset-14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DHFR_00019</td>\n",
       "      <td>chr5</td>\n",
       "      <td>79929796</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>guide_20503</td>\n",
       "      <td>+</td>\n",
       "      <td>79929793</td>\n",
       "      <td>79929790</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>REF2ALT:G&gt;C</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>DHFR#guide_20503#guide_20503:REF2ALT:offset14:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655</th>\n",
       "      <td>UROS_00020</td>\n",
       "      <td>chr10</td>\n",
       "      <td>127477526</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>guide_21154</td>\n",
       "      <td>+</td>\n",
       "      <td>127477531</td>\n",
       "      <td>127477528</td>\n",
       "      <td>-6</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-6</td>\n",
       "      <td>REF2ALT:G&gt;A</td>\n",
       "      <td>UROS</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>UROS#guide_21154#guide_21154:REF2ALT:offset14:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>656</th>\n",
       "      <td>UROS_00020</td>\n",
       "      <td>chr10</td>\n",
       "      <td>127477526</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>guide_21155</td>\n",
       "      <td>+</td>\n",
       "      <td>127477534</td>\n",
       "      <td>127477531</td>\n",
       "      <td>-9</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-9</td>\n",
       "      <td>REF2ALT:G&gt;A</td>\n",
       "      <td>UROS</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>UROS#guide_21155#guide_21155:REF2ALT:offset14:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>UROS_00029</td>\n",
       "      <td>chr10</td>\n",
       "      <td>127477556</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>guide_21156</td>\n",
       "      <td>+</td>\n",
       "      <td>127477556</td>\n",
       "      <td>127477553</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-1</td>\n",
       "      <td>REF2ALT:T&gt;C</td>\n",
       "      <td>UROS</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>UROS#guide_21156#guide_21156:REF2ALT:offset14:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>UROS_00029</td>\n",
       "      <td>chr10</td>\n",
       "      <td>127477556</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>guide_21157</td>\n",
       "      <td>+</td>\n",
       "      <td>127477557</td>\n",
       "      <td>127477554</td>\n",
       "      <td>-2</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-2</td>\n",
       "      <td>REF2ALT:T&gt;C</td>\n",
       "      <td>UROS</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>UROS#guide_21157#guide_21157:REF2ALT:offset14:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659</th>\n",
       "      <td>UROS_00029</td>\n",
       "      <td>chr10</td>\n",
       "      <td>127477556</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>guide_21158</td>\n",
       "      <td>+</td>\n",
       "      <td>127477564</td>\n",
       "      <td>127477561</td>\n",
       "      <td>-9</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>-9</td>\n",
       "      <td>REF2ALT:T&gt;C</td>\n",
       "      <td>UROS</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...</td>\n",
       "      <td>UROS#guide_21158#guide_21158:REF2ALT:offset14:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>660 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         var_id  chrom  SNP_chr_pos REF  ALT     guide_id guide_strand  \\\n",
       "0    DHFR_10007   DHFR          457   A  [T]  guide_20499            +   \n",
       "1    DHFR_10007   DHFR          457   A  [T]  guide_20500            -   \n",
       "2    DHFR_10008   DHFR          237   C  [T]  guide_20501            +   \n",
       "3    DHFR_10104   DHFR          186   G  [C]  guide_20502            -   \n",
       "4    DHFR_00019   chr5     79929796   G  [C]  guide_20503            +   \n",
       "..          ...    ...          ...  ..  ...          ...          ...   \n",
       "655  UROS_00020  chr10    127477526   G  [A]  guide_21154            +   \n",
       "656  UROS_00020  chr10    127477526   G  [A]  guide_21155            +   \n",
       "657  UROS_00029  chr10    127477556   T  [C]  guide_21156            +   \n",
       "658  UROS_00029  chr10    127477556   T  [C]  guide_21157            +   \n",
       "659  UROS_00029  chr10    127477556   T  [C]  guide_21158            +   \n",
       "\n",
       "     guide0_chr_pos  guide_cut_chr_pos  SNP_pos_in_guide  ...  \\\n",
       "0               460                457                -4  ...   \n",
       "1               449                453                -7  ...   \n",
       "2               241                238                -5  ...   \n",
       "3               180                184                -5  ...   \n",
       "4          79929793           79929790                 2  ...   \n",
       "..              ...                ...               ...  ...   \n",
       "655       127477531          127477528                -6  ...   \n",
       "656       127477534          127477531                -9  ...   \n",
       "657       127477556          127477553                -1  ...   \n",
       "658       127477557          127477554                -2  ...   \n",
       "659       127477564          127477561                -9  ...   \n",
       "\n",
       "     donor_seq_shift donor_mut_pos_in_guide donor_info_str set_name  \\\n",
       "0                 14                     -4    REF2ALT:A>T     DHFR   \n",
       "1                -14                     -7    REF2ALT:A>T     DHFR   \n",
       "2                 14                     -5    REF2ALT:C>T     DHFR   \n",
       "3                -14                     -5    REF2ALT:G>C     DHFR   \n",
       "4                 14                      2    REF2ALT:G>C     DHFR   \n",
       "..               ...                    ...            ...      ...   \n",
       "655               14                     -6    REF2ALT:G>A     UROS   \n",
       "656               14                     -9    REF2ALT:G>A     UROS   \n",
       "657               14                     -1    REF2ALT:T>C     UROS   \n",
       "658               14                     -2    REF2ALT:T>C     UROS   \n",
       "659               14                     -9    REF2ALT:T>C     UROS   \n",
       "\n",
       "    contains_excluded_seqs pool  barcode_seq  barcode_id  \\\n",
       "0                    False   -1          NaN        None   \n",
       "1                    False   -1          NaN        None   \n",
       "2                    False   -1          NaN        None   \n",
       "3                    False   -1          NaN        None   \n",
       "4                    False   -1          NaN        None   \n",
       "..                     ...  ...          ...         ...   \n",
       "655                  False   -1          NaN        None   \n",
       "656                  False   -1          NaN        None   \n",
       "657                  False   -1          NaN        None   \n",
       "658                  False   -1          NaN        None   \n",
       "659                  False   -1          NaN        None   \n",
       "\n",
       "                                             oligo_seq  \\\n",
       "0    GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "1    GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "2    GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "3    GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "4    GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "..                                                 ...   \n",
       "655  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "656  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "657  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "658  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "659  GTTGCAGTTAGCTAACAGGCCATGCNNNNNNNNNNNNGCATGCAGC...   \n",
       "\n",
       "                                              oligo_id  \n",
       "0    DHFR#guide_20499#guide_20499:REF2ALT:offset14:...  \n",
       "1    DHFR#guide_20500#guide_20500:REF2ALT:offset-14...  \n",
       "2    DHFR#guide_20501#guide_20501:REF2ALT:offset14:...  \n",
       "3    DHFR#guide_20502#guide_20502:REF2ALT:offset-14...  \n",
       "4    DHFR#guide_20503#guide_20503:REF2ALT:offset14:...  \n",
       "..                                                 ...  \n",
       "655  UROS#guide_21154#guide_21154:REF2ALT:offset14:...  \n",
       "656  UROS#guide_21155#guide_21155:REF2ALT:offset14:...  \n",
       "657  UROS#guide_21156#guide_21156:REF2ALT:offset14:...  \n",
       "658  UROS#guide_21157#guide_21157:REF2ALT:offset14:...  \n",
       "659  UROS#guide_21158#guide_21158:REF2ALT:offset14:...  \n",
       "\n",
       "[660 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# fritz roth set\n",
    "hyeast_oligos_priority_file = os.path.expanduser(\"~/crispey3/humanized/Output/all_SNPs_humanized_validated_GG_9bp_OLIGO.tab\")\n",
    "hyeast_oligos_priority = pd.read_csv(hyeast_oligos_priority_file, sep='\\t')\n",
    "\n",
    "# Shi-An's consolidated list from gnomAD\n",
    "hyeast_oligos_secondary_file = os.path.expanduser(\"~/crispey3/humanized/Output/all_SNPs_humanized_GG_9bp_OLIGO.tab\")\n",
    "hyeast_oligos_secondary = pd.read_csv(hyeast_oligos_secondary_file, sep='\\t')\n",
    "# calc number of oligos per var_id\n",
    "hyeast_oligos_secondary['num_of_oligos'] = hyeast_oligos_secondary['var_id'].map(hyeast_oligos_secondary.groupby('var_id').size())\n",
    "# sort by set_name, number of oligos, var_id\n",
    "hyeast_oligos_secondary = hyeast_oligos_secondary.sort_values(['set_name', 'num_of_oligos', 'var_id'], ascending=[True, False, True])\n",
    "\n",
    "\n",
    "# to select the final list of oligos for each gene, take everything from the priority list\n",
    "# and add secondary oligos till total number reaches 118 per gene\n",
    "\n",
    "final_oligos_df = []\n",
    "for gene in hyeast_oligos_priority.set_name.unique():\n",
    "    df = hyeast_oligos_priority.query('set_name==@gene')\n",
    "    secondary = hyeast_oligos_secondary.query('set_name==@gene')\n",
    "    df = pd.concat([df, secondary.drop('num_of_oligos', axis=1)])\n",
    "    # remove duplicate variant entries, and set cutoff for oligos\n",
    "    df = df.drop_duplicates(subset=['chrom', 'SNP_chr_pos', 'REF', 'ALT', 'guide0_chr_pos']).head(118)\n",
    "    final_oligos_df.append(df)\n",
    "    \n",
    "final_oligos_df = pd.concat(final_oligos_df).reset_index(drop=True)\n",
    "# relabel guide IDs\n",
    "guide_num = 20499\n",
    "for i, row in final_oligos_df.iterrows():\n",
    "    old_guide_id = row['guide_id']\n",
    "    new_guide_id = 'guide_{}'.format(guide_num)\n",
    "    final_oligos_df.loc[i,['guide_id', 'donor_id', 'oligo_id']] = final_oligos_df.loc[i,['guide_id', 'donor_id', 'oligo_id']].apply(str.replace, args=(old_guide_id, new_guide_id))\n",
    "    guide_num+=1\n",
    "\n",
    "display(final_oligos_df)\n",
    "\n",
    "# write to file\n",
    "final_oligos_df.to_csv(os.path.expanduser(\"~/crispey3/humanized/Output/all_SNPs_humanized_combined_GG_9bp_OLIGO.tab\"), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set_name\n",
       "DHFR       70\n",
       "DPAGT1    118\n",
       "NSDHL     118\n",
       "PGK1      118\n",
       "PKLR      118\n",
       "UROS      118\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# number of oligos per gene\n",
    "display(final_oligos_df.groupby('set_name').size())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write all variant records to new VCF file for final oligo synthesis into library\n",
    "- ERG/EG*: ergosterol set variants\n",
    "- GXG: GxG main set variants\n",
    "- TDH: TDH3 promoter variants\n",
    "- VAL: Significant hits in at least 3 strains (set 1) and large effect variants identified from at least one strain (set 2)\n",
    "- GXE: GxE main set variants\n",
    "- HSP/HSX: HSP90 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # REWRITE THIS\n",
    "# gxg_known_fitness_vcf_file = \"/home/users/rang/crispey3/costanzo_variants/Input/gxg_known_fitness_280820.vcf\"\n",
    "\n",
    "# candVars_allStrains_AFfilter = vcf.Reader(filename=candVars_allStrains_AFfilter_file)\n",
    "# gxg_known_fitness_vcf = vcf.Writer(open(gxg_known_fitness_vcf_file, 'w'), candVars_allStrains_AFfilter)\n",
    "# for record in pd.concat([prom_var_list, epi_var_list]):\n",
    "#     gxg_known_fitness_vcf.write_record(record)\n",
    "# gxg_known_fitness_vcf.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize other control oligos to design\n",
    "These oligos should have guide and donor sequences that have been pre-designed by \n",
    "- Neutral, non-targeting oligos (scrambled donors from old libraries, see \"tech_donor_scramble\")\n",
    "- Neutral, targeting oligos (???)\n",
    "- fitness ladder\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # from here we can try to focus on genes with at least 1 common variant\n",
    "# num_of_vars_cutoff = 1\n",
    "# genes_df_filtered = genes_df.query('num_of_vars_common>=@num_of_vars_cutoff')\n",
    "\n",
    "# # are the genes in either class comparable in terms of their number of variants, or their variant density?\n",
    "# display(genes_df_filtered.groupby('gene_class').apply(lambda x: x.num_of_vars.describe()))\n",
    "# display(genes_df_filtered.groupby('gene_class').apply(lambda x: x.region_length.describe()))\n",
    "# display(genes_df_filtered.groupby('gene_class').apply(lambda x: x.var_density.describe()))\n",
    "\n",
    "# # visualize genes by their number of variants relative to their length\n",
    "# fig, ax = plt.subplots(figsize=(8,6))\n",
    "# genes_df_filtered.query('num_of_vars>0').plot('region_length', 'num_of_vars', kind='scatter', ax=ax)\n",
    "# ax.tick_params(labelsize=14)\n",
    "# ax.set_xlabel('gene region length', fontsize=16)\n",
    "# ax.set_ylabel('number of variants', fontsize=16)\n",
    "# ax.set_title('number of variants vs gene region length', fontsize=16)\n",
    "# plt.show()\n",
    "\n",
    "# # distribution of allele frequencies\n",
    "# fig, ax = plt.subplots(figsize=(8,6))\n",
    "# ax.hist(np.log10(final_multiGuides.AF), bins=20)\n",
    "# ax.tick_params(labelsize=14)\n",
    "# ax.set_xlabel('log10(AF)', fontsize=16)\n",
    "# ax.set_ylabel('frequency', fontsize=16)\n",
    "# ax.set_title('asymmetrical distribution of variant AF', fontsize=16)\n",
    "# plt.show()\n",
    "\n",
    "# # distribution of variant annotation categories\n",
    "# display((final_multiGuides.region.value_counts()/len(final_multiGuides)).head())\n",
    "\n",
    "\n",
    "# # common variants median AF\n",
    "# display(final_vars_selected.sort_values('AF').head(1008).AF.median())\n",
    "# # rare variants median AF\n",
    "# display(final_vars_selected.sort_values('AF').tail(1008).AF.median())\n",
    "# # fraction essential\n",
    "# display(hub_genes.essential.sum()/len(hub_genes))\n",
    "# # fraction paralogs\n",
    "# display(hub_genes.paralog.sum()/len(hub_genes))\n",
    "\n",
    "# # variants in essential genes\n",
    "# ve = final_vars_selected[final_vars_selected.assoc_gene.apply(lambda x: any([g in genes_df.loc[genes_df.essential].index for g in x]))]\n",
    "# display(ve)\n",
    "# # distribution of variant annotation types\n",
    "# display((ve.region.value_counts()/len(ve)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # from here we can try to focus on genes with at least 1 common variant\n",
    "# num_of_vars_cutoff = 1\n",
    "# genes_df_filtered = genes_df.query('num_of_vars_common>=@num_of_vars_cutoff')\n",
    "\n",
    "# # are the genes in either class comparable in terms of their number of variants, or their variant density?\n",
    "# display(genes_df_filtered.groupby('gene_class').apply(lambda x: x.num_of_vars.describe()))\n",
    "# display(genes_df_filtered.groupby('gene_class').apply(lambda x: x.region_length.describe()))\n",
    "# display(genes_df_filtered.groupby('gene_class').apply(lambda x: x.var_density.describe()))\n",
    "\n",
    "# # visualize genes by their number of variants relative to their length\n",
    "# fig, ax = plt.subplots(figsize=(8,6))\n",
    "# genes_df_filtered.query('num_of_vars>0').plot('region_length', 'num_of_vars', kind='scatter', ax=ax)\n",
    "# ax.tick_params(labelsize=14)\n",
    "# ax.set_xlabel('gene region length', fontsize=16)\n",
    "# ax.set_ylabel('number of variants', fontsize=16)\n",
    "# ax.set_title('number of variants vs gene region length', fontsize=16)\n",
    "# plt.show()\n",
    "\n",
    "# # distribution of allele frequencies\n",
    "# fig, ax = plt.subplots(figsize=(8,6))\n",
    "# ax.hist(np.log10(final_multiGuides.AF), bins=20)\n",
    "# ax.tick_params(labelsize=14)\n",
    "# ax.set_xlabel('log10(AF)', fontsize=16)\n",
    "# ax.set_ylabel('frequency', fontsize=16)\n",
    "# ax.set_title('asymmetrical distribution of variant AF', fontsize=16)\n",
    "# plt.show()\n",
    "\n",
    "# # distribution of variant annotation categories\n",
    "# display((final_multiGuides.region.value_counts()/len(final_multiGuides)).head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 482,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cutoff = 1000\n",
    "# # hub genes\n",
    "# hub_genes = genes_df.query('gene_class==\"hub\" & num_of_vars_common<=0.5*num_of_vars').sort_values(['num_of_vars_common', 'num_of_vars_rare'], ascending=False)\n",
    "# common_vars_cumsum = np.cumsum(hub_genes.num_of_vars_common)\n",
    "# hub_genes = hub_genes[:(common_vars_cumsum>cutoff).idxmax()]\n",
    "# print('Number of hub genes selected:', len(hub_genes))\n",
    "\n",
    "# # peripheral genes\n",
    "# peripheral_genes = genes_df.query('gene_class==\"peripheral\"').sort_values(['num_of_vars_common', 'num_of_vars_rare'], ascending=False)\n",
    "# common_vars_cumsum = np.cumsum(peripheral_genes.num_of_vars_common)\n",
    "# peripheral_genes = peripheral_genes[:(common_vars_cumsum>cutoff).idxmax()]\n",
    "# print('Number of peripheral genes selected:', len(peripheral_genes))\n",
    "\n",
    "\n",
    "# # with selected hub genes and peripheral genes, we iterate over these genes to select variants\n",
    "# genes_list = hub_genes.index.tolist()# + peripheral_genes.index.tolist()\n",
    "# var_list = []\n",
    "# for gene in genes_list:\n",
    "#     df = final_multiGuides[final_multiGuides['assoc_gene'].apply(lambda x: gene in x)].sort_values(['AF', 'AC'], ascending=False)\n",
    "#     com_var = df.query('AF>=@af_cutoff_common').index.tolist()\n",
    "#     rar_var = df.tail(len(com_var)).index.tolist()\n",
    "    \n",
    "#     # add selected variants to var_list\n",
    "#     var_list += com_var\n",
    "#     var_list += rar_var\n",
    "\n",
    "# # select final list of variants that will be targeted\n",
    "# final_vars_selected = final_multiGuides.query('var_id.isin(@var_list)')\n",
    "# display(final_vars_selected)\n",
    "\n",
    "# # distribution of variant annotation types\n",
    "# display((final_vars_selected.region.value_counts()/len(final_vars_selected)))\n",
    "\n",
    "# # how many oligos are we expecting from this list of variants?\n",
    "# print(\"Number of oligos from hub-peripheral gene set:\", int(final_vars_selected.num_of_oligos.sum()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# v_list = [\"YOR341W\", \"YLR430W\"]\n",
    "# fig, axes = plt.subplots(figsize=(12,6), ncols=2, nrows=1)\n",
    "\n",
    "# for i in range(len(v_list)):\n",
    "#     v = v_list[i]\n",
    "#     ax = axes[i]\n",
    "    \n",
    "#     test = final[final['gene_group'].apply(lambda x: v in x)]\n",
    "#     display(test.AF.describe())\n",
    "\n",
    "#     ax.hist(np.log10(test.AF), bins=np.arange(-3, 0, 0.2))\n",
    "#     ax.axvline(np.log10(af_cutoff), color='gray')\n",
    "#     ax.tick_params(labelsize=14)\n",
    "#     ax.set_xlabel('log10(AF)', fontsize=14)\n",
    "#     ax.set_ylabel('frequency', fontsize=14)\n",
    "#     ax.set_title('AF distribution in {}'.format(v), fontsize=16)\n",
    "#     ax.set_xlim(-3, 0)\n",
    "#     ax.set_ylim(0, 50)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viz = genes_df_filtered[['gene_class', 'positive_interactions', 'negative_interactions', 'num_of_vars_common', 'num_of_vars_rare']].copy()\n",
    "# viz.loc[:,'num_interactions'] = viz['positive_interactions'] + viz['negative_interactions']\n",
    "# viz.loc[:,'common_rare_ratio'] = viz['num_of_vars_common'] / viz['num_of_vars_rare']\n",
    "# # viz.boxplot(column='common_rare_ratio', by='gene_class')\n",
    "\n",
    "# # visualize genes by their common-rare ratio relative to their number of interactions \n",
    "# fig, ax = plt.subplots(figsize=(8,6))\n",
    "# viz.plot('num_interactions', 'common_rare_ratio', kind='scatter', ax=ax)\n",
    "# ax.tick_params(labelsize=14)\n",
    "# ax.set_xlabel('number of interactions', fontsize=16)\n",
    "# ax.set_ylabel('common-rare ratio', fontsize=16)\n",
    "# ax.set_title('common-rare ratio vs number of interactions', fontsize=16)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # get annotation from annotated VCF from VEP\n",
    "# variants_final_annotated_file = candVars_allStrains_AFfilter_file.replace(\".vcf\", \"_annotated.vcf\")\n",
    "# variants_final_annotated = pd.Series([rec for rec in vcf.Reader(filename=variants_final_annotated_file)])\n",
    "# display(variants_final_annotated)\n",
    "\n",
    "\n",
    "# # how to handle multiple annotations?\n",
    "# # if coding annotation detected, always take coding annotation\n",
    "# # if noncoding, then select annotation associated with gene of interest\n",
    "# noncoding_annotation_type = ['upstream_gene_variant', 'downstream_gene_variant']\n",
    "# i=0\n",
    "# for rec in variants_final_annotated:\n",
    "#     if len(rec.INFO['CSQ'])>1:\n",
    "#         i+=1\n",
    "#         print(rec)\n",
    "#         print([x.split('|')[:8] for x in rec.INFO['CSQ']])\n",
    "#         var_anno_type = [x.split('|')[1] for x in rec.INFO['CSQ']]\n",
    "#         if all([anno in noncoding_annotation_type for anno in var_anno_type]):\n",
    "#             # intergenic\n",
    "#             # determine which two nearest genes upstream and downstream\n",
    "#         else:\n",
    "#             # take coding annotation\n",
    "#             rec.INFO['CSQ'] = [x for x in rec.INFO['CSQ'] if x.split('|')[1] not in noncoding_annotation_type]\n",
    "#         print()\n",
    "#         if i==5:\n",
    "#             break\n",
    "\n",
    "\n",
    "\n",
    "# noncoding_annotation_type = ['upstream_gene_variant', 'downstream_gene_variant']\n",
    "# for rec in variants_final_annotated:\n",
    "#     if len(rec.INFO['CSQ'])>4:\n",
    "#         print(rec)\n",
    "#         print(rec.INFO['CSQ'])\n",
    "#         print([x for x in rec.INFO['CSQ'] if x.split('|')[1] not in noncoding_annotation_type])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# candVars_allStrains_AFfilter = pd.Series([rec for rec in vcf.Reader(filename=candVars_allStrains_AFfilter_file)])\n",
    "# assemble = {}\n",
    "# assemble['CHROM'] = [rec.CHROM for rec in candVars_allStrains_AFfilter]\n",
    "# assemble['POS'] = [rec.POS for rec in candVars_allStrains_AFfilter]\n",
    "# assemble['REF'] = [rec.REF for rec in candVars_allStrains_AFfilter]\n",
    "# assemble['ALT'] = [rec.ALT for rec in candVars_allStrains_AFfilter]\n",
    "# assemble['AC'] = [rec.INFO['AC'] for rec in candVars_allStrains_AFfilter]\n",
    "# assemble['AN'] = [rec.INFO['AN'] for rec in candVars_allStrains_AFfilter]\n",
    "# assemble['AF'] = [rec.INFO['AF'] for rec in candVars_allStrains_AFfilter]\n",
    "\n",
    "# candVars_allStrains_AFfilter = pd.DataFrame.from_dict(assemble)\n",
    "# display(candVars_allStrains_AFfilter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # counters\n",
    "# total=0 # total num of variants processed\n",
    "# edgeVar_count=0 # num of variants skipped due to incomplete sequence window at chromosome edge\n",
    "# noNGG_count=0 # num of variants without NGG\n",
    "# nonunique_count=0 # num of variants skipped with nonunique sequence in reference genome\n",
    "# count_allStrains=0 # num of variants editable in all strains\n",
    "# count_refOnly=0 # num of variants editable in reference strain only\n",
    "\n",
    "# threshold = 10 # threshold value for searching NGGs\n",
    "# window = 91 # window size to check for unique sequence\n",
    "\n",
    "# # open starting set of potential candidate variants \n",
    "# gvcf = vcf.Reader(filename=candVars_file)\n",
    "# # open candVars_allStrains and candVars_refOnly files for writing\n",
    "# candVars_allStrains = vcf.Writer(open(candVars_allStrains_file, 'w'), gvcf)\n",
    "# candVars_refOnly = vcf.Writer(open(candVars_refOnly_file, 'w'), gvcf)\n",
    "\n",
    "# for record in gvcf:\n",
    "#     total+=1\n",
    "#     if total % 10 == 0:\n",
    "#         print(\"{} candidate variants parsed.\".format(total))\n",
    "#     edits_in_ref_strain = False\n",
    "#     edits_in_nonref_strain = False\n",
    "\n",
    "#     ##### FILTERING #####\n",
    "#     # 1. Skip variant if near chromosome edge\n",
    "#     seq_window = ref_genome[record.CHROM][record.POS-window//2-1:record.POS+window//2]\n",
    "#     if len(seq_window) != window:\n",
    "#         edgeVar_count+=1\n",
    "#         continue\n",
    "\n",
    "#     # 2. is there an 'NGG' sequence around the center of the seq_window?\n",
    "#     if not ngg_check(sequence=seq_window.seq, threshold=threshold):\n",
    "#         noNGG_count+=1\n",
    "#         continue\n",
    "\n",
    "#     # 3. check genomes if sequence flanking variant (seq_window) is unique\n",
    "#     unique_seq_status=[search_genome(seq_window.seq, genome) for genome in genomes_list]\n",
    "#     if unique_seq_status[0]==1: # reference genome\n",
    "#         edits_in_ref_strain = True\n",
    "#         if all([i==1 for i in unique_seq_status]): # reference genome + other genomes\n",
    "#             edits_in_nonref_strain = True\n",
    "#     else:\n",
    "#         nonunique_count+=1\n",
    "#         continue\n",
    "\n",
    "#     # 4. store variant info according to whether it can be edited in all strains, or just reference strain\n",
    "#     if edits_in_ref_strain:\n",
    "#         if edits_in_nonref_strain:\n",
    "#             count_allStrains+=1\n",
    "#             candVars_allStrains.write_record(record)\n",
    "#         else:\n",
    "#             count_refOnly+=1\n",
    "#             candVars_refOnly.write_record(record)\n",
    "#     else:\n",
    "#         print(\"Variant skipped: Cannot be edited in reference strain\") # sanity check step: should not appear in output as filters in previous steps should remove all such instances\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "# candVars_allStrains.close()\n",
    "# candVars_refOnly.close()\n",
    "\n",
    "# print()\n",
    "# print(\"Total number of variants parsed:\", total)\n",
    "# print(\"Variants skipped due to incomplete sequence window:\", edgeVar_count)\n",
    "# print(\"Variants without NGG site:\", noNGG_count)\n",
    "# print(\"Variants with non-unique sequence region:\", nonunique_count)\n",
    "# print(\"Number of variants editable in all strains:\", count_allStrains)\n",
    "# print(\"Number of variants editable in reference strain only:\", count_refOnly)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# as a sanity check, we can inspect the variants in candVar_allStrains to ensure\n",
    "# they do not appear in our list of known variants for non-reference strains\n",
    "\n",
    "# def ref_allele_check(record, var_df):\n",
    "#     '''\n",
    "#     checks var_df to see if a variant (alt allele) exists at the record's coordinates\n",
    "#     returns True if no alt alleles found (i.e. ref allele present)\n",
    "#     '''\n",
    "#     ref_allele = len(var_df.query('(CHROM==@record.CHROM) & (POS==@record.POS)'))==0\n",
    "#     return ref_allele\n",
    "\n",
    "# # consolidated variant files put together with consolidate_(strain_name)var.py\n",
    "# rm_var_file = \"/home/users/rang/yeast/consolidated_vars/RM_variants_complete.txt\"\n",
    "# yps_var_file = \"/home/users/rang/yeast/consolidated_vars/YPS128_variants_complete.txt\"\n",
    "# yjm_var_file = \"/home/users/rang/yeast/consolidated_vars/YJM789_variants_complete.txt\"\n",
    "\n",
    "# # list containing strain names (also the order which they are processed at each filtering step)\n",
    "# strains_list = [\"BY4742\", \"RM11-1a\", \"YPS128\", \"YJM789\"]\n",
    "\n",
    "# # store variants files in pandas dataframes\n",
    "# rm_var_df = pd.read_csv(rm_var_file, sep=\"\\t\", index_col=0)\n",
    "# yps_var_df = pd.read_csv(yps_var_file, sep=\"\\t\", index_col=0)\n",
    "# yjm_var_df = pd.read_csv(yjm_var_file, sep=\"\\t\", index_col=0)\n",
    "# variants_df_list = [rm_var_df, yps_var_df, yjm_var_df]\n",
    "\n",
    "\n",
    "#   # 0. Inspect if variant already exists in RM and/or YPS and/or YJM\t\t# REDUNDANT STEP\n",
    "#   check_ref_in_variants_dfs = [(record, df) for for df in variants_df_list]\n",
    "#   num_of_cores = min(len(os.sched_getaffinity(0)), len(check_ref_in_variants_dfs))\n",
    "#   with mp.Pool(num_of_cores) as pool:\n",
    "#      ref_allele_status = pool.starmap(ref_allele_check, check_ref_in_variants_dfs)\n",
    "#   if not all(ref_allele_status):\n",
    "#      #print(\"At least one strain has non-reference allele at chrom {}, pos {}\".format(record.CHROM, record.POS))\n",
    "#   continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # ras genes that Jose is working on\n",
    "# ras_genes = ['YOR101W',\n",
    "# 'YNL098C',\n",
    "# 'YLR310C',\n",
    "# 'YLL016W',\n",
    "# 'YBR140C',\n",
    "# 'YOL081W',\n",
    "# 'YJL005W',\n",
    "# 'YGL248W',\n",
    "# 'YOR360C',\n",
    "# 'YLR178C',\n",
    "# 'YIL033C',\n",
    "# 'YJL164C',\n",
    "# 'YPL203W',\n",
    "# 'YKL166C',\n",
    "# 'YER020W',\n",
    "# 'YOR107W',\n",
    "# 'YOR371C',\n",
    "# 'YAL056W',\n",
    "# 'YDL035C',\n",
    "# 'YHR186C',\n",
    "# 'YJR066W',\n",
    "# 'YKL203C',\n",
    "# 'YNL006W',\n",
    "# 'YPL180W',\n",
    "# 'YHR205W']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (crispey-bar)",
   "language": "python",
   "name": "crispey-bar"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
